{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Copy Number Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import os.path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from IPython.display import Image,display\n",
    "import dalmatian as dm\n",
    "from IPython.core.display import HTML \n",
    "from src.CCLE_postp_function import *\n",
    "import sys\n",
    "sys.path.insert(0, '../JKBio/')\n",
    "import TerraFunction as terra\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "import rpy2\n",
    "from taigapy import TaigaClient\n",
    "tc = TaigaClient()\n",
    "from gsheets import Sheets\n",
    "sheets = Sheets.from_files('~/.client_secret.json', '~/.storage.json')\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "workspace1=\"broad-genomics-delivery/Getz_IBM_CellLines_Exomes\"\n",
    "workspace2=\"broad-firecloud-ccle/CCLE_DepMap_WES\"\n",
    "workspace3=\"broad-genomics-delivery/CCLE_DepMap_WES\"\n",
    "refworkspace=\"broad-firecloud-ccle/DepMap_WES_CN_hg38\"\n",
    "source1=\"ibm\"\n",
    "source2=\"ccle\"\n",
    "source3=\"ccle\"\n",
    "sample_set_id = \"19Q4\"\n",
    "release = sample_set_id\n",
    "sheeturl = \"https://docs.google.com/spreadsheets/d/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "genome_version <- 'hg38'\n",
    "release <- '19Q4'\n",
    "hg38_cyto_band_reference <- '../JKBio/data/hg38_cytoband.gz'\n",
    "new_samples_copy_number_broad_wes <- 'temp/cnv_ccle.called.seg'\n",
    "prevfile <- 'wes.19Q3.segmented'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "wm1 = dm.WorkspaceManager(workspace1)\n",
    "wm2 = dm.WorkspaceManager(workspace2)\n",
    "wm3 = dm.WorkspaceManager(workspace3)\n",
    "\n",
    "refwm = dm.WorkspaceManager(refworkspace)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py:78: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  (samples[extract['id']].isin(forcekeep))]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(92)createDatasetWithNewCellLines()\n",
      "-> if gsfolderto is not None:\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(93)createDatasetWithNewCellLines()\n",
      "-> for i, val in sampless.iterrows():\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(94)createDatasetWithNewCellLines()\n",
      "-> res = os.system('gsutil cp ' + val[extract['bam']] + ' ' + val[extract['bai']] + ' ' + gsfolderto)\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(95)createDatasetWithNewCellLines()\n",
      "-> if res == 256:\n",
      "(Pdb) res\n",
      "0\n",
      "(Pdb) sampless\n",
      "                              sample_id       LC-SET        PDO aligned_reads  \\\n",
      "82          RP-1561_921_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "106       RP-1561_HOTHC_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "136      RP-1561_OCIP5X_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "139        RP-1561_P4E6_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "130     RP-1561_SUMB002_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "150  RP-1561_UPCISCC131_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "153      RP-1561_WM4235_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "\n",
      "    at_dropout bad_cycles                        bait_set category  \\\n",
      "82           0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "106          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "136          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "139          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "130          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "150          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "153          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "\n",
      "                                      crai_or_bai_path  \\\n",
      "82   gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...   \n",
      "106  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...   \n",
      "136  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...   \n",
      "139  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...   \n",
      "130  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...   \n",
      "150  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...   \n",
      "153  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...   \n",
      "\n",
      "                                      cram_or_bam_path  ...  \\\n",
      "82   gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...  ...   \n",
      "106  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...  ...   \n",
      "136  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...  ...   \n",
      "139  gs://fc-11136d2a-c57a-4e95-ac0e-ff6850e8db54/G...  ...   \n",
      "130  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...  ...   \n",
      "150  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...  ...   \n",
      "153  gs://fc-19945ed2-3eca-4a49-a1f2-9bf13d07d011/D...  ...   \n",
      "\n",
      "    library-3_estimated_library_size library-3_name  \\\n",
      "82                               NaN            NaN   \n",
      "106                              NaN            NaN   \n",
      "136                              NaN            NaN   \n",
      "139                              NaN            NaN   \n",
      "130                                                   \n",
      "150                                                   \n",
      "153                                                   \n",
      "\n",
      "    library-3_percent_duplication library-3_read_pair_duplicates  \\\n",
      "82                            NaN                            NaN   \n",
      "106                           NaN                            NaN   \n",
      "136                           NaN                            NaN   \n",
      "139                           NaN                            NaN   \n",
      "130                                                                \n",
      "150                                                                \n",
      "153                                                                \n",
      "\n",
      "    library-3_read_pair_optical_duplicates library-3_read_pairs_examined  \\\n",
      "82                                     NaN                           NaN   \n",
      "106                                    NaN                           NaN   \n",
      "136                                    NaN                           NaN   \n",
      "139                                    NaN                           NaN   \n",
      "130                                                                        \n",
      "150                                                                        \n",
      "153                                                                        \n",
      "\n",
      "    library-3_secondary_or_supplementary_rds library-3_unmapped_reads  \\\n",
      "82                                       NaN                      NaN   \n",
      "106                                      NaN                      NaN   \n",
      "136                                      NaN                      NaN   \n",
      "139                                      NaN                      NaN   \n",
      "130                                                                     \n",
      "150                                                                     \n",
      "153                                                                     \n",
      "\n",
      "    library-3_unpaired_read_duplicates library-3_unpaired_reads_examined  \n",
      "82                                 NaN                               NaN  \n",
      "106                                NaN                               NaN  \n",
      "136                                NaN                               NaN  \n",
      "139                                NaN                               NaN  \n",
      "130                                                                       \n",
      "150                                                                       \n",
      "153                                                                       \n",
      "\n",
      "[7 rows x 246 columns]\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(93)createDatasetWithNewCellLines()\n",
      "-> for i, val in sampless.iterrows():\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(94)createDatasetWithNewCellLines()\n",
      "-> res = os.system('gsutil cp ' + val[extract['bam']] + ' ' + val[extract['bai']] + ' ' + gsfolderto)\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(95)createDatasetWithNewCellLines()\n",
      "-> if res == 256:\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(93)createDatasetWithNewCellLines()\n",
      "-> for i, val in sampless.iterrows():\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(94)createDatasetWithNewCellLines()\n",
      "-> res = os.system('gsutil cp ' + val[extract['bam']] + ' ' + val[extract['bai']] + ' ' + gsfolderto)\n",
      "(Pdb) unt\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(95)createDatasetWithNewCellLines()\n",
      "-> if res == 256:\n",
      "(Pdb) unt\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(98)createDatasetWithNewCellLines()\n",
      "-> sampless[extract['bam']] = [gsfolderto + a.split('/')[-1] for a in sampless[extract['bam']]]\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(99)createDatasetWithNewCellLines()\n",
      "-> sampless[extract['bai']] = [gsfolderto + a.split('/')[-1] for a in sampless[extract['bai']]]\n",
      "(Pdb) n\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(100)createDatasetWithNewCellLines()\n",
      "-> print(sampless)\n",
      "(Pdb) sampless[extract['bai']]\n",
      "82            gs://ccle_bams/wes/921.bai\n",
      "106         gs://ccle_bams/wes/HOTHC.bai\n",
      "136        gs://ccle_bams/wes/OCIP5X.bai\n",
      "139          gs://ccle_bams/wes/P4E6.bai\n",
      "130       gs://ccle_bams/wes/SUMB002.bai\n",
      "150    gs://ccle_bams/wes/UPCISCC131.bai\n",
      "153        gs://ccle_bams/wes/WM4235.bai\n",
      "Name: crai_or_bai_path, dtype: object\n",
      "(Pdb) n\n",
      "                              sample_id       LC-SET        PDO aligned_reads  \\\n",
      "82          RP-1561_921_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "106       RP-1561_HOTHC_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "136      RP-1561_OCIP5X_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "139        RP-1561_P4E6_v1_Exome_OnPrem  LCSET-16758  PDO-19525             0   \n",
      "130     RP-1561_SUMB002_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "150  RP-1561_UPCISCC131_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "153      RP-1561_WM4235_v2_Exome_OnPrem  LCSET-16709  PDO-19393             0   \n",
      "\n",
      "    at_dropout bad_cycles                        bait_set category  \\\n",
      "82           0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "106          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "136          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "139          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "130          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "150          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "153          0          0  whole_exome_illumina_coding_v1     PAIR   \n",
      "\n",
      "                      crai_or_bai_path                   cram_or_bam_path  \\\n",
      "82          gs://ccle_bams/wes/921.bai         gs://ccle_bams/wes/921.bam   \n",
      "106       gs://ccle_bams/wes/HOTHC.bai       gs://ccle_bams/wes/HOTHC.bam   \n",
      "136      gs://ccle_bams/wes/OCIP5X.bai      gs://ccle_bams/wes/OCIP5X.bam   \n",
      "139        gs://ccle_bams/wes/P4E6.bai        gs://ccle_bams/wes/P4E6.bam   \n",
      "130     gs://ccle_bams/wes/SUMB002.bai     gs://ccle_bams/wes/SUMB002.bam   \n",
      "150  gs://ccle_bams/wes/UPCISCC131.bai  gs://ccle_bams/wes/UPCISCC131.bam   \n",
      "153      gs://ccle_bams/wes/WM4235.bai      gs://ccle_bams/wes/WM4235.bam   \n",
      "\n",
      "     ... library-3_estimated_library_size library-3_name  \\\n",
      "82   ...                              NaN            NaN   \n",
      "106  ...                              NaN            NaN   \n",
      "136  ...                              NaN            NaN   \n",
      "139  ...                              NaN            NaN   \n",
      "130  ...                                                   \n",
      "150  ...                                                   \n",
      "153  ...                                                   \n",
      "\n",
      "    library-3_percent_duplication library-3_read_pair_duplicates  \\\n",
      "82                            NaN                            NaN   \n",
      "106                           NaN                            NaN   \n",
      "136                           NaN                            NaN   \n",
      "139                           NaN                            NaN   \n",
      "130                                                                \n",
      "150                                                                \n",
      "153                                                                \n",
      "\n",
      "    library-3_read_pair_optical_duplicates library-3_read_pairs_examined  \\\n",
      "82                                     NaN                           NaN   \n",
      "106                                    NaN                           NaN   \n",
      "136                                    NaN                           NaN   \n",
      "139                                    NaN                           NaN   \n",
      "130                                                                        \n",
      "150                                                                        \n",
      "153                                                                        \n",
      "\n",
      "    library-3_secondary_or_supplementary_rds library-3_unmapped_reads  \\\n",
      "82                                       NaN                      NaN   \n",
      "106                                      NaN                      NaN   \n",
      "136                                      NaN                      NaN   \n",
      "139                                      NaN                      NaN   \n",
      "130                                                                     \n",
      "150                                                                     \n",
      "153                                                                     \n",
      "\n",
      "    library-3_unpaired_read_duplicates library-3_unpaired_reads_examined  \n",
      "82                                 NaN                               NaN  \n",
      "106                                NaN                               NaN  \n",
      "136                                NaN                               NaN  \n",
      "139                                NaN                               NaN  \n",
      "130                                                                       \n",
      "150                                                                       \n",
      "153                                                                       \n",
      "\n",
      "[7 rows x 246 columns]\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(101)createDatasetWithNewCellLines()\n",
      "-> for ind, val in sampless.iterrows():\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Pdb) unt\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(102)createDatasetWithNewCellLines()\n",
      "-> name = val[extract['source']] + '_' + val[extract['id']][:participantslicepos]\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(103)createDatasetWithNewCellLines()\n",
      "-> refsamples = refsamples.append(pd.Series(\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(105)createDatasetWithNewCellLines()\n",
      "-> \"CCLE_name\": val[extract['name']],\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(106)createDatasetWithNewCellLines()\n",
      "-> \"WES_bai\": val[extract['bai']],\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(107)createDatasetWithNewCellLines()\n",
      "-> \"WES_bam\": val[extract['bam']],\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(108)createDatasetWithNewCellLines()\n",
      "-> \"Source\": val[extract['source']],\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(109)createDatasetWithNewCellLines()\n",
      "-> \"participant\": val[extract['participant']][:participantslicepos],\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(111)createDatasetWithNewCellLines()\n",
      "-> }, name=name))\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(112)createDatasetWithNewCellLines()\n",
      "-> sample_ids.append(name)\n",
      "(Pdb) \n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(113)createDatasetWithNewCellLines()\n",
      "-> print(\"uploading new samples\")\n",
      "(Pdb) \n",
      "uploading new samples\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(114)createDatasetWithNewCellLines()\n",
      "-> wto.upload_samples(refsamples)\n",
      "(Pdb) refsamples\n",
      "                                                      CCLE_name  Source  \\\n",
      "sample_id                                                                 \n",
      "ccle2_ACH-000011                             253J_URINARY_TRACT   ccle2   \n",
      "ccle2_ACH-000013                                  ONCODG1_OVARY   ccle2   \n",
      "ccle2_ACH-000020    MHHCALL2_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000026                           253JBV_URINARY_TRACT   ccle2   \n",
      "ccle2_ACH-000027                    GOS3_CENTRAL_NERVOUS_SYSTEM   ccle2   \n",
      "ccle2_ACH-000031                              PANC0213_PANCREAS   ccle2   \n",
      "ccle2_ACH-000032    MHHCALL3_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000039                                     SKNMC_BONE   ccle2   \n",
      "ccle2_ACH-000041                                      RDES_BONE   ccle2   \n",
      "ccle2_ACH-000042                              PANC0203_PANCREAS   ccle2   \n",
      "ccle2_ACH-000043                              HS895T_FIBROBLAST   ccle2   \n",
      "ccle2_ACH-000051                             TE617T_SOFT_TISSUE   ccle2   \n",
      "ccle2_ACH-000063                              HS616T_FIBROBLAST   ccle2   \n",
      "ccle2_ACH-000065     OCIAML5_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000066                                   HCC4006_LUNG   ccle2   \n",
      "ccle2_ACH-000069      HS611T_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000073   GRANTA519_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000076        NCO2_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000077          MJ_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000079                              TE125T_FIBROBLAST   ccle2   \n",
      "ccle2_ACH-000080        BDCM_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000081        GDM1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "ccle2_ACH-000083                              HS281T_FIBROBLAST   ccle2   \n",
      "ccle2_ACH-000085                                  T3M4_PANCREAS   ccle2   \n",
      "ccle2_ACH-000086                                ACCMESO1_PLEURA   ccle2   \n",
      "ccle2_ACH-000089                                  NCIH684_LIVER   ccle2   \n",
      "ccle2_ACH-000093                              PANC0504_PANCREAS   ccle2   \n",
      "ccle2_ACH-000097                                   ZR751_BREAST   ccle2   \n",
      "ccle2_ACH-000099                         SIMA_AUTONOMIC_GANGLIA   ccle2   \n",
      "ccle2_ACH-000105      ALLSIL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE   ccle2   \n",
      "...                                                         ...     ...   \n",
      "sanger_ACH-002367               NCIBL2052_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002368               NCIBL2087_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002369                NCIBL209_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002370               NCIBL2122_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002371               NCIBL2126_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002372               NCIBL2171_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002373                 HCC1187_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002374                 HCC1599_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002375                 HCC1937_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002376                  LS1034_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002377                   HCC38_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002378                 HCC1143_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002379                     J82_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002380                 COLO829_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002381                 HCC2157_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002382                 HCC1395_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002383                 HCC2218_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002384                 HCC1954_MATCHED_NORMAL_TISSUE  sanger   \n",
      "sanger_ACH-002386                                   M00921_SKIN  sanger   \n",
      "sanger_ACH-002387                                 M1203273_SKIN  sanger   \n",
      "sanger_ACH-002388                                    MET2B_SKIN  sanger   \n",
      "sanger_ACH-002389                         ACN_AUTONOMIC_GANGLIA  sanger   \n",
      "sanger_ACH-002397                                  KMH2_THYROID  sanger   \n",
      "ibm_ACH-001441                                              921     ibm   \n",
      "ibm_ACH-002041                                            HOTHC     ibm   \n",
      "ibm_ACH-001370                                           OCIP5X     ibm   \n",
      "ibm_ACH-001627                                             P4E6     ibm   \n",
      "ccle_ACH-001201                                         SUMB002    ccle   \n",
      "ccle_ACH-001699                                      UPCISCC131    ccle   \n",
      "ccle_ACH-002509                                          WM4235    ccle   \n",
      "\n",
      "                                                             WES_bai  \\\n",
      "sample_id                                                              \n",
      "ccle2_ACH-000011              gs://ccle_bams/wes/C836.253J.1.bam.bai   \n",
      "ccle2_ACH-000013         gs://ccle_bams/wes/C836.ONCO-DG-1.2.bam.bai   \n",
      "ccle2_ACH-000020        gs://ccle_bams/wes/C836.MHH-CALL-2.1.bam.bai   \n",
      "ccle2_ACH-000026           gs://ccle_bams/wes/C836.253J-BV.4.bam.bai   \n",
      "ccle2_ACH-000027             gs://ccle_bams/wes/C836.GOS-3.2.bam.bai   \n",
      "ccle2_ACH-000031        gs://ccle_bams/wes/C836.Panc_02.13.2.bam.bai   \n",
      "ccle2_ACH-000032        gs://ccle_bams/wes/C836.MHH-CALL-3.1.bam.bai   \n",
      "ccle2_ACH-000039           gs://ccle_bams/wes/C836.SK-N-MC.2.bam.bai   \n",
      "ccle2_ACH-000041             gs://ccle_bams/wes/C836.RD-ES.2.bam.bai   \n",
      "ccle2_ACH-000042        gs://ccle_bams/wes/C836.Panc_02.03.2.bam.bai   \n",
      "ccle2_ACH-000043          gs://ccle_bams/wes/C836.Hs_895.T.1.bam.bai   \n",
      "ccle2_ACH-000051          gs://ccle_bams/wes/C836.TE_617.T.2.bam.bai   \n",
      "ccle2_ACH-000063          gs://ccle_bams/wes/C836.Hs_616.T.1.bam.bai   \n",
      "ccle2_ACH-000065          gs://ccle_bams/wes/C836.OCI-AML5.1.bam.bai   \n",
      "ccle2_ACH-000066           gs://ccle_bams/wes/C836.HCC4006.1.bam.bai   \n",
      "ccle2_ACH-000069          gs://ccle_bams/wes/C836.Hs_611.T.2.bam.bai   \n",
      "ccle2_ACH-000073        gs://ccle_bams/wes/C836.GRANTA-519.2.bam.bai   \n",
      "ccle2_ACH-000076              gs://ccle_bams/wes/C836.NCO2.2.bam.bai   \n",
      "ccle2_ACH-000077                gs://ccle_bams/wes/C836.MJ.2.bam.bai   \n",
      "ccle2_ACH-000079          gs://ccle_bams/wes/C836.TE_125.T.1.bam.bai   \n",
      "ccle2_ACH-000080              gs://ccle_bams/wes/C836.BDCM.2.bam.bai   \n",
      "ccle2_ACH-000081             gs://ccle_bams/wes/C836.GDM-1.2.bam.bai   \n",
      "ccle2_ACH-000083          gs://ccle_bams/wes/C836.Hs_281.T.1.bam.bai   \n",
      "ccle2_ACH-000085             gs://ccle_bams/wes/C836.T3M-4.2.bam.bai   \n",
      "ccle2_ACH-000086        gs://ccle_bams/wes/C836.ACC-MESO-1.2.bam.bai   \n",
      "ccle2_ACH-000089          gs://ccle_bams/wes/C836.NCI-H684.2.bam.bai   \n",
      "ccle2_ACH-000093        gs://ccle_bams/wes/C836.Panc_05.04.1.bam.bai   \n",
      "ccle2_ACH-000097           gs://ccle_bams/wes/C836.ZR-75-1.2.bam.bai   \n",
      "ccle2_ACH-000099              gs://ccle_bams/wes/C836.SIMA.1.bam.bai   \n",
      "ccle2_ACH-000105               gs://ccle_bams/wes/C836.ALL-SIL.1.bai   \n",
      "...                                                              ...   \n",
      "sanger_ACH-002367  gs://ccle_bams/wes/SANGER_NCIBL2052_MATCHED_NO...   \n",
      "sanger_ACH-002368  gs://ccle_bams/wes/SANGER_NCIBL2087_MATCHED_NO...   \n",
      "sanger_ACH-002369  gs://ccle_bams/wes/SANGER_NCIBL209_MATCHED_NOR...   \n",
      "sanger_ACH-002370  gs://ccle_bams/wes/SANGER_NCIBL2122_MATCHED_NO...   \n",
      "sanger_ACH-002371  gs://ccle_bams/wes/SANGER_NCIBL2126_MATCHED_NO...   \n",
      "sanger_ACH-002372  gs://ccle_bams/wes/SANGER_NCIBL2171_MATCHED_NO...   \n",
      "sanger_ACH-002373  gs://ccle_bams/wes/SANGER_HCC1187_MATCHED_NORM...   \n",
      "sanger_ACH-002374  gs://ccle_bams/wes/SANGER_HCC1599_MATCHED_NORM...   \n",
      "sanger_ACH-002375  gs://ccle_bams/wes/SANGER_HCC1937_MATCHED_NORM...   \n",
      "sanger_ACH-002376  gs://ccle_bams/wes/SANGER_LS1034_MATCHED_NORMA...   \n",
      "sanger_ACH-002377  gs://ccle_bams/wes/SANGER_HCC38_MATCHED_NORMAL...   \n",
      "sanger_ACH-002378  gs://ccle_bams/wes/SANGER_HCC1143_MATCHED_NORM...   \n",
      "sanger_ACH-002379  gs://ccle_bams/wes/SANGER_J82_MATCHED_NORMAL_T...   \n",
      "sanger_ACH-002380  gs://ccle_bams/wes/SANGER_COLO829_MATCHED_NORM...   \n",
      "sanger_ACH-002381  gs://ccle_bams/wes/SANGER_HCC2157_MATCHED_NORM...   \n",
      "sanger_ACH-002382  gs://ccle_bams/wes/SANGER_HCC1395_MATCHED_NORM...   \n",
      "sanger_ACH-002383  gs://ccle_bams/wes/SANGER_HCC2218_MATCHED_NORM...   \n",
      "sanger_ACH-002384  gs://ccle_bams/wes/SANGER_HCC1954_MATCHED_NORM...   \n",
      "sanger_ACH-002386           gs://ccle_bams/wes/SANGER_M00921.bam.bai   \n",
      "sanger_ACH-002387         gs://ccle_bams/wes/SANGER_M1203273.bam.bai   \n",
      "sanger_ACH-002388            gs://ccle_bams/wes/SANGER_MET2B.bam.bai   \n",
      "sanger_ACH-002389              gs://ccle_bams/wes/SANGER_ACN.bam.bai   \n",
      "sanger_ACH-002397     gs://ccle_bams/wes/SANGER_KMH2_THYROID.bam.bai   \n",
      "ibm_ACH-001441                            gs://ccle_bams/wes/921.bai   \n",
      "ibm_ACH-002041                          gs://ccle_bams/wes/HOTHC.bai   \n",
      "ibm_ACH-001370                         gs://ccle_bams/wes/OCIP5X.bai   \n",
      "ibm_ACH-001627                           gs://ccle_bams/wes/P4E6.bai   \n",
      "ccle_ACH-001201                       gs://ccle_bams/wes/SUMB002.bai   \n",
      "ccle_ACH-001699                    gs://ccle_bams/wes/UPCISCC131.bai   \n",
      "ccle_ACH-002509                        gs://ccle_bams/wes/WM4235.bai   \n",
      "\n",
      "                                                             WES_bam  \\\n",
      "sample_id                                                              \n",
      "ccle2_ACH-000011                  gs://ccle_bams/wes/C836.253J.1.bam   \n",
      "ccle2_ACH-000013             gs://ccle_bams/wes/C836.ONCO-DG-1.2.bam   \n",
      "ccle2_ACH-000020            gs://ccle_bams/wes/C836.MHH-CALL-2.1.bam   \n",
      "ccle2_ACH-000026               gs://ccle_bams/wes/C836.253J-BV.4.bam   \n",
      "ccle2_ACH-000027                 gs://ccle_bams/wes/C836.GOS-3.2.bam   \n",
      "ccle2_ACH-000031            gs://ccle_bams/wes/C836.Panc_02.13.2.bam   \n",
      "ccle2_ACH-000032            gs://ccle_bams/wes/C836.MHH-CALL-3.1.bam   \n",
      "ccle2_ACH-000039               gs://ccle_bams/wes/C836.SK-N-MC.2.bam   \n",
      "ccle2_ACH-000041                 gs://ccle_bams/wes/C836.RD-ES.2.bam   \n",
      "ccle2_ACH-000042            gs://ccle_bams/wes/C836.Panc_02.03.2.bam   \n",
      "ccle2_ACH-000043              gs://ccle_bams/wes/C836.Hs_895.T.1.bam   \n",
      "ccle2_ACH-000051              gs://ccle_bams/wes/C836.TE_617.T.2.bam   \n",
      "ccle2_ACH-000063              gs://ccle_bams/wes/C836.Hs_616.T.1.bam   \n",
      "ccle2_ACH-000065              gs://ccle_bams/wes/C836.OCI-AML5.1.bam   \n",
      "ccle2_ACH-000066               gs://ccle_bams/wes/C836.HCC4006.1.bam   \n",
      "ccle2_ACH-000069              gs://ccle_bams/wes/C836.Hs_611.T.2.bam   \n",
      "ccle2_ACH-000073            gs://ccle_bams/wes/C836.GRANTA-519.2.bam   \n",
      "ccle2_ACH-000076                  gs://ccle_bams/wes/C836.NCO2.2.bam   \n",
      "ccle2_ACH-000077                    gs://ccle_bams/wes/C836.MJ.2.bam   \n",
      "ccle2_ACH-000079              gs://ccle_bams/wes/C836.TE_125.T.1.bam   \n",
      "ccle2_ACH-000080                  gs://ccle_bams/wes/C836.BDCM.2.bam   \n",
      "ccle2_ACH-000081                 gs://ccle_bams/wes/C836.GDM-1.2.bam   \n",
      "ccle2_ACH-000083              gs://ccle_bams/wes/C836.Hs_281.T.1.bam   \n",
      "ccle2_ACH-000085                 gs://ccle_bams/wes/C836.T3M-4.2.bam   \n",
      "ccle2_ACH-000086            gs://ccle_bams/wes/C836.ACC-MESO-1.2.bam   \n",
      "ccle2_ACH-000089              gs://ccle_bams/wes/C836.NCI-H684.2.bam   \n",
      "ccle2_ACH-000093            gs://ccle_bams/wes/C836.Panc_05.04.1.bam   \n",
      "ccle2_ACH-000097               gs://ccle_bams/wes/C836.ZR-75-1.2.bam   \n",
      "ccle2_ACH-000099                  gs://ccle_bams/wes/C836.SIMA.1.bam   \n",
      "ccle2_ACH-000105               gs://ccle_bams/wes/C836.ALL-SIL.1.bam   \n",
      "...                                                              ...   \n",
      "sanger_ACH-002367  gs://ccle_bams/wes/SANGER_NCIBL2052_MATCHED_NO...   \n",
      "sanger_ACH-002368  gs://ccle_bams/wes/SANGER_NCIBL2087_MATCHED_NO...   \n",
      "sanger_ACH-002369  gs://ccle_bams/wes/SANGER_NCIBL209_MATCHED_NOR...   \n",
      "sanger_ACH-002370  gs://ccle_bams/wes/SANGER_NCIBL2122_MATCHED_NO...   \n",
      "sanger_ACH-002371  gs://ccle_bams/wes/SANGER_NCIBL2126_MATCHED_NO...   \n",
      "sanger_ACH-002372  gs://ccle_bams/wes/SANGER_NCIBL2171_MATCHED_NO...   \n",
      "sanger_ACH-002373  gs://ccle_bams/wes/SANGER_HCC1187_MATCHED_NORM...   \n",
      "sanger_ACH-002374  gs://ccle_bams/wes/SANGER_HCC1599_MATCHED_NORM...   \n",
      "sanger_ACH-002375  gs://ccle_bams/wes/SANGER_HCC1937_MATCHED_NORM...   \n",
      "sanger_ACH-002376  gs://ccle_bams/wes/SANGER_LS1034_MATCHED_NORMA...   \n",
      "sanger_ACH-002377  gs://ccle_bams/wes/SANGER_HCC38_MATCHED_NORMAL...   \n",
      "sanger_ACH-002378  gs://ccle_bams/wes/SANGER_HCC1143_MATCHED_NORM...   \n",
      "sanger_ACH-002379  gs://ccle_bams/wes/SANGER_J82_MATCHED_NORMAL_T...   \n",
      "sanger_ACH-002380  gs://ccle_bams/wes/SANGER_COLO829_MATCHED_NORM...   \n",
      "sanger_ACH-002381  gs://ccle_bams/wes/SANGER_HCC2157_MATCHED_NORM...   \n",
      "sanger_ACH-002382  gs://ccle_bams/wes/SANGER_HCC1395_MATCHED_NORM...   \n",
      "sanger_ACH-002383  gs://ccle_bams/wes/SANGER_HCC2218_MATCHED_NORM...   \n",
      "sanger_ACH-002384  gs://ccle_bams/wes/SANGER_HCC1954_MATCHED_NORM...   \n",
      "sanger_ACH-002386               gs://ccle_bams/wes/SANGER_M00921.bam   \n",
      "sanger_ACH-002387             gs://ccle_bams/wes/SANGER_M1203273.bam   \n",
      "sanger_ACH-002388                gs://ccle_bams/wes/SANGER_MET2B.bam   \n",
      "sanger_ACH-002389                  gs://ccle_bams/wes/SANGER_ACN.bam   \n",
      "sanger_ACH-002397         gs://ccle_bams/wes/SANGER_KMH2_THYROID.bam   \n",
      "ibm_ACH-001441                            gs://ccle_bams/wes/921.bam   \n",
      "ibm_ACH-002041                          gs://ccle_bams/wes/HOTHC.bam   \n",
      "ibm_ACH-001370                         gs://ccle_bams/wes/OCIP5X.bam   \n",
      "ibm_ACH-001627                           gs://ccle_bams/wes/P4E6.bam   \n",
      "ccle_ACH-001201                       gs://ccle_bams/wes/SUMB002.bam   \n",
      "ccle_ACH-001699                    gs://ccle_bams/wes/UPCISCC131.bam   \n",
      "ccle_ACH-002509                        gs://ccle_bams/wes/WM4235.bam   \n",
      "\n",
      "                   allele_fraction_legacy_segments_normal  \\\n",
      "sample_id                                                   \n",
      "ccle2_ACH-000011                                      NaN   \n",
      "ccle2_ACH-000013                                      NaN   \n",
      "ccle2_ACH-000020                                      NaN   \n",
      "ccle2_ACH-000026                                      NaN   \n",
      "ccle2_ACH-000027                                      NaN   \n",
      "ccle2_ACH-000031                                      NaN   \n",
      "ccle2_ACH-000032                                      NaN   \n",
      "ccle2_ACH-000039                                      NaN   \n",
      "ccle2_ACH-000041                                      NaN   \n",
      "ccle2_ACH-000042                                      NaN   \n",
      "ccle2_ACH-000043                                      NaN   \n",
      "ccle2_ACH-000051                                      NaN   \n",
      "ccle2_ACH-000063                                      NaN   \n",
      "ccle2_ACH-000065                                      NaN   \n",
      "ccle2_ACH-000066                                      NaN   \n",
      "ccle2_ACH-000069                                      NaN   \n",
      "ccle2_ACH-000073                                      NaN   \n",
      "ccle2_ACH-000076                                      NaN   \n",
      "ccle2_ACH-000077                                      NaN   \n",
      "ccle2_ACH-000079                                      NaN   \n",
      "ccle2_ACH-000080                                      NaN   \n",
      "ccle2_ACH-000081                                      NaN   \n",
      "ccle2_ACH-000083                                      NaN   \n",
      "ccle2_ACH-000085                                      NaN   \n",
      "ccle2_ACH-000086                                      NaN   \n",
      "ccle2_ACH-000089                                      NaN   \n",
      "ccle2_ACH-000093                                      NaN   \n",
      "ccle2_ACH-000097                                      NaN   \n",
      "ccle2_ACH-000099                                      NaN   \n",
      "ccle2_ACH-000105                                      NaN   \n",
      "...                                                   ...   \n",
      "sanger_ACH-002367                                     NaN   \n",
      "sanger_ACH-002368                                     NaN   \n",
      "sanger_ACH-002369                                     NaN   \n",
      "sanger_ACH-002370                                     NaN   \n",
      "sanger_ACH-002371                                     NaN   \n",
      "sanger_ACH-002372                                     NaN   \n",
      "sanger_ACH-002373                                     NaN   \n",
      "sanger_ACH-002374                                     NaN   \n",
      "sanger_ACH-002375                                     NaN   \n",
      "sanger_ACH-002376                                     NaN   \n",
      "sanger_ACH-002377                                     NaN   \n",
      "sanger_ACH-002378                                     NaN   \n",
      "sanger_ACH-002379                                     NaN   \n",
      "sanger_ACH-002380                                     NaN   \n",
      "sanger_ACH-002381                                     NaN   \n",
      "sanger_ACH-002382                                     NaN   \n",
      "sanger_ACH-002383                                     NaN   \n",
      "sanger_ACH-002384                                     NaN   \n",
      "sanger_ACH-002386                                     NaN   \n",
      "sanger_ACH-002387                                     NaN   \n",
      "sanger_ACH-002388                                     NaN   \n",
      "sanger_ACH-002389                                     NaN   \n",
      "sanger_ACH-002397                                     NaN   \n",
      "ibm_ACH-001441                                        NaN   \n",
      "ibm_ACH-002041                                        NaN   \n",
      "ibm_ACH-001370                                        NaN   \n",
      "ibm_ACH-001627                                        NaN   \n",
      "ccle_ACH-001201                                       NaN   \n",
      "ccle_ACH-001699                                       NaN   \n",
      "ccle_ACH-002509                                       NaN   \n",
      "\n",
      "                  allele_fraction_legacy_segments_tumor  \\\n",
      "sample_id                                                 \n",
      "ccle2_ACH-000011                                    NaN   \n",
      "ccle2_ACH-000013                                    NaN   \n",
      "ccle2_ACH-000020                                    NaN   \n",
      "ccle2_ACH-000026                                    NaN   \n",
      "ccle2_ACH-000027                                    NaN   \n",
      "ccle2_ACH-000031                                    NaN   \n",
      "ccle2_ACH-000032                                    NaN   \n",
      "ccle2_ACH-000039                                    NaN   \n",
      "ccle2_ACH-000041                                    NaN   \n",
      "ccle2_ACH-000042                                    NaN   \n",
      "ccle2_ACH-000043                                    NaN   \n",
      "ccle2_ACH-000051                                    NaN   \n",
      "ccle2_ACH-000063                                    NaN   \n",
      "ccle2_ACH-000065                                    NaN   \n",
      "ccle2_ACH-000066                                    NaN   \n",
      "ccle2_ACH-000069                                    NaN   \n",
      "ccle2_ACH-000073                                    NaN   \n",
      "ccle2_ACH-000076                                    NaN   \n",
      "ccle2_ACH-000077                                    NaN   \n",
      "ccle2_ACH-000079                                    NaN   \n",
      "ccle2_ACH-000080                                    NaN   \n",
      "ccle2_ACH-000081                                    NaN   \n",
      "ccle2_ACH-000083                                    NaN   \n",
      "ccle2_ACH-000085                                    NaN   \n",
      "ccle2_ACH-000086                                    NaN   \n",
      "ccle2_ACH-000089                                    NaN   \n",
      "ccle2_ACH-000093                                    NaN   \n",
      "ccle2_ACH-000097                                    NaN   \n",
      "ccle2_ACH-000099                                    NaN   \n",
      "ccle2_ACH-000105                                    NaN   \n",
      "...                                                 ...   \n",
      "sanger_ACH-002367                                   NaN   \n",
      "sanger_ACH-002368                                   NaN   \n",
      "sanger_ACH-002369                                   NaN   \n",
      "sanger_ACH-002370                                   NaN   \n",
      "sanger_ACH-002371                                   NaN   \n",
      "sanger_ACH-002372                                   NaN   \n",
      "sanger_ACH-002373                                   NaN   \n",
      "sanger_ACH-002374                                   NaN   \n",
      "sanger_ACH-002375                                   NaN   \n",
      "sanger_ACH-002376                                   NaN   \n",
      "sanger_ACH-002377                                   NaN   \n",
      "sanger_ACH-002378                                   NaN   \n",
      "sanger_ACH-002379                                   NaN   \n",
      "sanger_ACH-002380                                   NaN   \n",
      "sanger_ACH-002381                                   NaN   \n",
      "sanger_ACH-002382                                   NaN   \n",
      "sanger_ACH-002383                                   NaN   \n",
      "sanger_ACH-002384                                   NaN   \n",
      "sanger_ACH-002386                                   NaN   \n",
      "sanger_ACH-002387                                   NaN   \n",
      "sanger_ACH-002388                                   NaN   \n",
      "sanger_ACH-002389                                   NaN   \n",
      "sanger_ACH-002397                                   NaN   \n",
      "ibm_ACH-001441                                      NaN   \n",
      "ibm_ACH-002041                                      NaN   \n",
      "ibm_ACH-001370                                      NaN   \n",
      "ibm_ACH-001627                                      NaN   \n",
      "ccle_ACH-001201                                     NaN   \n",
      "ccle_ACH-001699                                     NaN   \n",
      "ccle_ACH-002509                                     NaN   \n",
      "\n",
      "                   allele_fraction_parameters_begin_normal  \\\n",
      "sample_id                                                    \n",
      "ccle2_ACH-000011                                       NaN   \n",
      "ccle2_ACH-000013                                       NaN   \n",
      "ccle2_ACH-000020                                       NaN   \n",
      "ccle2_ACH-000026                                       NaN   \n",
      "ccle2_ACH-000027                                       NaN   \n",
      "ccle2_ACH-000031                                       NaN   \n",
      "ccle2_ACH-000032                                       NaN   \n",
      "ccle2_ACH-000039                                       NaN   \n",
      "ccle2_ACH-000041                                       NaN   \n",
      "ccle2_ACH-000042                                       NaN   \n",
      "ccle2_ACH-000043                                       NaN   \n",
      "ccle2_ACH-000051                                       NaN   \n",
      "ccle2_ACH-000063                                       NaN   \n",
      "ccle2_ACH-000065                                       NaN   \n",
      "ccle2_ACH-000066                                       NaN   \n",
      "ccle2_ACH-000069                                       NaN   \n",
      "ccle2_ACH-000073                                       NaN   \n",
      "ccle2_ACH-000076                                       NaN   \n",
      "ccle2_ACH-000077                                       NaN   \n",
      "ccle2_ACH-000079                                       NaN   \n",
      "ccle2_ACH-000080                                       NaN   \n",
      "ccle2_ACH-000081                                       NaN   \n",
      "ccle2_ACH-000083                                       NaN   \n",
      "ccle2_ACH-000085                                       NaN   \n",
      "ccle2_ACH-000086                                       NaN   \n",
      "ccle2_ACH-000089                                       NaN   \n",
      "ccle2_ACH-000093                                       NaN   \n",
      "ccle2_ACH-000097                                       NaN   \n",
      "ccle2_ACH-000099                                       NaN   \n",
      "ccle2_ACH-000105                                       NaN   \n",
      "...                                                    ...   \n",
      "sanger_ACH-002367                                      NaN   \n",
      "sanger_ACH-002368                                      NaN   \n",
      "sanger_ACH-002369                                      NaN   \n",
      "sanger_ACH-002370                                      NaN   \n",
      "sanger_ACH-002371                                      NaN   \n",
      "sanger_ACH-002372                                      NaN   \n",
      "sanger_ACH-002373                                      NaN   \n",
      "sanger_ACH-002374                                      NaN   \n",
      "sanger_ACH-002375                                      NaN   \n",
      "sanger_ACH-002376                                      NaN   \n",
      "sanger_ACH-002377                                      NaN   \n",
      "sanger_ACH-002378                                      NaN   \n",
      "sanger_ACH-002379                                      NaN   \n",
      "sanger_ACH-002380                                      NaN   \n",
      "sanger_ACH-002381                                      NaN   \n",
      "sanger_ACH-002382                                      NaN   \n",
      "sanger_ACH-002383                                      NaN   \n",
      "sanger_ACH-002384                                      NaN   \n",
      "sanger_ACH-002386                                      NaN   \n",
      "sanger_ACH-002387                                      NaN   \n",
      "sanger_ACH-002388                                      NaN   \n",
      "sanger_ACH-002389                                      NaN   \n",
      "sanger_ACH-002397                                      NaN   \n",
      "ibm_ACH-001441                                         NaN   \n",
      "ibm_ACH-002041                                         NaN   \n",
      "ibm_ACH-001370                                         NaN   \n",
      "ibm_ACH-001627                                         NaN   \n",
      "ccle_ACH-001201                                        NaN   \n",
      "ccle_ACH-001699                                        NaN   \n",
      "ccle_ACH-002509                                        NaN   \n",
      "\n",
      "                  allele_fraction_parameters_begin_tumor  \\\n",
      "sample_id                                                  \n",
      "ccle2_ACH-000011                                     NaN   \n",
      "ccle2_ACH-000013                                     NaN   \n",
      "ccle2_ACH-000020                                     NaN   \n",
      "ccle2_ACH-000026                                     NaN   \n",
      "ccle2_ACH-000027                                     NaN   \n",
      "ccle2_ACH-000031                                     NaN   \n",
      "ccle2_ACH-000032                                     NaN   \n",
      "ccle2_ACH-000039                                     NaN   \n",
      "ccle2_ACH-000041                                     NaN   \n",
      "ccle2_ACH-000042                                     NaN   \n",
      "ccle2_ACH-000043                                     NaN   \n",
      "ccle2_ACH-000051                                     NaN   \n",
      "ccle2_ACH-000063                                     NaN   \n",
      "ccle2_ACH-000065                                     NaN   \n",
      "ccle2_ACH-000066                                     NaN   \n",
      "ccle2_ACH-000069                                     NaN   \n",
      "ccle2_ACH-000073                                     NaN   \n",
      "ccle2_ACH-000076                                     NaN   \n",
      "ccle2_ACH-000077                                     NaN   \n",
      "ccle2_ACH-000079                                     NaN   \n",
      "ccle2_ACH-000080                                     NaN   \n",
      "ccle2_ACH-000081                                     NaN   \n",
      "ccle2_ACH-000083                                     NaN   \n",
      "ccle2_ACH-000085                                     NaN   \n",
      "ccle2_ACH-000086                                     NaN   \n",
      "ccle2_ACH-000089                                     NaN   \n",
      "ccle2_ACH-000093                                     NaN   \n",
      "ccle2_ACH-000097                                     NaN   \n",
      "ccle2_ACH-000099                                     NaN   \n",
      "ccle2_ACH-000105                                     NaN   \n",
      "...                                                  ...   \n",
      "sanger_ACH-002367                                    NaN   \n",
      "sanger_ACH-002368                                    NaN   \n",
      "sanger_ACH-002369                                    NaN   \n",
      "sanger_ACH-002370                                    NaN   \n",
      "sanger_ACH-002371                                    NaN   \n",
      "sanger_ACH-002372                                    NaN   \n",
      "sanger_ACH-002373                                    NaN   \n",
      "sanger_ACH-002374                                    NaN   \n",
      "sanger_ACH-002375                                    NaN   \n",
      "sanger_ACH-002376                                    NaN   \n",
      "sanger_ACH-002377                                    NaN   \n",
      "sanger_ACH-002378                                    NaN   \n",
      "sanger_ACH-002379                                    NaN   \n",
      "sanger_ACH-002380                                    NaN   \n",
      "sanger_ACH-002381                                    NaN   \n",
      "sanger_ACH-002382                                    NaN   \n",
      "sanger_ACH-002383                                    NaN   \n",
      "sanger_ACH-002384                                    NaN   \n",
      "sanger_ACH-002386                                    NaN   \n",
      "sanger_ACH-002387                                    NaN   \n",
      "sanger_ACH-002388                                    NaN   \n",
      "sanger_ACH-002389                                    NaN   \n",
      "sanger_ACH-002397                                    NaN   \n",
      "ibm_ACH-001441                                       NaN   \n",
      "ibm_ACH-002041                                       NaN   \n",
      "ibm_ACH-001370                                       NaN   \n",
      "ibm_ACH-001627                                       NaN   \n",
      "ccle_ACH-001201                                      NaN   \n",
      "ccle_ACH-001699                                      NaN   \n",
      "ccle_ACH-002509                                      NaN   \n",
      "\n",
      "                   allele_fraction_parameters_normal  \\\n",
      "sample_id                                              \n",
      "ccle2_ACH-000011                                 NaN   \n",
      "ccle2_ACH-000013                                 NaN   \n",
      "ccle2_ACH-000020                                 NaN   \n",
      "ccle2_ACH-000026                                 NaN   \n",
      "ccle2_ACH-000027                                 NaN   \n",
      "ccle2_ACH-000031                                 NaN   \n",
      "ccle2_ACH-000032                                 NaN   \n",
      "ccle2_ACH-000039                                 NaN   \n",
      "ccle2_ACH-000041                                 NaN   \n",
      "ccle2_ACH-000042                                 NaN   \n",
      "ccle2_ACH-000043                                 NaN   \n",
      "ccle2_ACH-000051                                 NaN   \n",
      "ccle2_ACH-000063                                 NaN   \n",
      "ccle2_ACH-000065                                 NaN   \n",
      "ccle2_ACH-000066                                 NaN   \n",
      "ccle2_ACH-000069                                 NaN   \n",
      "ccle2_ACH-000073                                 NaN   \n",
      "ccle2_ACH-000076                                 NaN   \n",
      "ccle2_ACH-000077                                 NaN   \n",
      "ccle2_ACH-000079                                 NaN   \n",
      "ccle2_ACH-000080                                 NaN   \n",
      "ccle2_ACH-000081                                 NaN   \n",
      "ccle2_ACH-000083                                 NaN   \n",
      "ccle2_ACH-000085                                 NaN   \n",
      "ccle2_ACH-000086                                 NaN   \n",
      "ccle2_ACH-000089                                 NaN   \n",
      "ccle2_ACH-000093                                 NaN   \n",
      "ccle2_ACH-000097                                 NaN   \n",
      "ccle2_ACH-000099                                 NaN   \n",
      "ccle2_ACH-000105                                 NaN   \n",
      "...                                              ...   \n",
      "sanger_ACH-002367                                NaN   \n",
      "sanger_ACH-002368                                NaN   \n",
      "sanger_ACH-002369                                NaN   \n",
      "sanger_ACH-002370                                NaN   \n",
      "sanger_ACH-002371                                NaN   \n",
      "sanger_ACH-002372                                NaN   \n",
      "sanger_ACH-002373                                NaN   \n",
      "sanger_ACH-002374                                NaN   \n",
      "sanger_ACH-002375                                NaN   \n",
      "sanger_ACH-002376                                NaN   \n",
      "sanger_ACH-002377                                NaN   \n",
      "sanger_ACH-002378                                NaN   \n",
      "sanger_ACH-002379                                NaN   \n",
      "sanger_ACH-002380                                NaN   \n",
      "sanger_ACH-002381                                NaN   \n",
      "sanger_ACH-002382                                NaN   \n",
      "sanger_ACH-002383                                NaN   \n",
      "sanger_ACH-002384                                NaN   \n",
      "sanger_ACH-002386                                NaN   \n",
      "sanger_ACH-002387                                NaN   \n",
      "sanger_ACH-002388                                NaN   \n",
      "sanger_ACH-002389                                NaN   \n",
      "sanger_ACH-002397                                NaN   \n",
      "ibm_ACH-001441                                   NaN   \n",
      "ibm_ACH-002041                                   NaN   \n",
      "ibm_ACH-001370                                   NaN   \n",
      "ibm_ACH-001627                                   NaN   \n",
      "ccle_ACH-001201                                  NaN   \n",
      "ccle_ACH-001699                                  NaN   \n",
      "ccle_ACH-002509                                  NaN   \n",
      "\n",
      "                  allele_fraction_parameters_tumor  ...  readgroup_ubams  \\\n",
      "sample_id                                           ...                    \n",
      "ccle2_ACH-000011                               NaN  ...              NaN   \n",
      "ccle2_ACH-000013                               NaN  ...              NaN   \n",
      "ccle2_ACH-000020                               NaN  ...              NaN   \n",
      "ccle2_ACH-000026                               NaN  ...              NaN   \n",
      "ccle2_ACH-000027                               NaN  ...              NaN   \n",
      "ccle2_ACH-000031                               NaN  ...              NaN   \n",
      "ccle2_ACH-000032                               NaN  ...              NaN   \n",
      "ccle2_ACH-000039                               NaN  ...              NaN   \n",
      "ccle2_ACH-000041                               NaN  ...              NaN   \n",
      "ccle2_ACH-000042                               NaN  ...              NaN   \n",
      "ccle2_ACH-000043                               NaN  ...              NaN   \n",
      "ccle2_ACH-000051                               NaN  ...              NaN   \n",
      "ccle2_ACH-000063                               NaN  ...              NaN   \n",
      "ccle2_ACH-000065                               NaN  ...              NaN   \n",
      "ccle2_ACH-000066                               NaN  ...              NaN   \n",
      "ccle2_ACH-000069                               NaN  ...              NaN   \n",
      "ccle2_ACH-000073                               NaN  ...              NaN   \n",
      "ccle2_ACH-000076                               NaN  ...              NaN   \n",
      "ccle2_ACH-000077                               NaN  ...              NaN   \n",
      "ccle2_ACH-000079                               NaN  ...              NaN   \n",
      "ccle2_ACH-000080                               NaN  ...              NaN   \n",
      "ccle2_ACH-000081                               NaN  ...              NaN   \n",
      "ccle2_ACH-000083                               NaN  ...              NaN   \n",
      "ccle2_ACH-000085                               NaN  ...              NaN   \n",
      "ccle2_ACH-000086                               NaN  ...              NaN   \n",
      "ccle2_ACH-000089                               NaN  ...              NaN   \n",
      "ccle2_ACH-000093                               NaN  ...              NaN   \n",
      "ccle2_ACH-000097                               NaN  ...              NaN   \n",
      "ccle2_ACH-000099                               NaN  ...              NaN   \n",
      "ccle2_ACH-000105                               NaN  ...              NaN   \n",
      "...                                            ...  ...              ...   \n",
      "sanger_ACH-002367                              NaN  ...              NaN   \n",
      "sanger_ACH-002368                              NaN  ...              NaN   \n",
      "sanger_ACH-002369                              NaN  ...              NaN   \n",
      "sanger_ACH-002370                              NaN  ...              NaN   \n",
      "sanger_ACH-002371                              NaN  ...              NaN   \n",
      "sanger_ACH-002372                              NaN  ...              NaN   \n",
      "sanger_ACH-002373                              NaN  ...              NaN   \n",
      "sanger_ACH-002374                              NaN  ...              NaN   \n",
      "sanger_ACH-002375                              NaN  ...              NaN   \n",
      "sanger_ACH-002376                              NaN  ...              NaN   \n",
      "sanger_ACH-002377                              NaN  ...              NaN   \n",
      "sanger_ACH-002378                              NaN  ...              NaN   \n",
      "sanger_ACH-002379                              NaN  ...              NaN   \n",
      "sanger_ACH-002380                              NaN  ...              NaN   \n",
      "sanger_ACH-002381                              NaN  ...              NaN   \n",
      "sanger_ACH-002382                              NaN  ...              NaN   \n",
      "sanger_ACH-002383                              NaN  ...              NaN   \n",
      "sanger_ACH-002384                              NaN  ...              NaN   \n",
      "sanger_ACH-002386                              NaN  ...              NaN   \n",
      "sanger_ACH-002387                              NaN  ...              NaN   \n",
      "sanger_ACH-002388                              NaN  ...              NaN   \n",
      "sanger_ACH-002389                              NaN  ...              NaN   \n",
      "sanger_ACH-002397                              NaN  ...              NaN   \n",
      "ibm_ACH-001441                                 NaN  ...              NaN   \n",
      "ibm_ACH-002041                                 NaN  ...              NaN   \n",
      "ibm_ACH-001370                                 NaN  ...              NaN   \n",
      "ibm_ACH-001627                                 NaN  ...              NaN   \n",
      "ccle_ACH-001201                                NaN  ...              NaN   \n",
      "ccle_ACH-001699                                NaN  ...              NaN   \n",
      "ccle_ACH-002509                                NaN  ...              NaN   \n",
      "\n",
      "                  scaled_delta_MAD_normal  scaled_delta_MAD_tumor  \\\n",
      "sample_id                                                           \n",
      "ccle2_ACH-000011                      NaN                     NaN   \n",
      "ccle2_ACH-000013                      NaN                     NaN   \n",
      "ccle2_ACH-000020                      NaN                     NaN   \n",
      "ccle2_ACH-000026                      NaN                     NaN   \n",
      "ccle2_ACH-000027                      NaN                     NaN   \n",
      "ccle2_ACH-000031                      NaN                     NaN   \n",
      "ccle2_ACH-000032                      NaN                     NaN   \n",
      "ccle2_ACH-000039                      NaN                     NaN   \n",
      "ccle2_ACH-000041                      NaN                     NaN   \n",
      "ccle2_ACH-000042                      NaN                     NaN   \n",
      "ccle2_ACH-000043                      NaN                     NaN   \n",
      "ccle2_ACH-000051                      NaN                     NaN   \n",
      "ccle2_ACH-000063                      NaN                     NaN   \n",
      "ccle2_ACH-000065                      NaN                     NaN   \n",
      "ccle2_ACH-000066                      NaN                     NaN   \n",
      "ccle2_ACH-000069                      NaN                     NaN   \n",
      "ccle2_ACH-000073                      NaN                     NaN   \n",
      "ccle2_ACH-000076                      NaN                     NaN   \n",
      "ccle2_ACH-000077                      NaN                     NaN   \n",
      "ccle2_ACH-000079                      NaN                     NaN   \n",
      "ccle2_ACH-000080                      NaN                     NaN   \n",
      "ccle2_ACH-000081                      NaN                     NaN   \n",
      "ccle2_ACH-000083                      NaN                     NaN   \n",
      "ccle2_ACH-000085                      NaN                     NaN   \n",
      "ccle2_ACH-000086                      NaN                     NaN   \n",
      "ccle2_ACH-000089                      NaN                     NaN   \n",
      "ccle2_ACH-000093                      NaN                     NaN   \n",
      "ccle2_ACH-000097                      NaN                     NaN   \n",
      "ccle2_ACH-000099                      NaN                     NaN   \n",
      "ccle2_ACH-000105                      NaN                     NaN   \n",
      "...                                   ...                     ...   \n",
      "sanger_ACH-002367                     NaN                     NaN   \n",
      "sanger_ACH-002368                     NaN                     NaN   \n",
      "sanger_ACH-002369                     NaN                     NaN   \n",
      "sanger_ACH-002370                     NaN                     NaN   \n",
      "sanger_ACH-002371                     NaN                     NaN   \n",
      "sanger_ACH-002372                     NaN                     NaN   \n",
      "sanger_ACH-002373                     NaN                     NaN   \n",
      "sanger_ACH-002374                     NaN                     NaN   \n",
      "sanger_ACH-002375                     NaN                     NaN   \n",
      "sanger_ACH-002376                     NaN                     NaN   \n",
      "sanger_ACH-002377                     NaN                     NaN   \n",
      "sanger_ACH-002378                     NaN                     NaN   \n",
      "sanger_ACH-002379                     NaN                     NaN   \n",
      "sanger_ACH-002380                     NaN                     NaN   \n",
      "sanger_ACH-002381                     NaN                     NaN   \n",
      "sanger_ACH-002382                     NaN                     NaN   \n",
      "sanger_ACH-002383                     NaN                     NaN   \n",
      "sanger_ACH-002384                     NaN                     NaN   \n",
      "sanger_ACH-002386                     NaN                     NaN   \n",
      "sanger_ACH-002387                     NaN                     NaN   \n",
      "sanger_ACH-002388                     NaN                     NaN   \n",
      "sanger_ACH-002389                     NaN                     NaN   \n",
      "sanger_ACH-002397                     NaN                     NaN   \n",
      "ibm_ACH-001441                        NaN                     NaN   \n",
      "ibm_ACH-002041                        NaN                     NaN   \n",
      "ibm_ACH-001370                        NaN                     NaN   \n",
      "ibm_ACH-001627                        NaN                     NaN   \n",
      "ccle_ACH-001201                       NaN                     NaN   \n",
      "ccle_ACH-001699                       NaN                     NaN   \n",
      "ccle_ACH-002509                       NaN                     NaN   \n",
      "\n",
      "                  standardized_MAD_normal  standardized_MAD_tumor  \\\n",
      "sample_id                                                           \n",
      "ccle2_ACH-000011                      NaN                     NaN   \n",
      "ccle2_ACH-000013                      NaN                     NaN   \n",
      "ccle2_ACH-000020                      NaN                     NaN   \n",
      "ccle2_ACH-000026                      NaN                     NaN   \n",
      "ccle2_ACH-000027                      NaN                     NaN   \n",
      "ccle2_ACH-000031                      NaN                     NaN   \n",
      "ccle2_ACH-000032                      NaN                     NaN   \n",
      "ccle2_ACH-000039                      NaN                     NaN   \n",
      "ccle2_ACH-000041                      NaN                     NaN   \n",
      "ccle2_ACH-000042                      NaN                     NaN   \n",
      "ccle2_ACH-000043                      NaN                     NaN   \n",
      "ccle2_ACH-000051                      NaN                     NaN   \n",
      "ccle2_ACH-000063                      NaN                     NaN   \n",
      "ccle2_ACH-000065                      NaN                     NaN   \n",
      "ccle2_ACH-000066                      NaN                     NaN   \n",
      "ccle2_ACH-000069                      NaN                     NaN   \n",
      "ccle2_ACH-000073                      NaN                     NaN   \n",
      "ccle2_ACH-000076                      NaN                     NaN   \n",
      "ccle2_ACH-000077                      NaN                     NaN   \n",
      "ccle2_ACH-000079                      NaN                     NaN   \n",
      "ccle2_ACH-000080                      NaN                     NaN   \n",
      "ccle2_ACH-000081                      NaN                     NaN   \n",
      "ccle2_ACH-000083                      NaN                     NaN   \n",
      "ccle2_ACH-000085                      NaN                     NaN   \n",
      "ccle2_ACH-000086                      NaN                     NaN   \n",
      "ccle2_ACH-000089                      NaN                     NaN   \n",
      "ccle2_ACH-000093                      NaN                     NaN   \n",
      "ccle2_ACH-000097                      NaN                     NaN   \n",
      "ccle2_ACH-000099                      NaN                     NaN   \n",
      "ccle2_ACH-000105                      NaN                     NaN   \n",
      "...                                   ...                     ...   \n",
      "sanger_ACH-002367                     NaN                     NaN   \n",
      "sanger_ACH-002368                     NaN                     NaN   \n",
      "sanger_ACH-002369                     NaN                     NaN   \n",
      "sanger_ACH-002370                     NaN                     NaN   \n",
      "sanger_ACH-002371                     NaN                     NaN   \n",
      "sanger_ACH-002372                     NaN                     NaN   \n",
      "sanger_ACH-002373                     NaN                     NaN   \n",
      "sanger_ACH-002374                     NaN                     NaN   \n",
      "sanger_ACH-002375                     NaN                     NaN   \n",
      "sanger_ACH-002376                     NaN                     NaN   \n",
      "sanger_ACH-002377                     NaN                     NaN   \n",
      "sanger_ACH-002378                     NaN                     NaN   \n",
      "sanger_ACH-002379                     NaN                     NaN   \n",
      "sanger_ACH-002380                     NaN                     NaN   \n",
      "sanger_ACH-002381                     NaN                     NaN   \n",
      "sanger_ACH-002382                     NaN                     NaN   \n",
      "sanger_ACH-002383                     NaN                     NaN   \n",
      "sanger_ACH-002384                     NaN                     NaN   \n",
      "sanger_ACH-002386                     NaN                     NaN   \n",
      "sanger_ACH-002387                     NaN                     NaN   \n",
      "sanger_ACH-002388                     NaN                     NaN   \n",
      "sanger_ACH-002389                     NaN                     NaN   \n",
      "sanger_ACH-002397                     NaN                     NaN   \n",
      "ibm_ACH-001441                        NaN                     NaN   \n",
      "ibm_ACH-002041                        NaN                     NaN   \n",
      "ibm_ACH-001370                        NaN                     NaN   \n",
      "ibm_ACH-001627                        NaN                     NaN   \n",
      "ccle_ACH-001201                       NaN                     NaN   \n",
      "ccle_ACH-001699                       NaN                     NaN   \n",
      "ccle_ACH-002509                       NaN                     NaN   \n",
      "\n",
      "                  standardized_copy_ratios_normal  \\\n",
      "sample_id                                           \n",
      "ccle2_ACH-000011                              NaN   \n",
      "ccle2_ACH-000013                              NaN   \n",
      "ccle2_ACH-000020                              NaN   \n",
      "ccle2_ACH-000026                              NaN   \n",
      "ccle2_ACH-000027                              NaN   \n",
      "ccle2_ACH-000031                              NaN   \n",
      "ccle2_ACH-000032                              NaN   \n",
      "ccle2_ACH-000039                              NaN   \n",
      "ccle2_ACH-000041                              NaN   \n",
      "ccle2_ACH-000042                              NaN   \n",
      "ccle2_ACH-000043                              NaN   \n",
      "ccle2_ACH-000051                              NaN   \n",
      "ccle2_ACH-000063                              NaN   \n",
      "ccle2_ACH-000065                              NaN   \n",
      "ccle2_ACH-000066                              NaN   \n",
      "ccle2_ACH-000069                              NaN   \n",
      "ccle2_ACH-000073                              NaN   \n",
      "ccle2_ACH-000076                              NaN   \n",
      "ccle2_ACH-000077                              NaN   \n",
      "ccle2_ACH-000079                              NaN   \n",
      "ccle2_ACH-000080                              NaN   \n",
      "ccle2_ACH-000081                              NaN   \n",
      "ccle2_ACH-000083                              NaN   \n",
      "ccle2_ACH-000085                              NaN   \n",
      "ccle2_ACH-000086                              NaN   \n",
      "ccle2_ACH-000089                              NaN   \n",
      "ccle2_ACH-000093                              NaN   \n",
      "ccle2_ACH-000097                              NaN   \n",
      "ccle2_ACH-000099                              NaN   \n",
      "ccle2_ACH-000105                              NaN   \n",
      "...                                           ...   \n",
      "sanger_ACH-002367                             NaN   \n",
      "sanger_ACH-002368                             NaN   \n",
      "sanger_ACH-002369                             NaN   \n",
      "sanger_ACH-002370                             NaN   \n",
      "sanger_ACH-002371                             NaN   \n",
      "sanger_ACH-002372                             NaN   \n",
      "sanger_ACH-002373                             NaN   \n",
      "sanger_ACH-002374                             NaN   \n",
      "sanger_ACH-002375                             NaN   \n",
      "sanger_ACH-002376                             NaN   \n",
      "sanger_ACH-002377                             NaN   \n",
      "sanger_ACH-002378                             NaN   \n",
      "sanger_ACH-002379                             NaN   \n",
      "sanger_ACH-002380                             NaN   \n",
      "sanger_ACH-002381                             NaN   \n",
      "sanger_ACH-002382                             NaN   \n",
      "sanger_ACH-002383                             NaN   \n",
      "sanger_ACH-002384                             NaN   \n",
      "sanger_ACH-002386                             NaN   \n",
      "sanger_ACH-002387                             NaN   \n",
      "sanger_ACH-002388                             NaN   \n",
      "sanger_ACH-002389                             NaN   \n",
      "sanger_ACH-002397                             NaN   \n",
      "ibm_ACH-001441                                NaN   \n",
      "ibm_ACH-002041                                NaN   \n",
      "ibm_ACH-001370                                NaN   \n",
      "ibm_ACH-001627                                NaN   \n",
      "ccle_ACH-001201                               NaN   \n",
      "ccle_ACH-001699                               NaN   \n",
      "ccle_ACH-002509                               NaN   \n",
      "\n",
      "                   standardized_copy_ratios_tumor target_type  \\\n",
      "sample_id                                                       \n",
      "ccle2_ACH-000011                              NaN     AGILENT   \n",
      "ccle2_ACH-000013                              NaN     AGILENT   \n",
      "ccle2_ACH-000020                              NaN     AGILENT   \n",
      "ccle2_ACH-000026                              NaN     AGILENT   \n",
      "ccle2_ACH-000027                              NaN     AGILENT   \n",
      "ccle2_ACH-000031                              NaN     AGILENT   \n",
      "ccle2_ACH-000032                              NaN     AGILENT   \n",
      "ccle2_ACH-000039                              NaN     AGILENT   \n",
      "ccle2_ACH-000041                              NaN     AGILENT   \n",
      "ccle2_ACH-000042                              NaN     AGILENT   \n",
      "ccle2_ACH-000043                              NaN     AGILENT   \n",
      "ccle2_ACH-000051                              NaN     AGILENT   \n",
      "ccle2_ACH-000063                              NaN     AGILENT   \n",
      "ccle2_ACH-000065                              NaN     AGILENT   \n",
      "ccle2_ACH-000066                              NaN     AGILENT   \n",
      "ccle2_ACH-000069                              NaN     AGILENT   \n",
      "ccle2_ACH-000073                              NaN     AGILENT   \n",
      "ccle2_ACH-000076                              NaN     AGILENT   \n",
      "ccle2_ACH-000077                              NaN     AGILENT   \n",
      "ccle2_ACH-000079                              NaN     AGILENT   \n",
      "ccle2_ACH-000080                              NaN     AGILENT   \n",
      "ccle2_ACH-000081                              NaN     AGILENT   \n",
      "ccle2_ACH-000083                              NaN     AGILENT   \n",
      "ccle2_ACH-000085                              NaN     AGILENT   \n",
      "ccle2_ACH-000086                              NaN     AGILENT   \n",
      "ccle2_ACH-000089                              NaN     AGILENT   \n",
      "ccle2_ACH-000093                              NaN     AGILENT   \n",
      "ccle2_ACH-000097                              NaN     AGILENT   \n",
      "ccle2_ACH-000099                              NaN     AGILENT   \n",
      "ccle2_ACH-000105                              NaN     AGILENT   \n",
      "...                                           ...         ...   \n",
      "sanger_ACH-002367                             NaN     AGILENT   \n",
      "sanger_ACH-002368                             NaN     AGILENT   \n",
      "sanger_ACH-002369                             NaN     AGILENT   \n",
      "sanger_ACH-002370                             NaN     AGILENT   \n",
      "sanger_ACH-002371                             NaN     AGILENT   \n",
      "sanger_ACH-002372                             NaN     AGILENT   \n",
      "sanger_ACH-002373                             NaN     AGILENT   \n",
      "sanger_ACH-002374                             NaN     AGILENT   \n",
      "sanger_ACH-002375                             NaN     AGILENT   \n",
      "sanger_ACH-002376                             NaN     AGILENT   \n",
      "sanger_ACH-002377                             NaN     AGILENT   \n",
      "sanger_ACH-002378                             NaN     AGILENT   \n",
      "sanger_ACH-002379                             NaN     AGILENT   \n",
      "sanger_ACH-002380                             NaN     AGILENT   \n",
      "sanger_ACH-002381                             NaN     AGILENT   \n",
      "sanger_ACH-002382                             NaN     AGILENT   \n",
      "sanger_ACH-002383                             NaN     AGILENT   \n",
      "sanger_ACH-002384                             NaN     AGILENT   \n",
      "sanger_ACH-002386                             NaN     AGILENT   \n",
      "sanger_ACH-002387                             NaN     AGILENT   \n",
      "sanger_ACH-002388                             NaN     AGILENT   \n",
      "sanger_ACH-002389                             NaN     AGILENT   \n",
      "sanger_ACH-002397                             NaN     AGILENT   \n",
      "ibm_ACH-001441                                NaN         NaN   \n",
      "ibm_ACH-002041                                NaN         NaN   \n",
      "ibm_ACH-001370                                NaN         NaN   \n",
      "ibm_ACH-001627                                NaN         NaN   \n",
      "ccle_ACH-001201                               NaN         NaN   \n",
      "ccle_ACH-001699                               NaN         NaN   \n",
      "ccle_ACH-002509                               NaN         NaN   \n",
      "\n",
      "                   ubam_validation_reports  \\\n",
      "sample_id                                    \n",
      "ccle2_ACH-000011                       NaN   \n",
      "ccle2_ACH-000013                       NaN   \n",
      "ccle2_ACH-000020                       NaN   \n",
      "ccle2_ACH-000026                       NaN   \n",
      "ccle2_ACH-000027                       NaN   \n",
      "ccle2_ACH-000031                       NaN   \n",
      "ccle2_ACH-000032                       NaN   \n",
      "ccle2_ACH-000039                       NaN   \n",
      "ccle2_ACH-000041                       NaN   \n",
      "ccle2_ACH-000042                       NaN   \n",
      "ccle2_ACH-000043                       NaN   \n",
      "ccle2_ACH-000051                       NaN   \n",
      "ccle2_ACH-000063                       NaN   \n",
      "ccle2_ACH-000065                       NaN   \n",
      "ccle2_ACH-000066                       NaN   \n",
      "ccle2_ACH-000069                       NaN   \n",
      "ccle2_ACH-000073                       NaN   \n",
      "ccle2_ACH-000076                       NaN   \n",
      "ccle2_ACH-000077                       NaN   \n",
      "ccle2_ACH-000079                       NaN   \n",
      "ccle2_ACH-000080                       NaN   \n",
      "ccle2_ACH-000081                       NaN   \n",
      "ccle2_ACH-000083                       NaN   \n",
      "ccle2_ACH-000085                       NaN   \n",
      "ccle2_ACH-000086                       NaN   \n",
      "ccle2_ACH-000089                       NaN   \n",
      "ccle2_ACH-000093                       NaN   \n",
      "ccle2_ACH-000097                       NaN   \n",
      "ccle2_ACH-000099                       NaN   \n",
      "ccle2_ACH-000105                       NaN   \n",
      "...                                    ...   \n",
      "sanger_ACH-002367                      NaN   \n",
      "sanger_ACH-002368                      NaN   \n",
      "sanger_ACH-002369                      NaN   \n",
      "sanger_ACH-002370                      NaN   \n",
      "sanger_ACH-002371                      NaN   \n",
      "sanger_ACH-002372                      NaN   \n",
      "sanger_ACH-002373                      NaN   \n",
      "sanger_ACH-002374                      NaN   \n",
      "sanger_ACH-002375                      NaN   \n",
      "sanger_ACH-002376                      NaN   \n",
      "sanger_ACH-002377                      NaN   \n",
      "sanger_ACH-002378                      NaN   \n",
      "sanger_ACH-002379                      NaN   \n",
      "sanger_ACH-002380                      NaN   \n",
      "sanger_ACH-002381                      NaN   \n",
      "sanger_ACH-002382                      NaN   \n",
      "sanger_ACH-002383                      NaN   \n",
      "sanger_ACH-002384                      NaN   \n",
      "sanger_ACH-002386                      NaN   \n",
      "sanger_ACH-002387                      NaN   \n",
      "sanger_ACH-002388                      NaN   \n",
      "sanger_ACH-002389                      NaN   \n",
      "sanger_ACH-002397                      NaN   \n",
      "ibm_ACH-001441                         NaN   \n",
      "ibm_ACH-002041                         NaN   \n",
      "ibm_ACH-001370                         NaN   \n",
      "ibm_ACH-001627                         NaN   \n",
      "ccle_ACH-001201                        NaN   \n",
      "ccle_ACH-001699                        NaN   \n",
      "ccle_ACH-002509                        NaN   \n",
      "\n",
      "                                                  unmapped_bams_list  \n",
      "sample_id                                                             \n",
      "ccle2_ACH-000011   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000013   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000020   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000026   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000027   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000031   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000032   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000039   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000041   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000042   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000043   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000051   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000063   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000065   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000066   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000069   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000073   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000076   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000077   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000079   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000080   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000081   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000083   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000085   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000086   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000089   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000093   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000097   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000099   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ccle2_ACH-000105   gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "...                                                              ...  \n",
      "sanger_ACH-002367  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002368  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002369  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002370  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002371  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002372  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002373  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002374  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002375  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002376  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002377  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002378  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002379  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002380  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002381  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002382  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002383  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002384  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002386  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002387  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002388  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002389  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "sanger_ACH-002397  gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d2...  \n",
      "ibm_ACH-001441                                                   NaN  \n",
      "ibm_ACH-002041                                                   NaN  \n",
      "ibm_ACH-001370                                                   NaN  \n",
      "ibm_ACH-001627                                                   NaN  \n",
      "ccle_ACH-001201                                                  NaN  \n",
      "ccle_ACH-001699                                                  NaN  \n",
      "ccle_ACH-002509                                                  NaN  \n",
      "\n",
      "[1855 rows x 67 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Pdb) refsamples.loc[name]\n",
      "CCLE_name                                                         WM4235\n",
      "Source                                                              ccle\n",
      "WES_bai                                    gs://ccle_bams/wes/WM4235.bai\n",
      "WES_bam                                    gs://ccle_bams/wes/WM4235.bam\n",
      "allele_fraction_legacy_segments_normal                               NaN\n",
      "allele_fraction_legacy_segments_tumor                                NaN\n",
      "allele_fraction_parameters_begin_normal                              NaN\n",
      "allele_fraction_parameters_begin_tumor                               NaN\n",
      "allele_fraction_parameters_normal                                    NaN\n",
      "allele_fraction_parameters_tumor                                     NaN\n",
      "allelic_counts_entity_id_normal                                      NaN\n",
      "allelic_counts_entity_id_tumor                                       NaN\n",
      "allelic_counts_normal                                                NaN\n",
      "allelic_counts_tumor                                                 NaN\n",
      "called_copy_ratio_segments_normal                                    NaN\n",
      "called_copy_ratio_segments_tumor                                     NaN\n",
      "copy_ratio_legacy_segments_normal                                    NaN\n",
      "copy_ratio_legacy_segments_tumor                                     NaN\n",
      "copy_ratio_only_segments_normal                                      NaN\n",
      "copy_ratio_only_segments_tumor                                       NaN\n",
      "copy_ratio_parameters_begin_normal                                   NaN\n",
      "copy_ratio_parameters_begin_tumor                                    NaN\n",
      "copy_ratio_parameters_normal                                         NaN\n",
      "copy_ratio_parameters_tumor                                          NaN\n",
      "delta_MAD_normal                                                     NaN\n",
      "delta_MAD_tumor                                                      NaN\n",
      "denoised_MAD_normal                                                  NaN\n",
      "denoised_MAD_tumor                                                   NaN\n",
      "denoised_copy_ratios_lim_4_plot_normal                               NaN\n",
      "denoised_copy_ratios_lim_4_plot_tumor                                NaN\n",
      "                                                       ...              \n",
      "hg38_analysis_ready_bam_index                                        NaN\n",
      "hg38_analysis_ready_bam_md5                                          NaN\n",
      "hg38_bqsr_report                                                     NaN\n",
      "hg38_duplication_metrics                                             NaN\n",
      "modeled_segments_begin_normal                                        NaN\n",
      "modeled_segments_begin_tumor                                         NaN\n",
      "modeled_segments_normal                                              NaN\n",
      "modeled_segments_plot_normal                                         NaN\n",
      "modeled_segments_plot_tumor                                          NaN\n",
      "modeled_segments_tumor                                               NaN\n",
      "normal_het_allelic_counts_normal                                     NaN\n",
      "normal_het_allelic_counts_tumor                                      NaN\n",
      "oncotated_called_file_tumor                                          NaN\n",
      "oncotated_called_gene_list_file_tumor                                NaN\n",
      "participant                                                   ACH-002509\n",
      "preprocessed_intervals                                               NaN\n",
      "read_counts_entity_id_normal                                         NaN\n",
      "read_counts_entity_id_tumor                                          NaN\n",
      "read_counts_normal                                                   NaN\n",
      "read_counts_tumor                                                    NaN\n",
      "readgroup_ubams                                                      NaN\n",
      "scaled_delta_MAD_normal                                              NaN\n",
      "scaled_delta_MAD_tumor                                               NaN\n",
      "standardized_MAD_normal                                              NaN\n",
      "standardized_MAD_tumor                                               NaN\n",
      "standardized_copy_ratios_normal                                      NaN\n",
      "standardized_copy_ratios_tumor                                       NaN\n",
      "target_type                                                          NaN\n",
      "ubam_validation_reports                                              NaN\n",
      "unmapped_bams_list                                                   NaN\n",
      "Name: ccle_ACH-002509, Length: 67, dtype: object\n",
      "(Pdb) n\n",
      "Successfully imported 1630 participants.\n",
      "Successfully imported 1855 samples.\n",
      "Hound executing batch upload of> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(115)createDatasetWithNewCellLines()\n",
      "-> print(\"creating a sample set\")\n",
      " 252282 records\n",
      "(Pdb) n\n",
      "creating a sample set\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(116)createDatasetWithNewCellLines()\n",
      "-> wto.update_sample_set(sample_set_id=samplesetname, sample_ids=sample_ids)\n",
      "(Pdb) n\n",
      "Sample set \"19Q4\" (7 samples) successfully updated.\n",
      "> /Users/jeremie/Documents/Projects/BroadInstitute/ccle_processing/src/CCLE_postp_function.py(117)createDatasetWithNewCellLines()\n",
      "-> return sample_ids, refsamples['WES_bam'].values\n",
      "(Pdb) c\n"
     ]
    }
   ],
   "source": [
    "newsample = createDatasetWithNewCellLines(refwm, sample_set_id, wmfroms = [wm1, wm2, wm3], sources=[source1, source2, source3],gsfolderto='gs://ccle_bams/wes/',match='ACH', participantslicepos=10, accept_unknowntypes=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ReaLign in HG38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bamtoubam= \"BamToUnmappedRGBams_MC\"\n",
    "ubamtofilelist = \"Generate_uBAM_File_List\"\n",
    "realign=\"Realign_WES_GATK4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission 28d7be6b-c9dc-42d2-a4a8-e8c74e2dda3d.\n"
     ]
    },
    {
     "ename": "ConnectionError",
     "evalue": "('Connection aborted.', OSError(\"(60, 'ETIMEDOUT')\",))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    599\u001b[0m                                                   \u001b[0mbody\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 600\u001b[0;31m                                                   chunked=chunked)\n\u001b[0m\u001b[1;32m    601\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    383\u001b[0m                     \u001b[0;31m# otherwise it looks like a programming error was the cause.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m                     \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_from\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSocketTimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBaseSSLError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSocketError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/packages/six.py\u001b[0m in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    379\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 380\u001b[0;31m                     \u001b[0mhttplib_response\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    381\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1330\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1331\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1332\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    296\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 297\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    298\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    257\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 258\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    259\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    585\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 586\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    587\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    301\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 302\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mSocketError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    303\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOpenSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mZeroReturnError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: (60, 'ETIMEDOUT')",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mProtocolError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    448\u001b[0m                     \u001b[0mretries\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_retries\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 449\u001b[0;31m                     \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    450\u001b[0m                 )\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    637\u001b[0m             retries = retries.increment(method, url, error=e, _pool=self,\n\u001b[0;32m--> 638\u001b[0;31m                                         _stacktrace=sys.exc_info()[2])\n\u001b[0m\u001b[1;32m    639\u001b[0m             \u001b[0mretries\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/util/retry.py\u001b[0m in \u001b[0;36mincrement\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    367\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mread\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mFalse\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_method_retryable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 368\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreraise\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_stacktrace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    369\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mread\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/packages/six.py\u001b[0m in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    684\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    599\u001b[0m                                                   \u001b[0mbody\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 600\u001b[0;31m                                                   chunked=chunked)\n\u001b[0m\u001b[1;32m    601\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    383\u001b[0m                     \u001b[0;31m# otherwise it looks like a programming error was the cause.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m                     \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_from\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSocketTimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBaseSSLError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSocketError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/packages/six.py\u001b[0m in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    379\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 380\u001b[0;31m                     \u001b[0mhttplib_response\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    381\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1330\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1331\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1332\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    296\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 297\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    298\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    257\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 258\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    259\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    585\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 586\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    587\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    301\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 302\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mSocketError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    303\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOpenSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mZeroReturnError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mProtocolError\u001b[0m: ('Connection aborted.', OSError(\"(60, 'ETIMEDOUT')\",))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-3de5e6e1f935>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msubid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrefwm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_submission\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbamtoubam\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msample_set_id\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"sample_set\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"this.samples\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mterra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitForSubmission\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrefwm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/dalmatian/wmanager.py\u001b[0m in \u001b[0;36mcreate_submission\u001b[0;34m(self, reference, entity, etype, expression, use_callcache)\u001b[0m\n\u001b[1;32m   2115\u001b[0m                 ),\n\u001b[1;32m   2116\u001b[0m                 entities=[\n\u001b[0;32m-> 2117\u001b[0;31m                     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mentity\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2118\u001b[0m                 ]\n\u001b[1;32m   2119\u001b[0m             )\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/hound/client.py\u001b[0m in \u001b[0;36mwrite_log_entry\u001b[0;34m(self, log_type, text, entities)\u001b[0m\n\u001b[1;32m    389\u001b[0m             {\n\u001b[1;32m    390\u001b[0m                 \u001b[0;34m'entities'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mentities\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 391\u001b[0;31m                 \u001b[0;34m'text'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mtext\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    392\u001b[0m             }\n\u001b[1;32m    393\u001b[0m         )\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/hound/client.py\u001b[0m in \u001b[0;36mwrite\u001b[0;34m(self, path, data)\u001b[0m\n\u001b[1;32m    223\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_batch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m                 \u001b[0mstorage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mBlob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbucket\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupload_from_string\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/cloud/storage/blob.py\u001b[0m in \u001b[0;36mupload_from_string\u001b[0;34m(self, data, content_type, client, predefined_acl)\u001b[0m\n\u001b[1;32m   1255\u001b[0m             \u001b[0mcontent_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcontent_type\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m             \u001b[0mclient\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1257\u001b[0;31m             \u001b[0mpredefined_acl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpredefined_acl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1258\u001b[0m         )\n\u001b[1;32m   1259\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/cloud/storage/blob.py\u001b[0m in \u001b[0;36mupload_from_file\u001b[0;34m(self, file_obj, rewind, size, content_type, num_retries, client, predefined_acl)\u001b[0m\n\u001b[1;32m   1155\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1156\u001b[0m             created_json = self._do_upload(\n\u001b[0;32m-> 1157\u001b[0;31m                 \u001b[0mclient\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile_obj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontent_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_retries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredefined_acl\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1158\u001b[0m             )\n\u001b[1;32m   1159\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_properties\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcreated_json\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/cloud/storage/blob.py\u001b[0m in \u001b[0;36m_do_upload\u001b[0;34m(self, client, stream, content_type, size, num_retries, predefined_acl)\u001b[0m\n\u001b[1;32m   1061\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0msize\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0msize\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0m_MAX_MULTIPART_SIZE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1062\u001b[0m             response = self._do_multipart_upload(\n\u001b[0;32m-> 1063\u001b[0;31m                 \u001b[0mclient\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontent_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_retries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredefined_acl\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1064\u001b[0m             )\n\u001b[1;32m   1065\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/cloud/storage/blob.py\u001b[0m in \u001b[0;36m_do_multipart_upload\u001b[0;34m(self, client, stream, content_type, size, num_retries, predefined_acl)\u001b[0m\n\u001b[1;32m    855\u001b[0m             )\n\u001b[1;32m    856\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 857\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mupload\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransmit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtransport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobject_metadata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontent_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    858\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    859\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/resumable_media/requests/upload.py\u001b[0m in \u001b[0;36mtransmit\u001b[0;34m(self, transport, data, metadata, content_type)\u001b[0m\n\u001b[1;32m     95\u001b[0m         result = _helpers.http_request(\n\u001b[1;32m     96\u001b[0m             \u001b[0mtransport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpayload\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m             retry_strategy=self._retry_strategy)\n\u001b[0m\u001b[1;32m     98\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_process_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/resumable_media/requests/_helpers.py\u001b[0m in \u001b[0;36mhttp_request\u001b[0;34m(transport, method, url, data, headers, retry_strategy, **transport_kwargs)\u001b[0m\n\u001b[1;32m     99\u001b[0m         **transport_kwargs)\n\u001b[1;32m    100\u001b[0m     return _helpers.wait_and_retry(\n\u001b[0;32m--> 101\u001b[0;31m         func, RequestsMixin._get_status_code, retry_strategy)\n\u001b[0m",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/resumable_media/_helpers.py\u001b[0m in \u001b[0;36mwait_and_retry\u001b[0;34m(func, get_status_code, retry_strategy)\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0mobject\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mThe\u001b[0m \u001b[0;32mreturn\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0mof\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    145\u001b[0m     \"\"\"\n\u001b[0;32m--> 146\u001b[0;31m     \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    147\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mget_status_code\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mRETRYABLE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    148\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/google/auth/transport/requests.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, data, headers, **kwargs)\u001b[0m\n\u001b[1;32m    206\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m         response = super(AuthorizedSession, self).request(\n\u001b[0;32m--> 208\u001b[0;31m             method, url, data=data, headers=request_headers, **kwargs)\n\u001b[0m\u001b[1;32m    209\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m         \u001b[0;31m# If the response indicated that the credentials needed to be\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    531\u001b[0m         }\n\u001b[1;32m    532\u001b[0m         \u001b[0msend_kwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 533\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    534\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    535\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    644\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    645\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 646\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    648\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/envs/py36/lib/python3.6/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    496\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    497\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mProtocolError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 498\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    500\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mMaxRetryError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mConnectionError\u001b[0m: ('Connection aborted.', OSError(\"(60, 'ETIMEDOUT')\",))"
     ]
    }
   ],
   "source": [
    "subid = refwm.create_submission(bamtoubam,sample_set_id,\"sample_set\",\"this.samples\")\n",
    "terra.waitForSubmission(refwm, subid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission 8badb033-63f6-419f-8056-b9331f8bc64c.\n",
      "status is: Failed for 0 jobs in submission 0. 6 mn elapsed.\r"
     ]
    }
   ],
   "source": [
    "subid = refwm.create_submission(ubamtofilelist,sample_set_id,\"sample_set\",\"this.samples\")\n",
    "terra.waitForSubmission(refwm, subid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subid = refwm.create_submission(realign,sample_set_id,\"sample_set\",\"this.samples\")\n",
    "terra.waitForSubmission(refwm, subid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"CNV_sample_XX\",etype='sample_set',entity=sample_set_id,expression='this.samples')\n",
    "terra.waitForSubmission(refwm,submission_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"Aggregate_CN_seg_files\",entity=sample_set_id)\n",
    "terra.waitForSubmission(refwm,submission_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aggregated = refwm.get_entities('sample_set').loc[sample_set_id][\"combined_seg_file\"]\n",
    "aggregated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## deleting unmapped bams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "readgroup_ubams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_analysis_ready_bam\n",
    "hg38_analysis_ready_bam_index\n",
    "hg38_analysis_ready_bam_md5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://fc-secure-d2a2d895-a7af-4117-bdc7-652d7d268324/9b51ecda-b8d3-42bd-a56b-015b94533ee1/aggregate_CN_segments_wrkflw/a565c462-e783-47a4-85c5-a6eba786f03b/call-aggregate_CN_segments/19Q4.called.seg...\n",
      "- [1 files][110.8 KiB/110.8 KiB]      0.0 B/s                                   \n",
      "Operation completed over 1 objects/110.8 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "! gsutil cp $aggregated \"temp/cnv_ccle.called.seg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:07::WARNING  R[write to console]: ── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.2.1 ──\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.2.0     \u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 0.8.3\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 2.1.3     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.2.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.4.0\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: ── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m    masks \u001b[34mplyr\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m   masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m    masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m    masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m      masks \u001b[34mplyr\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m       masks \u001b[34mplyr\u001b[39m::desc(), \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m     masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m   masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m     masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m      masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m         masks \u001b[34mplyr\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m        masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m     masks \u001b[34mplyr\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m     masks \u001b[34mplyr\u001b[39m::rename(), \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m     masks \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m      masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m  masks \u001b[34mplyr\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m  masks \u001b[34mplyr\u001b[39m::summarize()\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘magrittr’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: Loading required package: lpSolve\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘zoo’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.Date, as.Date.numeric\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:readr’:\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘rlang’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    %@%, as_function, flatten, flatten_chr, flatten_dbl, flatten_int,\n",
      "    flatten_lgl, flatten_raw, invoke, list_along, modify, prepend,\n",
      "    splice\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    exprs\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: Loading required package: SnowballC\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘ggridges’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    scale_discrete_manual\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘GGally’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    nasa\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: \n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, mutate, rename, summarise\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: Loading required package: grid\n",
      "\n",
      "2019-10-08 11:41:07::WARNING  R[write to console]: Loading required package: futile.logger\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Registering fonts with R\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: \n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following object is masked from ‘package:ggthemes’:\n",
      "\n",
      "    theme_map\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    ggsave\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: .Keyboard already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: More than one version of regular/bold/italic found for System Font. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: .SF NS Rounded already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Andale Mono already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Anonymice Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: More than one version of regular/bold/italic found for Apple Braille. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: AppleMyungjo already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arial Black already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arial already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arial Narrow already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arial Rounded MT Bold already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arial Unicode MS already registered with pdfFonts().\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:08::WARNING  R[write to console]: Arimo for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Bodoni Ornaments already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Bodoni 72 Smallcaps already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: No regular (non-bold, non-italic) version of . Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: No regular (non-bold, non-italic) version of Brush Script MT. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Comic Sans MS already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Courier New already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Cousine for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: IBM 3270 already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: IBM 3270 Narrow already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: IBM 3270 Semi-Narrow already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: DejaVu Sans Mono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: No regular (non-bold, non-italic) version of DIN Alternate. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: No regular (non-bold, non-italic) version of DIN Condensed. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Droid Sans Mono Dotted for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Droid Sans Mono Slashed for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Georgia already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Go Mono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Hack already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Impact already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: No regular (non-bold, non-italic) version of Inconsolata for Powerline. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Khmer Sangam MN already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Lao Sangam MN already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Literation Mono Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Luminari already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG L DZ for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG L for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG M DZ for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG M for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG S DZ for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Meslo LG S for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Microsoft Sans Serif already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: More than one version of regular/bold/italic found for monofur for Powerline. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Noto Mono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: NovaMono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: ProFont for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: More than one version of regular/bold/italic found for Roboto. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: More than one version of regular/bold/italic found for Roboto Condensed. Skipping setup for this font.\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Light already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Medium already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Thin already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Light for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Medium for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Roboto Mono Thin for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Space Mono already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Space Mono for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Symbol Neu for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Tahoma already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Times New Roman already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Tinos for Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Trattatello already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Trebuchet MS already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Ubuntu Mono derivative Powerline already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Verdana already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Webdings already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Wingdings already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Wingdings 2 already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Wingdings 3 already registered with pdfFonts().\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: \n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following object is masked from ‘package:rlang’:\n",
      "\n",
      "    :=\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following objects are masked from ‘package:reshape2’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "2019-10-08 11:41:08::WARNING  R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: Loading required package: annotate\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: Loading required package: XML\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: \n",
      "Attaching package: ‘XML’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: The following object is masked from ‘package:tools’:\n",
      "\n",
      "    toHTML\n",
      "\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: Loading required package: graph\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: \n",
      "Attaching package: ‘graph’\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:09::WARNING  R[write to console]: The following object is masked from ‘package:XML’:\n",
      "\n",
      "    addNode\n",
      "\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: The following object is masked from ‘package:stringr’:\n",
      "\n",
      "    boundary\n",
      "\n",
      "\n",
      "2019-10-08 11:41:09::WARNING  R[write to console]: The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    join\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: Loading required package: RCurl\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: Loading required package: bitops\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: \n",
      "Attaching package: ‘RCurl’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    complete\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: Loading required package: httr\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: \n",
      "Attaching package: ‘httr’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: The following object is masked from ‘package:plotly’:\n",
      "\n",
      "    config\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    content\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: Package 'mclust' version 5.4.5\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: \n",
      "Attaching package: ‘mclust’\n",
      "\n",
      "\n",
      "2019-10-08 11:41:10::WARNING  R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    map\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "source('../gkugener/RScripts/load_libraries_and_annotations.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "source(\"src/CCLE_postp_function.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching https://cds.team/taiga/api/datafile?format=metadata&dataset_permaname=segmented-cn-wes-prioritzed-7fe1&datafile_name=wes.19Q3.segmented \n",
      "Status 200 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:29::WARNING  R[write to console]: No encoding supplied: defaulting to UTF-8.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not find segmented-cn-wes-prioritzed-7fe1 in cache, requesting from taiga...\n",
      "Fetching https://cds.team/taiga/api/datafile?format=rds&dataset_version_id=1113ef8f975b4d8c9137fdb09e405448&datafile_name=wes.19Q3.segmented \n",
      "Status 200 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:29::WARNING  R[write to console]: No encoding supplied: defaulting to UTF-8.\n",
      "\n",
      "2019-10-08 11:41:29::WARNING  R[write to console]: Downloading segmented-cn-wes-prioritzed-7fe1/v33/wes.19Q3.segmented ...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n",
      "Saving segmented-cn-wes-prioritzed-7fe1 in cache ( 1113ef8f975b4d8c9137fdb09e405448 wes.19Q3.segmented )...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 11:41:34::WARNING  R[write to console]: Saved to cache as 1113ef8f975b4d8c9137fdb09e405448_wes-19q3-segmented.rds\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing /Users/jeremie/.taiga/1113ef8f975b4d8c9137fdb09e405448_wes-19q3-segmented.idx /Users/jeremie/.taiga/segmented-cn-wes-prioritzed-7fe1_wes-19q3-segmented_33.idx \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# Previous release copy number profiles. This line will need to be updated as well\n",
    "wes.priority.cn.seg.profiles <- taigr::load.from.taiga(data.name='segmented-cn-wes-prioritzed-7fe1', data.file=prevfile) %>%\n",
    "  dplyr::select(DepMap_ID, Chromosome, Start, End, Num_Probes, Segment_Mean, Source)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## If want to reprocess something"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%R\n",
    "#segments_unfiltered <- readr::read_csv('temp/wes.19Q3.segmented.cn')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## post Procesing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### getting blacklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 12:03:51::INFO  URL being requested: GET https://sheets.googleapis.com/v4/spreadsheets/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4?alt=json\n",
      "2019-10-08 12:03:52::INFO  URL being requested: GET https://sheets.googleapis.com/v4/spreadsheets/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4/values:batchGet?majorDimension=ROWS&valueRenderOption=UNFORMATTED_VALUE&dateTimeRenderOption=FORMATTED_STRING&ranges=Guide&ranges=RNAseq+Embargo&ranges=19Q4&ranges=WES+Embargo&ranges=Blacklist&ranges=Allie%27s+list+-+public+in+19Q2&ranges=full_list&alt=json\n"
     ]
    }
   ],
   "source": [
    "gsheets = sheets.get(sheeturl).sheets[6].to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ibm_embargo = [i for i in gsheets['RNAseq_IBM_embargo'].values.tolist() if i is not np.nan]\n",
    "rna_dmc_embargo = [i for i in gsheets['RNAseq_DMC_embargo'].values.tolist() if i is not np.nan]\n",
    "wes_dmc_embargo = [i for i in gsheets['WES_DMC_embargo'].values.tolist() if i is not np.nan]\n",
    "blacklist = [i for i in gsheets['blacklist'].values.tolist() if i is not np.nan]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-10-08 12:18:27::WARNING  R[write to console]: Parsed with column specification:\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m,\n",
      "  X3 = \u001b[32mcol_double()\u001b[39m,\n",
      "  X4 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X5 = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# A tibble: 2,094 x 7\n",
      "   DepMap_ID       seqnames     start       end Num_Probes Segment_Mean Source  \n",
      "   <chr>           <fct>        <dbl>     <dbl>      <dbl>        <dbl> <chr>   \n",
      " 1 ccle_ACH-002510 chr1             1  21518734       2734     9.76e- 1 Broad W…\n",
      " 2 ccle_ACH-002510 chr1      21518735 101921860       6429     9.26e- 1 Broad W…\n",
      " 3 ccle_ACH-002510 chr1     101921861 132724462       1386     6.34e- 1 Broad W…\n",
      " 4 ccle_ACH-002510 chr1     132724463 148537132        250     1.12e+ 0 Broad W…\n",
      " 5 ccle_ACH-002510 chr1     148537133 148549687          1     7.40e-10 Broad W…\n",
      " 6 ccle_ACH-002510 chr1     148549688 161569704       2475     1.06e+ 0 Broad W…\n",
      " 7 ccle_ACH-002510 chr1     161569705 161651296          9     1.49e- 7 Broad W…\n",
      " 8 ccle_ACH-002510 chr1     161651297 216391800       3824     1.07e+ 0 Broad W…\n",
      " 9 ccle_ACH-002510 chr1     216391801 222131173        199     1.87e+ 0 Broad W…\n",
      "10 ccle_ACH-002510 chr1     222131174 232963176        962     1.04e+ 0 Broad W…\n",
      "# … with 2,084 more rows\n",
      "   DepMap_ID Chromosome    Start      End Num_Probes Segment_Mean     Source\n",
      "1 ACH-000001          1        1  1969745        286     2.546065 Sanger WES\n",
      "2 ACH-000001          1  1969746  6354345        365     2.175759 Sanger WES\n",
      "3 ACH-000001          1  6354346  6958256        100     3.109430 Sanger WES\n",
      "4 ACH-000001          1  6958257 15977206        884     2.134831 Sanger WES\n",
      "5 ACH-000001          1 15977207 16174774         57     2.952592 Sanger WES\n",
      "6 ACH-000001          1 16174775 16275770         19     1.710129 Sanger WES\n",
      "# A tibble: 2,094 x 7\n",
      "   DepMap_ID      Chromosome     Start      End Num_Probes Segment_Mean Source  \n",
      "   <chr>          <fct>          <dbl>    <dbl>      <dbl>        <dbl> <chr>   \n",
      " 1 ccle_ACH-0025… chr1               1   2.15e7       2734     9.76e- 1 Broad W…\n",
      " 2 ccle_ACH-0025… chr1        21518735   1.02e8       6429     9.26e- 1 Broad W…\n",
      " 3 ccle_ACH-0025… chr1       101921861   1.33e8       1386     6.34e- 1 Broad W…\n",
      " 4 ccle_ACH-0025… chr1       132724463   1.49e8        250     1.12e+ 0 Broad W…\n",
      " 5 ccle_ACH-0025… chr1       148537133   1.49e8          1     7.40e-10 Broad W…\n",
      " 6 ccle_ACH-0025… chr1       148549688   1.62e8       2475     1.06e+ 0 Broad W…\n",
      " 7 ccle_ACH-0025… chr1       161569705   1.62e8          9     1.49e- 7 Broad W…\n",
      " 8 ccle_ACH-0025… chr1       161651297   2.16e8       3824     1.07e+ 0 Broad W…\n",
      " 9 ccle_ACH-0025… chr1       216391801   2.22e8        199     1.87e+ 0 Broad W…\n",
      "10 ccle_ACH-0025… chr1       222131174   2.33e8        962     1.04e+ 0 Broad W…\n",
      "# … with 2,084 more rows\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "segments <- processSegments(new_samples_copy_number_broad_wes)\n",
    "segments <- filterForCCLE(segments)\n",
    "segments <- interpolateGapsInSegmented(segments)$segs\n",
    "segments <- extendEndsOfSegments(segments,hg38_cyto_band_reference)\n",
    "print(segments)\n",
    "print(head(wes.priority.cn.seg.profiles))\n",
    "segments_unfiltered <- reprioritizeData(segments, wes.priority.cn.seg.profiles)\n",
    "# Fill in the gaps on the entire dataset\n",
    "# Extend start sites to 1, end sites to the end of the chromosome?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i wes_dmc_embargo,blacklist\n",
    "\n",
    "segments_blacklisted <- filterBlackListedLine(blacklist,segments_unfiltered)\n",
    "\n",
    "# TODO: IF seqnames (CHR) are 1-9 values, append \"chr\" in front of each! (bulk)\n",
    "# ifelse(grepl(\"[0-9]+\", new_copy_number$ChrChromosome), \"chr\"+, \"no\")\n",
    "segments_embargo <- filterBlackListedLine(wes_dmc_embargo,segments_unfiltered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "write.table(segments_unfiltered, file = paste0(\"temp/wes.\",release,\".segmented.cn\"), sep = ',', quote = F, row.names = F) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "entrezgenes <- generateEntrezGenes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   EGID SYMBOL CHR    CHRLOC CHRLOCEND\n",
      "1     1   A1BG  19  58346805  58353499\n",
      "2    10   NAT2   8  18391281  18401215\n",
      "3   100    ADA  20  44619518  44651758\n",
      "4  1000   CDH2  18  27950962  28177130\n",
      "5 10000   AKT3   1 243488232 243843584\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "head(corner(entrezgenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "res <- generateGeneLevelMatrixFromSegments(entrezgenes, segments_unfiltered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           A1BG (1) NAT2 (10) ADA (100) CDH2 (1000) AKT3 (10000)\n",
      "ACH-000001 1.265172 1.1415254 1.2958359   0.6627515     1.012791\n",
      "ACH-000002 1.013633 0.9741737 1.0175121   1.4893275     1.012618\n",
      "ACH-000003 1.031898 1.0958118 1.8768326   0.5364814     1.002498\n",
      "ACH-000004 1.349290 1.0865089 0.7855092   1.0770610     1.086057\n",
      "ACH-000005 1.222033 1.1622070 0.7612536   1.0724579     1.120937\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "genematrix_unfiltered <- res$gene_level_data_hg38\n",
    "corner(genematrix_unfiltered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "write.table(genematrix_unfiltered, file = paste0('temp/wes.',release,'.gene.cn'), \n",
    "sep = ',', quote = F, row.names = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn = pd.read_csv('temp/wes.'+release+'.gene.cn', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevsegmentcn = pd.read_csv('temp/cnv_ccle.called.seg', sep = '\\t')\n",
    "segmentcn = pd.read_csv('temp/wes.'+release+'.segmented.cn', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn.index = [i[-1] for i in genecn.index.str.split('_').tolist()]\n",
    "segmentcn[\"DepMap_ID\"] = [i[-1] for i in segmentcn[\"DepMap_ID\"].str.split('_').tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embargoed = \"WES_embargoed.txt\"\n",
    "embargoed = [line.rstrip('\\n') for line in open(embargoed)]\n",
    "blacklisted = \"blacklist.txt\"\n",
    "blacklisted = [line.rstrip('\\n') for line in open(blacklisted)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prev = set(tc.get(name='segmented-cn-wes-prioritzed-7fe1', version=27, file='wes.19Q3interim.gene').index.values.tolist())\n",
    "new1 = set(genecn.index.values.tolist())\n",
    "new2 = set(segmentcn['Sample'].values.tolist())\n",
    "print(len(prev), len(prev & new1), len(new1), len(new1 & new2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "checkAmountOfSegments(segmentcn,thresh = 750)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "checkGeneChangeAccrossAll(genecn, thresh=1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newsamples =  list(set(segmentcn[\"Sample\"].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "segmentcn[segmentcn[\"DepMap_ID\"]==\"ACH-000658\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newsamples= refwm.get_sample_sets().loc[sample_set_id].samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "samples = refwm.get_samples()\n",
    "plots = samples.loc[samples.index.isin(newsamples)][\"modeled_segments_plot_tumor\"].values\n",
    "for plot in plots:\n",
    "    ! gsutil cp $plot temp/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for plot in plots:\n",
    "    display(Image('temp/'+plot.split('/')[-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn.to_csv('temp/wes.'+release+'.gene.cn')\n",
    "segmentcn.to_csv('temp/wes.'+release+'.segmented.cn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for genecn removing first blacklisted, then embargoed, to create two datasets\n",
    "genecn = genecn.apply(lambda x: np.log2(1+x))\n",
    "for i in genecn.index:\n",
    "    if i in blacklisted:\n",
    "        genecn = genecn.drop(i)\n",
    "genecn.to_csv('temp/wes.'+release+'.blacklisted.gene.cn')\n",
    "for i in genecn.index:\n",
    "    if i in embargoed:\n",
    "        genecn = genecn.drop(i)\n",
    "genecn.to_csv('temp/wes.'+release+'.embargoed.gene.cn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for segment removing first blacklisted, then embargoed, to create two datasets\n",
    "segmentcn['Segment_Mean'] = segmentcn['Segment_Mean'].apply(lambda x: np.log2(1+x))\n",
    "for i, val in segmentcn.iterrows():\n",
    "    if val['DepMap_ID'] in blacklisted:\n",
    "        segmentcn = segmentcn.drop(i)\n",
    "segmentcn.to_csv('temp/wes.'+release+'.blacklisted.segmented.cn')\n",
    "for i, val in segmentcn.iterrows():\n",
    "    if val['DepMap_ID'] in embargoed:\n",
    "        segmentcn = segmentcn.drop(i)\n",
    "segmentcn.to_csv('temp/wes.'+release+'.embargoed.segmented.cn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segmentcn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to taiga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tc.update_dataset(dataset_permaname=\"segmented-cn-wes-prioritzed-7fe1\", \n",
    "                  upload_file_path_dict={'temp/wes.'+release+'.gene.cn': 'NumericMatrixCSV',\n",
    "                                        'temp/wes.'+release+'.segmented.cn': 'TableCSV',\n",
    "                                        'temp/wes.'+release+'.embargoed.gene.cn': 'NumericMatrixCSV','temp/wes.'+release+'.embargoed.segmented.cn': 'TableCSV',\n",
    "                                         'temp/wes.'+release+'.blacklisted.gene.cn': 'NumericMatrixCSV',\n",
    "                                         'temp/wes.'+release+'.blacklisted.segmented.cn': 'TableCSV'},\n",
    "                dataset_description=\"updating to \"+release+\". Some cells lines have been flagged as: \\n \\\n",
    "                - having bad looking copy ration plots = [] \\n \\\n",
    "                 - having too many segments = [] \\n \\\n",
    "                  - Genes having a similar CN value accross all []\",\n",
    "                 )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
