{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro & Loading "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T19:03:10.160960Z",
     "start_time": "2021-01-22T19:02:56.119553Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have JKBio in your path:\n",
      "e.g. have installed JKBio in the same folder as ccle_processing\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:14: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n",
       "  var css_urls = [];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n  var css_urls = [];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import pandas as pd\n",
    "\n",
    "from genepy.utils import helper as h\n",
    "from gsheets import Sheets\n",
    "from taigapy import TaigaClient\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "tc = TaigaClient()\n",
    "output_notebook()\n",
    "\n",
    "my_id = '~/.client_secret.json'\n",
    "mystorage_id = \"~/.storage.json\"\n",
    "sheets = Sheets.from_files(my_id, mystorage_id)\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T19:03:11.322504Z",
     "start_time": "2021-01-22T19:03:11.290334Z"
    }
   },
   "outputs": [],
   "source": [
    "samplesetname = \"21Q1\"\n",
    "\n",
    "virtual= {}\n",
    "virtual['public'] = 'public-21q1-4b39'\n",
    "virtual['ibm'] = 'ibm-21q1-abd9'\n",
    "virtual['dmc'] = 'dmc-21q1-0e11'\n",
    "virtual['internal']='internal-21q1-4fc4'\n",
    "\n",
    "\n",
    "taiga_mutation = {}\n",
    "taiga_mutation['internal'] = \"depmap-mutation-calls-9be3\"\n",
    "taiga_mutation['ibm'] = \"mutations-b05c\"\n",
    "taiga_mutation['dmc'] = \"depmap-mutation-calls-dfce\"\n",
    "taiga_mutation['public'] =\"depmap-mutation-calls-9a1a\"\n",
    "\n",
    "taiga_expression = {}\n",
    "taiga_expression['internal'] =\"depmap-rnaseq-expression-data-363a\"\n",
    "taiga_expression['ibm'] = \"expression-9d97\"\n",
    "taiga_expression['dmc'] = \"depmap-rnaseq-expression-data-80ef\"\n",
    "taiga_expression['public'] =\"depmap-rnaseq-expression-data-ccd0\"\n",
    "\n",
    "taiga_fusion = {}\n",
    "taiga_fusion['internal'] = \"gene-fusions-8b7a\"\n",
    "taiga_fusion['ibm'] = \"fusion-dc08\"\n",
    "taiga_fusion['dmc'] = \"gene-fusions-375f\"\n",
    "taiga_fusion['public'] =\"gene-fusions-6212\"\n",
    "\n",
    "taiga_copynumber = {}\n",
    "taiga_copynumber['internal'] =\"depmap-wes-cn-data-81a7\"\n",
    "taiga_copynumber['ibm'] = \"cn-e20f\"\n",
    "taiga_copynumber['dmc'] = \"depmap-cn-data-9b9d\"\n",
    "taiga_copynumber['public'] = \"depmap-wes-cn-data-97cc\"\n",
    "\n",
    "prev_virtual = {}\n",
    "#prev_virtual['public'] = 'public-20q3-3d35'\n",
    "#prev_virtual['dmc'] = 'dmc-20q3-deprecated-never-released--5f55'\n",
    "#prev_virtual['internal'] = 'internal-20q3-00d0'\n",
    "\n",
    "prev_virtual['public'] = 'public-20q4-a4b3'\n",
    "prev_virtual['dmc'] = 'dmc-20q4-fcf4'\n",
    "# prev_virtual['ibm'] = 'ibm-20q4-269f'\n",
    "prev_virtual['internal'] = 'internal-20q4-2540'\n",
    "potential_list_url = \"https://docs.google.com/spreadsheets/d/1YuKEgZ1pFKRYzydvncQt9Y_BKToPlHP-oDB-0CAv3gE\"\n",
    "\n",
    "release = samplesetname.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:11:30.687633Z",
     "start_time": "2021-01-21T19:11:29.357957Z"
    }
   },
   "outputs": [],
   "source": [
    "new = {}\n",
    "gsheets = sheets.get(potential_list_url).sheets[0].to_frame()\n",
    "new['internal'] = set([i for i in gsheets['Internal'].values.tolist() if str(i) != \"nan\"])\n",
    "new['dmc'] = set([i for i in gsheets['DMC'].values.tolist() if str(i) != \"nan\"])\n",
    "new['ibm'] = set([i for i in gsheets['IBM'].values.tolist() if str(i) != \"nan\"])\n",
    "new['public'] = set([i for i in gsheets['Public'].values.tolist() if str(i) != \"nan\"])\n",
    "\n",
    "\n",
    "new[\"internal\"] = new[\"internal\"] | new[\"ibm\"] | new[\"dmc\"] | new[\"public\"]\n",
    "new[\"ibm\"] = new[\"ibm\"] | new[\"dmc\"] | new[\"public\"]\n",
    "new[\"dmc\"] = new[\"dmc\"] | new[\"public\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting what was released before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:12:13.384112Z",
     "start_time": "2021-01-21T19:11:35.577929Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "internal\n",
      "dmc\n",
      "public\n"
     ]
    }
   ],
   "source": [
    "prevmut = {}\n",
    "prevrna = {}\n",
    "prevcn = {}\n",
    "prevwes = {}\n",
    "prev = {}\n",
    "for val in ['internal', 'dmc', 'public']:\n",
    "    print(val)\n",
    "    prevmut[val] = set(tc.get(name=prev_virtual[val], file='CCLE_mutations').DepMap_ID)\n",
    "    prevrna[val] = set(tc.get(name=prev_virtual[val], file='CCLE_expression').index)\n",
    "    prevcn[val] = set(tc.get(name=prev_virtual[val], file='CCLE_segment_cn').DepMap_ID)\n",
    "    prev[val] = prevmut[val] | prevrna[val] | prevcn[val]\n",
    "    prevwes[val] = prevmut[val] | prevcn[val]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:12:46.306287Z",
     "start_time": "2021-01-21T19:12:46.273424Z"
    }
   },
   "outputs": [],
   "source": [
    "prevmut[\"dmc\"] = prevmut[\"dmc\"] | prevmut[\"public\"]\n",
    "prevrna[\"dmc\"] = prevrna[\"dmc\"] | prevrna[\"public\"]\n",
    "prevcn[\"dmc\"] = prevcn[\"dmc\"] | prevcn[\"public\"]\n",
    "prev[\"dmc\"] = prev[\"dmc\"] | prev[\"public\"]\n",
    "prevwes[\"dmc\"] = prevwes[\"dmc\"] | prevwes[\"public\"]\n",
    "\n",
    "prevmut[\"internal\"] = prevmut[\"internal\"] | prevmut[\"dmc\"]\n",
    "prevrna[\"internal\"] = prevrna[\"internal\"] | prevrna[\"dmc\"]\n",
    "prevcn[\"internal\"] = prevcn[\"internal\"] | prevcn[\"dmc\"]\n",
    "prev[\"internal\"] = prev[\"internal\"] | prev[\"dmc\"]\n",
    "prevwes[\"internal\"] = prevwes[\"internal\"] | prevwes[\"dmc\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:20:27.118455Z",
     "start_time": "2021-01-21T19:20:27.083805Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in cn but not mut\n",
      "{'ACH-001175', 'ACH-001037', 'ACH-001198', 'ACH-001225', 'ACH-001171', 'ACH-001079', 'ACH-002335', 'ACH-001045', 'ACH-001088', 'ACH-001087', 'ACH-001071', 'ACH-001015', 'ACH-001224', 'ACH-001018', 'ACH-001000', 'ACH-001101', 'ACH-002303', 'ACH-001017', 'ACH-001249', 'ACH-001121'}\n",
      "in mut but not cn\n",
      "{'ACH-002014', 'ACH-001971', 'ACH-000033', 'ACH-001854', 'ACH-001187', 'ACH-002512', 'ACH-002394', 'ACH-002392', 'ACH-001349', 'ACH-001293', 'ACH-002396', 'ACH-000629', 'ACH-002395', 'ACH-001693', 'ACH-001210', 'ACH-001449', 'ACH-002010', 'ACH-001855', 'ACH-002390', 'ACH-001493', 'ACH-001438', 'ACH-001537', 'ACH-000712', 'ACH-001669', 'ACH-002393', 'ACH-001696', 'ACH-001131', 'ACH-001078', 'ACH-001676', 'ACH-002391', 'ACH-001662', 'ACH-001502', 'ACH-000084', 'ACH-001437', 'ACH-001108', 'ACH-000010', 'ACH-001011', 'ACH-001109', 'ACH-001672'}\n",
      "in rna but no wes/wgs\n",
      "{'ACH-001743', 'ACH-001429', 'ACH-001712', 'ACH-001316', 'ACH-001708', 'ACH-001393', 'ACH-001741'}\n",
      "in wes/wgs but not rna\n",
      "{'ACH-002304', 'ACH-001489', 'ACH-002133', 'ACH-002144', 'ACH-001002', 'ACH-002245', 'ACH-001107', 'ACH-002191', 'ACH-002093', 'ACH-002252', 'ACH-002105', 'ACH-002175', 'ACH-002346', 'ACH-002097', 'ACH-002204', 'ACH-002161', 'ACH-001143', 'ACH-002190', 'ACH-002192', 'ACH-002337', 'ACH-001639', 'ACH-002376', 'ACH-001130', 'ACH-002122', 'ACH-002180', 'ACH-001125', 'ACH-002291', 'ACH-002285', 'ACH-002089', 'ACH-002316', 'ACH-002387', 'ACH-001171', 'ACH-002315', 'ACH-002241', 'ACH-002173', 'ACH-001955', 'ACH-002312', 'ACH-001350', 'ACH-002142', 'ACH-002292', 'ACH-002141', 'ACH-001063', 'ACH-002389', 'ACH-002382', 'ACH-002279', 'ACH-001233', 'ACH-001064', 'ACH-002248', 'ACH-001502', 'ACH-002107', 'ACH-001137', 'ACH-001249', 'ACH-002178', 'ACH-001108', 'ACH-002183', 'ACH-002260', 'ACH-001047', 'ACH-002121', 'ACH-002340', 'ACH-002077', 'ACH-002102', 'ACH-000992', 'ACH-002395', 'ACH-001138', 'ACH-001234', 'ACH-002176', 'ACH-002143', 'ACH-001224', 'ACH-002296', 'ACH-002282', 'ACH-001090', 'ACH-001331', 'ACH-001101', 'ACH-001230', 'ACH-002391', 'ACH-002299', 'ACH-002243', 'ACH-002262', 'ACH-002215', 'ACH-001355', 'ACH-002020', 'ACH-002269', 'ACH-002228', 'ACH-002096', 'ACH-001208', 'ACH-002365', 'ACH-002110', 'ACH-002378', 'ACH-002374', 'ACH-002246', 'ACH-001293', 'ACH-001079', 'ACH-001599', 'ACH-002280', 'ACH-002263', 'ACH-002381', 'ACH-002195', 'ACH-001641', 'ACH-002211', 'ACH-002249', 'ACH-002179', 'ACH-002271', 'ACH-002352', 'ACH-001206', 'ACH-001490', 'ACH-002229', 'ACH-002106', 'ACH-002295', 'ACH-002217', 'ACH-002377', 'ACH-000979', 'ACH-002052', 'ACH-001021', 'ACH-002375', 'ACH-002154', 'ACH-002223', 'ACH-002350', 'ACH-002373', 'ACH-002170', 'ACH-002113', 'ACH-002099', 'ACH-002259', 'ACH-001198', 'ACH-002353', 'ACH-002264', 'ACH-002251', 'ACH-002090', 'ACH-002160', 'ACH-001738', 'ACH-001203', 'ACH-002051', 'ACH-002226', 'ACH-001189', 'ACH-001081', 'ACH-002198', 'ACH-002237', 'ACH-002236', 'ACH-002270', 'ACH-002164', 'ACH-002275', 'ACH-001094', 'ACH-001767', 'ACH-002401', 'ACH-002397', 'ACH-001023', 'ACH-002233', 'ACH-002221', 'ACH-002369', 'ACH-001045', 'ACH-002366', 'ACH-002206', 'ACH-001957', 'ACH-001150', 'ACH-002231', 'ACH-002290', 'ACH-002383', 'ACH-002171', 'ACH-002388', 'ACH-002162', 'ACH-001042', 'ACH-002281', 'ACH-002286', 'ACH-002174', 'ACH-002349', 'ACH-002276', 'ACH-002058', 'ACH-002134', 'ACH-002301', 'ACH-002372', 'ACH-002131', 'ACH-001188', 'ACH-002127', 'ACH-002272', 'ACH-001044', 'ACH-001146', 'ACH-001182', 'ACH-001131', 'ACH-000795', 'ACH-002277', 'ACH-002361', 'ACH-002202', 'ACH-002118', 'ACH-002135', 'ACH-001606', 'ACH-002194', 'ACH-002120', 'ACH-002214', 'ACH-001126', 'ACH-001136', 'ACH-002124', 'ACH-002209', 'ACH-002224', 'ACH-002370', 'ACH-001162', 'ACH-001225', 'ACH-002338', 'ACH-002043', 'ACH-002399', 'ACH-001958', 'ACH-002390', 'ACH-001071', 'ACH-002343', 'ACH-001362', 'ACH-000047', 'ACH-001015', 'ACH-002384', 'ACH-001091', 'ACH-001017', 'ACH-001089', 'ACH-002336', 'ACH-002253', 'ACH-002103', 'ACH-002257', 'ACH-001118', 'ACH-002287', 'ACH-002177', 'ACH-002100', 'ACH-002151', 'ACH-002348', 'ACH-002095', 'ACH-001088', 'ACH-002132', 'ACH-002356', 'ACH-002200', 'ACH-002111', 'ACH-001383', 'ACH-002116', 'ACH-001358', 'ACH-002205', 'ACH-001127', 'ACH-002145', 'ACH-002364', 'ACH-002213', 'ACH-001093', 'ACH-002218', 'ACH-002196', 'ACH-002258', 'ACH-002146', 'ACH-001011', 'ACH-002152', 'ACH-002256', 'ACH-002308', 'ACH-002379', 'ACH-002339', 'ACH-002360', 'ACH-001452', 'ACH-002289', 'ACH-002240', 'ACH-002148', 'ACH-001039', 'ACH-002186', 'ACH-002136', 'ACH-002254', 'ACH-002250', 'ACH-002167', 'ACH-001359', 'ACH-002309', 'ACH-002153', 'ACH-001365', 'ACH-002119', 'ACH-002357', 'ACH-002317', 'ACH-002108', 'ACH-001214', 'ACH-002266', 'ACH-001175', 'ACH-002293', 'ACH-001037', 'ACH-002189', 'ACH-002166', 'ACH-002344', 'ACH-002306', 'ACH-002386', 'ACH-001205', 'ACH-002297', 'ACH-002396', 'ACH-002181', 'ACH-002112', 'ACH-001024', 'ACH-002126', 'ACH-001018', 'ACH-002163', 'ACH-002230', 'ACH-002393', 'ACH-002311', 'ACH-002185', 'ACH-002232', 'ACH-001591', 'ACH-002092', 'ACH-002225', 'ACH-002305', 'ACH-002400', 'ACH-002222', 'ACH-002083', 'ACH-001072', 'ACH-002300', 'ACH-002380', 'ACH-002371', 'ACH-002139', 'ACH-002091', 'ACH-002363', 'ACH-002123', 'ACH-002394', 'ACH-002159', 'ACH-002341', 'ACH-001364', 'ACH-002104', 'ACH-002182', 'ACH-002040', 'ACH-001417', 'ACH-002150', 'ACH-002193', 'ACH-002188', 'ACH-002298', 'ACH-002294', 'ACH-002101', 'ACH-001199', 'ACH-002117', 'ACH-001704', 'ACH-002125', 'ACH-002342', 'ACH-002273', 'ACH-002184', 'ACH-002210', 'ACH-002169', 'ACH-002345', 'ACH-001121', 'ACH-002238', 'ACH-002358', 'ACH-002242', 'ACH-002129', 'ACH-002244', 'ACH-001187', 'ACH-002234', 'ACH-002392', 'ACH-002128', 'ACH-001092', 'ACH-001680', 'ACH-002207', 'ACH-002149', 'ACH-002115', 'ACH-002199', 'ACH-000727', 'ACH-002274', 'ACH-001000', 'ACH-000309', 'ACH-001956', 'ACH-002165', 'ACH-000426', 'ACH-002227', 'ACH-002314', 'ACH-002303', 'ACH-001049', 'ACH-002155', 'ACH-002168', 'ACH-002147', 'ACH-002212', 'ACH-002156', 'ACH-002351', 'ACH-002359', 'ACH-001066', 'ACH-002283', 'ACH-002114', 'ACH-002367', 'ACH-002140', 'ACH-001282', 'ACH-001087', 'ACH-002197', 'ACH-002208', 'ACH-002302', 'ACH-002354', 'ACH-002310', 'ACH-001363', 'ACH-002094', 'ACH-002288', 'ACH-002109', 'ACH-002265', 'ACH-002216', 'ACH-002267', 'ACH-002235', 'ACH-002201', 'ACH-002247', 'ACH-002239', 'ACH-002313', 'ACH-001065', 'ACH-002098', 'ACH-002268', 'ACH-002278', 'ACH-002335', 'ACH-002368', 'ACH-001338', 'ACH-002347', 'ACH-002255', 'ACH-002284', 'ACH-002130', 'ACH-002220', 'ACH-001043', 'ACH-002307', 'ACH-002362', 'ACH-002355', 'ACH-001544', 'ACH-002261', 'ACH-002157', 'ACH-002187'}\n"
     ]
    }
   ],
   "source": [
    "print('in cn but not mut')\n",
    "print(prevwes[\"internal\"] - prevmut[\"internal\"])\n",
    "print('in mut but not cn')\n",
    "print(prevwes[\"internal\"] - prevcn[\"internal\"])\n",
    "print('in rna but no wes/wgs')\n",
    "print(prev[\"internal\"] - prevwes[\"internal\"])\n",
    "print('in wes/wgs but not rna')\n",
    "print(prev[\"internal\"] - prevrna[\"internal\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:21:00.654528Z",
     "start_time": "2021-01-21T19:21:00.625877Z"
    }
   },
   "outputs": [],
   "source": [
    "prev['ibm'] = prev['dmc']\n",
    "prevcn['ibm'] = prevcn['dmc']\n",
    "prevmut['ibm'] = prevmut['dmc']\n",
    "prevrna['ibm'] = prevrna['dmc']\n",
    "prevwes['ibm'] = prevwes['dmc']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## managing the readmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fatal: destination path 'depmap-release-readmes' already exists and is not an empty directory.\r\n"
     ]
    }
   ],
   "source": [
    "#! cd .. && git clone https://github.com/broadinstitute/depmap-release-readmes.git && cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-05T18:24:04.999781Z",
     "start_time": "2021-01-05T18:24:01.456862Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating bb67687..fa199c0\n",
      "Fast-forward\n",
      " release-20q4/dmc-20q4.txt      | 41 \u001b[32m+\u001b[m\u001b[31m---------------------\u001b[m\n",
      " release-20q4/internal-20q4.txt | 77 \u001b[32m++++++++++++++++++++\u001b[m\u001b[31m----------------------\u001b[m\n",
      " release-20q4/public-20q4.txt   | 74 \u001b[32m++++++++++++++++++++\u001b[m\u001b[31m--------------------\u001b[m\n",
      " 3 files changed, 74 insertions(+), 118 deletions(-)\n"
     ]
    }
   ],
   "source": [
    "! cd ../depmap-release-readmes && git pull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-12T17:22:20.622719Z",
     "start_time": "2021-01-12T17:22:14.856061Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making public\n",
      "Making internal\n",
      "Making dmc\n",
      "[master 87f7b1c] 21Q1\n",
      " 4 files changed, 140 insertions(+), 454 deletions(-)\n",
      " delete mode 100644 release-21q1/ibm-21q1.txt\n",
      "Counting objects: 6, done.\n",
      "Delta compression using up to 12 threads.\n",
      "Compressing objects: 100% (6/6), done.\n",
      "Writing objects: 100% (6/6), 2.63 KiB | 0 bytes/s, done.\n",
      "Total 6 (delta 4), reused 0 (delta 0)\n",
      "remote: Resolving deltas: 100% (4/4), completed with 2 local objects.\u001b[K\n",
      "To https://github.com/broadinstitute/depmap-release-readmes.git\n",
      "   8cb35f2..87f7b1c  master -> master\n"
     ]
    }
   ],
   "source": [
    "!cd ../depmap-release-readmes/ && python3 make_new_release.py $release && git add . && git commit -m $release && git push "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOW UPDATE THE READMEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-12T17:25:38.148489Z",
     "start_time": "2021-01-12T17:25:37.349110Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘temp/README/’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "! mkdir temp/README/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T17:47:46.216279Z",
     "start_time": "2021-01-22T17:47:45.801806Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On branch master\r\n",
      "Your branch is ahead of 'origin/master' by 3 commits.\r\n",
      "  (use \"git push\" to publish your local commits)\r\n",
      "nothing to commit, working tree clean\r\n"
     ]
    }
   ],
   "source": [
    "! cd ../depmap-release-readmes && cp release-$release/* ../ccle_processing/readmes/ && git add . && git commit -m \"Omics: updating readmes to new release\" && git push"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T17:44:49.047820Z",
     "start_time": "2021-01-22T17:44:38.574077Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Auto-merging release-21q1/public-21q1.txt\n",
      "\u001b[?1000h\u001b7\u001b[?47h\u001b[?1h\u001b=\u001b[2;1H▽\u001b[6n\u001b[2;1H  \u001b[1;1H\u001b]11;?\u0007\u001b[1;24r\u001b[m\u001b[H\u001b[2J\u001b[24;1H\"~/depmap-release-readmes/.git/MERGE_MSG\" 7L, 309C\u001b[>c\u001b[1;1H\u001b[33mMerge branch 'master' of https://github.com/broadi\u001b[mnstitute/depmap-release-readmee\u001b[2;1Hs\n",
      "\n",
      "\u001b[34m# Please enter a commit message to explain why this merge is necessary,\n",
      "# especially if it merges an updated upstream into a topic branch.\n",
      "#\n",
      "# Lines starting with '#' will be ignored, and an empty message aborts\n",
      "# the commit.\u001b[m\n",
      "\u001b[1m\u001b[34m~                                                                               \u001b[10;1H~                                                                               \u001b[11;1H~                                                                               \u001b[12;1H~                                                                               \u001b[13;1H~                                                                               \u001b[14;1H~                                                                               \u001b[15;1H~                                                                               \u001b[16;1H~                                                                               \u001b[17;1H~                                                                               \u001b[18;1H~                                                                               \u001b[19;1H~                                                                               \u001b[20;1H~                                                                               \u001b[21;1H~                                                                               \u001b[22;1H~                                                                               \u001b[23;1H~                                                                               \u001b[m\u001b[24;63H1,1\u001b[11CAll\u001b[1;1H\u001b[24;1HType  :quit<Enter>  to exit Vim\u001b[24;32H\u001b[K\u0007\u001b[24;63H1,1\u001b[11CAll\u001b[1;1H"
     ]
    }
   ],
   "source": [
    "! cd ../depmap-release-readmes && git pull &&  cp release-$release/* ../ccle_processing/readmes/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for val in ['internal', 'ibm', 'dmc', 'public']:\n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README').read()\n",
    "    AddToVirtual(virtual[val], files = [('Raw','temp/README')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Somatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:22:25.057047Z",
     "start_time": "2021-01-21T19:21:34.453533Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (3,22,25,28,29,31,33,34,35,36) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "mutations = pd.read_csv(\"temp/wes_somatic_mutations_withlegacy_\"+samplesetname+\".csv\")\n",
    "damaging = pd.read_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv', index_col=0)\n",
    "othercons = pd.read_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv', index_col=0)\n",
    "othernoncons = pd.read_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv', index_col=0)\n",
    "hotspot = pd.read_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv', index_col=0)\n",
    "hotspot=hotspot.astype(int)\n",
    "damaging=damaging.astype(int)\n",
    "othercons=othercons.astype(int)\n",
    "othernoncons=othernoncons.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:23:05.233815Z",
     "start_time": "2021-01-21T19:23:03.165509Z"
    }
   },
   "outputs": [],
   "source": [
    "#reverting to previous versions\n",
    "mutations = mutations[mutations.is_likely_immortalization!=True]\n",
    "mutations = mutations[['Hugo_Symbol', 'Entrez_Gene_Id', 'NCBI_Build', 'Chromosome',\n",
    "       'Start_position', 'End_position', 'Strand', 'Variant_Classification',\n",
    "       'Variant_Type', 'Reference_Allele', 'Tumor_Allele', 'dbSNP_RS',\n",
    "       'dbSNP_Val_Status', 'Genome_Change', 'Annotation_Transcript',\n",
    "       'DepMap_ID', 'cDNA_Change', 'Codon_Change', 'Protein_Change', 'isDeleterious',\n",
    "       'isTCGAhotspot', 'TCGAhsCnt', 'isCOSMIChotspot', 'COSMIChsCnt',\n",
    "       'ExAC_AF',\"Variant_annotation\", 'CGA_WES_AC', 'HC_AC',\n",
    "       'RD_AC', 'RNAseq_AC', 'SangerWES_AC', 'WGS_AC']].rename(columns={\"Tumor_Allele\":\"Tumor_Seq_Allele1\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-19T21:43:37.191425Z",
     "start_time": "2021-01-19T21:43:36.824489Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "temp/README\r\n"
     ]
    }
   ],
   "source": [
    "! ls temp/README"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T17:00:05.364682Z",
     "start_time": "2021-01-22T17:00:05.332207Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"ACH-001502\" in prevmut[\"ibm\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:45:22.019825Z",
     "start_time": "2021-01-21T19:27:36.111119Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________\n",
      "internal\n",
      "nott present\n",
      "{'ACH-001175', 'ACH-001037', 'ACH-001198', 'ACH-002378', 'ACH-002341', 'ACH-001225', 'ACH-001171', 'ACH-002315', 'ACH-002335', 'ACH-001079', 'ACH-001045', 'ACH-001088', 'ACH-001429', 'ACH-001087', 'ACH-001712', 'ACH-001071', 'ACH-001316', 'ACH-001393', 'ACH-001224', 'ACH-001015', 'ACH-001018', 'ACH-001000', 'ACH-001101', 'ACH-002303', 'ACH-001741', 'ACH-001017', 'ACH-001743', 'ACH-001189', 'ACH-001249', 'ACH-001121', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002378', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "{'ACH-001691', 'ACH-002828', 'ACH-001983', 'ACH-001978', 'ACH-001986', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001683', 'ACH-001977', 'ACH-001981', 'ACH-001686', 'ACH-001512', 'ACH-002080', 'ACH-001759', 'ACH-002504', 'ACH-001975', 'ACH-002825', 'ACH-001980', 'ACH-001701', 'ACH-002851', 'ACH-001705', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-001974', 'ACH-002013', 'ACH-001984', 'ACH-001985', 'ACH-001503', 'ACH-002699', 'ACH-001447', 'ACH-001553', 'ACH-003210', 'ACH-001828', 'ACH-001046', 'ACH-002074', 'ACH-001700', 'ACH-001227', 'ACH-002822', 'ACH-001758', 'ACH-001953', 'ACH-001994', 'ACH-001982', 'ACH-002476', 'ACH-001817', 'ACH-001697', 'ACH-002047', 'ACH-002034', 'ACH-001760', 'ACH-002138', 'ACH-001571', 'ACH-001949', 'ACH-002060', 'ACH-001756', 'ACH-002061', 'ACH-002810'}\n",
      "42406\n",
      "56\n",
      "0\n",
      "57\n",
      "54\n",
      "Uploading all_somatic_mutations_withlegacy...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_withlegacy properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_damaging...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_damaging properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othernoncons...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othernoncons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othercons...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othercons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_hotspot...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_hotspot properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/eba60ae917c244a58c6e29a8f7b08028\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dataset version with id 7b5de4830b9143729456bbc060a2bbd1 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/7b5de4830b9143729456bbc060a2bbd1\n",
      "[('CCLE_mutations', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_withlegacy'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README', 'depmap-mutation-calls-9be3.36/README')]\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datafile/c90aa3040856420daac306d007ff16f1\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id f907a3e9eb434cd9b22dbc2e9efd9526 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/f907a3e9eb434cd9b22dbc2e9efd9526\n",
      "_________________________________________________\n",
      "ibm\n",
      "nott present\n",
      "{'ACH-001175', 'ACH-001037', 'ACH-001198', 'ACH-002378', 'ACH-002341', 'ACH-001225', 'ACH-001171', 'ACH-002315', 'ACH-002335', 'ACH-001079', 'ACH-001045', 'ACH-001088', 'ACH-001429', 'ACH-001087', 'ACH-001712', 'ACH-001071', 'ACH-001316', 'ACH-001224', 'ACH-001015', 'ACH-001018', 'ACH-001000', 'ACH-001101', 'ACH-002303', 'ACH-001741', 'ACH-001017', 'ACH-001743', 'ACH-001189', 'ACH-001249', 'ACH-001121', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002378', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "{'ACH-001983', 'ACH-002828', 'ACH-001978', 'ACH-001981', 'ACH-001512', 'ACH-002080', 'ACH-001759', 'ACH-002504', 'ACH-001974', 'ACH-001984', 'ACH-001985', 'ACH-001447', 'ACH-001553', 'ACH-003210', 'ACH-001046', 'ACH-002074', 'ACH-001227', 'ACH-002822', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-001697', 'ACH-002138', 'ACH-002060', 'ACH-001756', 'ACH-001691', 'ACH-001986', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001683', 'ACH-001977', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-001980', 'ACH-001701', 'ACH-002851', 'ACH-001705', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-002699', 'ACH-001828', 'ACH-001700', 'ACH-001994', 'ACH-001953', 'ACH-001982', 'ACH-002034', 'ACH-002047', 'ACH-001760', 'ACH-001571', 'ACH-001949', 'ACH-002061', 'ACH-002810'}\n",
      "42406\n",
      "56\n",
      "0\n",
      "57\n",
      "54\n",
      "Uploading all_somatic_mutations_withlegacy...\n",
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_withlegacy properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_damaging...\n",
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_damaging properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othernoncons...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othernoncons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othercons...\n",
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othercons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_hotspot...\n",
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_hotspot properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/6cd93f556ba24e46a73ad8e2754f619a\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 8d24f52b49964264a78ec5fb5d94dfd0 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/8d24f52b49964264a78ec5fb5d94dfd0\n",
      "[('CCLE_mutations', 'mutations-b05c.10/all_somatic_mutations_withlegacy'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'mutations-b05c.10/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'mutations-b05c.10/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'mutations-b05c.10/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'mutations-b05c.10/all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README', 'mutations-b05c.10/README')]\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datafile/4abbbb6d3e1f40e3a1779cf7f8816a73\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id dbc65a056d66406caa0cd53c6224f5e3 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/dbc65a056d66406caa0cd53c6224f5e3\n",
      "_________________________________________________\n",
      "dmc\n",
      "nott present\n",
      "{'ACH-001175', 'ACH-001037', 'ACH-001198', 'ACH-002378', 'ACH-002341', 'ACH-001225', 'ACH-001171', 'ACH-002315', 'ACH-002335', 'ACH-001079', 'ACH-001045', 'ACH-001088', 'ACH-001429', 'ACH-001087', 'ACH-001712', 'ACH-001071', 'ACH-001316', 'ACH-001224', 'ACH-001015', 'ACH-001018', 'ACH-001000', 'ACH-001101', 'ACH-002303', 'ACH-001741', 'ACH-001017', 'ACH-001743', 'ACH-001189', 'ACH-001249', 'ACH-001121', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002378', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "{'ACH-001983', 'ACH-002828', 'ACH-002055', 'ACH-001978', 'ACH-001981', 'ACH-001512', 'ACH-002080', 'ACH-001759', 'ACH-002504', 'ACH-001974', 'ACH-001984', 'ACH-001985', 'ACH-001447', 'ACH-001553', 'ACH-003210', 'ACH-001046', 'ACH-002074', 'ACH-001227', 'ACH-002822', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-001697', 'ACH-002138', 'ACH-002060', 'ACH-001756', 'ACH-001691', 'ACH-001986', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001683', 'ACH-001977', 'ACH-001675', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-001980', 'ACH-001701', 'ACH-002851', 'ACH-001705', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-002699', 'ACH-001828', 'ACH-001700', 'ACH-001994', 'ACH-001953', 'ACH-001982', 'ACH-002034', 'ACH-002047', 'ACH-001760', 'ACH-001571', 'ACH-001949', 'ACH-002061', 'ACH-002810'}\n",
      "43020\n",
      "58\n",
      "0\n",
      "59\n",
      "56\n",
      "Uploading all_somatic_mutations_withlegacy...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_withlegacy properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_damaging...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_damaging properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othernoncons...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othernoncons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othercons...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othercons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_hotspot...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_hotspot properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/07c1897cdf19481f8e469606947faebb\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 9e4645071747411a9cf09a054024ec50 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/9e4645071747411a9cf09a054024ec50\n",
      "[('CCLE_mutations', 'depmap-mutation-calls-dfce.27/all_somatic_mutations_withlegacy'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-dfce.27/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-dfce.27/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-dfce.27/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-dfce.27/all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README', 'depmap-mutation-calls-dfce.27/README')]\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datafile/124b8c8dd3a944e0a6efa5ec2e5770dc\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id ebd5af0caf2047a28d534fa7b3da5a6e created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/ebd5af0caf2047a28d534fa7b3da5a6e\n",
      "_________________________________________________\n",
      "public\n",
      "nott present\n",
      "{'ACH-001175', 'ACH-001037', 'ACH-001198', 'ACH-002378', 'ACH-002341', 'ACH-001225', 'ACH-001171', 'ACH-002315', 'ACH-002335', 'ACH-001079', 'ACH-001045', 'ACH-001088', 'ACH-001087', 'ACH-001712', 'ACH-001071', 'ACH-001316', 'ACH-001224', 'ACH-001015', 'ACH-001018', 'ACH-001000', 'ACH-001101', 'ACH-002303', 'ACH-001741', 'ACH-001017', 'ACH-001743', 'ACH-001189', 'ACH-001249', 'ACH-001121', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002378', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "{'ACH-001983', 'ACH-002828', 'ACH-002055', 'ACH-001978', 'ACH-002014', 'ACH-001981', 'ACH-001693', 'ACH-001449', 'ACH-002010', 'ACH-001855', 'ACH-001512', 'ACH-002504', 'ACH-001493', 'ACH-001537', 'ACH-001547', 'ACH-001696', 'ACH-001984', 'ACH-001678', 'ACH-001553', 'ACH-001992', 'ACH-001662', 'ACH-001046', 'ACH-001227', 'ACH-002822', 'ACH-001697', 'ACH-002060', 'ACH-001756', 'ACH-001672', 'ACH-001971', 'ACH-001986', 'ACH-002021', 'ACH-001683', 'ACH-001349', 'ACH-001973', 'ACH-001980', 'ACH-001701', 'ACH-001669', 'ACH-001705', 'ACH-002699', 'ACH-001700', 'ACH-001502', 'ACH-001953', 'ACH-001982', 'ACH-002047', 'ACH-001437', 'ACH-001571', 'ACH-001679', 'ACH-002810', 'ACH-002401', 'ACH-001854', 'ACH-002512', 'ACH-002080', 'ACH-001759', 'ACH-001438', 'ACH-002084', 'ACH-001974', 'ACH-001985', 'ACH-001447', 'ACH-001676', 'ACH-003210', 'ACH-002074', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-002048', 'ACH-002138', 'ACH-001691', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001293', 'ACH-001977', 'ACH-001675', 'ACH-002465', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-002851', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-001828', 'ACH-001994', 'ACH-001970', 'ACH-001707', 'ACH-002400', 'ACH-001533', 'ACH-001847', 'ACH-002034', 'ACH-001760', 'ACH-001949', 'ACH-002061'}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59311\n",
      "93\n",
      "0\n",
      "94\n",
      "91\n",
      "Uploading all_somatic_mutations_withlegacy...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_withlegacy properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_damaging...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_damaging properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othernoncons...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othernoncons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_othercons...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_othercons properly converted and uploaded\n",
      "Uploading all_somatic_mutations_boolmatrix_fordepmap_hotspot...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_somatic_mutations_boolmatrix_fordepmap_hotspot properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/581e3dd77d674dd2aeef999425a53977\n",
      "Conversion and upload...:\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id c7f10c35bbda409fb58e8a016b3f335f created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/c7f10c35bbda409fb58e8a016b3f335f\n",
      "[('CCLE_mutations', 'depmap-mutation-calls-9a1a.35/all_somatic_mutations_withlegacy'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-9a1a.35/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-9a1a.35/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-9a1a.35/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-9a1a.35/all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README', 'depmap-mutation-calls-9a1a.35/README')]\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datafile/9a4674254cc64d8fb40bcc6de67caa9c\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 65995cb0278947ff857a576894e34f0e created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/65995cb0278947ff857a576894e34f0e\n"
     ]
    }
   ],
   "source": [
    "blacklist = set()\n",
    "for val in ['internal', 'ibm', \"dmc\", \"public\"]:\n",
    "    print('_________________________________________________')\n",
    "    print(val)\n",
    "    missing = set(new[val]) - set(mutations.DepMap_ID)\n",
    "    print('nott present')\n",
    "    removed = set(prev[val]) - set(mutations.DepMap_ID)\n",
    "    print(removed)\n",
    "    print('removed')\n",
    "    removed = set(prevmut[val]) - set(mutations.DepMap_ID)\n",
    "    print(removed)\n",
    "    blacklist = (set(mutations.DepMap_ID) - (prevmut[val] | set(new[val]))) | blacklist\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(blacklist)\n",
    "    # adding files\n",
    "    a = mutations[~mutations.DepMap_ID.isin(blacklist)]\n",
    "    print(len(mutations)- len(a))\n",
    "    a.to_csv('temp/all_somatic_mutations_withlegacy.csv', index=False)\n",
    "    a = damaging[~damaging.index.isin(blacklist)]\n",
    "    print(len(damaging) - len(a))\n",
    "    a.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv')\n",
    "    a = othercons[~othercons.index.isin(blacklist)]\n",
    "    print(len(othercons) - len(a))\n",
    "    a.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv')\n",
    "    a = othernoncons[~othernoncons.index.isin(blacklist)]\n",
    "    print(len(othernoncons) - len(a))\n",
    "    a.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv')\n",
    "    a = hotspot[~hotspot.index.isin(blacklist)]\n",
    "    print(len(hotspot) - len(a))\n",
    "    a.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv')\n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README').read()\n",
    "    \n",
    "    # updating on taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_mutation[val],\n",
    "                     upload_file_path_dict={\n",
    "    'temp/all_somatic_mutations_withlegacy.csv': 'TableCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv': 'NumericMatrixCSV',\n",
    "    'temp/README': 'Raw'},\n",
    "                     dataset_description=\"\"\"\n",
    "    #Mutations Omics:\n",
    "\n",
    "    for informations, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    BLACKLISTED:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # To add to a virtual dataset\n",
    "    AddToVirtual(virtual[val], taiga_mutation[val], [('CCLE_mutations', 'all_somatic_mutations_withlegacy'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'all_somatic_mutations_boolmatrix_fordepmap_damaging'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'all_somatic_mutations_boolmatrix_fordepmap_othernoncons'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'all_somatic_mutations_boolmatrix_fordepmap_othercons'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T21:32:24.735668Z",
     "start_time": "2021-01-21T21:32:12.616453Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_mutations', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_withlegacy'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-9be3.36/all_somatic_mutations_boolmatrix_fordepmap_hotspot')]\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datafile/59f4074ce9c24fe793f7b5ac1bde2f64\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 2fb8c75a532d449ab14c88140cfdfe27 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/2fb8c75a532d449ab14c88140cfdfe27\n"
     ]
    }
   ],
   "source": [
    "# To add to a eternal dataset\n",
    "AddToVirtual('depmap-a0ab', taiga_mutation['internal'], [('CCLE_mutations', 'all_somatic_mutations_withlegacy'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'all_somatic_mutations_boolmatrix_fordepmap_damaging'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'all_somatic_mutations_boolmatrix_fordepmap_othernoncons'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'all_somatic_mutations_boolmatrix_fordepmap_othercons'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'all_somatic_mutations_boolmatrix_fordepmap_hotspot')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Copy Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-19T18:07:11.415298Z",
     "start_time": "2021-01-19T18:05:28.336773Z"
    }
   },
   "outputs": [],
   "source": [
    "genecn= pd.read_csv('temp/all_'+samplesetname+'_gene_cn.csv',index_col=0)\n",
    "segmentcn = pd.read_csv('temp/all_'+samplesetname+'_segment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-19T16:04:29.112579Z",
     "start_time": "2021-01-19T16:04:29.005774Z"
    }
   },
   "outputs": [],
   "source": [
    "#genecn = genecn.apply(lambda x: (x**2)-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T20:09:39.499808Z",
     "start_time": "2021-01-20T19:53:52.876640Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________\n",
      "internal\n",
      "not present\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "81 {'ACH-001983', 'ACH-002828', 'ACH-001978', 'ACH-002014', 'ACH-001854', 'ACH-002512', 'ACH-001981', 'ACH-001693', 'ACH-001449', 'ACH-001210', 'ACH-002010', 'ACH-001855', 'ACH-001512', 'ACH-002080', 'ACH-001759', 'ACH-002504', 'ACH-001493', 'ACH-001438', 'ACH-001537', 'ACH-001696', 'ACH-001974', 'ACH-001984', 'ACH-001985', 'ACH-001447', 'ACH-001676', 'ACH-003210', 'ACH-001662', 'ACH-001046', 'ACH-002074', 'ACH-001227', 'ACH-002822', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-001697', 'ACH-002138', 'ACH-000010', 'ACH-002060', 'ACH-001756', 'ACH-001672', 'ACH-001691', 'ACH-001971', 'ACH-001986', 'ACH-001708', 'ACH-002818', 'ACH-002392', 'ACH-001683', 'ACH-001349', 'ACH-001976', 'ACH-001293', 'ACH-001977', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-000712', 'ACH-001980', 'ACH-001701', 'ACH-001669', 'ACH-001705', 'ACH-002851', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001078', 'ACH-002699', 'ACH-001503', 'ACH-001828', 'ACH-001700', 'ACH-001502', 'ACH-001994', 'ACH-001953', 'ACH-001982', 'ACH-002034', 'ACH-002047', 'ACH-001437', 'ACH-001760', 'ACH-001571', 'ACH-001949', 'ACH-002061', 'ACH-002810'}\n",
      "777644\n",
      "740978\n",
      "1846\n",
      "1765\n",
      "Uploading all_merged_genes_cn...\n",
      "hitting https://cds.team/taiga/api/datafile/43e276a203934304a53a45f839abe979\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_genes_cn properly converted and uploaded\n",
      "Uploading all_merged_segments...\n",
      "hitting https://cds.team/taiga/api/datafile/43e276a203934304a53a45f839abe979\n",
      "Conversion and upload...:\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Waiting in the task queue\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_segments properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/43e276a203934304a53a45f839abe979\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 5120c0c6aa5b4b0383f7ebe4e4bc4e0f created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/5120c0c6aa5b4b0383f7ebe4e4bc4e0f\n",
      "[('CCLE_gene_cn', 'depmap-wes-cn-data-81a7.35/all_merged_genes_cn'), ('CCLE_segment_cn', 'depmap-wes-cn-data-81a7.35/all_merged_segments')]\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datafile/7542d56da292406fb599c716fa2ba4a4\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id ad8e4146fb7f45e0928ddf696c7f766d created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/ad8e4146fb7f45e0928ddf696c7f766d\n",
      "_________________________________________________\n",
      "ibm\n",
      "not present\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "81 {'ACH-001983', 'ACH-002828', 'ACH-001978', 'ACH-002014', 'ACH-001981', 'ACH-001693', 'ACH-001449', 'ACH-002010', 'ACH-001855', 'ACH-001512', 'ACH-002504', 'ACH-001493', 'ACH-001537', 'ACH-001696', 'ACH-001984', 'ACH-001662', 'ACH-001046', 'ACH-001227', 'ACH-002822', 'ACH-001697', 'ACH-000010', 'ACH-002060', 'ACH-001756', 'ACH-001672', 'ACH-001971', 'ACH-001986', 'ACH-002392', 'ACH-001683', 'ACH-001349', 'ACH-001980', 'ACH-001701', 'ACH-001669', 'ACH-001705', 'ACH-001078', 'ACH-002699', 'ACH-001700', 'ACH-001502', 'ACH-001953', 'ACH-001982', 'ACH-002047', 'ACH-001437', 'ACH-001571', 'ACH-002810', 'ACH-001854', 'ACH-002512', 'ACH-001210', 'ACH-002080', 'ACH-001759', 'ACH-001438', 'ACH-001974', 'ACH-001985', 'ACH-001447', 'ACH-001676', 'ACH-003210', 'ACH-002074', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-002138', 'ACH-001691', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001293', 'ACH-001977', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-000712', 'ACH-002851', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-001828', 'ACH-001994', 'ACH-002034', 'ACH-001760', 'ACH-001949', 'ACH-002061'}\n",
      "777644\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "740978\n",
      "1846\n",
      "1765\n",
      "Uploading all_merged_genes_cn...\n",
      "hitting https://cds.team/taiga/api/datafile/d086b8cb8581490a96955d1a73808e7a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1750\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_genes_cn properly converted and uploaded\n",
      "Uploading all_merged_segments...\n",
      "hitting https://cds.team/taiga/api/datafile/d086b8cb8581490a96955d1a73808e7a\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_segments properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/d086b8cb8581490a96955d1a73808e7a\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 51c3ba4a1c0245d4a614a939139ab3c1 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/51c3ba4a1c0245d4a614a939139ab3c1\n",
      "[('CCLE_gene_cn', 'cn-e20f.5/all_merged_genes_cn'), ('CCLE_segment_cn', 'cn-e20f.5/all_merged_segments')]\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datafile/07d9d1eee5974bad85232b00fffccb40\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 4feb964089e64853a31634fab8d40c31 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/4feb964089e64853a31634fab8d40c31\n",
      "_________________________________________________\n",
      "dmc\n",
      "not present\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "83 {'ACH-001983', 'ACH-002828', 'ACH-002055', 'ACH-001978', 'ACH-002014', 'ACH-001981', 'ACH-001693', 'ACH-001449', 'ACH-002010', 'ACH-001855', 'ACH-001512', 'ACH-002504', 'ACH-001493', 'ACH-001537', 'ACH-001696', 'ACH-001984', 'ACH-001662', 'ACH-001046', 'ACH-001227', 'ACH-002822', 'ACH-001697', 'ACH-000010', 'ACH-002060', 'ACH-001756', 'ACH-001672', 'ACH-001971', 'ACH-001986', 'ACH-002392', 'ACH-001683', 'ACH-001349', 'ACH-001980', 'ACH-001701', 'ACH-001669', 'ACH-001705', 'ACH-001078', 'ACH-002699', 'ACH-001700', 'ACH-001502', 'ACH-001953', 'ACH-001982', 'ACH-002047', 'ACH-001437', 'ACH-001571', 'ACH-002810', 'ACH-001854', 'ACH-002512', 'ACH-001210', 'ACH-002080', 'ACH-001759', 'ACH-001438', 'ACH-001974', 'ACH-001985', 'ACH-001447', 'ACH-001676', 'ACH-003210', 'ACH-002074', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-002138', 'ACH-001691', 'ACH-001708', 'ACH-002818', 'ACH-001976', 'ACH-001293', 'ACH-001977', 'ACH-001675', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-000712', 'ACH-002851', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-001828', 'ACH-001994', 'ACH-002034', 'ACH-001760', 'ACH-001949', 'ACH-002061'}\n",
      "777644\n",
      "739842\n",
      "1846\n",
      "1763\n",
      "Uploading all_merged_genes_cn...\n",
      "hitting https://cds.team/taiga/api/datafile/f632f567cfa949d7b7b90b7c530aa18d\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_genes_cn properly converted and uploaded\n",
      "Uploading all_merged_segments...\n",
      "hitting https://cds.team/taiga/api/datafile/f632f567cfa949d7b7b90b7c530aa18d\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_segments properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/f632f567cfa949d7b7b90b7c530aa18d\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 6b66fc53bd114bdc912af40f04040d36 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/6b66fc53bd114bdc912af40f04040d36\n",
      "[('CCLE_gene_cn', 'depmap-cn-data-9b9d.22/all_merged_genes_cn'), ('CCLE_segment_cn', 'depmap-cn-data-9b9d.22/all_merged_segments')]\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datafile/91629515b47746129ead41814fa3c1fd\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 95c1da2d176c40aa98b101f13afd4116 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/95c1da2d176c40aa98b101f13afd4116\n",
      "_________________________________________________\n",
      "public\n",
      "not present\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "removed\n",
      "{'ACH-002315', 'ACH-001189', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "106 {'ACH-001983', 'ACH-002828', 'ACH-002055', 'ACH-001978', 'ACH-002014', 'ACH-001981', 'ACH-002399', 'ACH-001693', 'ACH-001449', 'ACH-002010', 'ACH-001855', 'ACH-001574', 'ACH-001512', 'ACH-002504', 'ACH-001493', 'ACH-001537', 'ACH-001547', 'ACH-001696', 'ACH-001984', 'ACH-001678', 'ACH-001992', 'ACH-001662', 'ACH-001046', 'ACH-002463', 'ACH-001227', 'ACH-002822', 'ACH-001697', 'ACH-000010', 'ACH-002060', 'ACH-001756', 'ACH-001672', 'ACH-001971', 'ACH-002065', 'ACH-001986', 'ACH-002021', 'ACH-002467', 'ACH-002392', 'ACH-001683', 'ACH-001349', 'ACH-001973', 'ACH-002466', 'ACH-001980', 'ACH-001701', 'ACH-001669', 'ACH-001705', 'ACH-001078', 'ACH-002699', 'ACH-001700', 'ACH-001502', 'ACH-001953', 'ACH-001982', 'ACH-002047', 'ACH-001437', 'ACH-002462', 'ACH-001571', 'ACH-001679', 'ACH-002810', 'ACH-002401', 'ACH-001854', 'ACH-002512', 'ACH-001210', 'ACH-002080', 'ACH-001759', 'ACH-001438', 'ACH-002084', 'ACH-001974', 'ACH-001985', 'ACH-001447', 'ACH-001676', 'ACH-003210', 'ACH-002074', 'ACH-001758', 'ACH-002476', 'ACH-001817', 'ACH-002048', 'ACH-002138', 'ACH-001691', 'ACH-001708', 'ACH-002818', 'ACH-002464', 'ACH-001976', 'ACH-001293', 'ACH-001977', 'ACH-001675', 'ACH-002465', 'ACH-001686', 'ACH-001975', 'ACH-002825', 'ACH-000712', 'ACH-002851', 'ACH-001813', 'ACH-001492', 'ACH-001434', 'ACH-002013', 'ACH-001503', 'ACH-001828', 'ACH-001994', 'ACH-001970', 'ACH-001707', 'ACH-002400', 'ACH-001533', 'ACH-001847', 'ACH-002034', 'ACH-001760', 'ACH-001949', 'ACH-002061'}\n",
      "777644\n",
      "734230\n",
      "1846\n",
      "1740\n",
      "Uploading all_merged_genes_cn...\n",
      "hitting https://cds.team/taiga/api/datafile/6f25509e229543b4babfbaf2576f56cc\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_genes_cn properly converted and uploaded\n",
      "Uploading all_merged_segments...\n",
      "hitting https://cds.team/taiga/api/datafile/6f25509e229543b4babfbaf2576f56cc\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: all_merged_segments properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/6f25509e229543b4babfbaf2576f56cc\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id f2b6fa65da684596a5f6a2289c315a32 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/f2b6fa65da684596a5f6a2289c315a32\n",
      "[('CCLE_gene_cn', 'depmap-wes-cn-data-97cc.42/all_merged_genes_cn'), ('CCLE_segment_cn', 'depmap-wes-cn-data-97cc.42/all_merged_segments')]\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datafile/bb188db64c2042e8ad692cfbba7c95af\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id af07aeb9b6db4192ba3ab60878e27d12 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/af07aeb9b6db4192ba3ab60878e27d12\n"
     ]
    }
   ],
   "source": [
    "blacklist = set()\n",
    "for val in ['internal', 'ibm', 'dmc', 'public']:\n",
    "    print('_________________________________________________')\n",
    "    print(val)\n",
    "    print('not present')\n",
    "    removed = set(prevcn[val]) - set(segmentcn.DepMap_ID)\n",
    "    print(removed)\n",
    "    print('removed')\n",
    "    removed = set(prevcn[val]) - set(segmentcn.DepMap_ID)\n",
    "    print(removed)\n",
    "    missing = set(new[val]) - set(segmentcn.DepMap_ID)\n",
    "    blacklist = (set(segmentcn.DepMap_ID) - (prevcn[val] | set(new[val]))) | blacklist\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "    ## for segment removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(segmentcn))\n",
    "    a = segmentcn[~segmentcn.DepMap_ID.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/all_merged_segments.csv', index=False)\n",
    "    print(len(genecn))\n",
    "    a = genecn[~genecn.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/all_merged_genes_cn.csv')\n",
    "    \n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # Add to Taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_copynumber[val], \n",
    "                      upload_file_path_dict={\n",
    "                        'temp/all_merged_genes_cn.csv': 'NumericMatrixCSV',\n",
    "                        'temp/all_merged_segments.csv': 'TableCSV',\n",
    "                        'temp/README': 'Raw'},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    #Copy Number Omics:\n",
    "\n",
    "    for informations, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "    # To add to a virtual dataset\n",
    "    AddToVirtual(virtual[val], taiga_copynumber[val], [('CCLE_gene_cn', 'all_merged_genes_cn'), ('CCLE_segment_cn','all_merged_segments')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T20:10:06.385906Z",
     "start_time": "2021-01-20T20:09:54.334227Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_gene_cn', 'depmap-wes-cn-data-81a7.35/all_merged_genes_cn'), ('CCLE_segment_cn', 'depmap-wes-cn-data-81a7.35/all_merged_segments')]\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datafile/44b3c702a1ad4fbb9b462936c0687f1e\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 24debf6ac7364144877e830c89d8e3a1 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/24debf6ac7364144877e830c89d8e3a1\n"
     ]
    }
   ],
   "source": [
    "# To add to a eternal dataset\n",
    "AddToVirtual('depmap-a0ab', taiga_copynumber['internal'], [('CCLE_gene_cn', 'all_merged_genes_cn'), ('CCLE_segment_cn', 'all_merged_segments')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T19:53:41.866886Z",
     "start_time": "2021-01-20T19:53:41.393276Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34madditionals\u001b[0m/                Mutations_CCLE.ipynb\r\n",
      "\u001b[01;34mccle_tasks\u001b[0m/                 Mutations_CCLF.ipynb\r\n",
      "CDS-NbGq6q.candidateSV.vcf  Mutations_prev_new_comparison.ipynb\r\n",
      "\u001b[01;34mCNV_pipeline\u001b[0m/               preprocessing_Achilles.ipynb\r\n",
      "Copy_Number_CCLE.ipynb      README.md\r\n",
      "Copy_Number_CCLF.ipynb      \u001b[01;34mreadmes\u001b[0m/\r\n",
      "\u001b[01;34mdata\u001b[0m/                       requirements.txt\r\n",
      "DepMap Upload-Copy1.ipynb   RNA_CCLE.ipynb\r\n",
      "DepMap Upload-Copy2.ipynb   RNA_CCLF.ipynb\r\n",
      "DepMap Upload.ipynb         \u001b[01;34mRNA_pipeline\u001b[0m/\r\n",
      "\u001b[01;34mdocumentation\u001b[0m/              \u001b[01;34msrc\u001b[0m/\r\n",
      "firecloud_documentation.md  \u001b[01;34mtemp\u001b[0m/\r\n",
      "\u001b[01;34mlegacy\u001b[0m/                     Untitled.ipynb\r\n",
      "\u001b[01;34mmutation_pipeline\u001b[0m/          WGS_CCLE.ipynb\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-13T19:27:05.643601Z",
     "start_time": "2021-01-13T18:28:03.336480Z"
    }
   },
   "outputs": [],
   "source": [
    "transcripts_tpm = pd.read_csv('temp/expression_' + samplesetname + '_transcripts_tpm.csv',index_col=0)\n",
    "genes_tpm = pd.read_csv('temp/expression_' + samplesetname + '_genes_tpm.csv',index_col=0)\n",
    "genes_expected_count = pd.read_csv('temp/expression_' + samplesetname + '_genes_expected_count.csv',index_col=0)\n",
    "proteincoding_genes_expected_count = pd.read_csv('temp/expression_' + samplesetname + '_proteincoding_genes_expected_count.csv',index_col=0)\n",
    "proteincoding_genes_tpm = pd.read_csv('temp/expression_' + samplesetname + '_proteincoding_genes_tpm.csv',index_col=0)\n",
    "transcripts_expected_count = pd.read_csv('temp/expression_' + samplesetname + '_transcripts_expected_count.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-13T19:31:40.311849Z",
     "start_time": "2021-01-13T19:29:29.593413Z"
    }
   },
   "outputs": [],
   "source": [
    "#putting it back to what it was before\n",
    "\n",
    "# log transforming tpm data\n",
    "transcripts_tpm=transcripts_tpm.apply(lambda x: np.log2(x+1))\n",
    "genes_tpm=genes_tpm.apply(lambda x: np.log2(x+1))\n",
    "proteincoding_genes_tpm=proteincoding_genes_tpm.apply(lambda x: np.log2(x+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-13T19:33:53.141962Z",
     "start_time": "2021-01-13T19:33:52.992591Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no stored variable or alias rnafailed\n"
     ]
    }
   ],
   "source": [
    "%store -r failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-15T16:00:15.829685Z",
     "start_time": "2021-01-15T16:00:11.888788Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________\n",
      "internal\n",
      "not present\n",
      "{'ACH-002162', 'ACH-001662', 'ACH-001101', 'ACH-002116', 'ACH-002152', 'ACH-002247', 'ACH-002221', 'ACH-002105', 'ACH-001091', 'ACH-002220', 'ACH-001044', 'ACH-001047', 'ACH-002100', 'ACH-001384', 'ACH-001224', 'ACH-002316', 'ACH-002021', 'ACH-002185', 'ACH-002356', 'ACH-002374', 'ACH-002348', 'ACH-002106', 'ACH-002289', 'ACH-002096', 'ACH-002224', 'ACH-002341', 'ACH-002394', 'ACH-001137', 'ACH-002344', 'ACH-001130', 'ACH-002092', 'ACH-002400', 'ACH-002161', 'ACH-001678', 'ACH-001233', 'ACH-002155', 'ACH-001364', 'ACH-002258', 'ACH-002243', 'ACH-002458', 'ACH-002306', 'ACH-001605', 'ACH-002300', 'ACH-001704', 'ACH-001092', 'ACH-001093', 'ACH-002168', 'ACH-002357', 'ACH-002465', 'ACH-001528', 'ACH-002244', 'ACH-001182', 'ACH-002304', 'ACH-002313', 'ACH-002122', 'ACH-001107', 'ACH-002358', 'ACH-002359', 'ACH-002192', 'ACH-001547', 'ACH-001649', 'ACH-002298', 'ACH-002040', 'ACH-002178', 'ACH-000047', 'ACH-002099', 'ACH-002102', 'ACH-001090', 'ACH-002144', 'ACH-001708', 'ACH-002146', 'ACH-001429', 'ACH-002274', 'ACH-002233', 'ACH-001855', 'ACH-002275', 'ACH-002094', 'ACH-001002', 'ACH-002135', 'ACH-002252', 'ACH-002210', 'ACH-001121', 'ACH-001991', 'ACH-002153', 'ACH-001042', 'ACH-001066', 'ACH-001043', 'ACH-002027', 'ACH-001171', 'ACH-001205', 'ACH-002287', 'ACH-001362', 'ACH-001089', 'ACH-002267', 'ACH-002234', 'ACH-002225', 'ACH-002389', 'ACH-002209', 'ACH-002133', 'ACH-002354', 'ACH-001150', 'ACH-001626', 'ACH-000309', 'ACH-002222', 'ACH-002379', 'ACH-001018', 'ACH-002238', 'ACH-002058', 'ACH-002151', 'ACH-001131', 'ACH-001709', 'ACH-002261', 'ACH-001230', 'ACH-001679', 'ACH-002141', 'ACH-002241', 'ACH-002214', 'ACH-001630', 'ACH-002242', 'ACH-001533', 'ACH-002292', 'ACH-001136', 'ACH-002154', 'ACH-001573', 'ACH-002459', 'ACH-002368', 'ACH-002212', 'ACH-002199', 'ACH-001616', 'ACH-002149', 'ACH-002171', 'ACH-002366', 'ACH-002363', 'ACH-002175', 'ACH-002265', 'ACH-001347', 'ACH-002390', 'ACH-001087', 'ACH-002142', 'ACH-001146', 'ACH-001489', 'ACH-002062', 'ACH-002131', 'ACH-002235', 'ACH-002022', 'ACH-001389', 'ACH-002169', 'ACH-002461', 'ACH-002193', 'ACH-001071', 'ACH-001126', 'ACH-002239', 'ACH-002401', 'ACH-001696', 'ACH-002215', 'ACH-001225', 'ACH-001591', 'ACH-002104', 'ACH-001365', 'ACH-002378', 'ACH-001518', 'ACH-002140', 'ACH-001282', 'ACH-002174', 'ACH-001449', 'ACH-001703', 'ACH-002173', 'ACH-002095', 'ACH-002020', 'ACH-002180', 'ACH-002386', 'ACH-002196', 'ACH-001537', 'ACH-002107', 'ACH-001331', 'ACH-002299', 'ACH-002362', 'ACH-002277', 'ACH-002091', 'ACH-002314', 'ACH-001693', 'ACH-002217', 'ACH-001834', 'ACH-001081', 'ACH-001208', 'ACH-002347', 'ACH-001203', 'ACH-002112', 'ACH-001063', 'ACH-002382', 'ACH-001632', 'ACH-002395', 'ACH-002167', 'ACH-002218', 'ACH-001692', 'ACH-002124', 'ACH-002709', 'ACH-001437', 'ACH-002346', 'ACH-002268', 'ACH-002282', 'ACH-002343', 'ACH-002103', 'ACH-002291', 'ACH-001606', 'ACH-001648', 'ACH-002051', 'ACH-001973', 'ACH-002111', 'ACH-002373', 'ACH-002108', 'ACH-002213', 'ACH-001599', 'ACH-000795', 'ACH-001680', 'ACH-002024', 'ACH-002270', 'ACH-001578', 'ACH-000426', 'ACH-002190', 'ACH-001452', 'ACH-001539', 'ACH-002511', 'ACH-001393', 'ACH-002191', 'ACH-002114', 'ACH-002276', 'ACH-002369', 'ACH-002364', 'ACH-002177', 'ACH-001358', 'ACH-002198', 'ACH-002083', 'ACH-002023', 'ACH-002189', 'ACH-001143', 'ACH-002014', 'ACH-002109', 'ACH-001079', 'ACH-002113', 'ACH-002123', 'ACH-002201', 'ACH-001094', 'ACH-002046', 'ACH-001206', 'ACH-002301', 'ACH-001088', 'ACH-002342', 'ACH-001194', 'ACH-002115', 'ACH-001349', 'ACH-002294', 'ACH-002043', 'ACH-002281', 'ACH-001453', 'ACH-001577', 'ACH-001422', 'ACH-002132', 'ACH-002048', 'ACH-002365', 'ACH-001690', 'ACH-002097', 'ACH-001108', 'ACH-002337', 'ACH-001338', 'ACH-002119', 'ACH-002253', 'ACH-002383', 'ACH-002052', 'ACH-002245', 'ACH-001669', 'ACH-002286', 'ACH-001687', 'ACH-002396', 'ACH-002305', 'ACH-001544', 'ACH-001249', 'ACH-002145', 'ACH-002296', 'ACH-001957', 'ACH-002371', 'ACH-002188', 'ACH-001125', 'ACH-002361', 'ACH-002339', 'ACH-001036', 'ACH-002272', 'ACH-002375', 'ACH-002204', 'ACH-002207', 'ACH-002353', 'ACH-001214', 'ACH-001970', 'ACH-001493', 'ACH-001639', 'ACH-002336', 'ACH-001045', 'ACH-002026', 'ACH-002285', 'ACH-002110', 'ACH-001024', 'ACH-001955', 'ACH-001694', 'ACH-001198', 'ACH-002143', 'ACH-002117', 'ACH-002216', 'ACH-002269', 'ACH-001625', 'ACH-002126', 'ACH-001516', 'ACH-001162', 'ACH-001417', 'ACH-000979', 'ACH-002194', 'ACH-001641', 'ACH-001716', 'ACH-002392', 'ACH-001199', 'ACH-002231', 'ACH-001039', 'ACH-002157', 'ACH-002251', 'ACH-002170', 'ACH-002317', 'ACH-002208', 'ACH-002372', 'ACH-001956', 'ACH-002098', 'ACH-001502', 'ACH-001719', 'ACH-002025', 'ACH-002179', 'ACH-002303', 'ACH-002240', 'ACH-002351', 'ACH-002257', 'ACH-002164', 'ACH-002069', 'ACH-001189', 'ACH-002150', 'ACH-001395', 'ACH-002019', 'ACH-002118', 'ACH-000727', 'ACH-002355', 'ACH-002148', 'ACH-002136', 'ACH-001958', 'ACH-002309', 'ACH-001541', 'ACH-001672', 'ACH-002246', 'ACH-002279', 'ACH-002205', 'ACH-002250', 'ACH-002128', 'ACH-001359', 'ACH-002159', 'ACH-002349', 'ACH-002090', 'ACH-002460', 'ACH-002284', 'ACH-002377', 'ACH-002223', 'ACH-002308', 'ACH-001234', 'ACH-002248', 'ACH-001017', 'ACH-001698', 'ACH-001971', 'ACH-001049', 'ACH-001627', 'ACH-002262', 'ACH-002160', 'ACH-002255', 'ACH-002121', 'ACH-002387', 'ACH-002397', 'ACH-001820', 'ACH-001015', 'ACH-002307', 'ACH-002226', 'ACH-002041', 'ACH-002352', 'ACH-001383', 'ACH-002184', 'ACH-002039', 'ACH-002130', 'ACH-002293', 'ACH-001676', 'ACH-002302', 'ACH-002206', 'ACH-002237', 'ACH-002256', 'ACH-002229', 'ACH-001854', 'ACH-002335', 'ACH-002263', 'ACH-001738', 'ACH-002340', 'ACH-002370', 'ACH-002228', 'ACH-002384', 'ACH-002166', 'ACH-000992', 'ACH-002163', 'ACH-002388', 'ACH-001037', 'ACH-002129', 'ACH-002134', 'ACH-002211', 'ACH-001065', 'ACH-002399', 'ACH-001138', 'ACH-002182', 'ACH-001419', 'ACH-002176', 'ACH-002367', 'ACH-002147', 'ACH-002376', 'ACH-001363', 'ACH-002165', 'ACH-001293', 'ACH-002288', 'ACH-002067', 'ACH-002295', 'ACH-002249', 'ACH-001438', 'ACH-002183', 'ACH-002278', 'ACH-002038', 'ACH-001519', 'ACH-002089', 'ACH-002260', 'ACH-001388', 'ACH-002266', 'ACH-002181', 'ACH-002345', 'ACH-001011', 'ACH-002197', 'ACH-001512', 'ACH-002195', 'ACH-002350', 'ACH-002290', 'ACH-002200', 'ACH-002310', 'ACH-001441', 'ACH-002475', 'ACH-002283', 'ACH-001021', 'ACH-002264', 'ACH-001603', 'ACH-002236', 'ACH-002512', 'ACH-002315', 'ACH-002127', 'ACH-001151', 'ACH-002202', 'ACH-002312', 'ACH-002230', 'ACH-001339', 'ACH-001847', 'ACH-002259', 'ACH-002042', 'ACH-001187', 'ACH-002311', 'ACH-002186', 'ACH-002232', 'ACH-001064', 'ACH-001127', 'ACH-002271', 'ACH-002338', 'ACH-001767', 'ACH-002010', 'ACH-002381', 'ACH-001188', 'ACH-002280', 'ACH-002254', 'ACH-002360', 'ACH-002139', 'ACH-002463', 'ACH-002101', 'ACH-001072', 'ACH-001355', 'ACH-001398', 'ACH-001175', 'ACH-001118', 'ACH-001350', 'ACH-002380', 'ACH-002120', 'ACH-001688', 'ACH-002125', 'ACH-001490', 'ACH-002187', 'ACH-001000', 'ACH-002093', 'ACH-002391', 'ACH-001023', 'ACH-001370', 'ACH-002273', 'ACH-002393', 'ACH-002077', 'ACH-002156', 'ACH-001569', 'ACH-001385', 'ACH-001229', 'ACH-002227', 'ACH-002297'}\n",
      "removed\n",
      "missing\n",
      "{'ACH-001648', 'ACH-002014', 'ACH-001694', 'ACH-001973', 'ACH-001992', 'ACH-002055', 'ACH-001709', 'ACH-001688', 'ACH-001820', 'ACH-001675', 'ACH-002465', 'ACH-002021', 'ACH-001641', 'ACH-001036', 'ACH-001528', 'ACH-001991', 'ACH-002062', 'ACH-001533', 'ACH-001389', 'ACH-001422', 'ACH-002461', 'ACH-002039', 'ACH-001370', 'ACH-001419', 'ACH-001573', 'ACH-001970', 'ACH-001547', 'ACH-001707', 'ACH-002084', 'ACH-002026', 'ACH-002512', 'ACH-001385', 'ACH-002067', 'ACH-002023'}\n",
      "blacklist\n",
      "0 set()\n",
      "1304\n",
      "1304\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m--------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m      Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-202-698b27ea5701>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0mgenes_expected_count\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenes_expected_count\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mgenes_expected_count\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblacklist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes_expected_count\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m     \u001b[0mgenes_expected_count\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'temp/expression_genes_expected_count.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes_tpm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0mgenes_tpm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenes_tpm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mgenes_tpm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblacklist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal, errors)\u001b[0m\n\u001b[1;32m   3168\u001b[0m             \u001b[0mdecimal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdecimal\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3169\u001b[0m         )\n\u001b[0;32m-> 3170\u001b[0;31m         \u001b[0mformatter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3172\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpath_or_buf\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    204\u001b[0m             )\n\u001b[1;32m    205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    326\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    344\u001b[0m                 \u001b[0mdecimal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecimal\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    345\u001b[0m                 \u001b[0mdate_format\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdate_format\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 346\u001b[0;31m                 \u001b[0mquoting\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquoting\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    347\u001b[0m             )\n\u001b[1;32m    348\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mto_native_types\u001b[0;34m(self, na_rep, float_format, decimal, quoting, **kwargs)\u001b[0m\n\u001b[1;32m   1949\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1950\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mquoting\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1951\u001b[0;31m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1952\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1953\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"object\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "blacklist = set()\n",
    "for val in ['internal','ibm', 'dmc','public']:\n",
    "    print('_________________________________________________')\n",
    "    print(val)\n",
    "    print('not present')\n",
    "    removed = set(prev[val]) - set(genes_tpm.index)\n",
    "    print(removed)\n",
    "    print('removed for QC reasons')\n",
    "    print(failed)\n",
    "    print('removed')\n",
    "    removed = set(prevrna[val]) - set(genes_tpm.index)\n",
    "    #print(removed - set(rename.keys()))\n",
    "    missing = set(new[val]) - set(genes_tpm.index)\n",
    "    blacklist = (set(genes_tpm.index) - (prevrna[val] | set(new[val]))) | blacklist\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "\n",
    "    ## removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(genes_expected_count))\n",
    "    a = genes_expected_count[~genes_expected_count.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_genes_expected_count.csv')\n",
    "    print(len(genes_tpm))\n",
    "    a = genes_tpm[~genes_tpm.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_genes_tpm.csv')\n",
    "    print(len(proteincoding_genes_tpm))\n",
    "    a = proteincoding_genes_tpm[~proteincoding_genes_tpm.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_proteincoding_genes_tpm.csv')\n",
    "    print(len(transcripts_tpm))\n",
    "    a = transcripts_tpm[~transcripts_tpm.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_transcripts_tpm.csv')\n",
    "    print(len(proteincoding_genes_expected_count))\n",
    "    a = proteincoding_genes_expected_count[~proteincoding_genes_expected_count.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_proteincoding_genes_expected_count.csv')\n",
    "    print(len(transcripts_expected_count))\n",
    "    a = transcripts_expected_count[~transcripts_expected_count.index.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/expression_transcripts_expected_count.csv')\n",
    "    \n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # adding to taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_expression[val],\n",
    "                     upload_file_path_dict={\n",
    "                       'temp/expression_genes_expected_count.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_transcripts_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_genes_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_proteincoding_genes_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_proteincoding_genes_expected_count.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_transcripts_expected_count.csv': 'NumericMatrixCSV',\n",
    "                     \"temp/README\": \"Raw\"},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    # INTERNAL RNA\n",
    "\n",
    "    for information, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    REMOVED FOR QC REASONS:\n",
    "    \"\"\"+str(failed)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # add to virtual \n",
    "    AddToVirtual(virtual[val], taiga_expression[val], files=[\n",
    "    ('CCLE_expression_full', 'expression_genes_tpm'), \n",
    "    ('CCLE_RNAseq_transcripts', 'expression_transcripts_tpm'),\n",
    "    ('CCLE_RNAseq_reads', 'expression_genes_expected_count'),\n",
    "    ('CCLE_expression', 'expression_proteincoding_genes_tpm'),\n",
    "    ('CCLE_expression_proteincoding_genes_expected_count', 'expression_proteincoding_genes_expected_count'),\n",
    "    ('CCLE_expression_transcripts_expected_count', 'expression_transcripts_expected_count')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T13:54:11.063956Z",
     "start_time": "2021-01-22T13:53:58.945745Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_expression_full', 'depmap-rnaseq-expression-data-363a.36/expression_genes_tpm'), ('CCLE_RNAseq_transcripts', 'depmap-rnaseq-expression-data-363a.36/expression_transcripts_tpm'), ('CCLE_RNAseq_reads', 'depmap-rnaseq-expression-data-363a.36/expression_genes_expected_count'), ('CCLE_expression', 'depmap-rnaseq-expression-data-363a.36/expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'depmap-rnaseq-expression-data-363a.36/expression_proteincoding_genes_expected_count'), ('CCLE_expression_transcripts_expected_count', 'depmap-rnaseq-expression-data-363a.36/expression_transcripts_expected_count')]\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datafile/982a8b4e984b4b9fbf91d54d3fb4cc6d\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id d22ed5b476cd479fa49dc78c74e2501c created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/d22ed5b476cd479fa49dc78c74e2501c\n"
     ]
    }
   ],
   "source": [
    "AddToVirtual('depmap-a0ab', taiga_expression['internal'], files=[\n",
    "('CCLE_expression_full', 'expression_genes_tpm'), \n",
    "('CCLE_RNAseq_transcripts', 'expression_transcripts_tpm'),\n",
    "('CCLE_RNAseq_reads', 'expression_genes_expected_count'),\n",
    "('CCLE_expression', 'expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'expression_proteincoding_genes_expected_count'),('CCLE_expression_transcripts_expected_count', 'expression_transcripts_expected_count')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T13:40:21.097164Z",
     "start_time": "2021-01-22T13:40:19.704349Z"
    }
   },
   "outputs": [],
   "source": [
    "fusions = pd.read_csv('temp/fusions_'+samplesetname+'.csv')\n",
    "filtered = pd.read_csv('temp/filtered_fusions_'+samplesetname+'.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%store -r failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T14:00:44.614814Z",
     "start_time": "2021-01-22T14:00:44.524505Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "set()\n"
     ]
    }
   ],
   "source": [
    "print(set(fusions.DepMap_ID) - set(filtered.DepMap_ID))\n",
    "print(set(filtered.DepMap_ID) - set(fusions.DepMap_ID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T13:46:04.639343Z",
     "start_time": "2021-01-22T13:42:59.906017Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________\n",
      "internal\n",
      "not present\n",
      "{'ACH-002304', 'ACH-001489', 'ACH-002133', 'ACH-002144', 'ACH-001002', 'ACH-002245', 'ACH-001107', 'ACH-002191', 'ACH-002093', 'ACH-002252', 'ACH-002105', 'ACH-002175', 'ACH-002346', 'ACH-002097', 'ACH-002204', 'ACH-002161', 'ACH-001143', 'ACH-002190', 'ACH-002192', 'ACH-002337', 'ACH-001639', 'ACH-002376', 'ACH-001130', 'ACH-002122', 'ACH-002180', 'ACH-001125', 'ACH-002291', 'ACH-002285', 'ACH-002089', 'ACH-002316', 'ACH-002387', 'ACH-001171', 'ACH-002315', 'ACH-002241', 'ACH-002173', 'ACH-001955', 'ACH-002312', 'ACH-002142', 'ACH-001350', 'ACH-002292', 'ACH-002141', 'ACH-001063', 'ACH-002389', 'ACH-002382', 'ACH-002279', 'ACH-001233', 'ACH-001064', 'ACH-002248', 'ACH-001502', 'ACH-002107', 'ACH-001137', 'ACH-001249', 'ACH-002178', 'ACH-001108', 'ACH-002183', 'ACH-002260', 'ACH-001047', 'ACH-002121', 'ACH-002340', 'ACH-002102', 'ACH-000992', 'ACH-002395', 'ACH-001138', 'ACH-001234', 'ACH-002176', 'ACH-002143', 'ACH-001224', 'ACH-002296', 'ACH-002282', 'ACH-001090', 'ACH-001331', 'ACH-001101', 'ACH-001230', 'ACH-002391', 'ACH-002299', 'ACH-002243', 'ACH-002262', 'ACH-002215', 'ACH-001355', 'ACH-002020', 'ACH-002269', 'ACH-002228', 'ACH-002096', 'ACH-001208', 'ACH-002365', 'ACH-002110', 'ACH-002378', 'ACH-002374', 'ACH-002246', 'ACH-001293', 'ACH-001079', 'ACH-001599', 'ACH-002280', 'ACH-002263', 'ACH-002381', 'ACH-002195', 'ACH-002211', 'ACH-002249', 'ACH-002179', 'ACH-002271', 'ACH-002352', 'ACH-001206', 'ACH-001490', 'ACH-002229', 'ACH-002106', 'ACH-002295', 'ACH-002217', 'ACH-002377', 'ACH-000979', 'ACH-002052', 'ACH-001021', 'ACH-002375', 'ACH-002223', 'ACH-002154', 'ACH-002350', 'ACH-002373', 'ACH-002170', 'ACH-002113', 'ACH-002099', 'ACH-002259', 'ACH-001198', 'ACH-002353', 'ACH-002264', 'ACH-002251', 'ACH-002090', 'ACH-002160', 'ACH-001738', 'ACH-001203', 'ACH-002051', 'ACH-002226', 'ACH-001189', 'ACH-001081', 'ACH-002198', 'ACH-002237', 'ACH-002236', 'ACH-002270', 'ACH-002164', 'ACH-002275', 'ACH-001094', 'ACH-001767', 'ACH-002401', 'ACH-002397', 'ACH-001023', 'ACH-002233', 'ACH-002221', 'ACH-001045', 'ACH-002369', 'ACH-002366', 'ACH-002206', 'ACH-001957', 'ACH-001150', 'ACH-002231', 'ACH-002290', 'ACH-002383', 'ACH-002171', 'ACH-002388', 'ACH-002162', 'ACH-001042', 'ACH-002281', 'ACH-002286', 'ACH-002174', 'ACH-002349', 'ACH-002276', 'ACH-002058', 'ACH-002134', 'ACH-002301', 'ACH-002372', 'ACH-002131', 'ACH-001188', 'ACH-002127', 'ACH-002272', 'ACH-001044', 'ACH-001146', 'ACH-001182', 'ACH-001131', 'ACH-000795', 'ACH-002277', 'ACH-002361', 'ACH-002202', 'ACH-002118', 'ACH-002135', 'ACH-001606', 'ACH-002194', 'ACH-002120', 'ACH-002214', 'ACH-001126', 'ACH-001136', 'ACH-002124', 'ACH-002209', 'ACH-002224', 'ACH-002370', 'ACH-001162', 'ACH-001225', 'ACH-002338', 'ACH-002399', 'ACH-002043', 'ACH-001958', 'ACH-002390', 'ACH-001071', 'ACH-002343', 'ACH-001362', 'ACH-000047', 'ACH-001015', 'ACH-002384', 'ACH-001091', 'ACH-001017', 'ACH-001089', 'ACH-002336', 'ACH-002253', 'ACH-002103', 'ACH-002257', 'ACH-001118', 'ACH-002287', 'ACH-002177', 'ACH-002100', 'ACH-002151', 'ACH-002348', 'ACH-001088', 'ACH-002095', 'ACH-002132', 'ACH-002356', 'ACH-002200', 'ACH-002111', 'ACH-001383', 'ACH-002116', 'ACH-001358', 'ACH-002205', 'ACH-001127', 'ACH-002145', 'ACH-002364', 'ACH-002213', 'ACH-001093', 'ACH-002218', 'ACH-002196', 'ACH-002258', 'ACH-002146', 'ACH-001011', 'ACH-002152', 'ACH-002256', 'ACH-002308', 'ACH-002379', 'ACH-002339', 'ACH-002360', 'ACH-001452', 'ACH-002289', 'ACH-002240', 'ACH-002148', 'ACH-001039', 'ACH-002186', 'ACH-002136', 'ACH-002254', 'ACH-002250', 'ACH-002167', 'ACH-001359', 'ACH-002309', 'ACH-002153', 'ACH-001365', 'ACH-002119', 'ACH-002357', 'ACH-002317', 'ACH-002108', 'ACH-001214', 'ACH-002266', 'ACH-001175', 'ACH-002293', 'ACH-001037', 'ACH-002189', 'ACH-002166', 'ACH-002344', 'ACH-002306', 'ACH-002386', 'ACH-001205', 'ACH-002297', 'ACH-002396', 'ACH-002181', 'ACH-002112', 'ACH-001024', 'ACH-002126', 'ACH-001018', 'ACH-002163', 'ACH-002230', 'ACH-002393', 'ACH-002311', 'ACH-002185', 'ACH-002232', 'ACH-001591', 'ACH-002092', 'ACH-002225', 'ACH-002305', 'ACH-002222', 'ACH-002400', 'ACH-002083', 'ACH-001072', 'ACH-002300', 'ACH-002380', 'ACH-002371', 'ACH-002139', 'ACH-002091', 'ACH-002363', 'ACH-002123', 'ACH-002394', 'ACH-002159', 'ACH-002341', 'ACH-001364', 'ACH-002104', 'ACH-002182', 'ACH-002040', 'ACH-001417', 'ACH-002150', 'ACH-002193', 'ACH-002188', 'ACH-002298', 'ACH-002294', 'ACH-002101', 'ACH-001199', 'ACH-002117', 'ACH-001704', 'ACH-002125', 'ACH-002342', 'ACH-002273', 'ACH-002184', 'ACH-002210', 'ACH-002169', 'ACH-002345', 'ACH-001121', 'ACH-002238', 'ACH-002358', 'ACH-002242', 'ACH-002129', 'ACH-002244', 'ACH-001187', 'ACH-002234', 'ACH-002392', 'ACH-002128', 'ACH-001092', 'ACH-001680', 'ACH-002207', 'ACH-002149', 'ACH-002115', 'ACH-002199', 'ACH-000727', 'ACH-002274', 'ACH-001000', 'ACH-000309', 'ACH-001956', 'ACH-002165', 'ACH-000426', 'ACH-002227', 'ACH-002314', 'ACH-002303', 'ACH-001049', 'ACH-002155', 'ACH-002168', 'ACH-002147', 'ACH-002212', 'ACH-002156', 'ACH-002351', 'ACH-002359', 'ACH-001066', 'ACH-002283', 'ACH-002114', 'ACH-002140', 'ACH-002367', 'ACH-001282', 'ACH-001087', 'ACH-002197', 'ACH-002208', 'ACH-002302', 'ACH-002354', 'ACH-002310', 'ACH-001363', 'ACH-002094', 'ACH-002288', 'ACH-002109', 'ACH-002265', 'ACH-002216', 'ACH-002267', 'ACH-002235', 'ACH-002201', 'ACH-002247', 'ACH-002239', 'ACH-002313', 'ACH-001065', 'ACH-002098', 'ACH-002268', 'ACH-002278', 'ACH-002335', 'ACH-002368', 'ACH-002347', 'ACH-002255', 'ACH-002284', 'ACH-002130', 'ACH-002220', 'ACH-001043', 'ACH-002307', 'ACH-002362', 'ACH-002355', 'ACH-001544', 'ACH-002261', 'ACH-002157', 'ACH-002187'}\n",
      "removed for QC reasons\n",
      "{'PEDS117'}\n",
      "removed\n",
      "set()\n",
      "missing\n",
      "{'ACH-001675'}\n",
      "blacklist\n",
      "20 {'ACH-002077', 'ACH-001434', 'ACH-002013', 'ACH-001553', 'ACH-002687', 'ACH-001633', 'ACH-001996', 'ACH-001046', 'ACH-001227', 'ACH-001758', 'ACH-002476', 'ACH-001686', 'ACH-002034', 'ACH-002047', 'ACH-002080', 'ACH-001338', 'ACH-001759', 'ACH-001760', 'ACH-002707', 'ACH-001756'}\n",
      "359333\n",
      "356874\n",
      "44802\n",
      "44119\n",
      "Uploading fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/450ceeae5de444de97c5333d2586b839\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: fusions properly converted and uploaded\n",
      "Uploading filtered_fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/450ceeae5de444de97c5333d2586b839\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: filtered_fusions properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/450ceeae5de444de97c5333d2586b839\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id c081c03e4fe343ce9dfd37f80fe00450 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/c081c03e4fe343ce9dfd37f80fe00450\n",
      "[('CCLE_fusions_unfiltered', 'gene-fusions-8b7a.20/fusions'), ('CCLE_fusions', 'gene-fusions-8b7a.20/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datafile/833379f114c54ecda9ec0f73a2ed2370\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 92850e4bba2044b5bd10f36f0112d20c created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/92850e4bba2044b5bd10f36f0112d20c\n",
      "_________________________________________________\n",
      "ibm\n",
      "not present\n",
      "{'ACH-002304', 'ACH-001489', 'ACH-002133', 'ACH-002144', 'ACH-001002', 'ACH-002245', 'ACH-001107', 'ACH-002191', 'ACH-002093', 'ACH-002252', 'ACH-002105', 'ACH-002175', 'ACH-002346', 'ACH-002097', 'ACH-002204', 'ACH-002161', 'ACH-001143', 'ACH-002190', 'ACH-002192', 'ACH-002337', 'ACH-001639', 'ACH-002376', 'ACH-001130', 'ACH-002122', 'ACH-002180', 'ACH-001125', 'ACH-002291', 'ACH-002285', 'ACH-002089', 'ACH-002316', 'ACH-002387', 'ACH-001171', 'ACH-002315', 'ACH-002241', 'ACH-002173', 'ACH-001955', 'ACH-002312', 'ACH-002142', 'ACH-001350', 'ACH-002292', 'ACH-002141', 'ACH-001063', 'ACH-002389', 'ACH-002382', 'ACH-002279', 'ACH-001233', 'ACH-001064', 'ACH-002248', 'ACH-001502', 'ACH-002107', 'ACH-001137', 'ACH-001249', 'ACH-002178', 'ACH-001108', 'ACH-002183', 'ACH-002260', 'ACH-001047', 'ACH-002121', 'ACH-002340', 'ACH-002102', 'ACH-000992', 'ACH-002395', 'ACH-001138', 'ACH-001234', 'ACH-002176', 'ACH-002143', 'ACH-001224', 'ACH-002296', 'ACH-002282', 'ACH-001090', 'ACH-001331', 'ACH-001101', 'ACH-001230', 'ACH-002391', 'ACH-002299', 'ACH-002243', 'ACH-002262', 'ACH-002215', 'ACH-001355', 'ACH-002020', 'ACH-002269', 'ACH-002228', 'ACH-002096', 'ACH-001208', 'ACH-002365', 'ACH-002110', 'ACH-002378', 'ACH-002374', 'ACH-001293', 'ACH-002246', 'ACH-001079', 'ACH-001599', 'ACH-002280', 'ACH-002263', 'ACH-002381', 'ACH-002195', 'ACH-002211', 'ACH-002249', 'ACH-002179', 'ACH-002271', 'ACH-002352', 'ACH-001206', 'ACH-001490', 'ACH-002229', 'ACH-002106', 'ACH-002295', 'ACH-002217', 'ACH-002377', 'ACH-000979', 'ACH-002052', 'ACH-001021', 'ACH-002375', 'ACH-002223', 'ACH-002154', 'ACH-002350', 'ACH-002373', 'ACH-002170', 'ACH-002113', 'ACH-002099', 'ACH-002259', 'ACH-001198', 'ACH-002353', 'ACH-002264', 'ACH-002251', 'ACH-002090', 'ACH-002160', 'ACH-001738', 'ACH-001203', 'ACH-002051', 'ACH-002226', 'ACH-001189', 'ACH-001081', 'ACH-002198', 'ACH-002237', 'ACH-002236', 'ACH-002270', 'ACH-002164', 'ACH-002275', 'ACH-001094', 'ACH-001767', 'ACH-002401', 'ACH-002397', 'ACH-001023', 'ACH-002233', 'ACH-002221', 'ACH-001045', 'ACH-002369', 'ACH-002366', 'ACH-002206', 'ACH-001957', 'ACH-001150', 'ACH-002231', 'ACH-002290', 'ACH-002383', 'ACH-002171', 'ACH-002388', 'ACH-002162', 'ACH-001042', 'ACH-002281', 'ACH-002286', 'ACH-002174', 'ACH-002349', 'ACH-002276', 'ACH-002058', 'ACH-002134', 'ACH-002301', 'ACH-002372', 'ACH-002131', 'ACH-001188', 'ACH-002127', 'ACH-002272', 'ACH-001044', 'ACH-001146', 'ACH-001182', 'ACH-001131', 'ACH-000795', 'ACH-002277', 'ACH-002361', 'ACH-002202', 'ACH-002118', 'ACH-002135', 'ACH-001606', 'ACH-002194', 'ACH-002120', 'ACH-002214', 'ACH-001126', 'ACH-001136', 'ACH-002124', 'ACH-002209', 'ACH-002224', 'ACH-002370', 'ACH-001162', 'ACH-001225', 'ACH-002338', 'ACH-002399', 'ACH-002043', 'ACH-001958', 'ACH-002390', 'ACH-001071', 'ACH-002343', 'ACH-001362', 'ACH-000047', 'ACH-001015', 'ACH-002384', 'ACH-001091', 'ACH-001017', 'ACH-001089', 'ACH-002336', 'ACH-002253', 'ACH-002103', 'ACH-002257', 'ACH-001118', 'ACH-002287', 'ACH-002177', 'ACH-002100', 'ACH-002151', 'ACH-002348', 'ACH-001088', 'ACH-002095', 'ACH-002132', 'ACH-002356', 'ACH-002200', 'ACH-002111', 'ACH-001383', 'ACH-002116', 'ACH-001358', 'ACH-002205', 'ACH-001127', 'ACH-002145', 'ACH-002364', 'ACH-002213', 'ACH-001093', 'ACH-002218', 'ACH-002196', 'ACH-002258', 'ACH-002146', 'ACH-001011', 'ACH-002152', 'ACH-002256', 'ACH-002308', 'ACH-002379', 'ACH-002339', 'ACH-002360', 'ACH-001452', 'ACH-002289', 'ACH-002240', 'ACH-002148', 'ACH-001039', 'ACH-002186', 'ACH-002136', 'ACH-002254', 'ACH-002250', 'ACH-002167', 'ACH-001359', 'ACH-002309', 'ACH-002153', 'ACH-001365', 'ACH-002119', 'ACH-002357', 'ACH-002317', 'ACH-002108', 'ACH-001214', 'ACH-002266', 'ACH-001175', 'ACH-002293', 'ACH-001037', 'ACH-002189', 'ACH-002166', 'ACH-002344', 'ACH-002306', 'ACH-002386', 'ACH-001205', 'ACH-002297', 'ACH-002396', 'ACH-002181', 'ACH-002112', 'ACH-001024', 'ACH-002126', 'ACH-001018', 'ACH-002163', 'ACH-002230', 'ACH-002393', 'ACH-002311', 'ACH-002185', 'ACH-002232', 'ACH-001591', 'ACH-002092', 'ACH-002225', 'ACH-002305', 'ACH-002222', 'ACH-002400', 'ACH-002083', 'ACH-001072', 'ACH-002300', 'ACH-002380', 'ACH-002371', 'ACH-002139', 'ACH-002091', 'ACH-002363', 'ACH-002123', 'ACH-002394', 'ACH-002159', 'ACH-002341', 'ACH-001364', 'ACH-002104', 'ACH-002182', 'ACH-002040', 'ACH-001417', 'ACH-002150', 'ACH-002193', 'ACH-002188', 'ACH-002298', 'ACH-002294', 'ACH-002101', 'ACH-001199', 'ACH-002117', 'ACH-001704', 'ACH-002125', 'ACH-002342', 'ACH-002273', 'ACH-002184', 'ACH-002210', 'ACH-002169', 'ACH-002345', 'ACH-001121', 'ACH-002238', 'ACH-002358', 'ACH-002242', 'ACH-002129', 'ACH-002244', 'ACH-001187', 'ACH-002234', 'ACH-002392', 'ACH-002128', 'ACH-001092', 'ACH-001680', 'ACH-002207', 'ACH-002149', 'ACH-002115', 'ACH-002199', 'ACH-000727', 'ACH-002274', 'ACH-001000', 'ACH-000309', 'ACH-001956', 'ACH-002165', 'ACH-000426', 'ACH-002227', 'ACH-002314', 'ACH-002303', 'ACH-001049', 'ACH-002155', 'ACH-002168', 'ACH-002147', 'ACH-002212', 'ACH-002156', 'ACH-002351', 'ACH-002359', 'ACH-001066', 'ACH-002283', 'ACH-002114', 'ACH-002140', 'ACH-002367', 'ACH-001282', 'ACH-001087', 'ACH-002197', 'ACH-002208', 'ACH-002302', 'ACH-002354', 'ACH-002310', 'ACH-001363', 'ACH-002094', 'ACH-002288', 'ACH-002109', 'ACH-002265', 'ACH-002216', 'ACH-002267', 'ACH-002235', 'ACH-002201', 'ACH-002247', 'ACH-002239', 'ACH-002313', 'ACH-001065', 'ACH-002098', 'ACH-002268', 'ACH-002278', 'ACH-002335', 'ACH-002368', 'ACH-002347', 'ACH-002255', 'ACH-002284', 'ACH-002130', 'ACH-002220', 'ACH-001043', 'ACH-002307', 'ACH-002362', 'ACH-002355', 'ACH-001544', 'ACH-002261', 'ACH-002157', 'ACH-002187'}\n",
      "removed for QC reasons\n",
      "{'PEDS117'}\n",
      "removed\n",
      "set()\n",
      "missing\n",
      "{'ACH-001675'}\n",
      "blacklist\n",
      "22 {'ACH-002077', 'ACH-001708', 'ACH-002687', 'ACH-001633', 'ACH-001686', 'ACH-002080', 'ACH-001338', 'ACH-001759', 'ACH-001393', 'ACH-001434', 'ACH-002013', 'ACH-001553', 'ACH-001996', 'ACH-001046', 'ACH-001227', 'ACH-001758', 'ACH-002476', 'ACH-002034', 'ACH-002047', 'ACH-001760', 'ACH-002707', 'ACH-001756'}\n",
      "359333\n",
      "356553\n",
      "44802\n",
      "44071\n",
      "Uploading fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/cd0b98dd5c9b45bbb386a0809388bcf9\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: fusions properly converted and uploaded\n",
      "Uploading filtered_fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/cd0b98dd5c9b45bbb386a0809388bcf9\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: filtered_fusions properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/cd0b98dd5c9b45bbb386a0809388bcf9\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 85aaa68f08c94e5b94a39408480c8f75 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/85aaa68f08c94e5b94a39408480c8f75\n",
      "[('CCLE_fusions_unfiltered', 'fusion-dc08.5/fusions'), ('CCLE_fusions', 'fusion-dc08.5/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datafile/c18a6f78a1e74d22b78478db33a786be\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 8fbea4526633429f807a401271585183 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/8fbea4526633429f807a401271585183\n",
      "_________________________________________________\n",
      "dmc\n",
      "not present\n",
      "{'ACH-002304', 'ACH-001489', 'ACH-002133', 'ACH-002144', 'ACH-001002', 'ACH-002245', 'ACH-001107', 'ACH-002191', 'ACH-002093', 'ACH-002252', 'ACH-002105', 'ACH-002175', 'ACH-002346', 'ACH-002097', 'ACH-002204', 'ACH-002161', 'ACH-001143', 'ACH-002190', 'ACH-002192', 'ACH-002337', 'ACH-001639', 'ACH-002376', 'ACH-001130', 'ACH-002122', 'ACH-002180', 'ACH-001125', 'ACH-002291', 'ACH-002285', 'ACH-002089', 'ACH-002316', 'ACH-002387', 'ACH-001171', 'ACH-002315', 'ACH-002241', 'ACH-002173', 'ACH-001955', 'ACH-002312', 'ACH-002142', 'ACH-001350', 'ACH-002292', 'ACH-002141', 'ACH-001063', 'ACH-002389', 'ACH-002382', 'ACH-002279', 'ACH-001233', 'ACH-001064', 'ACH-002248', 'ACH-001502', 'ACH-002107', 'ACH-001137', 'ACH-001249', 'ACH-002178', 'ACH-001108', 'ACH-002183', 'ACH-002260', 'ACH-001047', 'ACH-002121', 'ACH-002340', 'ACH-002102', 'ACH-000992', 'ACH-002395', 'ACH-001138', 'ACH-001234', 'ACH-002176', 'ACH-002143', 'ACH-001224', 'ACH-002296', 'ACH-002282', 'ACH-001090', 'ACH-001331', 'ACH-001101', 'ACH-001230', 'ACH-002391', 'ACH-002299', 'ACH-002243', 'ACH-002262', 'ACH-002215', 'ACH-001355', 'ACH-002020', 'ACH-002269', 'ACH-002228', 'ACH-002096', 'ACH-001208', 'ACH-002365', 'ACH-002110', 'ACH-002378', 'ACH-002374', 'ACH-001293', 'ACH-002246', 'ACH-001079', 'ACH-001599', 'ACH-002280', 'ACH-002263', 'ACH-002381', 'ACH-002195', 'ACH-002211', 'ACH-002249', 'ACH-002179', 'ACH-002271', 'ACH-002352', 'ACH-001206', 'ACH-001490', 'ACH-002229', 'ACH-002106', 'ACH-002295', 'ACH-002217', 'ACH-002377', 'ACH-000979', 'ACH-002052', 'ACH-001021', 'ACH-002375', 'ACH-002223', 'ACH-002154', 'ACH-002350', 'ACH-002373', 'ACH-002170', 'ACH-002113', 'ACH-002099', 'ACH-002259', 'ACH-001198', 'ACH-002353', 'ACH-002264', 'ACH-002251', 'ACH-002090', 'ACH-002160', 'ACH-001738', 'ACH-001203', 'ACH-002051', 'ACH-002226', 'ACH-001189', 'ACH-001081', 'ACH-002198', 'ACH-002237', 'ACH-002236', 'ACH-002270', 'ACH-002164', 'ACH-002275', 'ACH-001094', 'ACH-001767', 'ACH-002401', 'ACH-002397', 'ACH-001023', 'ACH-002233', 'ACH-002221', 'ACH-001045', 'ACH-002369', 'ACH-002366', 'ACH-002206', 'ACH-001957', 'ACH-001150', 'ACH-002231', 'ACH-002290', 'ACH-002383', 'ACH-002171', 'ACH-002388', 'ACH-002162', 'ACH-001042', 'ACH-002281', 'ACH-002286', 'ACH-002174', 'ACH-002349', 'ACH-002276', 'ACH-002058', 'ACH-002134', 'ACH-002301', 'ACH-002372', 'ACH-002131', 'ACH-001188', 'ACH-002127', 'ACH-002272', 'ACH-001044', 'ACH-001146', 'ACH-001182', 'ACH-001131', 'ACH-000795', 'ACH-002277', 'ACH-002361', 'ACH-002202', 'ACH-002118', 'ACH-002135', 'ACH-001606', 'ACH-002194', 'ACH-002120', 'ACH-002214', 'ACH-001126', 'ACH-001136', 'ACH-002124', 'ACH-002209', 'ACH-002224', 'ACH-002370', 'ACH-001162', 'ACH-001225', 'ACH-002338', 'ACH-002399', 'ACH-002043', 'ACH-001958', 'ACH-002390', 'ACH-001071', 'ACH-002343', 'ACH-001362', 'ACH-000047', 'ACH-001015', 'ACH-002384', 'ACH-001091', 'ACH-001017', 'ACH-001089', 'ACH-002336', 'ACH-002253', 'ACH-002103', 'ACH-002257', 'ACH-001118', 'ACH-002287', 'ACH-002177', 'ACH-002100', 'ACH-002151', 'ACH-002348', 'ACH-001088', 'ACH-002095', 'ACH-002132', 'ACH-002356', 'ACH-002200', 'ACH-002111', 'ACH-001383', 'ACH-002116', 'ACH-001358', 'ACH-002205', 'ACH-001127', 'ACH-002145', 'ACH-002364', 'ACH-002213', 'ACH-001093', 'ACH-002218', 'ACH-002196', 'ACH-002258', 'ACH-002146', 'ACH-001011', 'ACH-002152', 'ACH-002256', 'ACH-002308', 'ACH-002379', 'ACH-002339', 'ACH-002360', 'ACH-001452', 'ACH-002289', 'ACH-002240', 'ACH-002148', 'ACH-001039', 'ACH-002186', 'ACH-002136', 'ACH-002254', 'ACH-002250', 'ACH-002167', 'ACH-001359', 'ACH-002309', 'ACH-002153', 'ACH-001365', 'ACH-002119', 'ACH-002357', 'ACH-002317', 'ACH-002108', 'ACH-001214', 'ACH-002266', 'ACH-001175', 'ACH-002293', 'ACH-001037', 'ACH-002189', 'ACH-002166', 'ACH-002344', 'ACH-002306', 'ACH-002386', 'ACH-001205', 'ACH-002297', 'ACH-002396', 'ACH-002181', 'ACH-002112', 'ACH-001024', 'ACH-002126', 'ACH-001018', 'ACH-002163', 'ACH-002230', 'ACH-002393', 'ACH-002311', 'ACH-002185', 'ACH-002232', 'ACH-001591', 'ACH-002092', 'ACH-002225', 'ACH-002305', 'ACH-002222', 'ACH-002400', 'ACH-002083', 'ACH-001072', 'ACH-002300', 'ACH-002380', 'ACH-002371', 'ACH-002139', 'ACH-002091', 'ACH-002363', 'ACH-002123', 'ACH-002394', 'ACH-002159', 'ACH-002341', 'ACH-001364', 'ACH-002104', 'ACH-002182', 'ACH-002040', 'ACH-001417', 'ACH-002150', 'ACH-002193', 'ACH-002188', 'ACH-002298', 'ACH-002294', 'ACH-002101', 'ACH-001199', 'ACH-002117', 'ACH-001704', 'ACH-002125', 'ACH-002342', 'ACH-002273', 'ACH-002184', 'ACH-002210', 'ACH-002169', 'ACH-002345', 'ACH-001121', 'ACH-002238', 'ACH-002358', 'ACH-002242', 'ACH-002129', 'ACH-002244', 'ACH-001187', 'ACH-002234', 'ACH-002392', 'ACH-002128', 'ACH-001092', 'ACH-001680', 'ACH-002207', 'ACH-002149', 'ACH-002115', 'ACH-002199', 'ACH-000727', 'ACH-002274', 'ACH-001000', 'ACH-000309', 'ACH-001956', 'ACH-002165', 'ACH-000426', 'ACH-002227', 'ACH-002314', 'ACH-002303', 'ACH-001049', 'ACH-002155', 'ACH-002168', 'ACH-002147', 'ACH-002212', 'ACH-002156', 'ACH-002351', 'ACH-002359', 'ACH-001066', 'ACH-002283', 'ACH-002114', 'ACH-002140', 'ACH-002367', 'ACH-001282', 'ACH-001087', 'ACH-002197', 'ACH-002208', 'ACH-002302', 'ACH-002354', 'ACH-002310', 'ACH-001363', 'ACH-002094', 'ACH-002288', 'ACH-002109', 'ACH-002265', 'ACH-002216', 'ACH-002267', 'ACH-002235', 'ACH-002201', 'ACH-002247', 'ACH-002239', 'ACH-002313', 'ACH-001065', 'ACH-002098', 'ACH-002268', 'ACH-002278', 'ACH-002335', 'ACH-002368', 'ACH-002347', 'ACH-002255', 'ACH-002284', 'ACH-002130', 'ACH-002220', 'ACH-001043', 'ACH-002307', 'ACH-002362', 'ACH-002355', 'ACH-001544', 'ACH-002261', 'ACH-002157', 'ACH-002187'}\n",
      "removed for QC reasons\n",
      "{'PEDS117'}\n",
      "removed\n",
      "set()\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "23 {'ACH-002077', 'ACH-002055', 'ACH-001708', 'ACH-002687', 'ACH-001633', 'ACH-001686', 'ACH-002080', 'ACH-001338', 'ACH-001759', 'ACH-001393', 'ACH-001434', 'ACH-002013', 'ACH-001553', 'ACH-001996', 'ACH-001046', 'ACH-001227', 'ACH-001758', 'ACH-002476', 'ACH-002034', 'ACH-002047', 'ACH-001760', 'ACH-002707', 'ACH-001756'}\n",
      "359333\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "356396\n",
      "44802\n",
      "44057\n",
      "Uploading fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/a6f8e263ff904f529dd77ea151b26a4d\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: fusions properly converted and uploaded\n",
      "Uploading filtered_fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/a6f8e263ff904f529dd77ea151b26a4d\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: filtered_fusions properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/a6f8e263ff904f529dd77ea151b26a4d\n",
      "Conversion and upload...:\n",
      "\t Uploading the DataFileFormat.Raw to S3\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 1db9fa6f7ce44b0ba4f35e1a73f1076a created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/1db9fa6f7ce44b0ba4f35e1a73f1076a\n",
      "[('CCLE_fusions_unfiltered', 'gene-fusions-375f.18/fusions'), ('CCLE_fusions', 'gene-fusions-375f.18/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datafile/ec12ceef2ac44481ac4e04138990aef5\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id b9c1fe9bdc6e41c2badce084c5686081 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/b9c1fe9bdc6e41c2badce084c5686081\n",
      "_________________________________________________\n",
      "public\n",
      "not present\n",
      "{'ACH-002304', 'ACH-001489', 'ACH-002133', 'ACH-002144', 'ACH-001002', 'ACH-002245', 'ACH-001107', 'ACH-002191', 'ACH-002093', 'ACH-002252', 'ACH-002105', 'ACH-002175', 'ACH-002346', 'ACH-002097', 'ACH-002204', 'ACH-002161', 'ACH-001143', 'ACH-002190', 'ACH-002192', 'ACH-002337', 'ACH-001639', 'ACH-002376', 'ACH-001130', 'ACH-002122', 'ACH-002180', 'ACH-001125', 'ACH-002291', 'ACH-002285', 'ACH-002089', 'ACH-002316', 'ACH-002387', 'ACH-001171', 'ACH-002315', 'ACH-002241', 'ACH-002173', 'ACH-001955', 'ACH-002312', 'ACH-002142', 'ACH-001350', 'ACH-002292', 'ACH-002141', 'ACH-001063', 'ACH-002389', 'ACH-002382', 'ACH-002279', 'ACH-001233', 'ACH-001064', 'ACH-002248', 'ACH-002107', 'ACH-001137', 'ACH-001249', 'ACH-002178', 'ACH-001108', 'ACH-002183', 'ACH-002260', 'ACH-001047', 'ACH-002121', 'ACH-002340', 'ACH-002102', 'ACH-000992', 'ACH-002395', 'ACH-001138', 'ACH-001234', 'ACH-002176', 'ACH-002143', 'ACH-001224', 'ACH-002296', 'ACH-002282', 'ACH-001090', 'ACH-001331', 'ACH-001101', 'ACH-001230', 'ACH-002391', 'ACH-002299', 'ACH-002243', 'ACH-002262', 'ACH-002215', 'ACH-001355', 'ACH-002020', 'ACH-002269', 'ACH-002228', 'ACH-002096', 'ACH-001208', 'ACH-002365', 'ACH-002110', 'ACH-002378', 'ACH-002374', 'ACH-002246', 'ACH-001079', 'ACH-001599', 'ACH-002280', 'ACH-002263', 'ACH-002381', 'ACH-002195', 'ACH-002211', 'ACH-002249', 'ACH-002179', 'ACH-002271', 'ACH-002352', 'ACH-001206', 'ACH-001490', 'ACH-002229', 'ACH-002106', 'ACH-002295', 'ACH-002217', 'ACH-002377', 'ACH-000979', 'ACH-002052', 'ACH-001021', 'ACH-002375', 'ACH-002223', 'ACH-002154', 'ACH-002350', 'ACH-002373', 'ACH-002170', 'ACH-002113', 'ACH-002099', 'ACH-002259', 'ACH-001198', 'ACH-002353', 'ACH-002264', 'ACH-002251', 'ACH-002090', 'ACH-002160', 'ACH-001738', 'ACH-001203', 'ACH-002051', 'ACH-002226', 'ACH-001189', 'ACH-001081', 'ACH-002198', 'ACH-002237', 'ACH-002236', 'ACH-002270', 'ACH-002164', 'ACH-002275', 'ACH-001094', 'ACH-001767', 'ACH-002397', 'ACH-001023', 'ACH-002233', 'ACH-002221', 'ACH-001045', 'ACH-002369', 'ACH-002366', 'ACH-002206', 'ACH-001957', 'ACH-001150', 'ACH-002231', 'ACH-002290', 'ACH-002383', 'ACH-002171', 'ACH-002388', 'ACH-002162', 'ACH-001042', 'ACH-002281', 'ACH-002286', 'ACH-002174', 'ACH-002349', 'ACH-002276', 'ACH-002058', 'ACH-002134', 'ACH-002301', 'ACH-002372', 'ACH-002131', 'ACH-001188', 'ACH-002127', 'ACH-002272', 'ACH-001044', 'ACH-001146', 'ACH-001182', 'ACH-001131', 'ACH-000795', 'ACH-002277', 'ACH-002361', 'ACH-002202', 'ACH-002118', 'ACH-002135', 'ACH-001606', 'ACH-002194', 'ACH-002120', 'ACH-002214', 'ACH-001126', 'ACH-001136', 'ACH-002124', 'ACH-002209', 'ACH-002224', 'ACH-002370', 'ACH-001162', 'ACH-001225', 'ACH-002338', 'ACH-002399', 'ACH-002043', 'ACH-001958', 'ACH-002390', 'ACH-001071', 'ACH-002343', 'ACH-001362', 'ACH-000047', 'ACH-001015', 'ACH-002384', 'ACH-001091', 'ACH-001017', 'ACH-001089', 'ACH-002336', 'ACH-002253', 'ACH-002103', 'ACH-002257', 'ACH-001118', 'ACH-002287', 'ACH-002177', 'ACH-002100', 'ACH-002151', 'ACH-002348', 'ACH-001088', 'ACH-002095', 'ACH-002132', 'ACH-002356', 'ACH-002200', 'ACH-002111', 'ACH-001383', 'ACH-002116', 'ACH-001358', 'ACH-002205', 'ACH-001127', 'ACH-002145', 'ACH-002364', 'ACH-002213', 'ACH-001093', 'ACH-002218', 'ACH-002196', 'ACH-002258', 'ACH-002146', 'ACH-001011', 'ACH-002152', 'ACH-002256', 'ACH-002308', 'ACH-002379', 'ACH-002339', 'ACH-002360', 'ACH-001452', 'ACH-002289', 'ACH-002240', 'ACH-002148', 'ACH-001039', 'ACH-002186', 'ACH-002136', 'ACH-002254', 'ACH-002250', 'ACH-002167', 'ACH-001359', 'ACH-002309', 'ACH-002153', 'ACH-001365', 'ACH-002119', 'ACH-002357', 'ACH-002317', 'ACH-002108', 'ACH-001214', 'ACH-002266', 'ACH-001175', 'ACH-002293', 'ACH-001037', 'ACH-002189', 'ACH-002166', 'ACH-002344', 'ACH-002306', 'ACH-002386', 'ACH-001205', 'ACH-002297', 'ACH-002396', 'ACH-002181', 'ACH-002112', 'ACH-001024', 'ACH-002126', 'ACH-001018', 'ACH-002163', 'ACH-002230', 'ACH-002393', 'ACH-002311', 'ACH-002185', 'ACH-002232', 'ACH-001591', 'ACH-002092', 'ACH-002225', 'ACH-002305', 'ACH-002222', 'ACH-002083', 'ACH-001072', 'ACH-002300', 'ACH-002380', 'ACH-002371', 'ACH-002139', 'ACH-002091', 'ACH-002363', 'ACH-002123', 'ACH-002394', 'ACH-002159', 'ACH-002341', 'ACH-001364', 'ACH-002104', 'ACH-002182', 'ACH-002040', 'ACH-001417', 'ACH-002150', 'ACH-002193', 'ACH-002188', 'ACH-002298', 'ACH-002294', 'ACH-002101', 'ACH-001199', 'ACH-002117', 'ACH-001704', 'ACH-002125', 'ACH-002342', 'ACH-002273', 'ACH-002184', 'ACH-002210', 'ACH-002169', 'ACH-002345', 'ACH-001121', 'ACH-002238', 'ACH-002358', 'ACH-002242', 'ACH-002129', 'ACH-002244', 'ACH-001187', 'ACH-002234', 'ACH-002392', 'ACH-002128', 'ACH-001092', 'ACH-001680', 'ACH-002207', 'ACH-002149', 'ACH-002115', 'ACH-002199', 'ACH-000727', 'ACH-002274', 'ACH-001000', 'ACH-000309', 'ACH-001956', 'ACH-002165', 'ACH-000426', 'ACH-002227', 'ACH-002314', 'ACH-002303', 'ACH-001049', 'ACH-002155', 'ACH-002168', 'ACH-002147', 'ACH-002212', 'ACH-002156', 'ACH-002351', 'ACH-002359', 'ACH-001066', 'ACH-002283', 'ACH-002114', 'ACH-002140', 'ACH-002367', 'ACH-001282', 'ACH-001087', 'ACH-002197', 'ACH-002208', 'ACH-002302', 'ACH-002354', 'ACH-002310', 'ACH-001363', 'ACH-002094', 'ACH-002288', 'ACH-002109', 'ACH-002265', 'ACH-002216', 'ACH-002267', 'ACH-002235', 'ACH-002201', 'ACH-002247', 'ACH-002239', 'ACH-002313', 'ACH-001065', 'ACH-002098', 'ACH-002268', 'ACH-002278', 'ACH-002335', 'ACH-002368', 'ACH-002347', 'ACH-002255', 'ACH-002284', 'ACH-002130', 'ACH-002220', 'ACH-001043', 'ACH-002307', 'ACH-002362', 'ACH-002355', 'ACH-001544', 'ACH-002261', 'ACH-002157', 'ACH-002187'}\n",
      "removed for QC reasons\n",
      "{'PEDS117'}\n",
      "removed\n",
      "set()\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "56 {'ACH-002077', 'ACH-002055', 'ACH-001641', 'ACH-002014', 'ACH-001854', 'ACH-002512', 'ACH-001633', 'ACH-001693', 'ACH-001449', 'ACH-001429', 'ACH-002010', 'ACH-001855', 'ACH-002080', 'ACH-001759', 'ACH-001493', 'ACH-001438', 'ACH-001537', 'ACH-001547', 'ACH-002084', 'ACH-001696', 'ACH-001678', 'ACH-001676', 'ACH-001553', 'ACH-001992', 'ACH-001662', 'ACH-001046', 'ACH-001227', 'ACH-001758', 'ACH-002476', 'ACH-002048', 'ACH-001756', 'ACH-001672', 'ACH-001971', 'ACH-001708', 'ACH-002021', 'ACH-002687', 'ACH-001349', 'ACH-002465', 'ACH-001686', 'ACH-001338', 'ACH-001393', 'ACH-001973', 'ACH-001669', 'ACH-001434', 'ACH-002013', 'ACH-001996', 'ACH-001970', 'ACH-001707', 'ACH-001533', 'ACH-001847', 'ACH-002047', 'ACH-002034', 'ACH-001437', 'ACH-001760', 'ACH-002707', 'ACH-001679'}\n",
      "359333\n",
      "352533\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44802\n",
      "43388\n",
      "Uploading fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/91fbefe314124aa8930faf6b9350a7ee\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: fusions properly converted and uploaded\n",
      "Uploading filtered_fusions...\n",
      "hitting https://cds.team/taiga/api/datafile/91fbefe314124aa8930faf6b9350a7ee\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: filtered_fusions properly converted and uploaded\n",
      "Uploading README...\n",
      "hitting https://cds.team/taiga/api/datafile/91fbefe314124aa8930faf6b9350a7ee\n",
      "Conversion and upload...:\n",
      "\t Waiting in the task queue\n",
      "\n",
      "\t Done: README properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id ffb573a74d2a41d0afb076241f86aee2 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/ffb573a74d2a41d0afb076241f86aee2\n",
      "[('CCLE_fusions_unfiltered', 'gene-fusions-6212.22/fusions'), ('CCLE_fusions', 'gene-fusions-6212.22/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datafile/45eed812ab7f47c887cba3b585efcb89\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 85ba27b4ca544389a31162eeff674135 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/85ba27b4ca544389a31162eeff674135\n"
     ]
    }
   ],
   "source": [
    "blacklist = set()\n",
    "for val in ['internal', 'ibm', 'dmc', 'public']:\n",
    "    print('_________________________________________________')\n",
    "    print(val)\n",
    "    print('not present')\n",
    "    removed = set(prev[val]) - set(fusions.DepMap_ID)\n",
    "    print(removed)\n",
    "    print('removed for QC reasons')\n",
    "    print(failed)\n",
    "    print('removed')\n",
    "    removed = set(prevrna[val]) - set(fusions.DepMap_ID)\n",
    "    print(removed - set(rename.keys()))\n",
    "    missing = set(new[val]) - set(fusions.DepMap_ID)\n",
    "    blacklist = (set(fusions.DepMap_ID) - (prevrna[val] | set(new[val]))) | blacklist\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "    ## removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(fusions))\n",
    "    a = fusions[~fusions.DepMap_ID.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/fusions.csv', index=False)\n",
    "    print(len(filtered))\n",
    "    a= filtered[~filtered.DepMap_ID.isin(blacklist)]\n",
    "    print(len(a))\n",
    "    a.to_csv('temp/filtered_fusions.csv', index=False)\n",
    "\n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # uploading to taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_fusion[val],\n",
    "                     upload_file_path_dict={\n",
    "                         'temp/fusions.csv': 'TableCSV',\n",
    "                         'temp/filtered_fusions.csv': 'TableCSV',\n",
    "                         'temp/README': 'Raw'},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    # Internal Fusions\n",
    "\n",
    "    for more information, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    REMOVED FOR QC REASONS:\n",
    "    \"\"\"+str(failed)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # virtual datasets\n",
    "    AddToVirtual(virtual[val], taiga_fusion[val], files=[('CCLE_fusions_unfiltered', 'fusions'),('CCLE_fusions', 'filtered_fusions')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T13:55:10.030697Z",
     "start_time": "2021-01-22T13:54:57.965327Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_fusions_unfiltered', 'gene-fusions-8b7a.20/fusions'), ('CCLE_fusions', 'gene-fusions-8b7a.20/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datafile/dca7ae8e8142431897ee0d8e4e4377ed\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id ae82bea2f4e3457191d774c98da4c5fd created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/ae82bea2f4e3457191d774c98da4c5fd\n"
     ]
    }
   ],
   "source": [
    "AddToVirtual('depmap-a0ab', \"gene-fusions-8b7a\", files=[('CCLE_fusions_unfiltered', 'fusions'),('CCLE_fusions', 'filtered_fusions')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "206px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
