{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro & Loading "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-05T13:08:16.348113Z",
     "start_time": "2021-01-05T13:08:15.741375Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1003\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1003\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1003\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1003' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n",
       "  var css_urls = [];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1003\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1003\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1003\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1003' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n  var css_urls = [];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1003\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import os.path\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "sys.path.insert(0, '..')\n",
    "\n",
    "from src.CCLE_postp_function import *\n",
    "from JKBio import Datanalytics as da \n",
    "from JKBio import TerraFunction as terra\n",
    "from JKBio import Helper as h\n",
    "from gsheets import Sheets\n",
    "from taigapy import TaigaClient\n",
    "import dalmatian as dm\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "from bokeh.plotting import *\n",
    "from bokeh.models import HoverTool\n",
    "from collections import OrderedDict\n",
    "from IPython.display import Image,display\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "tc = TaigaClient()\n",
    "output_notebook()\n",
    "\n",
    "my_id = '~/.client_secret.json'\n",
    "mystorage_id = \"~/.storage.json\"\n",
    "sheets = Sheets.from_files(my_id, mystorage_id)\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-05T16:40:15.677048Z",
     "start_time": "2021-01-05T16:40:15.545624Z"
    }
   },
   "outputs": [],
   "source": [
    "samplesetname = \"21Q1\"\n",
    "\n",
    "virtual= {}\n",
    "virtual['public'] = 'public-21q1-4b39'\n",
    "virtual['ibm'] = 'ibm-21q1-abd9'\n",
    "virtual['dmc'] = 'dmc-21q1-0e11'\n",
    "virtual['internal']='internal-21q1-4fc4'\n",
    "\n",
    "taiga_mutation = {}\n",
    "taiga_mutation['internal'] = \"depmap-mutation-calls-9be3\"\n",
    "taiga_mutation['ibm'] = \"\"\n",
    "taiga_mutation['dmc'] = \"depmap-mutation-calls-dfce\"\n",
    "taiga_mutation['public'] =\"depmap-mutation-calls-9a1a\"\n",
    "\n",
    "taiga_expression = {}\n",
    "taiga_expression['internal'] =\"depmap-rnaseq-expression-data-363a\"\n",
    "taiga_expression['ibm'] = \"\"\n",
    "taiga_expression['dmc'] = \"depmap-rnaseq-expression-data-80ef\"\n",
    "taiga_expression['public'] =\"depmap-rnaseq-expression-data-ccd0\"\n",
    "\n",
    "taiga_fusion = {}\n",
    "taiga_fusion['internal'] = \"gene-fusions-8b7a\"\n",
    "taiga_fusion['ibm'] = \"\"\n",
    "taiga_fusion['dmc'] = \"gene-fusions-375f\"\n",
    "taiga_fusion['public'] =\"gene-fusions-6212\"\n",
    "\n",
    "taiga_copynumber = {}\n",
    "taiga_copynumber['internal'] =\"depmap-wes-cn-data-81a7\"\n",
    "taiga_copynumber['ibm'] = \"\"\n",
    "taiga_copynumber['dmc'] = \"depmap-cn-data-9b9d\"\n",
    "taiga_copynumber['public'] = \"depmap-wes-cn-data-97cc\"\n",
    "\n",
    "prevname=\"20Q4\"\n",
    "prev_virtual = {}\n",
    "prev_virtual['public'] = 'public-20q4-a4b3'\n",
    "prev_virtual['dmc'] = 'ibm-20q4-269f'\n",
    "prev_virtual['ibm'] = 'dmc-20q4-fcf4'\n",
    "prev_virtual['internal'] = 'internal-20q4-2540'\n",
    "potential_list_url = \"https://docs.google.com/spreadsheets/d/1YuKEgZ1pFKRYzydvncQt9Y_BKToPlHP-oDB-0CAv3gE\"\n",
    "\n",
    "release = samplesetname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-05T17:15:45.972331Z",
     "start_time": "2021-01-05T17:15:44.215516Z"
    }
   },
   "outputs": [],
   "source": [
    "new = {}\n",
    "gsheets = sheets.get(potential_list_url).sheets[0].to_frame()\n",
    "new['internal'] = [i for i in gsheets['Internal'].values.tolist() if str(i) != \"nan\"]\n",
    "new['dmc'] = [i for i in gsheets['DMC'].values.tolist() if str(i) != \"nan\"]\n",
    "new['ibm'] = [i for i in gsheets['IBM'].values.tolist() if str(i) != \"nan\"]\n",
    "new['public'] = [i for i in gsheets['Public'].values.tolist() if str(i) != \"nan\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting what was released before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-05T16:40:45.805410Z",
     "start_time": "2021-01-05T16:40:40.315Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for val in ['internal', 'dmc', 'public']:\n",
    "    print(val)\n",
    "    prevmut[val] = set(tc.get(name=prev_virtual[val], file='CCLE_mutations').DepMap_ID)\n",
    "    prevrna[val] = set(tc.get(name=prev_virtual[val], file='CCLE_expression').index)\n",
    "    prevcn[val] = set(tc.get(name=prev_virtual[val], file='CCLE_segment_cn').DepMap_ID)\n",
    "    prev[val] = prevmut[val] | prevrna[val] | prevcn[val]\n",
    "    print('mismatch cn/mut')\n",
    "    print(prevmut[val] ^ prevcn[val])\n",
    "    print('mismatch rna+cn/mut')\n",
    "    print(prev[val] - prevmut[val])\n",
    "    print('mismatch mut+cn/rna')\n",
    "    print(prev[val] - prevrna[val])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## managing the readmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fatal: destination path 'depmap-release-readmes' already exists and is not an empty directory.\r\n"
     ]
    }
   ],
   "source": [
    "! cd .. && git clone https://github.com/broadinstitute/depmap-release-readmes.git && cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "remote: Enumerating objects: 27, done.\u001b[K\n",
      "remote: Counting objects: 100% (27/27), done.\u001b[K\n",
      "remote: Compressing objects: 100% (16/16), done.\u001b[K\n",
      "remote: Total 26 (delta 14), reused 19 (delta 9), pack-reused 0\u001b[K\n",
      "Unpacking objects: 100% (26/26), done.\n",
      "From https://github.com/broadinstitute/depmap-release-readmes\n",
      "   5e66713..32bbcb4  master     -> origin/master\n",
      "Updating 5e66713..32bbcb4\n",
      "Fast-forward\n",
      " .gitignore                   |   1 \u001b[32m+\u001b[m\n",
      " release-20q3/.DS_Store       | Bin \u001b[31m0\u001b[m -> \u001b[32m6148\u001b[m bytes\n",
      " release-20q3/dmc-20q3.txt    | 348 \u001b[32m+++++++++++++++++++++++++++++++++++++++++++\u001b[m\n",
      " release-20q3/public-20q3.txt | 342 \u001b[32m++++++++++++++++++++++++++++++++++++++++++\u001b[m\n",
      " 4 files changed, 691 insertions(+)\n",
      " create mode 100644 .gitignore\n",
      " create mode 100644 release-20q3/.DS_Store\n",
      " create mode 100644 release-20q3/dmc-20q3.txt\n",
      " create mode 100644 release-20q3/public-20q3.txt\n"
     ]
    }
   ],
   "source": [
    "! cd ../depmap-release-readmes && git pull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making public\n",
      "Making internal\n",
      "Making dmc\n",
      "[master c7fb45e] 20Q4\n",
      " 4 files changed, 1386 insertions(+)\n",
      " create mode 100644 release-20q3/internal-20q3.txt\n",
      " create mode 100644 release-20q4/dmc-20q4.txt\n",
      " create mode 100644 release-20q4/internal-20q4.txt\n",
      " create mode 100644 release-20q4/public-20q4.txt\n",
      "Counting objects: 6, done.\n",
      "Delta compression using up to 12 threads.\n",
      "Compressing objects: 100% (6/6), done.\n",
      "Writing objects: 100% (6/6), 6.32 KiB | 0 bytes/s, done.\n",
      "Total 6 (delta 3), reused 0 (delta 0)\n",
      "remote: Resolving deltas: 100% (3/3), completed with 2 local objects.\u001b[K\n",
      "To https://github.com/broadinstitute/depmap-release-readmes.git\n",
      "   32bbcb4..c7fb45e  master -> master\n"
     ]
    }
   ],
   "source": [
    "!cd ../depmap-release-readmes/ && python3 make_new_release.py $release && git add . && git commit -m $release && git push "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir temp/README/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already up-to-date.\n",
      "/home/jeremie/ccle_processing\n"
     ]
    }
   ],
   "source": [
    "! cd ../depmap-release-readmes && git pull && cp release-$release/* ../ccle_processing/readmes/ && cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOW UPDATE THE READMEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! git add . && git commit -m \"Omics: updating readmes to new release\" && git push"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Somatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (4,23,26,32,34,35,36,37) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "mutations = pd.read_csv(\"temp/wes_somatic_mutations_withlegacy_\"+release+\".csv\")\n",
    "damaging = pd.read_csv('temp/wes_somatic_mutations_boolmatrix_fordepmap_damaging_' + samplesetname + \".csv\", index_col=0)\n",
    "othercons = pd.read_csv('temp/wes_somatic_mutations_boolmatrix_fordepmap_othercons_' + samplesetname + \".csv\", index_col=0)\n",
    "othernoncons = pd.read_csv('temp/wes_somatic_mutations_boolmatrix_fordepmap_othernoncons_' + samplesetname + \".csv\", index_col=0)\n",
    "hotspot = pd.read_csv('temp/wes_somatic_mutations_boolmatrix_fordepmap_hotspot_' + samplesetname + '.csv', index_col=0)\n",
    "hotspot=hotspot.astype(int)\n",
    "damaging=damaging.astype(int)\n",
    "othercons=othercons.astype(int)\n",
    "othernoncons=othernoncons.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reverting to previous versions\n",
    "mutations = mutations[mutations.is_likely_immortalization!=True]\n",
    "mutations = mutations[['Hugo_Symbol', 'Entrez_Gene_Id', 'NCBI_Build', 'Chromosome',\n",
    "       'Start_position', 'End_position', 'Strand', 'Variant_Classification',\n",
    "       'Variant_Type', 'Reference_Allele', 'Tumor_Allele', 'dbSNP_RS',\n",
    "       'dbSNP_Val_Status', 'Genome_Change', 'Annotation_Transcript',\n",
    "       'DepMap_ID', 'cDNA_Change', 'Codon_Change', 'Protein_Change', 'isDeleterious',\n",
    "       'isTCGAhotspot', 'TCGAhsCnt', 'isCOSMIChotspot', 'COSMIChsCnt',\n",
    "       'ExAC_AF',\"Variant_annotation\", 'CGA_WES_AC', 'HC_AC',\n",
    "       'RD_AC', 'RNAseq_AC', 'SangerWES_AC', 'WGS_AC']].rename(columns={\"Tumor_Allele\":\"Tumor_Seq_Allele1\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = mutations[~mutations.DepMap_ID.isin([\"ACH-001714\",\n",
    "\"ACH-002709\",\n",
    "\"ACH-002874\",\n",
    "\"ACH-002875\",\n",
    "\"ACH-001189\",\n",
    "\"ACH-002303\",\n",
    "\"ACH-002315\",\n",
    "\"ACH-002341\"])]\n",
    "damaging = damaging[set(damaging.columns)-set([\"ACH-001714\",\n",
    "\"ACH-002709\",\n",
    "\"ACH-002874\",\n",
    "\"ACH-002875\",\n",
    "\"ACH-001189\",\n",
    "\"ACH-002303\",\n",
    "\"ACH-002315\",\n",
    "\"ACH-002341\"])]\n",
    "othercons = othercons[set(othercons.columns)-set([\"ACH-001714\",\n",
    "\"ACH-002709\",\n",
    "\"ACH-002874\",\n",
    "\"ACH-002875\",\n",
    "\"ACH-001189\",\n",
    "\"ACH-002303\",\n",
    "\"ACH-002315\",\n",
    "\"ACH-002341\"])]\n",
    "othernoncons = othernoncons[set(othernoncons.columns)-set([\"ACH-001714\",\n",
    "\"ACH-002709\",\n",
    "\"ACH-002874\",\n",
    "\"ACH-002875\",\n",
    "\"ACH-001189\",\n",
    "\"ACH-002303\",\n",
    "\"ACH-002315\",\n",
    "\"ACH-002341\"])]\n",
    "hotspot = hotspot[set(hotspot.columns)-set([\"ACH-001714\",\n",
    "\"ACH-002709\",\n",
    "\"ACH-002874\",\n",
    "\"ACH-002875\",\n",
    "\"ACH-001189\",\n",
    "\"ACH-002303\",\n",
    "\"ACH-002315\",\n",
    "\"ACH-002341\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('nott present')\n",
    "removed = set(prev[val]) - set(mutations.DepMap_ID)\n",
    "print(removed)\n",
    "print('removed')\n",
    "removed = set(prevmut[val]) - set(mutations.DepMap_ID)\n",
    "print(removed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "internal\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'new' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m-------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-95-daf4ce239c11>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"internal\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dmc\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"public\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mmissing\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDepMap_ID\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mblacklist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDepMap_ID\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mprev\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'missing'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'new' is not defined"
     ]
    }
   ],
   "source": [
    "for val in [\"internal\", \"dmc\", \"public\"]:\n",
    "    print(val)\n",
    "    missing = set(new[val]) - set(mutations.DepMap_ID)\n",
    "    blacklist = set(mutations.DepMap_ID) - (prev[val] | set(new[val]))\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "    a = len(mutations)\n",
    "    mutations = mutations[~mutations.DepMap_ID.isin(blacklist)]\n",
    "    print(a - len(mutations))\n",
    "    mutations.to_csv('temp/all_somatic_mutations_withlegacy.csv', index=False)\n",
    "    a = len(damaging.columns)\n",
    "    damaging = damaging[set(damaging.columns) -blacklist]\n",
    "    print(a - len(damaging.columns))\n",
    "    damaging.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv')\n",
    "    a = len(othercons.columns)\n",
    "    othercons = othercons[set(othercons.columns) -blacklist]\n",
    "    print(a - len(othercons.columns))\n",
    "    othercons.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv',)\n",
    "    a = len(othernoncons.columns)\n",
    "    othernoncons = othernoncons[set(othernoncons.columns) -blacklist]\n",
    "    print(a - len(othernoncons.columns))\n",
    "    othernoncons.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv',)\n",
    "    a = len(hotspot.columns)\n",
    "    hotspot = hotspot[set(hotspot.columns) -blacklist]\n",
    "    print(a - len(hotspot.columns))\n",
    "    hotspot.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv',)\n",
    "    # adding files\n",
    "    a = len(mutations)\n",
    "    mutations = mutations[~mutations.DepMap_ID.isin(blacklist)]\n",
    "    print(a - len(mutations))\n",
    "    mutations.to_csv('temp/all_somatic_mutations_withlegacy.csv', index=False)\n",
    "    a = len(damaging.columns)\n",
    "    damaging = damaging[set(damaging.columns) -blacklist]\n",
    "    print(a - len(damaging.columns))\n",
    "    damaging.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv')\n",
    "    a = len(othercons.columns)\n",
    "    othercons = othercons[set(othercons.columns) -blacklist]\n",
    "    print(a - len(othercons.columns))\n",
    "    othercons.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv',)\n",
    "    a = len(othernoncons.columns)\n",
    "    othernoncons = othernoncons[set(othernoncons.columns) -blacklist]\n",
    "    print(a - len(othernoncons.columns))\n",
    "    othernoncons.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv',)\n",
    "    a = len(hotspot.columns)\n",
    "    hotspot = hotspot[set(hotspot.columns) -blacklist]\n",
    "    print(a - len(hotspot.columns))\n",
    "    hotspot.to_csv('temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv',)\n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "    \n",
    "    # updating on taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_mutation[val],\n",
    "                     upload_file_path_dict={\n",
    "    'temp/all_somatic_mutations_withlegacy.csv': 'TableCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_damaging.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_othernoncons.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_othercons.csv': 'NumericMatrixCSV',\n",
    "    'temp/all_somatic_mutations_boolmatrix_fordepmap_hotspot.csv': 'NumericMatrixCSV',\n",
    "    'temp/README': 'Raw'},\n",
    "                     dataset_description=\"\"\"\n",
    "    #Mutations Omics:\n",
    "\n",
    "    for informations, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    BLACKLISTED:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # To add to a virtual dataset\n",
    "    AddToVirtual(virtual[val], taiga_mutation[val], [('CCLE_all_somatic_mutations_withlegacy', 'all_somatic_mutations_withlegacy'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'all_somatic_mutations_boolmatrix_fordepmap_damaging'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'all_somatic_mutations_boolmatrix_fordepmap_othernoncons'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'all_somatic_mutations_boolmatrix_fordepmap_othercons'),\n",
    "    ('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'all_somatic_mutations_boolmatrix_fordepmap_hotspot'), ('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_mutations', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_withlegacy'), ('all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_hotspot')]\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datafile/cf78b026ca274b6d8ae86c787012c606\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id e3cd3dafab9d4946b0ee6cfe179e56a3 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/e3cd3dafab9d4946b0ee6cfe179e56a3\n",
      "[('CCLE_mutations', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_withlegacy'), ('all_somatic_mutations_boolmatrix_fordepmap_damaging', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_damaging'), ('all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_othernoncons'), ('all_somatic_mutations_boolmatrix_fordepmap_othercons', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_othercons'), ('all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'depmap-mutation-calls-9be3.30/all_somatic_mutations_boolmatrix_fordepmap_hotspot')]\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datafile/cc3395ea162f47c4a7035ef510c5d877\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id b53a973b68e34705a05996b04208d8ff created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/b53a973b68e34705a05996b04208d8ff\n"
     ]
    }
   ],
   "source": [
    "# To add to a eternal dataset\n",
    "AddToVirtual('depmap-a0ab', taiga_mutation['internal'], [('CCLE_all_somatic_mutations_withlegacy', 'all_somatic_mutations_withlegacy'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_damaging', 'all_somatic_mutations_boolmatrix_fordepmap_damaging'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othernoncons', 'all_somatic_mutations_boolmatrix_fordepmap_othernoncons'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_othercons', 'all_somatic_mutations_boolmatrix_fordepmap_othercons'),\n",
    "('CCLE_all_somatic_mutations_boolmatrix_fordepmap_hotspot', 'all_somatic_mutations_boolmatrix_fordepmap_hotspot')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Copy Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn= pd.read_csv('temp/all_'+release+'_gene_cn.csv',index_col=0)\n",
    "segmentcn = pd.read_csv('temp/all_'+release+'_segment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genecn = genecn.apply(lambda x: (x**2)-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not present\n",
      "{'ACH-000033', 'ACH-000084', 'ACH-001189', 'ACH-002390', 'ACH-002396', 'ACH-001429', 'ACH-001131', 'ACH-001187', 'ACH-001011', 'ACH-001712', 'ACH-001393', 'ACH-002391', 'ACH-001109', 'ACH-001741', 'ACH-001316', 'ACH-002393', 'ACH-002341', 'ACH-002359', 'ACH-002394', 'ACH-000629', 'ACH-001108', 'ACH-002395', 'ACH-002709', 'ACH-002303', 'ACH-002315', 'ACH-001743'}\n",
      "removed\n",
      "{'ACH-001189', 'ACH-002315', 'ACH-002303', 'ACH-002341', 'ACH-002359'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "16 {'ACH-001434', 'ACH-001707', 'ACH-001705', 'ACH-002013', 'ACH-001759', 'ACH-001756', 'ACH-001758', 'ACH-001553', 'ACH-002476', 'ACH-001686', 'ACH-001760', 'ACH-001046', 'ACH-001227', 'ACH-001828', 'ACH-002138', 'ACH-002055'}\n"
     ]
    }
   ],
   "source": [
    "for val in ['internal','dmc','public']:\n",
    "    print('not present')\n",
    "    removed = set(prev[val]) - set(segmentcn.DepMap_ID)\n",
    "    print(removed)\n",
    "    print('removed')\n",
    "    removed = set(prevcn[val]) - set(segmentcn.DepMap_ID)\n",
    "    print(removed)\n",
    "    missing = set(new[val]) - set(segmentcn.DepMap_ID)\n",
    "    blacklist = set(segmentcn.DepMap_ID) - (prev[val] | set(new[val]))\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "    ## for segment removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(segmentcn))\n",
    "    segmentcn = segmentcn[~segmentcn.DepMap_ID.isin(blacklist)]\n",
    "    print(len(segmentcn))\n",
    "    segmentcn.to_csv('temp/all_merged_segments.csv', index=False)\n",
    "    print(len(genecn))\n",
    "    genecn = genecn[~genecn.index.isin(blacklist)]\n",
    "    print(len(genecn))\n",
    "    genecn.to_csv('temp/all_merged_genes_cn.csv')\n",
    "    \n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # Add to Taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_copynumber[val], \n",
    "                      upload_file_path_dict={\n",
    "                        'temp/all_merged_genes_cn.csv': 'NumericMatrixCSV',\n",
    "                        'temp/all_merged_segments.csv': 'TableCSV',\n",
    "                        'temp/README': 'raw'},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    #Copy Number Omics:\n",
    "\n",
    "    for informations, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "    # To add to a virtual dataset\n",
    "    AddToVirtual(virtual[val], taiga_copynumber[val], [('CCLE_all_merged_genes_cn', 'all_merged_genes_cn'), ('CCLE_all_merged_segments','all_merged_segments'), ('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To add to a eternal dataset\n",
    "AddToVirtual('depmap-a0ab', taiga_copynumber['internal'], [('CCLE_all_merged_genes_cn', 'all_merged_genes_cn'), ('CCLE_all_merged_segments', 'all_merged_segments')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts_tpm = pd.read_csv('temp/expression_' + release + '_transcripts_tpm.csv',index_col=0)\n",
    "genes_tpm = pd.read_csv('temp/expression_' + release + '_genes_tpm.csv',index_col=0)\n",
    "genes_expected_count = pd.read_csv('temp/expression_' + release + '_genes_expected_count.csv',index_col=0)\n",
    "proteincoding_genes_expected_count = pd.read_csv('temp/expression_' + release + '_proteincoding_genes_expected_count.csv',index_col=0)\n",
    "proteincoding_genes_tpm = pd.read_csv('temp/expression_' + release + '_proteincoding_genes_tpm.csv',index_col=0)\n",
    "transcripts_expected_count = pd.read_csv('temp/expression_' + release + '_transcripts_expected_count.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "cm = \"anvil-datastorage/AnVIL_GTEx_V8_hg38\"\n",
    "bm = \"broad-firecloud-ccle/DepMap_WGS_PON_hg38\"\n",
    "import dalmatian as dm\n",
    "wm = dm.WorkspaceManager(cm)\n",
    "sam = wm.get_participants()\n",
    "sam = sam[(sam.sex==\"Female\") & (~sam.wgs_cram_file.isna())]\n",
    "sam.index.name=\"sample_id\"\n",
    "sam['participant_id']=sam.index\n",
    "dm.WorkspaceManager(bm).disable_hound().upload_samples(sam[['wgs_cram_file','wgs_cram_index', \"participant_id\"]])\n",
    "dm.WorkspaceManager(bm).update_sample_set('WGS_XX_GTEX', sam.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully imported 296 participants.\n",
      "Successfully imported 296 samples.\n",
      "Successfully imported 1 sample sets:\n",
      "  * WGS_XX_GTEX (296 samples)\n"
     ]
    }
   ],
   "source": [
    "dm.WorkspaceManager(bm).disable_hound().upload_samples(sam[['wgs_cram_file','wgs_cram_index', \"participant_id\"]])\n",
    "dm.WorkspaceManager(bm).update_sample_set('WGS_XX_GTEX', sam.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "wm = dm.WorkspaceManager(bm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample set \"WGS_XX_GTEX\" (0 samples) successfully updated.\n",
      "Sample(s) ['GTEX-1117F', 'GTEX-1122O', 'GTEX-1128S', 'GTEX-113JC', 'GTEX-11DXX', 'GTEX-11EM3', 'GTEX-11EMC', 'GTEX-11GSP', 'GTEX-11I78', 'GTEX-11ILO', 'GTEX-11P81', 'GTEX-11TTK', 'GTEX-11UD1', 'GTEX-11VI4', 'GTEX-11XUK', 'GTEX-11ZTS', 'GTEX-11ZTT', 'GTEX-11ZVC', 'GTEX-1211K', 'GTEX-1269C', 'GTEX-12WS9', 'GTEX-12WSB', 'GTEX-12WSD', 'GTEX-12WSG', 'GTEX-12WSJ', 'GTEX-12WSK', 'GTEX-12ZZX', 'GTEX-13113', 'GTEX-1313W', 'GTEX-131XG', 'GTEX-131XW', 'GTEX-131YS', 'GTEX-132AR', 'GTEX-133LE', 'GTEX-1399S', 'GTEX-1399U', 'GTEX-139D8', 'GTEX-139T4', 'GTEX-13CF3', 'GTEX-13D11', 'GTEX-13FH7', 'GTEX-13FTX', 'GTEX-13FTY', 'GTEX-13JUV', 'GTEX-13N11', 'GTEX-13NZ8', 'GTEX-13O3O', 'GTEX-13OVI', 'GTEX-13OVJ', 'GTEX-13PL6', 'GTEX-13PL7', 'GTEX-13PLJ', 'GTEX-13PVR', 'GTEX-13QBU', 'GTEX-13QIC', 'GTEX-13QJC', 'GTEX-13S7M', 'GTEX-13SLX', 'GTEX-13U4I', 'GTEX-13VXT', 'GTEX-13W3W', 'GTEX-13X6H', 'GTEX-13X6K', 'GTEX-145MI', 'GTEX-146FH', 'GTEX-146FR', 'GTEX-147F3', 'GTEX-148VI', 'GTEX-14A5I', 'GTEX-14AS3', 'GTEX-14BIM', 'GTEX-14BIN', 'GTEX-14BMU', 'GTEX-14DAQ', 'GTEX-14H4A', 'GTEX-14JG6', 'GTEX-14LLW', 'GTEX-14LZ3', 'GTEX-14PHW', 'GTEX-14PHY', 'GTEX-14PJ5', 'GTEX-14PJ6', 'GTEX-14PJM', 'GTEX-14PK6', 'GTEX-14PKU', 'GTEX-14PKV', 'GTEX-14PN3', 'GTEX-14PQA', 'GTEX-14XAO', 'GTEX-15CHC', 'GTEX-15DCD', 'GTEX-15DCZ', 'GTEX-15DDE', 'GTEX-15DYW', 'GTEX-15DZA', 'GTEX-15EOM', 'GTEX-15ER7', 'GTEX-15ETS', 'GTEX-15FZZ', 'GTEX-15SB6', 'GTEX-15SDE', 'GTEX-15SHU', 'GTEX-15UF6', 'GTEX-16A39', 'GTEX-16BQI', 'GTEX-16MT8', 'GTEX-16MT9', 'GTEX-16NGA', 'GTEX-16NPV', 'GTEX-17F98', 'GTEX-17F9Y', 'GTEX-17JCI', 'GTEX-183WM', 'GTEX-18A7A', 'GTEX-18D9A', 'GTEX-19HZE', 'GTEX-1A32A', 'GTEX-1A3MW', 'GTEX-1A8FM', 'GTEX-1AMEY', 'GTEX-1AMFI', 'GTEX-1B932', 'GTEX-1B933', 'GTEX-1B97J', 'GTEX-1C475', 'GTEX-1CAMS', 'GTEX-1CB4H', 'GTEX-1EWIQ', 'GTEX-1F48J', 'GTEX-1F52S', 'GTEX-1F5PL', 'GTEX-1F6RS', 'GTEX-1F7RK', 'GTEX-1F88E', 'GTEX-1F88F', 'GTEX-1GF9U', 'GTEX-1GF9X', 'GTEX-1GMR8', 'GTEX-1GN1U', 'GTEX-1GN1W', 'GTEX-1GZ4H', 'GTEX-1GZHY', 'GTEX-1I4MK', 'GTEX-1ICG6', 'GTEX-1IDJE', 'GTEX-1JMOU', 'GTEX-1JMQL', 'GTEX-1JN1M', 'GTEX-1JN76', 'GTEX-1K9T9', 'GTEX-1KANC', 'GTEX-1LGOU', 'GTEX-1LGRB', 'GTEX-1LH75', 'GTEX-1LKK1', 'GTEX-1LNCM', 'GTEX-1LVAM', 'GTEX-1LVAN', 'GTEX-1MA7W', 'GTEX-1MCC2', 'GTEX-1MCYP', 'GTEX-1N7R6', 'GTEX-1NT2E', 'GTEX-1NUQO', 'GTEX-1NV5F', 'GTEX-1OJC4', 'GTEX-1PBJI', 'GTEX-1PIEJ', 'GTEX-1POEN', 'GTEX-1PPH6', 'GTEX-1QCLZ', 'GTEX-1QEPI', 'GTEX-1QP28', 'GTEX-1QP67', 'GTEX-1QPFJ', 'GTEX-1R46S', 'GTEX-1R9JW', 'GTEX-1R9PN', 'GTEX-1R9PO', 'GTEX-1RNTQ', 'GTEX-1S5ZU', 'GTEX-1S82P', 'GTEX-1S82U', 'GTEX-1S82Z', 'GTEX-N7MT', 'GTEX-NL3G', 'GTEX-NPJ7', 'GTEX-O5YV', 'GTEX-OHPK', 'GTEX-OHPL', 'GTEX-OHPN', 'GTEX-OXRK', 'GTEX-OXRO', 'GTEX-OXRP', 'GTEX-P4PP', 'GTEX-P4QT', 'GTEX-P78B', 'GTEX-PLZ4', 'GTEX-POMQ', 'GTEX-PWCY', 'GTEX-PWN1', 'GTEX-PWO3', 'GTEX-PWOO', 'GTEX-PX3G', 'GTEX-Q2AG', 'GTEX-Q734', 'GTEX-QCQG', 'GTEX-QDT8', 'GTEX-QVJO', 'GTEX-QVUS', 'GTEX-R3RS', 'GTEX-R53T', 'GTEX-R55G', 'GTEX-RN5K', 'GTEX-RNOR', 'GTEX-RTLS', 'GTEX-RU1J', 'GTEX-RU72', 'GTEX-RWS6', 'GTEX-S32W', 'GTEX-S341', 'GTEX-S4P3', 'GTEX-S4UY', 'GTEX-S7SF', 'GTEX-SE5C', 'GTEX-SN8G', 'GTEX-T2IS', 'GTEX-T5JW', 'GTEX-T6MO', 'GTEX-TML8', 'GTEX-TMMY', 'GTEX-TSE9', 'GTEX-U3ZN', 'GTEX-UJHI', 'GTEX-UPIC', 'GTEX-VJWN', 'GTEX-W5WG', 'GTEX-W5X1', 'GTEX-WEY5', 'GTEX-WI4N', 'GTEX-WRHK', 'GTEX-WRHU', 'GTEX-WWYW', 'GTEX-WXYG', 'GTEX-WYBS', 'GTEX-WYVS', 'GTEX-X15G', 'GTEX-X4EO', 'GTEX-X4EP', 'GTEX-X638', 'GTEX-X8HC', 'GTEX-XMD2', 'GTEX-XMD3', 'GTEX-XOT4', 'GTEX-XUJ4', 'GTEX-XUW1', 'GTEX-XUZC', 'GTEX-XV7Q', 'GTEX-XYKS', 'GTEX-Y114', 'GTEX-Y3IK', 'GTEX-Y5LM', 'GTEX-Y5V5', 'GTEX-Y8LW', 'GTEX-YB5K', 'GTEX-YBZK', 'GTEX-YFC4', 'GTEX-YJ8O', 'GTEX-ZAJG', 'GTEX-ZAK1', 'GTEX-ZC5H', 'GTEX-ZE7O', 'GTEX-ZF29', 'GTEX-ZF2S', 'GTEX-ZF3C', 'GTEX-ZGAY', 'GTEX-ZLV1', 'GTEX-ZLWG', 'GTEX-ZP4G', 'GTEX-ZPCL', 'GTEX-ZPIC', 'GTEX-ZQG8', 'GTEX-ZTPG', 'GTEX-ZV68', 'GTEX-ZV6S', 'GTEX-ZVE1', 'GTEX-ZVT2', 'GTEX-ZVT3', 'GTEX-ZVT4', 'GTEX-ZVZQ', 'GTEX-ZXES', 'GTEX-ZYFG', 'GTEX-ZYVF', 'GTEX-ZYWO', 'GTEX-ZYY3', 'GTEX-ZZPU'] successfully deleted.\n"
     ]
    }
   ],
   "source": [
    "wm.delete_sample(sam.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "#putting it back to what it was before\n",
    "\n",
    "# log transforming tpm data\n",
    "transcripts_tpm=transcripts_tpm.apply(lambda x: np.log2(x+1))\n",
    "genes_tpm=genes_tpm.apply(lambda x: np.log2(x+1))\n",
    "proteincoding_genes_tpm=proteincoding_genes_tpm.apply(lambda x: np.log2(x+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "store -r rename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not present\n",
      "{'ACH-002162', 'ACH-001101', 'ACH-002116', 'ACH-002152', 'ACH-002247', 'ACH-002221', 'ACH-002105', 'ACH-001091', 'ACH-002220', 'ACH-001044', 'ACH-001047', 'ACH-002100', 'ACH-001224', 'ACH-002316', 'ACH-002185', 'ACH-002356', 'ACH-002374', 'ACH-002348', 'ACH-002106', 'ACH-002289', 'ACH-002096', 'ACH-002224', 'ACH-002341', 'ACH-002394', 'ACH-001137', 'ACH-002344', 'ACH-001130', 'ACH-002092', 'ACH-002400', 'ACH-002161', 'ACH-001233', 'ACH-002155', 'ACH-001364', 'ACH-002258', 'ACH-002243', 'ACH-002306', 'ACH-002300', 'ACH-001704', 'ACH-001092', 'ACH-001093', 'ACH-002168', 'ACH-002357', 'ACH-002244', 'ACH-001182', 'ACH-002304', 'ACH-002313', 'ACH-002122', 'ACH-001107', 'ACH-002358', 'ACH-002359', 'ACH-002192', 'ACH-002298', 'ACH-002040', 'ACH-002178', 'ACH-000047', 'ACH-002099', 'ACH-002102', 'ACH-001090', 'ACH-002144', 'ACH-002146', 'ACH-002274', 'ACH-002233', 'ACH-002275', 'ACH-002094', 'ACH-001002', 'ACH-002135', 'ACH-002252', 'ACH-002210', 'ACH-001121', 'ACH-002153', 'ACH-001042', 'ACH-001066', 'ACH-001043', 'ACH-001171', 'ACH-001205', 'ACH-002287', 'ACH-001362', 'ACH-001089', 'ACH-002267', 'ACH-002234', 'ACH-002225', 'ACH-002389', 'ACH-002209', 'ACH-002133', 'ACH-002354', 'ACH-001150', 'ACH-000309', 'ACH-002222', 'ACH-002379', 'ACH-001018', 'ACH-002238', 'ACH-002058', 'ACH-002151', 'ACH-001131', 'ACH-002261', 'ACH-001230', 'ACH-002141', 'ACH-002241', 'ACH-002214', 'ACH-002242', 'ACH-002292', 'ACH-001136', 'ACH-002154', 'ACH-002368', 'ACH-002212', 'ACH-002199', 'ACH-002149', 'ACH-002171', 'ACH-002366', 'ACH-002363', 'ACH-002175', 'ACH-002265', 'ACH-002390', 'ACH-001087', 'ACH-002142', 'ACH-001146', 'ACH-001489', 'ACH-002131', 'ACH-002235', 'ACH-002169', 'ACH-002193', 'ACH-001071', 'ACH-001126', 'ACH-002239', 'ACH-002401', 'ACH-002215', 'ACH-001225', 'ACH-001591', 'ACH-002104', 'ACH-001365', 'ACH-002378', 'ACH-002140', 'ACH-001282', 'ACH-002174', 'ACH-002173', 'ACH-002095', 'ACH-002020', 'ACH-002180', 'ACH-002386', 'ACH-002196', 'ACH-002107', 'ACH-001331', 'ACH-002299', 'ACH-002362', 'ACH-002277', 'ACH-002091', 'ACH-002314', 'ACH-002217', 'ACH-001081', 'ACH-001208', 'ACH-002347', 'ACH-001203', 'ACH-002112', 'ACH-001063', 'ACH-002382', 'ACH-002395', 'ACH-002167', 'ACH-002218', 'ACH-002124', 'ACH-002709', 'ACH-002346', 'ACH-002268', 'ACH-002282', 'ACH-002343', 'ACH-002103', 'ACH-002291', 'ACH-001606', 'ACH-002051', 'ACH-002111', 'ACH-002373', 'ACH-002108', 'ACH-002213', 'ACH-001599', 'ACH-000795', 'ACH-001680', 'ACH-002270', 'ACH-000426', 'ACH-002190', 'ACH-001452', 'ACH-002191', 'ACH-002114', 'ACH-002276', 'ACH-002369', 'ACH-002364', 'ACH-002177', 'ACH-001358', 'ACH-002198', 'ACH-002083', 'ACH-002189', 'ACH-001143', 'ACH-002109', 'ACH-001079', 'ACH-002113', 'ACH-002123', 'ACH-002201', 'ACH-001094', 'ACH-001206', 'ACH-002301', 'ACH-001088', 'ACH-002342', 'ACH-002115', 'ACH-002294', 'ACH-002043', 'ACH-002281', 'ACH-002132', 'ACH-002365', 'ACH-002097', 'ACH-001108', 'ACH-002337', 'ACH-001338', 'ACH-002119', 'ACH-002253', 'ACH-002383', 'ACH-002052', 'ACH-002245', 'ACH-002286', 'ACH-002396', 'ACH-002305', 'ACH-001544', 'ACH-001249', 'ACH-002145', 'ACH-002296', 'ACH-001957', 'ACH-002371', 'ACH-002188', 'ACH-001125', 'ACH-002361', 'ACH-002339', 'ACH-002272', 'ACH-002375', 'ACH-002204', 'ACH-002207', 'ACH-002353', 'ACH-001214', 'ACH-001639', 'ACH-002336', 'ACH-001045', 'ACH-002285', 'ACH-002110', 'ACH-001024', 'ACH-001955', 'ACH-001198', 'ACH-002143', 'ACH-002117', 'ACH-002216', 'ACH-002269', 'ACH-002126', 'ACH-001162', 'ACH-001417', 'ACH-000979', 'ACH-002194', 'ACH-001641', 'ACH-002392', 'ACH-001199', 'ACH-002231', 'ACH-001039', 'ACH-002157', 'ACH-002251', 'ACH-002170', 'ACH-002317', 'ACH-002208', 'ACH-002372', 'ACH-001956', 'ACH-002098', 'ACH-001502', 'ACH-002179', 'ACH-002303', 'ACH-002240', 'ACH-002351', 'ACH-002257', 'ACH-002164', 'ACH-001189', 'ACH-002150', 'ACH-002118', 'ACH-000727', 'ACH-002355', 'ACH-002148', 'ACH-002136', 'ACH-001958', 'ACH-002309', 'ACH-002246', 'ACH-002279', 'ACH-002205', 'ACH-002250', 'ACH-002128', 'ACH-001359', 'ACH-002159', 'ACH-002349', 'ACH-002090', 'ACH-002284', 'ACH-002377', 'ACH-002223', 'ACH-002308', 'ACH-001234', 'ACH-002248', 'ACH-001017', 'ACH-001049', 'ACH-002262', 'ACH-002160', 'ACH-002255', 'ACH-002121', 'ACH-002387', 'ACH-002397', 'ACH-001015', 'ACH-002307', 'ACH-002226', 'ACH-002352', 'ACH-001383', 'ACH-002184', 'ACH-002130', 'ACH-002293', 'ACH-002302', 'ACH-002206', 'ACH-002237', 'ACH-002256', 'ACH-002229', 'ACH-002335', 'ACH-002263', 'ACH-001738', 'ACH-002340', 'ACH-002370', 'ACH-002228', 'ACH-002384', 'ACH-002166', 'ACH-000992', 'ACH-002163', 'ACH-002388', 'ACH-001037', 'ACH-002129', 'ACH-002134', 'ACH-002211', 'ACH-001065', 'ACH-002399', 'ACH-001138', 'ACH-002182', 'ACH-002176', 'ACH-002367', 'ACH-002147', 'ACH-002376', 'ACH-001363', 'ACH-002165', 'ACH-001293', 'ACH-002288', 'ACH-002295', 'ACH-002249', 'ACH-002183', 'ACH-002278', 'ACH-002089', 'ACH-002260', 'ACH-002266', 'ACH-002181', 'ACH-002345', 'ACH-001011', 'ACH-002197', 'ACH-001512', 'ACH-002195', 'ACH-002350', 'ACH-002290', 'ACH-002200', 'ACH-002310', 'ACH-002283', 'ACH-001021', 'ACH-002264', 'ACH-002236', 'ACH-002315', 'ACH-002127', 'ACH-002202', 'ACH-002312', 'ACH-002230', 'ACH-002259', 'ACH-001187', 'ACH-002311', 'ACH-002186', 'ACH-002232', 'ACH-001064', 'ACH-001127', 'ACH-002271', 'ACH-002338', 'ACH-001767', 'ACH-002381', 'ACH-001188', 'ACH-002280', 'ACH-002254', 'ACH-002360', 'ACH-002139', 'ACH-002101', 'ACH-001072', 'ACH-001355', 'ACH-001175', 'ACH-001118', 'ACH-001350', 'ACH-002380', 'ACH-002120', 'ACH-002125', 'ACH-001490', 'ACH-002187', 'ACH-001000', 'ACH-002093', 'ACH-002391', 'ACH-001023', 'ACH-002273', 'ACH-002393', 'ACH-002077', 'ACH-002156', 'ACH-002227', 'ACH-002297'}\n",
      "removed for QC reasons\n",
      "{'PEDS117'}\n",
      "removed\n",
      "{'ACH-001502', 'ACH-001143', 'ACH-000795', 'ACH-002709', 'ACH-000309', 'ACH-000426', 'ACH-001293', 'ACH-000979', 'ACH-001512', 'ACH-000992', 'ACH-000047', 'ACH-000727', 'ACH-001205'}\n",
      "missing\n",
      "set()\n",
      "blacklist\n",
      "11 {'ACH-001434', 'ACH-001759', 'ACH-001756', 'ACH-001758', 'ACH-001686', 'ACH-002476', 'ACH-001227', 'ACH-001760', 'ACH-001046', 'ACH-002047', 'ACH-002055'}\n"
     ]
    }
   ],
   "source": [
    "for val in ['internal','dmc','public']:\n",
    "    print('not present')\n",
    "    removed = set(prev[val]) - set(genes_tpm.index)\n",
    "    print(removed)\n",
    "    print('removed for QC reasons')\n",
    "    print(set(rename.keys()))\n",
    "    print('removed')\n",
    "    removed = set(prevrna[val]) - set(genes_tpm.index)\n",
    "    print(removed - set(rename.keys()))\n",
    "    missing = set(new[val]) - set(genes_tpm.index)\n",
    "    blacklist = set(genes_tpm.index) - (prevrna[val] | set(new[val]))\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "\n",
    "    ## removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(genes_expected_count))\n",
    "    genes_expected_count = genes_expected_count[~genes_expected_count.index.isin(blacklist)]\n",
    "    print(len(genes_expected_count))\n",
    "    genes_expected_count.to_csv('temp/expression_genes_expected_count.csv')\n",
    "    print(len(genes_tpm))\n",
    "    genes_tpm = genes_tpm[~genes_tpm.index.isin(blacklist)]\n",
    "    print(len(genes_tpm))\n",
    "    genes_tpm.to_csv('temp/expression_genes_tpm.csv')\n",
    "    print(len(proteincoding_genes_tpm))\n",
    "    proteincoding_genes_tpm = proteincoding_genes_tpm[~proteincoding_genes_tpm.index.isin(blacklist)]\n",
    "    print(len(proteincoding_genes_tpm))\n",
    "    proteincoding_genes_tpm.to_csv('temp/expression_proteincoding_genes_tpm.csv')\n",
    "    print(len(transcripts_tpm))\n",
    "    transcripts_tpm = transcripts_tpm[~transcripts_tpm.index.isin(blacklist)]\n",
    "    print(len(transcripts_tpm))\n",
    "    transcripts_tpm.to_csv('temp/expression_transcripts_tpm.csv')\n",
    "    print(len(proteincoding_genes_expected_count))\n",
    "    proteincoding_genes_expected_count = proteincoding_genes_expected_count[~proteincoding_genes_expected_count.index.isin(blacklist)]\n",
    "    print(len(proteincoding_genes_expected_count))\n",
    "    proteincoding_genes_expected_count.to_csv('temp/expression_proteincoding_genes_expected_count.csv')\n",
    "    print(len(transcripts_expected_count))\n",
    "    transcripts_expected_count = transcripts_expected_count[~transcripts_expected_count.index.isin(blacklist)]\n",
    "    print(len(transcripts_expected_count))\n",
    "    transcripts_expected_count.to_csv('temp/expression_transcripts_expected_count.csv')\n",
    "    \n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # adding to taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_rna[val],\n",
    "                     upload_file_path_dict={\n",
    "                       'temp/expression_genes_expected_count.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_transcripts_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_genes_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_proteincoding_genes_tpm.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_proteincoding_genes_expected_count.csv': 'NumericMatrixCSV',\n",
    "                       'temp/expression_transcripts_expected_count.csv': 'NumericMatrixCSV',\n",
    "                     \"temp/README\": \"README\"},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    # INTERNAL RNA\n",
    "\n",
    "    for information, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    REMOVED FOR QC REASONS:\n",
    "    \"\"\"+str(rename)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # add to virtual \n",
    "    AddToVirtual(virtual[val], taiga_rna[val], files=[\n",
    "    ('CCLE_expression_genes_tpm', 'expression_genes_tpm'), \n",
    "    ('CCLE_expression_transcripts_tpm', 'expression_transcripts_tpm'),\n",
    "    ('CCLE_expression_genes_expected_count', 'expression_genes_expected_count'),\n",
    "    ('CCLE_expression_proteincoding_genes_tpm', 'expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'expression_proteincoding_genes_expected_count'),('CCLE_expression_transcripts_expected_count', 'expression_transcripts_expected_count'), ('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_expression_full', 'depmap-rnaseq-expression-data-363a.32/expression_genes_tpm'), ('CCLE_RNAseq_transcripts', 'depmap-rnaseq-expression-data-363a.32/expression_transcripts_tpm'), ('CCLE_RNAseq_reads', 'depmap-rnaseq-expression-data-363a.32/expression_genes_expected_count'), ('CCLE_expression', 'depmap-rnaseq-expression-data-363a.32/expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'depmap-rnaseq-expression-data-363a.32/expression_proteincoding_genes_expected_count'), ('CCLE_expression_transcripts_expected_count', 'depmap-rnaseq-expression-data-363a.32/expression_transcripts_expected_count')]\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datafile/766a8a1003394a1da6d478547f106c23\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id cd248c30d945490eab91db75ee0dd4ac created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/cd248c30d945490eab91db75ee0dd4ac\n",
      "[('CCLE_expression_full', 'depmap-rnaseq-expression-data-363a.32/expression_genes_tpm'), ('CCLE_RNAseq_transcripts', 'depmap-rnaseq-expression-data-363a.32/expression_transcripts_tpm'), ('CCLE_RNAseq_reads', 'depmap-rnaseq-expression-data-363a.32/expression_genes_expected_count'), ('CCLE_expression', 'depmap-rnaseq-expression-data-363a.32/expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'depmap-rnaseq-expression-data-363a.32/expression_proteincoding_genes_expected_count'), ('CCLE_expression_transcripts_expected_count', 'depmap-rnaseq-expression-data-363a.32/expression_transcripts_expected_count')]\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hitting https://cds.team/taiga/api/datafile/c3b3f1a82b7a4e29b40bf4cd2b64b96f\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 1685b01832ce49cf8b58ec9c253ec01b created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/1685b01832ce49cf8b58ec9c253ec01b\n"
     ]
    }
   ],
   "source": [
    "AddToVirtual('depmap-a0ab', taiga_rna['internal'], files=[\n",
    "('CCLE_expression_genes_tpm', 'expression_genes_tpm'), \n",
    "('CCLE_expression_transcripts_tpm', 'expression_transcripts_tpm'),\n",
    "('CCLE_expression_genes_expected_count', 'expression_genes_expected_count'),\n",
    "('CCLE_expression_proteincoding_genes_tpm', 'expression_proteincoding_genes_tpm'), ('CCLE_expression_proteincoding_genes_expected_count', 'expression_proteincoding_genes_expected_count'),('CCLE_expression_transcripts_expected_count', 'expression_transcripts_expected_count')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'temp/unfiltered_fusions_20Q4.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m-------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-55-e248d4212cc5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfusions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'temp/unfiltered_fusions_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mrelease\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mfiltered\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'temp/fusions_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mrelease\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    684\u001b[0m     )\n\u001b[1;32m    685\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 686\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    687\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    688\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 452\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    453\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    454\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    944\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    945\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 946\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    947\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    948\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1176\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1177\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1178\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1179\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1180\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   2006\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2007\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2008\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2009\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2010\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'temp/unfiltered_fusions_20Q4.csv'"
     ]
    }
   ],
   "source": [
    "fusions=pd.read_csv('temp/unfiltered_fusions_'+release+'.csv')\n",
    "filtered=pd.read_csv('temp/fusions_'+release+'.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not present\n",
      "{'ACH-002196', 'ACH-002205', 'ACH-002289', 'ACH-001064', 'ACH-001641', 'ACH-002127', 'ACH-002344', 'ACH-002384', 'ACH-002383', 'ACH-002157', 'ACH-001037', 'ACH-001143', 'ACH-002216', 'ACH-001680', 'ACH-000727', 'ACH-002351', 'ACH-002207', 'ACH-002114', 'ACH-002300', 'ACH-000795', 'ACH-002377', 'ACH-002204', 'ACH-002360', 'ACH-002292', 'ACH-002058', 'ACH-002104', 'ACH-002314', 'ACH-002183', 'ACH-002213', 'ACH-002168', 'ACH-002382', 'ACH-002260', 'ACH-001208', 'ACH-002095', 'ACH-002200', 'ACH-002269', 'ACH-002387', 'ACH-002375', 'ACH-002129', 'ACH-001199', 'ACH-002312', 'ACH-000561', 'ACH-002282', 'ACH-002395', 'ACH-001338', 'ACH-002232', 'ACH-002267', 'ACH-002190', 'ACH-001502', 'ACH-001093', 'ACH-002336', 'ACH-002281', 'ACH-001137', 'ACH-002199', 'ACH-002208', 'ACH-002185', 'ACH-002276', 'ACH-002374', 'ACH-002396', 'ACH-001383', 'ACH-002237', 'ACH-001151', 'ACH-002119', 'ACH-002136', 'ACH-002359', 'ACH-001182', 'ACH-002399', 'ACH-002106', 'ACH-001187', 'ACH-002394', 'ACH-002180', 'ACH-001045', 'ACH-002291', 'ACH-002230', 'ACH-002193', 'ACH-001089', 'ACH-001331', 'ACH-002401', 'ACH-002147', 'ACH-002242', 'ACH-002361', 'ACH-001205', 'ACH-002201', 'ACH-002191', 'ACH-002143', 'ACH-001203', 'ACH-002290', 'ACH-002188', 'ACH-001018', 'ACH-002294', 'ACH-002247', 'ACH-002167', 'ACH-002391', 'ACH-001125', 'ACH-001591', 'ACH-001118', 'ACH-002159', 'ACH-002197', 'ACH-002253', 'ACH-002102', 'ACH-001023', 'ACH-002238', 'ACH-002182', 'ACH-002198', 'ACH-002175', 'ACH-001599', 'ACH-002250', 'ACH-002284', 'ACH-002091', 'ACH-001233', 'ACH-002299', 'ACH-001092', 'ACH-002369', 'ACH-002123', 'ACH-002305', 'ACH-002141', 'ACH-001767', 'ACH-002341', 'ACH-002350', 'ACH-001011', 'ACH-002296', 'ACH-002169', 'ACH-001101', 'ACH-002128', 'ACH-002235', 'ACH-002093', 'ACH-002367', 'ACH-001225', 'ACH-001127', 'ACH-002125', 'ACH-002225', 'ACH-002297', 'ACH-002224', 'ACH-002315', 'ACH-001044', 'ACH-002266', 'ACH-002400', 'ACH-001350', 'ACH-001648', 'ACH-002380', 'ACH-002231', 'ACH-002096', 'ACH-002274', 'ACH-002379', 'ACH-002135', 'ACH-002043', 'ACH-001042', 'ACH-002275', 'ACH-001072', 'ACH-002146', 'ACH-002302', 'ACH-001065', 'ACH-002392', 'ACH-002277', 'ACH-002152', 'ACH-002217', 'ACH-002113', 'ACH-002288', 'ACH-002378', 'ACH-002099', 'ACH-001364', 'ACH-002163', 'ACH-002293', 'ACH-002246', 'ACH-002124', 'ACH-002372', 'ACH-002174', 'ACH-002243', 'ACH-002162', 'ACH-002342', 'ACH-002339', 'ACH-002210', 'ACH-001189', 'ACH-002212', 'ACH-001091', 'ACH-001066', 'ACH-001214', 'ACH-002256', 'ACH-002121', 'ACH-002228', 'ACH-002221', 'ACH-002134', 'ACH-002280', 'ACH-002349', 'ACH-002255', 'ACH-002337', 'ACH-002287', 'ACH-002132', 'ACH-002283', 'ACH-000426', 'ACH-001094', 'ACH-002209', 'ACH-001150', 'ACH-001000', 'ACH-002107', 'ACH-002089', 'ACH-001017', 'ACH-001224', 'ACH-002279', 'ACH-002270', 'ACH-002273', 'ACH-002112', 'ACH-001108', 'ACH-000992', 'ACH-002229', 'ACH-002150', 'ACH-001490', 'ACH-002338', 'ACH-002186', 'ACH-002239', 'ACH-002202', 'ACH-002097', 'ACH-002286', 'ACH-002272', 'ACH-001206', 'ACH-001738', 'ACH-002397', 'ACH-002131', 'ACH-001087', 'ACH-002304', 'ACH-001071', 'ACH-002365', 'ACH-001021', 'ACH-001230', 'ACH-002108', 'ACH-001363', 'ACH-001136', 'ACH-002259', 'ACH-002051', 'ACH-002101', 'ACH-002153', 'ACH-002020', 'ACH-002301', 'ACH-001131', 'ACH-000309', 'ACH-000979', 'ACH-002120', 'ACH-002346', 'ACH-002358', 'ACH-001126', 'ACH-001039', 'ACH-002393', 'ACH-001452', 'ACH-002262', 'ACH-002040', 'ACH-002189', 'ACH-002268', 'ACH-002094', 'ACH-002381', 'ACH-002092', 'ACH-002241', 'ACH-002370', 'ACH-002165', 'ACH-002144', 'ACH-002245', 'ACH-002187', 'ACH-002285', 'ACH-002116', 'ACH-001171', 'ACH-001958', 'ACH-002356', 'ACH-002130', 'ACH-002233', 'ACH-002376', 'ACH-002364', 'ACH-002098', 'ACH-001146', 'ACH-002264', 'ACH-002178', 'ACH-002171', 'ACH-001512', 'ACH-001606', 'ACH-001234', 'ACH-001002', 'ACH-002149', 'ACH-002308', 'ACH-002122', 'ACH-002220', 'ACH-002317', 'ACH-001049', 'ACH-002316', 'ACH-002252', 'ACH-002118', 'ACH-002140', 'ACH-001198', 'ACH-002109', 'ACH-002126', 'ACH-002340', 'ACH-001090', 'ACH-001015', 'ACH-001121', 'ACH-001957', 'ACH-002306', 'ACH-001079', 'ACH-002160', 'ACH-002148', 'ACH-001047', 'ACH-002353', 'ACH-002311', 'ACH-001358', 'ACH-002257', 'ACH-002313', 'ACH-001293', 'ACH-001417', 'ACH-001081', 'ACH-001359', 'ACH-002117', 'ACH-002388', 'ACH-001955', 'ACH-002166', 'ACH-002192', 'ACH-002386', 'ACH-002145', 'ACH-002184', 'ACH-002348', 'ACH-002176', 'ACH-002310', 'ACH-002362', 'ACH-002347', 'ACH-001639', 'ACH-001704', 'ACH-002248', 'ACH-001024', 'ACH-002278', 'ACH-002139', 'ACH-002254', 'ACH-002249', 'ACH-002103', 'ACH-002105', 'ACH-002240', 'ACH-002155', 'ACH-001130', 'ACH-002389', 'ACH-002223', 'ACH-001175', 'ACH-002181', 'ACH-002307', 'ACH-002390', 'ACH-002100', 'ACH-002211', 'ACH-000047', 'ACH-001489', 'ACH-002354', 'ACH-002357', 'ACH-002335', 'ACH-002709', 'ACH-002343', 'ACH-001162', 'ACH-001088', 'ACH-002298', 'ACH-002214', 'ACH-002265', 'ACH-002194', 'ACH-002111', 'ACH-001355', 'ACH-001362', 'ACH-002215', 'ACH-002206', 'ACH-002371', 'ACH-002352', 'ACH-002156', 'ACH-002177', 'ACH-002355', 'ACH-002234', 'ACH-002244', 'ACH-001249', 'ACH-001063', 'ACH-001107', 'ACH-002345', 'ACH-002251', 'ACH-002090', 'ACH-002368', 'ACH-001365', 'ACH-002236', 'ACH-002052', 'ACH-002303', 'ACH-002173', 'ACH-001544', 'ACH-002170', 'ACH-002195', 'ACH-001138', 'ACH-002366', 'ACH-002133', 'ACH-002110', 'ACH-002222', 'ACH-002295', 'ACH-002271', 'ACH-002363', 'ACH-002154', 'ACH-001043', 'ACH-002115', 'ACH-002226', 'ACH-002142', 'ACH-001956', 'ACH-002309', 'ACH-001188', 'ACH-002083', 'ACH-002261', 'ACH-002161', 'ACH-002179', 'ACH-002263', 'ACH-002258', 'ACH-002227', 'ACH-002164', 'ACH-002218', 'ACH-002373', 'ACH-001282', 'ACH-002151', 'ACH-002077'}\n",
      "removed for QC reasons\n",
      "{'ACH-000992', 'ACH-001293', 'ACH-002709', 'ACH-001502', 'ACH-001143', 'ACH-000309', 'ACH-001205', 'ACH-000979', 'ACH-000426', 'ACH-000047', 'ACH-001512', 'ACH-000795', 'ACH-000727'}\n",
      "removed\n",
      "{'ACH-000561'}\n",
      "missing\n",
      "{'ACH-001648'}\n",
      "blacklist\n",
      "8 {'ACH-001756', 'ACH-002055', 'ACH-001760', 'ACH-002476', 'ACH-001686', 'ACH-001758', 'ACH-001759', 'ACH-001434'}\n"
     ]
    }
   ],
   "source": [
    "for val in ['internal', 'dmc', 'public']:\n",
    "    print('not present')\n",
    "    removed = set(prev[val]) - set(fusions.DepMap_ID)\n",
    "    print(removed)\n",
    "    print('removed for QC reasons')\n",
    "    print(set(rename.keys()))\n",
    "    print('removed')\n",
    "    removed = set(prevrna[val]) - set(fusions.DepMap_ID)\n",
    "    print(removed - set(rename.keys()))\n",
    "    missing = set(new[val]) - set(fusions.DepMap_ID)\n",
    "    blacklist = set(fusions.DepMap_ID) - (prev[val] | set(new[val]))\n",
    "    print('missing')\n",
    "    print(missing)\n",
    "    newlines = set(new[val]) \n",
    "    print('blacklist')\n",
    "    print(len(blacklist), blacklist)\n",
    "    ## removing first blacklisted, then embargoed, to create two datasets\n",
    "    print(len(fusions))\n",
    "    fusions = fusions[~fusions.DepMap_ID.isin(blacklist)]\n",
    "    print(len(fusions))\n",
    "    fusions.to_csv('temp/fusions.csv', index=False)\n",
    "    print(len(filtered))\n",
    "    filtered= filtered[~filtered.DepMap_ID.isin(blacklist)]\n",
    "    print(len(filtered))\n",
    "    filtered.to_csv('temp/filtered_fusions.csv', index=False)\n",
    "\n",
    "    os.popen('cp readmes/'+val+'-'+release+'.txt temp/README')\n",
    "\n",
    "    # uploading to taiga\n",
    "    tc.update_dataset(dataset_permaname=taiga_fusion[val],\n",
    "                     upload_file_path_dict={\n",
    "                         'temp/fusions.csv': 'TableCSV',\n",
    "                         'temp/filtered_fusions.csv': 'TableCSV',\n",
    "                         'temp/README', 'raw'},\n",
    "                      dataset_description=\n",
    "    \"\"\"\n",
    "    # Internal Fusions\n",
    "\n",
    "    for more information, see README\n",
    "\n",
    "    NEW LINES:\n",
    "    \"\"\"+str(newlines)+\"\"\"\n",
    "\n",
    "    REMOVED FOR QC REASONS:\n",
    "    \"\"\"+str(rename)+\"\"\"\n",
    "\n",
    "    BLACKLIST:\n",
    "    \"\"\"+str(blacklist))\n",
    "\n",
    "    # virtual datasets\n",
    "    AddToVirtual(virtual[val], taiga_fusion[val], files=[('CCLE_fusions', 'fusions'),('CCLE_filtered_fusions', 'filtered_fusions'), ('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_fusions_unfiltered', 'gene-fusions-8b7a.16/fusions'), ('CCLE_fusions', 'gene-fusions-8b7a.16/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datafile/e9e5878dd5db4d1585d0184db1854b46\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 7dc5923dd29d4cda930829a534d629f5 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/7dc5923dd29d4cda930829a534d629f5\n",
      "[('CCLE_fusions_unfiltered', 'gene-fusions-8b7a.16/fusions'), ('CCLE_fusions', 'gene-fusions-8b7a.16/filtered_fusions')]\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datafile/a73c2a4dc4d6405d921482142de04ee1\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 23bf1b099ff342c58aef5f3d27551c37 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/23bf1b099ff342c58aef5f3d27551c37\n"
     ]
    }
   ],
   "source": [
    "AddToVirtual('depmap-a0ab', \"gene-fusions-8b7a\", files=[('CCLE_fusions', 'fusions'),('CCLE_filtered_fusions', 'filtered_fusions')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/google/auth/_default.py:69: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a \"quota exceeded\" or \"API not enabled\" error. We recommend you rerun `gcloud auth application-default login` and make sure a quota project is added. Or you can use service accounts instead. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n",
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/google/auth/_default.py:69: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a \"quota exceeded\" or \"API not enabled\" error. We recommend you rerun `gcloud auth application-default login` and make sure a quota project is added. Or you can use service accounts instead. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n",
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/google/auth/_default.py:69: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a \"quota exceeded\" or \"API not enabled\" error. We recommend you rerun `gcloud auth application-default login` and make sure a quota project is added. Or you can use service accounts instead. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "status is: Failed for 0 jobs in submission 0. 774 mn elapsed.\r"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(refworkspace, \"847eef61-01af-496d-8d8e-cb1cbaa48f7c\")\n",
    "submission_id2 = refwm.create_submission(\"CNV_Somatic_Workflow_on_Sample\", samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "206px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
