{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n",
       "  var css_urls = [];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n  var css_urls = [];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import os.path\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "sys.path.insert(0, '..')\n",
    "\n",
    "from src.CCLE_postp_function import *\n",
    "from JKBio import Datanalytics as da \n",
    "from JKBio import TerraFunction as terra\n",
    "from JKBio import Helper as h\n",
    "from gsheets import Sheets\n",
    "from taigapy import TaigaClient\n",
    "import dalmatian as dm\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "from bokeh.plotting import *\n",
    "from bokeh.models import HoverTool\n",
    "from collections import OrderedDict\n",
    "from IPython.display import Image,display\n",
    "\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "tc = TaigaClient()\n",
    "output_notebook()\n",
    "sheets = Sheets.from_files('~/.client_secret.json', '~/.storage.json')\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## boot up\n",
    "\n",
    "we are instanciating all the parameters needed for this pipeline to run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesetname = \"20Q3\"\n",
    "prevname=\"20Q2\"\n",
    "prevversion = 22\n",
    "prevprevname ='20Q1'\n",
    "prevprevversion= 20\n",
    "virtual_internal='internal-20q3-00d0'\n",
    "\n",
    "refworkspace=\"broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline\"\n",
    "\n",
    "\n",
    "\n",
    "refsheet_url = \"https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE\"\n",
    "sheeturl = \"https://docs.google.com/spreadsheets/d/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4\"\n",
    "\n",
    "release = samplesetname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "release <- '20Q3'\n",
    "prevname <- '20Q2'\n",
    "genome_version <- 'hg19'\n",
    "taiga_version <- 10\n",
    "prevversion <-13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "refwm = dm.WorkspaceManager(refworkspace)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "filtered_CGA_MAF_aggregated    gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b7...\n",
       "samples                        [CDS-00rz9N, CDS-01bI6z, CDS-02waxZ, CDS-04TUV...\n",
       "merged_mutations               gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b7...\n",
       "Name: all, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = refwm.get_sample_sets().loc[\"all\"]\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/b49f7c3a-3f14-49e4-a464-006490d4a37b/aggregateMAFs_selectFields_workflow/66076582-bcd7-44d9-bfde-ee81bcb677e9/call-aggregateMAFs_selectFields/all.mergedMAF.txt...\n",
      "/ [1 files][253.5 MiB/253.5 MiB]                                                \n",
      "Operation completed over 1 objects/253.5 MiB.                                    \n"
     ]
    }
   ],
   "source": [
    "filtered = res['filtered_CGA_MAF_aggregated']\n",
    "! gsutil cp $filtered \"temp/mutation_filtered_terra_merged.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (27) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build', 'Chromosome',\n",
      "       'Start_position', 'End_position', 'Strand', 'Variant_Classification',\n",
      "       'Variant_Type'],\n",
      "      dtype='object')\n",
      "removed 247 duplicate samples\n",
      "Empty DataFrame\n",
      "Columns: [Hugo_Symbol, Entrez_Gene_Id, Center, NCBI_Build, Chromosome, Start_position, End_position, Strand, Variant_Classification, Variant_Type, Reference_Allele, Tumor_Seq_Allele1, Tumor_Seq_Allele2, dbSNP_RS, dbSNP_Val_Status, Genome_Change, Annotation_Transcript, Tumor_Sample_Barcode, cDNA_Change, Codon_Change, Protein_Change, t_alt_count, t_ref_count, tumor_f, isDeleterious, isTCGAhotspot, TCGAhsCnt, isCOSMIChotspot, COSMIChsCnt, i_ExAC_AF, PASS]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 31 columns]\n"
     ]
    }
   ],
   "source": [
    "file = pd.read_csv('temp/mutation_filtered_terra_merged.txt',sep='\\t') \n",
    "print(file.columns[:10])\n",
    "renaming = removeOlderVersions(names = set(file['Tumor_Sample_Barcode']), refsamples = refwm.get_samples(), arxspan_id = \"arxspan_id\", version=\"version\")\n",
    "print(file[file['Chromosome']=='0'])\n",
    "file[file['Tumor_Sample_Barcode'].isin(renaming.keys())].replace({'Tumor_Sample_Barcode':renaming}).reset_index(drop=True).to_csv('temp/mutation_filtered_terra_merged.txt',sep='\\t',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCLF_processing.py         \u001b[0m\u001b[01;34mhelper\u001b[0m/                requirements.txt\r\n",
      "\u001b[01;34mcell_line_mapping-master\u001b[0m/  Helper.py              RScript.R\r\n",
      "\u001b[01;34mdata\u001b[0m/                      __init__.py            setup.py\r\n",
      "Datanalytics.py            LICENSE                \u001b[01;34mtaigr\u001b[0m/\r\n",
      "\u001b[01;34mepigenetics\u001b[0m/               my_custom_commands.sh  TerraFunction.py\r\n",
      "GCPFunction.py             \u001b[01;34m__pycache__\u001b[0m/\r\n",
      "generate_sample_info.sh    README.md\r\n"
     ]
    }
   ],
   "source": [
    "ls ../JKBio/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#source('src/load_libraries_and_annotations.R')\n",
    "load('src/Annotations.rdata') \n",
    "# There are some cell lines the celllinemapr does not know how to map so we need to load this data object for now (from old datasets)\n",
    "source('src/CCLE_postp_function.R')\n",
    "library('cdsomics')\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(taigr)\n",
    "library(cdsomics)\n",
    "library(celllinemapr) # To pull out DepMap_IDs from CCLE_names where needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"Hugo_Symbol\"            \"Entrez_Gene_Id\"         \"NCBI_Build\"            \n",
      " [4] \"Chromosome\"             \"Start_position\"         \"End_position\"          \n",
      " [7] \"Strand\"                 \"Variant_Classification\" \"Variant_Type\"          \n",
      "[10] \"Reference_Allele\"       \"Tumor_Seq_Allele2\"      \"dbSNP_RS\"              \n",
      "[13] \"dbSNP_Val_Status\"       \"Genome_Change\"          \"Annotation_Transcript\" \n",
      "[16] \"Tumor_Sample_Barcode\"   \"cDNA_Change\"            \"Codon_Change\"          \n",
      "[19] \"Protein_Change\"         \"isDeleterious\"          \"isTCGAhotspot\"         \n",
      "[22] \"TCGAhsCnt\"              \"isCOSMIChotspot\"        \"COSMIChsCnt\"           \n",
      "[25] \"t_alt_count\"            \"t_ref_count\"            \"ExAC_AF\"               \n",
      "[28] \"pass\"                  \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "newly_merged_maf <- readMutations('temp/mutation_filtered_terra_merged.txt')\n",
    "new_release <- createSNPs(newly_merged_maf)\n",
    "names(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "new_release <- renameAsInMainMutation(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1002964      29\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "filtered <- filterAllelicFraction(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "write.table(\n",
    "  filtered$merged, \n",
    "  paste0('temp/newmutations.', release, '.all.csv'), sep = ',', quote = F, row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in `[.data.frame`(merged_latest_release, , c(\"CGA_WES_AC\", \"WGS_AC\",  : \n",
      "  undefined columns selected\n",
      "Calls: <Anonymous> ... withVisible -> filterMinCoverage -> apply -> [ -> [.data.frame\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In FUN(newX[, i], ...) :\n",
      "R[write to console]:  NAs introduced by coercion\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In FUN(newX[, i], ...) :\n",
      "R[write to console]:  NAs introduced by coercion\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in `[.data.frame`(merged_latest_release, , c(\"CGA_WES_AC\", \"WGS_AC\",  : \n",
      "  undefined columns selected\n",
      "Calls: <Anonymous> ... withVisible -> filterMinCoverage -> apply -> [ -> [.data.frame\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "filtered <- filterMinCoverage(filtered$merged, filtered$removed_from_maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "head(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "clean_annotations <- mergeAnnotations(merged,previous.release.maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Allie's version\n",
    "new_release <- addAnnotation(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# some matric files that does get used internaly and might be useful\n",
    "damaging_mutation <- mutation_maf_to_binary_matrix(new_release, damaging =  TRUE)\n",
    "other_mutation <- mutation_maf_to_binary_matrix(new_release, other = TRUE)\n",
    "hotspot_mutation <- mutation_maf_to_binary_matrix(new_release, hotspot = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  new_release, \n",
    "  paste0('temp/mutations.', release, '.all.csv'), sep = ',', quote = F, row.names = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  damaging_mutation, \n",
    "  paste0('temp/damaging_mutation.', release, '.all.csv'), sep = ',', quote = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  other_mutation, \n",
    "  paste0('temp/other_mutation.', release, '.all.csv'), sep = ',', quote = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  hotspot_mutation, \n",
    "  paste0('temp/hotspot_mutation.', release, '.all.csv'), sep = ',', quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "newmutations = pd.read_csv('temp/mutations.'+release+'.all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_20Q2_all = tc.get(name='depmap-mutations-maf-35fe', version=14, file='mutations.20Q2.all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#removing all of the WGS and snp arrays\n",
    "mutations_20Q2_all = mutations_20Q2_all[~mutations_20Q2_all.DepMap_ID.isin(set(mutations_20Q2_all.DepMap_ID) - set(newmutations.Tumor_Sample_Barcode))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACH-001061',\n",
       " 'ACH-001098',\n",
       " 'ACH-001194',\n",
       " 'ACH-001227',\n",
       " 'ACH-001678',\n",
       " 'ACH-001847',\n",
       " 'ACH-002048',\n",
       " 'ACH-002138',\n",
       " 'ACH-002204'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Too recent? removed by what means? why?\n",
    "set(newmutations.Tumor_Sample_Barcode) - set(mutations_20Q2_all.DepMap_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "newmutations = newmutations[~newmutations.Tumor_Sample_Barcode.isin(set(newmutations.Tumor_Sample_Barcode) - set(mutations_20Q2_all.DepMap_ID))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "newmutations = newmutations.sort_values(by=['Tumor_Sample_Barcode','Chromosome','Start_position','End_position'])\n",
    "mutations_20Q2_all = mutations_20Q2_all.sort_values(by=['DepMap_ID','Chromosome','Start_position','End_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "newmutations[['Tumor_Sample_Barcode','Chromosome','Start_position','End_position']] = newmutations[['Tumor_Sample_Barcode','Chromosome','Start_position','End_position']].astype(str)\n",
    "newmutationsset = newmutations['Tumor_Sample_Barcode']+'_'+newmutations['Chromosome']+':'+newmutations['Start_position']+'-'+newmutations['End_position']\n",
    "newmutations['grouped'] = newmutationsset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "986724"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(newmutationsset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "dups = h.dups(newmutationsset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "newmutationsset = set(newmutationsset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_20Q2_all[['DepMap_ID','Chromosome','Start_position','End_position']] = mutations_20Q2_all[['DepMap_ID','Chromosome','Start_position','End_position']].astype(str)\n",
    "mutations_20Q2_allset = mutations_20Q2_all['DepMap_ID']+'_'+mutations_20Q2_all['Chromosome']+':'+mutations_20Q2_all['Start_position']+'-'+mutations_20Q2_all['End_position']\n",
    "mutations_20Q2_all['grouped'] = mutations_20Q2_allset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1265853"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mutations_20Q2_allset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1258486"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mutations_20Q2_allset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "dups = h.dups(mutations_20Q2_allset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7367"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## How come?? the merging did not work well\n",
    "len(dups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>grouped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1020703</td>\n",
       "      <td>SLC25A32</td>\n",
       "      <td>81034</td>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>104414198</td>\n",
       "      <td>104414198</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>95:107</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-001528</td>\n",
       "      <td>ACH-001528_8:104414198-104414198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1020704</td>\n",
       "      <td>SLC25A32</td>\n",
       "      <td>81034</td>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>104414198</td>\n",
       "      <td>104414198</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>23:39</td>\n",
       "      <td>22:37</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-001528</td>\n",
       "      <td>ACH-001528_8:104414198-104414198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id  NCBI_Build Chromosome Start_position  \\\n",
       "1020703    SLC25A32           81034          37          8      104414198   \n",
       "1020704    SLC25A32           81034          37          8      104414198   \n",
       "\n",
       "        End_position Strand Variant_Classification Variant_Type  \\\n",
       "1020703    104414198      +      Missense_Mutation          SNP   \n",
       "1020704    104414198      +      Missense_Mutation          SNP   \n",
       "\n",
       "        Reference_Allele  ... CGA_WES_AC SangerWES_AC SangerRecalibWES_AC  \\\n",
       "1020703                G  ...     95:107         None                None   \n",
       "1020704                G  ...       None        23:39               22:37   \n",
       "\n",
       "        RNAseq_AC HC_AC RD_AC WGS_AC    Variant_annotation   DepMap_ID  \\\n",
       "1020703      None  None  None   None  other non-conserving  ACH-001528   \n",
       "1020704      None  None  None   None  other non-conserving  ACH-001528   \n",
       "\n",
       "                                  grouped  \n",
       "1020703  ACH-001528_8:104414198-104414198  \n",
       "1020704  ACH-001528_8:104414198-104414198  \n",
       "\n",
       "[2 rows x 35 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_20Q2_all[mutations_20Q2_all['grouped']==dups[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "submutations_20Q2_all = mutations_20Q2_all[~mutations_20Q2_all.CGA_WES_AC.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "462966"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(submutations_20Q2_all) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "submutations_20Q2_allset = submutations_20Q2_all.grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_20Q2_allset = mutations_20Q2_all.grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9918136537024317"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(submutations_20Q2_allset) & set(newmutationsset))/ len(submutations_20Q2_allset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.46535404023820237"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(submutations_20Q2_allset) & set(newmutationsset))/ len(newmutationsset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "with everything from all pipelines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9639747284955064"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mutations_20Q2_allset) & set(newmutationsset))/ len(newmutationsset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.751411893798095"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mutations_20Q2_allset) & set(newmutationsset))/ len(mutations_20Q2_allset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>grouped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1165855</td>\n",
       "      <td>TRIM51</td>\n",
       "      <td>84767</td>\n",
       "      <td>37</td>\n",
       "      <td>11</td>\n",
       "      <td>55653610</td>\n",
       "      <td>55653610</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>54:59</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_11:55653610-55653610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>273033</td>\n",
       "      <td>DACH1</td>\n",
       "      <td>1602</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>72440538</td>\n",
       "      <td>72440539</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>26:98</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_13:72440538-72440539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>82171</td>\n",
       "      <td>ARIH1</td>\n",
       "      <td>25820</td>\n",
       "      <td>37</td>\n",
       "      <td>15</td>\n",
       "      <td>72767214</td>\n",
       "      <td>72767215</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>46:32</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_15:72767214-72767215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>212898</td>\n",
       "      <td>CHRNE</td>\n",
       "      <td>1145</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>4802798</td>\n",
       "      <td>4802799</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>107:107</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>13:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_17:4802798-4802799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>157133</td>\n",
       "      <td>CACNB1</td>\n",
       "      <td>782</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>37341119</td>\n",
       "      <td>37341120</td>\n",
       "      <td>+</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>51:48</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_17:37341119-37341120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1156165</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>7579533</td>\n",
       "      <td>7579533</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>83:154</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-002461</td>\n",
       "      <td>ACH-002461_17:7579533-7579533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>344631</td>\n",
       "      <td>ENOSF1</td>\n",
       "      <td>55556</td>\n",
       "      <td>37</td>\n",
       "      <td>18</td>\n",
       "      <td>691068</td>\n",
       "      <td>691069</td>\n",
       "      <td>+</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CT</td>\n",
       "      <td>...</td>\n",
       "      <td>69:55</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-002461</td>\n",
       "      <td>ACH-002461_18:691068-691069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1008336</td>\n",
       "      <td>SIGLEC9</td>\n",
       "      <td>27180</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>51628387</td>\n",
       "      <td>51628387</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>111:270</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-002461</td>\n",
       "      <td>ACH-002461_19:51628387-51628387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30658</td>\n",
       "      <td>ADNP</td>\n",
       "      <td>23394</td>\n",
       "      <td>37</td>\n",
       "      <td>20</td>\n",
       "      <td>49509245</td>\n",
       "      <td>49509245</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>86:434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-002461</td>\n",
       "      <td>ACH-002461_20:49509245-49509245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>916787</td>\n",
       "      <td>RABEPK</td>\n",
       "      <td>10244</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>127990317</td>\n",
       "      <td>127990318</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>19:44</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-002461</td>\n",
       "      <td>ACH-002461_9:127990317-127990318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3191 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id  NCBI_Build Chromosome Start_position  \\\n",
       "1165855      TRIM51           84767          37         11       55653610   \n",
       "273033        DACH1            1602          37         13       72440538   \n",
       "82171         ARIH1           25820          37         15       72767214   \n",
       "212898        CHRNE            1145          37         17        4802798   \n",
       "157133       CACNB1             782          37         17       37341119   \n",
       "...             ...             ...         ...        ...            ...   \n",
       "1156165        TP53            7157          37         17        7579533   \n",
       "344631       ENOSF1           55556          37         18         691068   \n",
       "1008336     SIGLEC9           27180          37         19       51628387   \n",
       "30658          ADNP           23394          37         20       49509245   \n",
       "916787       RABEPK           10244          37          9      127990317   \n",
       "\n",
       "        End_position Strand Variant_Classification Variant_Type  \\\n",
       "1165855     55653610      +        Frame_Shift_Del          DEL   \n",
       "273033      72440539      +           In_Frame_Ins          INS   \n",
       "82171       72767215      +           In_Frame_Ins          INS   \n",
       "212898       4802799      +        Frame_Shift_Ins          INS   \n",
       "157133      37341120      +            Splice_Site          INS   \n",
       "...              ...    ...                    ...          ...   \n",
       "1156165      7579533      +        Frame_Shift_Del          DEL   \n",
       "344631        691069      +            Splice_Site          DEL   \n",
       "1008336     51628387      +        Frame_Shift_Del          DEL   \n",
       "30658       49509245      +        Frame_Shift_Del          DEL   \n",
       "916787     127990318      +        Frame_Shift_Ins          INS   \n",
       "\n",
       "        Reference_Allele  ... CGA_WES_AC SangerWES_AC SangerRecalibWES_AC  \\\n",
       "1165855                A  ...      54:59         None                None   \n",
       "273033                 -  ...      26:98         None                None   \n",
       "82171                  -  ...      46:32         None                None   \n",
       "212898                 -  ...    107:107         None                None   \n",
       "157133                 -  ...      51:48         None                None   \n",
       "...                  ...  ...        ...          ...                 ...   \n",
       "1156165                G  ...     83:154         None                None   \n",
       "344631                CT  ...      69:55         None                None   \n",
       "1008336                C  ...    111:270         None                None   \n",
       "30658                  C  ...     86:434         None                None   \n",
       "916787                 -  ...      19:44         None                None   \n",
       "\n",
       "        RNAseq_AC HC_AC RD_AC WGS_AC    Variant_annotation   DepMap_ID  \\\n",
       "1165855      None  None  None   None              damaging  ACH-000007   \n",
       "273033       None  None  None   None  other non-conserving  ACH-000007   \n",
       "82171        None  None  None   None  other non-conserving  ACH-000007   \n",
       "212898       None  None  None  13:17              damaging  ACH-000007   \n",
       "157133       None  None  None   None              damaging  ACH-000007   \n",
       "...           ...   ...   ...    ...                   ...         ...   \n",
       "1156165      None  None  None   None              damaging  ACH-002461   \n",
       "344631       None  None  None   None              damaging  ACH-002461   \n",
       "1008336      None  None  None   None              damaging  ACH-002461   \n",
       "30658        None  None  None   None              damaging  ACH-002461   \n",
       "916787       None  None  None   None              damaging  ACH-002461   \n",
       "\n",
       "                                  grouped  \n",
       "1165855   ACH-000007_11:55653610-55653610  \n",
       "273033    ACH-000007_13:72440538-72440539  \n",
       "82171     ACH-000007_15:72767214-72767215  \n",
       "212898      ACH-000007_17:4802798-4802799  \n",
       "157133    ACH-000007_17:37341119-37341120  \n",
       "...                                   ...  \n",
       "1156165     ACH-002461_17:7579533-7579533  \n",
       "344631        ACH-002461_18:691068-691069  \n",
       "1008336   ACH-002461_19:51628387-51628387  \n",
       "30658     ACH-002461_20:49509245-49509245  \n",
       "916787   ACH-002461_9:127990317-127990318  \n",
       "\n",
       "[3191 rows x 35 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submutations_20Q2_all[submutations_20Q2_all.grouped.isin(set(submutations_20Q2_allset) - set(newmutationsset))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "notfound = submutations_20Q2_all[submutations_20Q2_all.grouped.isin(set(submutations_20Q2_allset) - set(newmutationsset))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "177"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(notfound.DepMap_ID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>grouped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>212898</td>\n",
       "      <td>CHRNE</td>\n",
       "      <td>1145</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>4802798</td>\n",
       "      <td>4802799</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>107:107</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>13:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_17:4802798-4802799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1244462</td>\n",
       "      <td>YIPF2</td>\n",
       "      <td>78992</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>11036411</td>\n",
       "      <td>11036411</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>80:113</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>20:144</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12:21</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_19:11036411-11036411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>468372</td>\n",
       "      <td>GYS1</td>\n",
       "      <td>2997</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>49494632</td>\n",
       "      <td>49494636</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TGGGT</td>\n",
       "      <td>...</td>\n",
       "      <td>85:130</td>\n",
       "      <td>None</td>\n",
       "      <td>8:33</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>15:18</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_19:49494632-49494636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>740436</td>\n",
       "      <td>NLRP5</td>\n",
       "      <td>126206</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>56515231</td>\n",
       "      <td>56515239</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AGCTAGACA</td>\n",
       "      <td>...</td>\n",
       "      <td>67:118</td>\n",
       "      <td>None</td>\n",
       "      <td>53:125</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>13:16</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_19:56515231-56515239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1252308</td>\n",
       "      <td>ZC3HAV1L</td>\n",
       "      <td>92092</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>138719301</td>\n",
       "      <td>138719309</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>ACAGGGGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>92:69</td>\n",
       "      <td>None</td>\n",
       "      <td>40:52</td>\n",
       "      <td>21:27</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26:19</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>ACH-000007_7:138719301-138719309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>333381</td>\n",
       "      <td>EFS</td>\n",
       "      <td>10278</td>\n",
       "      <td>37</td>\n",
       "      <td>14</td>\n",
       "      <td>23830009</td>\n",
       "      <td>23830010</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>66:0</td>\n",
       "      <td>None</td>\n",
       "      <td>8:11</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>ACH-001715_14:23830009-23830010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>948776</td>\n",
       "      <td>RNF165</td>\n",
       "      <td>494470</td>\n",
       "      <td>37</td>\n",
       "      <td>18</td>\n",
       "      <td>44013214</td>\n",
       "      <td>44013214</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>47:35</td>\n",
       "      <td>None</td>\n",
       "      <td>3:6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>ACH-001715_18:44013214-44013214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>465296</td>\n",
       "      <td>GTF2E2</td>\n",
       "      <td>2961</td>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>30464626</td>\n",
       "      <td>30464626</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>138:122</td>\n",
       "      <td>111:35</td>\n",
       "      <td>151:141</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>ACH-001715_8:30464626-30464626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1057872</td>\n",
       "      <td>SPAG1</td>\n",
       "      <td>6674</td>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>101178125</td>\n",
       "      <td>101178126</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>151:100</td>\n",
       "      <td>18:25</td>\n",
       "      <td>44:49</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>ACH-001715_8:101178125-101178126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2136</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>19</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>107583767</td>\n",
       "      <td>107583769</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAC</td>\n",
       "      <td>...</td>\n",
       "      <td>159:178</td>\n",
       "      <td>13:2</td>\n",
       "      <td>18:43</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>ACH-001715_9:107583767-107583769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1335 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id  NCBI_Build Chromosome Start_position  \\\n",
       "212898        CHRNE            1145          37         17        4802798   \n",
       "1244462       YIPF2           78992          37         19       11036411   \n",
       "468372         GYS1            2997          37         19       49494632   \n",
       "740436        NLRP5          126206          37         19       56515231   \n",
       "1252308    ZC3HAV1L           92092          37          7      138719301   \n",
       "...             ...             ...         ...        ...            ...   \n",
       "333381          EFS           10278          37         14       23830009   \n",
       "948776       RNF165          494470          37         18       44013214   \n",
       "465296       GTF2E2            2961          37          8       30464626   \n",
       "1057872       SPAG1            6674          37          8      101178125   \n",
       "2136          ABCA1              19          37          9      107583767   \n",
       "\n",
       "        End_position Strand Variant_Classification Variant_Type  \\\n",
       "212898       4802799      +        Frame_Shift_Ins          INS   \n",
       "1244462     11036411      +        Frame_Shift_Del          DEL   \n",
       "468372      49494636      +        Frame_Shift_Del          DEL   \n",
       "740436      56515239      +           In_Frame_Del          DEL   \n",
       "1252308    138719309      +           In_Frame_Del          DEL   \n",
       "...              ...    ...                    ...          ...   \n",
       "333381      23830010      +        Frame_Shift_Ins          INS   \n",
       "948776      44013214      +        Frame_Shift_Del          DEL   \n",
       "465296      30464626      +        Frame_Shift_Del          DEL   \n",
       "1057872    101178126      +        Frame_Shift_Ins          INS   \n",
       "2136       107583769      +           In_Frame_Del          DEL   \n",
       "\n",
       "        Reference_Allele  ... CGA_WES_AC SangerWES_AC SangerRecalibWES_AC  \\\n",
       "212898                 -  ...    107:107         None                None   \n",
       "1244462                A  ...     80:113         None                None   \n",
       "468372             TGGGT  ...     85:130         None                8:33   \n",
       "740436         AGCTAGACA  ...     67:118         None              53:125   \n",
       "1252308        ACAGGGGTC  ...      92:69         None               40:52   \n",
       "...                  ...  ...        ...          ...                 ...   \n",
       "333381                 -  ...       66:0         None                8:11   \n",
       "948776                 C  ...      47:35         None                 3:6   \n",
       "465296                 C  ...    138:122       111:35             151:141   \n",
       "1057872                -  ...    151:100        18:25               44:49   \n",
       "2136                 AAC  ...    159:178         13:2               18:43   \n",
       "\n",
       "        RNAseq_AC HC_AC RD_AC WGS_AC    Variant_annotation   DepMap_ID  \\\n",
       "212898       None  None  None  13:17              damaging  ACH-000007   \n",
       "1244462    20:144  None  None  12:21              damaging  ACH-000007   \n",
       "468372       None  None  None  15:18              damaging  ACH-000007   \n",
       "740436       None  None  None  13:16  other non-conserving  ACH-000007   \n",
       "1252308     21:27  None  None  26:19  other non-conserving  ACH-000007   \n",
       "...           ...   ...   ...    ...                   ...         ...   \n",
       "333381       None  None  None   None              damaging  ACH-001715   \n",
       "948776       None  None  None   None              damaging  ACH-001715   \n",
       "465296       None  None  None   None              damaging  ACH-001715   \n",
       "1057872      None  None  None   None              damaging  ACH-001715   \n",
       "2136         None  None  None   None  other non-conserving  ACH-001715   \n",
       "\n",
       "                                  grouped  \n",
       "212898      ACH-000007_17:4802798-4802799  \n",
       "1244462   ACH-000007_19:11036411-11036411  \n",
       "468372    ACH-000007_19:49494632-49494636  \n",
       "740436    ACH-000007_19:56515231-56515239  \n",
       "1252308  ACH-000007_7:138719301-138719309  \n",
       "...                                   ...  \n",
       "333381    ACH-001715_14:23830009-23830010  \n",
       "948776    ACH-001715_18:44013214-44013214  \n",
       "465296     ACH-001715_8:30464626-30464626  \n",
       "1057872  ACH-001715_8:101178125-101178126  \n",
       "2136     ACH-001715_9:107583767-107583769  \n",
       "\n",
       "[1335 rows x 35 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notfound[~(notfound.SangerRecalibWES_AC.isna() & notfound.SangerWES_AC.isna() & notfound.HC_AC.isna() & notfound.WGS_AC.isna() & notfound.RNAseq_AC.isna() &notfound.RD_AC.isna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "only one third of the notfound mutations were backed by any other analysis, showing an FP rate of 2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35547"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(newmutationsset) - set(mutations_20Q2_allset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGAhsCnt</th>\n",
       "      <th>isCOSMIChotspot</th>\n",
       "      <th>COSMIChsCnt</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>PERC_CGA_WES_AC</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>PERC_PERC_CGA_WES_AC</th>\n",
       "      <th>INCLUDE</th>\n",
       "      <th>grouped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>959821</td>\n",
       "      <td>ALPK2</td>\n",
       "      <td>115701</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>18</td>\n",
       "      <td>56205373</td>\n",
       "      <td>56205374</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_18:56205373-56205374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189948</td>\n",
       "      <td>TBXA2R</td>\n",
       "      <td>6915</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>19</td>\n",
       "      <td>3600465</td>\n",
       "      <td>3600465</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:9</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_19:3600465-3600465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>959822</td>\n",
       "      <td>TICAM1</td>\n",
       "      <td>148022</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>19</td>\n",
       "      <td>4817288</td>\n",
       "      <td>4817289</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8:11</td>\n",
       "      <td>0.421053</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_19:4817288-4817289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189958</td>\n",
       "      <td>PPFIA3</td>\n",
       "      <td>8541</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>19</td>\n",
       "      <td>49641575</td>\n",
       "      <td>49641575</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:25</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_19:49641575-49641575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>959814</td>\n",
       "      <td>INPP1</td>\n",
       "      <td>3628</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>2</td>\n",
       "      <td>191231419</td>\n",
       "      <td>191231419</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39:78</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_2:191231419-191231419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>903175</td>\n",
       "      <td>WASF2</td>\n",
       "      <td>10163</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>27736334</td>\n",
       "      <td>27736334</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:5</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-002467_1:27736334-27736334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>112764</td>\n",
       "      <td>HAO1</td>\n",
       "      <td>54363</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>20</td>\n",
       "      <td>7894923</td>\n",
       "      <td>7894923</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>21:689</td>\n",
       "      <td>0.029577</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-002508_20:7894923-7894923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>634800</td>\n",
       "      <td>TCHHL1</td>\n",
       "      <td>126637</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>152060019</td>\n",
       "      <td>152060019</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:2</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-002509_1:152060019-152060019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>191073</td>\n",
       "      <td>OR5AS1</td>\n",
       "      <td>219447</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>11</td>\n",
       "      <td>55798056</td>\n",
       "      <td>55798056</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:4</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-002511_11:55798056-55798056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>191110</td>\n",
       "      <td>ARIH1</td>\n",
       "      <td>25820</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>15</td>\n",
       "      <td>72767235</td>\n",
       "      <td>72767235</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:27</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-002511_15:72767235-72767235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35547 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Hugo_Symbol  Entrez_Gene_Id   NCBI_Build Chromosome Start_position  \\\n",
       "959821       ALPK2          115701  __UNKNOWN__         18       56205373   \n",
       "189948      TBXA2R            6915  __UNKNOWN__         19        3600465   \n",
       "959822      TICAM1          148022  __UNKNOWN__         19        4817288   \n",
       "189958      PPFIA3            8541  __UNKNOWN__         19       49641575   \n",
       "959814       INPP1            3628  __UNKNOWN__          2      191231419   \n",
       "...            ...             ...          ...        ...            ...   \n",
       "903175       WASF2           10163  __UNKNOWN__          1       27736334   \n",
       "112764        HAO1           54363  __UNKNOWN__         20        7894923   \n",
       "634800      TCHHL1          126637  __UNKNOWN__          1      152060019   \n",
       "191073      OR5AS1          219447  __UNKNOWN__         11       55798056   \n",
       "191110       ARIH1           25820  __UNKNOWN__         15       72767235   \n",
       "\n",
       "       End_position       Strand Variant_Classification Variant_Type  \\\n",
       "959821     56205374  __UNKNOWN__        Frame_Shift_Ins          INS   \n",
       "189948      3600465  __UNKNOWN__                 Silent          SNP   \n",
       "959822      4817289  __UNKNOWN__           In_Frame_Ins          INS   \n",
       "189958     49641575  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "959814    191231419  __UNKNOWN__            Splice_Site          DEL   \n",
       "...             ...          ...                    ...          ...   \n",
       "903175     27736334  __UNKNOWN__                 Silent          SNP   \n",
       "112764      7894923  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "634800    152060019  __UNKNOWN__            Splice_Site          SNP   \n",
       "191073     55798056  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "191110     72767235  __UNKNOWN__                 Silent          SNP   \n",
       "\n",
       "       Reference_Allele  ... TCGAhsCnt isCOSMIChotspot COSMIChsCnt   ExAC_AF  \\\n",
       "959821                -  ...       NaN           False         0.0       NaN   \n",
       "189948                A  ...       NaN           False         0.0       NaN   \n",
       "959822                -  ...       NaN           False         0.0       NaN   \n",
       "189958                C  ...       NaN           False         0.0       NaN   \n",
       "959814                A  ...       NaN           False         0.0       NaN   \n",
       "...                 ...  ...       ...             ...         ...       ...   \n",
       "903175                A  ...       NaN           False         0.0       NaN   \n",
       "112764                G  ...       5.0           False         0.0  0.000058   \n",
       "634800                G  ...       NaN           False         0.0       NaN   \n",
       "191073                C  ...       NaN           False         0.0       NaN   \n",
       "191110                C  ...       NaN           False         0.0       NaN   \n",
       "\n",
       "       CGA_WES_AC PERC_CGA_WES_AC    MAX_AF PERC_PERC_CGA_WES_AC INCLUDE  \\\n",
       "959821       40:0        1.000000  1.000000                  0.5    True   \n",
       "189948        5:9        0.357143  0.500000                  0.5    True   \n",
       "959822       8:11        0.421053  0.500000                  0.5    True   \n",
       "189958       3:25        0.107143  0.500000                  0.5    True   \n",
       "959814      39:78        0.333333  0.500000                  0.5    True   \n",
       "...           ...             ...       ...                  ...     ...   \n",
       "903175        3:5        0.375000  0.500000                  0.5    True   \n",
       "112764     21:689        0.029577  0.500000                  0.5    True   \n",
       "634800        5:2        0.714286  0.714286                  0.5    True   \n",
       "191073        3:4        0.428571  0.500000                  0.5    True   \n",
       "191110       3:27        0.100000  0.500000                  0.5    True   \n",
       "\n",
       "                                 grouped  \n",
       "959821   ACH-000001_18:56205373-56205374  \n",
       "189948     ACH-000001_19:3600465-3600465  \n",
       "959822     ACH-000001_19:4817288-4817289  \n",
       "189958   ACH-000001_19:49641575-49641575  \n",
       "959814  ACH-000001_2:191231419-191231419  \n",
       "...                                  ...  \n",
       "903175    ACH-002467_1:27736334-27736334  \n",
       "112764     ACH-002508_20:7894923-7894923  \n",
       "634800  ACH-002509_1:152060019-152060019  \n",
       "191073   ACH-002511_11:55798056-55798056  \n",
       "191110   ACH-002511_15:72767235-72767235  \n",
       "\n",
       "[35547 rows x 31 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newmutations[newmutations.grouped.isin(set(newmutationsset) - set(mutations_20Q2_allset))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1343"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(newmutations[newmutations.grouped.isin(set(newmutationsset) - set(mutations_20Q2_allset))].Tumor_Sample_Barcode))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "alternate = np.array(newmutations[newmutations.grouped.isin(set(newmutationsset) - set(mutations_20Q2_allset))]['CGA_WES_AC'].str.split(':').tolist()).astype(int)[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46.00326328522801"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alternate.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alternate.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlYAAAI/CAYAAAC1XpeNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAdS0lEQVR4nO3dfZBsd1kn8O+PxCigDAhJiAkYrISXlOsKXtGtAswCyyaBmxDCW1Z00cg1LFEQLBdWVrB8BUVciizxAoFdXN4Mb7kQSBAXwSpEEl5iMEQii+ZCQgKsw25RJUv52z+6gWHoPn36zjP0zL2fT9XU7e45z5zn9jzT/Z3z6z7Teu8BAGDrbrfqBgAADheCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQ5etUNJMnd7na3fvLJJ6+6DQCAha655prP996PnfW5HRGsTj755Fx99dWrbgMAYKHW2t/P+5ylQACAIoIVAEARwQoAoIhgBQBQRLACACiy0mDVWtvbWtu/vr6+yjYAAEqsNFj13g/03vetra2tsg0AgBKWAgEAighWAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCL+CDMAQBF/hBkAoIilQACAIoIVAEARwQoAoMjRq25gVW552W+O3vbuT33uNnYCABwuHLECACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFFlpsGqt7W2t7V9fX19lGwAAJVYarHrvB3rv+9bW1lbZBgBACUuBAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCIrDVattb2ttf3r6+urbAMAoMRKg1Xv/UDvfd/a2toq2wAAKGEpEACgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBkpcGqtba3tbZ/fX19lW0AAJRYabDqvR/ove9bW1tbZRsAACUsBQIAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEWOXnUDu83Blz5l9LYnXfTybewEANhpHLECACjiiNW3yScuPmf0tvd92tu2sRMAYLsIVjvch/5o7+htf/TnD2xjJwDAIpYCAQCKCFYAAEUEKwCAIoIVAECR8mDVWrtfa+2S1tplrbWnVn99AICdalSwaq1d2lq7tbV23abbz2it3dBau7G19uwk6b1f33u/MMnjk+ypbxkAYGcae8Tq1UnO2HhDa+2oJBcnOTPJaUnOb62dNv3c2Un+Isl7yjoFANjhRp3Hqvf+vtbayZtufmCSG3vvn0qS1trrk5yT5G9675cnuby19o4kr61rl7He+/JHjt729Ke8Yxs7AYAjx1ZOEHpikps2XD+Y5Mdaa6cneUyS70xyxbzi1tq+JPuS5J73vOcW2gAA2Bm2EqzajNt67/29Sd67qLj3vj/J/iTZs2dP30IfAAA7wlbeFXgwyT02XD8pyWe31g4AwO61lWD1oSSnttbu1Vo7JskTk1xe0xYAwO4z9nQLr0vygST3aa0dbK1d0Hv/apKLklyZ5Pokb+y9f3z7WgUA2NnGvivw/Dm3X5GBF6gDABxJ/EkbAIAiKw1WrbW9rbX96+vrq2wDAKDESoNV7/1A733f2traKtsAAChhKRAAoIhgBQBQRLACACgiWAEAFNnK3wrkMHTFK88ave1ZFziFGQBs5IgVAEAR57ECACjiPFYAAEUsBQIAFBGsAACKCFYAAEUEKwCAIs5jRYnLXnXG6G0f+zPv2sZOAGB1HLECACgiWAEAFLEUyEr991f/29Hb/vSTr9zGTgBg65x5HQCgiDOvAwAUsRTIrvRHrxm/hPjzP2UJEYBvDy9eBwAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKDISs9j1Vrbm2TvKaecsso2OIL8wWvHn//qmf/O+a8AWI4zrwMAFLEUCABQRLACACgiWAEAFPFHmGGE579x/Iven/94L3oHOFI5YgUAUMQRK9hGT3vzGaO3vfgx79rGTgD4dhCsYAc6823njd72nee8aRs7AWAZlgIBAIoIVgAARQQrAIAighUAQJGVBqvW2t7W2v719fVVtgEAUMIfYQYAKOJ0C3AYOeutzxq97RWPftE2dgJwZPIaKwCAIoIVAEARS4FAznrLb47e9opzn7uNnQDsbo5YAQAUEawAAIpYCgQOySPf/OKltn/HY35pmzoB2DkEK+Cw9qjL/sdS27/9sT+5TZ0ARwJLgQAARRyxAr6tHvnmly21/Tse89Rt6gSgniNWAABFBCsAgCKCFQBAEa+xApjhUX/yJ6O3ffvjHreNnQC7yUqPWLXW9rbW9q+vr6+yDQCAEis9YtV7P5DkwJ49e56yyj4AVu3Rl7179LZvfey/2cZOgK2wFAiwi537pr8Yve1bznvQNnYCJF68DgBQxhErgEJnX3Zg9LaXP3bvNnYCrIIjVgAARQQrAIAilgIBjkCPfdOHR2972XkP2MZO4PAiWAG7wqPe9Kqltn/7eT+zTZ0c2Z7w5htHb/uGx5yyjZ3AzmQpEACgiGAFAFBEsAIAKCJYAQAUEawAAIp4VyAAbNH7X3Pb6G0f/FPHbmMnrJojVgAARRyxAmDb/ae3fGb0tr997onb2Mnh4ZMv/dzobU+96Pht7ITNBCsAdqyL3zI+QDztXAGC1bMUCABQxBErAJh65xs+P3rbM59wt23s5PDwuT+8ZvS2xz/jR7axk2+flR6xaq3tba3tX19fX2UbAAAlVhqseu8Heu/71tbWVtkGAECJXb8UeNsll4ze9tgLL9zGTgBgOR95xa2jt73/zx23jZ1QZdcHKwDY7A1vGv9aqSec57VS1PGuQACAIo5YAQCDbnnRJ0Zve/dn3XcbO9n5HLECACgiWAEAFLEUCADsGJ97yfuX2v74X3xwkuTWl161VN1xFz1iqe3HcsQKAKCII1YAwBHr1ovfutT2xz3t0YOfd8QKAKCIYAUAUESwAgAo4jVWAHCEuPmFnxm97Qm/cuI2dnL4csQKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAiqw0WLXW9rbW9q+vr6+yDQCAEisNVr33A733fWtra6tsAwCghKVAAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgSHmwaq09urX28tba21prj6j++gAAO9WoYNVau7S1dmtr7bpNt5/RWruhtXZja+3ZSdJ7f2vv/SlJnpzkCeUdAwDsUGOPWL06yRkbb2itHZXk4iRnJjktyfmttdM2bPLc6ecBAI4Io4JV7/19Sb646eYHJrmx9/6p3vtXkrw+yTlt4gVJ3tl7/3BtuwAAO9dWXmN1YpKbNlw/OL3tF5I8PMljW2sXziture1rrV3dWrv6tttu20IbAAA7w9FbqG0zbuu995ckecmi4t77/iT7k2TPnj19C30AAOwIWzlidTDJPTZcPynJZ7fWDgDA7rWVYPWhJKe21u7VWjsmyROTXF7TFgDA7jP2dAuvS/KBJPdprR1srV3Qe/9qkouSXJnk+iRv7L1/fPtaBQDY2Ua9xqr3fv6c269IckVpRwAAu5Q/aQMAUGSlwaq1tre1tn99fX2VbQAAlFhpsOq9H+i971tbW1tlGwAAJSwFAgAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiPNYAQAUcR4rAIAilgIBAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUceZ1AIAizrwOAFDEUiAAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBF/0gYAoIg/aQMAUMRSIABAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKCIvxUIAFDE3woEAChiKRAAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEVWGqxaa3tba/vX19dX2QYAQImVBqve+4He+761tbVVtgEAUMJSIABAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKDISoNVa21va23/+vr6KtsAACix0mDVez/Qe9+3tra2yjYAAEpYCgQAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUWWmwaq3tba3tX19fX2UbAAAlVhqseu8Heu/71tbWVtkGAEAJS4EAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAECR8mDVWvuB1torW2uXVX9tAICdbFSwaq1d2lq7tbV23abbz2it3dBau7G19uwk6b1/qvd+wXY0CwCwk409YvXqJGdsvKG1dlSSi5OcmeS0JOe31k4r7Q4AYBcZFax67+9L8sVNNz8wyY3TI1RfSfL6JOcU9wcAsGts5TVWJya5acP1g0lObK3dtbV2SZL7t9aeM6+4tbavtXZ1a+3q2267bQttAADsDEdvobbNuK333r+Q5MJFxb33/Un2J8mePXv6FvoAANgRtnLE6mCSe2y4flKSz26tHQCA3WsrwepDSU5trd2rtXZMkicmubymLQCA3Wfs6RZel+QDSe7TWjvYWrug9/7VJBcluTLJ9Une2Hv/+Pa1CgCws416jVXv/fw5t1+R5IrSjgAAdil/0gYAoMhKg1VrbW9rbf/6+voq2wAAKLHSYNV7P9B737e2trbKNgAASlgKBAAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAEcEKAKCIE4QCABRxglAAgCKWAgEAighWAABFBCsAgCKCFQBAEcEKAKCIYAUAUESwAgAo4gShAABFnCAUAKCIpUAAgCKCFQBAEcEKAKCIYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCL+pA0AQBF/0gYAoIilQACAIoIVAEARwQoAoIhgBQBQRLACACgiWAEAFBGsAACKCFYAAEUEKwCAIoIVAEARfysQAKCIvxUIAFDEUiAAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIocvcqdt9b2Jtl7yimn5LaX/fFStcc+9Unb0xQAwCFa6RGr3vuB3vu+tbW1VbYBAFDCUiAAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgiGAFAFBEsAIAKCJYAQAUEawAAIoIVgAARQQrAIAighUAQBHBCgCgyEqDVWttb2tt//r6+irbAAAosdJg1Xs/0Hvft7a2tso2AABKWAoEACgiWAEAFBGsAACKtN77qntIa+22JH8/59N3S/L5Q/iy6nZv3W7oUZ06dbuvbjf0qG531H1/7/3YmZ/pve/ojyRXqzuy6nZDj+rUqdt9dbuhR3W7v85SIABAEcEKAKDIbghW+9UdcXW7oUd16tTtvrrd0KO6XV63I168DgBwONgNR6wAAHaHQ3nF+7fjI8mlSW5Nct0SNfdI8j+TXJ/k40mePrLuu5L8VZKPTet+fclej0rykSRvX6Lm00n+OslHs8Q7D5LcOcllST4x/X/+qxE195nu52sfX0ryjJH7+6XpfXJdktcl+a6RdU+f1nx8aF+zvs9JvjfJu5N8cvrvXUbWPW66v39OsmeJ/f3e9P68Nslbktx5ZN1vTGs+muSqJN+3zBwn+eUkPcndRu7v+Uk+s+H7eNbY/SX5hSQ3TO+fF47c3xs27OvTST46su6Hk/zl12Y7yQNH1v3LJB+Y/lwcSHKnTTUzf74XzctA3eC8DNQNzstA3eC8zKtbNC8D+xucl6H9Dc3LwP4G52WgbnBeBuoWzcvMx/UR8zKvbtG8zKtbNC/z6hbNy+Dz1sC8zNvfonmZu7958zKwr0WzMq9u0azMqxuclQ313/RcvmhW5n0s3GBVH0kekuQBWS5YnZDkAdPL35Pkb5OcNqKuJfnu6eXvSPLBJD++xH6fmeS1WT5YfcsT6oi6/5bk56aXj9n8Qzqi/qgkt2RyDo5F256Y5H8luf30+huTPHlE3Q9mEqrukOToJH+a5NSx3+ckL0zy7OnlZyd5wci6+2USIt+b+cFqVt0jkhw9vfyCJfZ3pw2XfzHJJWPnOJMniyszOX/brGA1a3/PT/LLC+77WXX/evo9+M7p9ePG9rnh8y9K8msj93dVkjOnl89K8t6RdR9K8hPTyz+b5Dc21cz8+V40LwN1g/MyUDc4LwN1g/Myr27RvAzsb3BeBuoG52Woz6F5Gdjf4LwM1C2al5mP6yPmZV7donmZV7doXubVLZqXuc9bC+Zl3v4Wzcu8urnzMtTjglmZt69FszKvbnBWNtR/03P5olmZ97FjlwJ77+9L8sUla27uvX94evn/ZPIbzokj6nrv/f9Or37H9KOP2Wdr7aQkj0zyimV6PRSttTtl8oT0yiTpvX+l9/6PS36ZhyX5u977vBOybnZ0ktu31o7OJCh9dkTN/ZL8Ze/9y733ryb58yTnztpwzvf5nEwCZKb/PnpMXe/9+t77DUONzam7atpnMvlt6KSRdV/acPWOmTEzA3P84iS/MqtmQd2gOXVPTfK7vfd/mm5z6zL7a621JI/P5IjlmLqe5E7Ty2uZMTNz6u6T5H3Ty+9Oct6mmnk/34PzMq9u0bwM1A3Oy0Dd4LwsePyaOy9beNybVzc4L4v2N29eBuoG52WgbtG8zHtcXzQvM+tGzMu8ukXzMq9u0bwMPW8NzcshPd8N1M2dl0X7GpiVeXWLZmVe3eCsTHuZ9Vy+8Llolh0brLaqtXZykvtnkljHbH9Ua+2jmSxPvLv3PqouyR9mMsD/vGSLPclVrbVrWmv7Rtb8QJLbkryqtfaR1torWmt3XHK/T8yMJ8iZDfb+mSS/n+QfktycZL33ftWI0uuSPKS1dtfW2h0y+c3iHkv0eHzv/eZpDzcnOW6J2q362STvHLtxa+23Wms3JfnJJL82subsJJ/pvX/sEPq7qLV2bWvt0tbaXUbW3DvJg1trH2yt/Xlr7UeX3OeDk3yu9/7Jkds/I8nvTe+X30/ynJF11yU5e3r5cRmYmU0/36PnZdnHhRF1g/OyuW7svGysW2ZeZvQ5al421Y2elzn3y8J52VQ3el421S2clzmP6wvn5VCfD0bUzZyXeXWL5mVW3Zh5GehzcF7m1A3Oy4L7ZO6szKlbOCtz6sY8tsx6Lj+056I+4rDWqj6SnJwllgI31H13kmuSPOYQau+cyXr+D47Y9lFJ/uv08ulZbinw+6b/HpfJevBDRtTsSfLVJD82vf5fMueQ5pz6YzI5Pf/xI7e/S5I/S3JsJsn/rUmeNLL2giQfzuS3hEuSvHjs9znJP276/P9eZj4ysBS4oO5XM3kNRFt2HjP5AZ/52ryNdZkc9ftgkrXp9U9nzpLwjPvl+EyWcm+X5LeSXDqy7rokL8nkMPkDM1ne/Zb/48D98rIkz1ri+/eSJOdNLz8+yZ+OrLtvJof6r0nyvCRfmFP3TT/fS8zLzMeFEfMyr27RvMx9HFowL1+vW3JeNt8vY+dlc93YeZl3vyyal837Gzsvm+tGzct0268/ro+dl811Y+dloG5wXubVLZqXTXU/NHZeZtwvo+ZlRt3YeZl1nwzOyox9jZqVGXWDs5I5z+XLzMo3bTdmo1V95BCCVSYB4Mokz9zCfp+XBa9lmW73O0kOTof3liRfTvLHh7C/54/c392TfHrD9QcneccS+zknyVVLbP+4JK/ccP2nvzZ8S/7/fjvJfxj7fc7kRZAnTC+fkOSGZeYjhxCskvz7TF7ceIdDmcck3z/wua/XJfkXmfwm9enpx1czOSJ49yX3N/pzSd6V5PQN1/8uybEj75ejk3wuyUlLfP/W841TubQkXzqE/8O9k/zVjNu/5ed7zLzMqhszL/PqFs3L0P6G5mVz3dh5GbG/mff1nPtz4bwM3C+D8zJnfwvnZcT/b+a8bNrmeZm8oHvU48vmujHzMq9u0bwM7W9oXmbU/ecx8zJifzPnZc79OerxZcZ9svCxZca+Rj22LPi/fcusZM5z+bKz8rWPw2opcLpe+8ok1/fe/2CJumNba3eeXr59kodn8i6OQb335/TeT+q9n5zJEtuf9d6fNGJ/d2ytfc/XLmfy4sbrRuzvliQ3tdbuM73pYUn+ZlHdBudn5DLg1D8k+fHW2h2m9+3DMnl9w0KtteOm/94zk9+8l9nv5Zk8EGX679uWqF1aa+2MJP8xydm99y8vUXfqhqtnZ9zM/HXv/bje+8nTuTmYyQtzbxmxvxM2XD03I2Zm6q1JHjr9GvfON45cjvHwJJ/ovR8cuX0yed3DT0wvPzSTd9QstGFmbpfkuZkc6dz4+Xk/34PzsoXHhZl1i+ZloG5wXmbVjZmXgf0NzsvA/TI4Lwvuz7nzMlA3OC8D/79F8zLvcX3RvBzS88G8uhHzMq9u0bzMqvvIiHmZt79F8zLvfpk7Lwvuy6FZmVe3aFbm/d8GZ2XgufzQnovGpK9VfGTyRHxzkv+XyXBcMKLmQZm8dulrb1Gd+Zb0GXU/lMlbLK/NZJi+5d1PI77G6Rm5FJjJa6U+lm+8JfRXl9jPD2fyNtNrMxnocW//nCwpfCHTQ8RL7O/XpwN9XZLXZPrOjxF1788k9H0sycOW+T4nuWuS92TyQ/OeJN87su7c6eV/yuQ3oStH1t2Y5KYNMzPr3X2z6t40vV+uzeQtvCcuO8eZc6h+zv5ek8nbha/N5Af+hJF1x2Ty29d1mSzPPnRsn0leneTCJb9/D8rkkPvHMlmW+JGRdU/P5B1ff5vkd7NpOSFzfr4XzctA3eC8DNQNzstA3eC8zKtbNC8D+xucl4G6wXkZ6nNoXgb2NzgvA3WL5mXm4/qIeZlXt2he5tUtmpd5dYvmZeHz1px5mbe/RfMyr27uvAz1uGBW5u1r0azMqxuclU1f4/R8Yylw4XPRrA9nXgcAKHJYLQUCAKySYAUAUESwAgAoIlgBABQRrAAAighWAABFBCsAgCKCFQBAkf8PqmipxybO/FwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "a,b = np.unique(alternate, return_counts=True)\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "ax=sns.barplot(x=a[:40],y=b[:40], ci=None, ax=ax)\n",
    "ax.set_yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9917292598531522"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alternate[alternate>2]) / len(alternate) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8605226882718654"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alternate[alternate>3]) / len(alternate) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, ...,  True,  True, False])"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "onlyinnew = newmutations[newmutations.grouped.isin(set(newmutationsset) - set(mutations_20Q2_allset))]\n",
    "highconf = onlyinnew[(onlyinnew['CGA_WES_AC'].str.split(':').str[0].values.astype(int)>3) & (onlyinnew['PERC_CGA_WES_AC'].values.astype(float)>0.3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5861535431963316"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(highconf)/len(onlyinnew)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGAhsCnt</th>\n",
       "      <th>isCOSMIChotspot</th>\n",
       "      <th>COSMIChsCnt</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>PERC_CGA_WES_AC</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>PERC_PERC_CGA_WES_AC</th>\n",
       "      <th>INCLUDE</th>\n",
       "      <th>grouped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>189786</td>\n",
       "      <td>VPS13D</td>\n",
       "      <td>55187</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>12359347</td>\n",
       "      <td>12359347</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34:213</td>\n",
       "      <td>0.137652</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_1:12359347-12359347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>959813</td>\n",
       "      <td>AADACL4</td>\n",
       "      <td>343066</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>12726308</td>\n",
       "      <td>12726322</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CTGGCGTGACGCCAT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57:141</td>\n",
       "      <td>0.287879</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_1:12726308-12726322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189787</td>\n",
       "      <td>IFNLR1</td>\n",
       "      <td>163702</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>24484172</td>\n",
       "      <td>24484172</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>118:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_1:24484172-24484172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189788</td>\n",
       "      <td>ZSCAN20</td>\n",
       "      <td>7579</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>33954141</td>\n",
       "      <td>33954141</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28:62</td>\n",
       "      <td>0.311111</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_1:33954141-33954141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>189789</td>\n",
       "      <td>GBP4</td>\n",
       "      <td>115361</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>1</td>\n",
       "      <td>89657103</td>\n",
       "      <td>89657103</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>96:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-000001_1:89657103-89657103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>901963</td>\n",
       "      <td>CYLC1</td>\n",
       "      <td>1538</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>X</td>\n",
       "      <td>83126495</td>\n",
       "      <td>83126495</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34:38</td>\n",
       "      <td>0.472222</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-003000_X:83126495-83126495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>901964</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>1288</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>X</td>\n",
       "      <td>107400439</td>\n",
       "      <td>107400439</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-003000_X:107400439-107400439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>901965</td>\n",
       "      <td>COL4A5</td>\n",
       "      <td>1287</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>X</td>\n",
       "      <td>107821607</td>\n",
       "      <td>107821607</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-003000_X:107821607-107821607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>901966</td>\n",
       "      <td>GPR112</td>\n",
       "      <td>139378</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>X</td>\n",
       "      <td>135426613</td>\n",
       "      <td>135426613</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-003000_X:135426613-135426613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>901967</td>\n",
       "      <td>SLITRK4</td>\n",
       "      <td>139065</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>X</td>\n",
       "      <td>142717252</td>\n",
       "      <td>142717252</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>105:0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>ACH-003000_X:142717252-142717252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>986724 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Hugo_Symbol  Entrez_Gene_Id   NCBI_Build Chromosome Start_position  \\\n",
       "189786      VPS13D           55187  __UNKNOWN__          1       12359347   \n",
       "959813     AADACL4          343066  __UNKNOWN__          1       12726308   \n",
       "189787      IFNLR1          163702  __UNKNOWN__          1       24484172   \n",
       "189788     ZSCAN20            7579  __UNKNOWN__          1       33954141   \n",
       "189789        GBP4          115361  __UNKNOWN__          1       89657103   \n",
       "...            ...             ...          ...        ...            ...   \n",
       "901963       CYLC1            1538  __UNKNOWN__          X       83126495   \n",
       "901964      COL4A6            1288  __UNKNOWN__          X      107400439   \n",
       "901965      COL4A5            1287  __UNKNOWN__          X      107821607   \n",
       "901966      GPR112          139378  __UNKNOWN__          X      135426613   \n",
       "901967     SLITRK4          139065  __UNKNOWN__          X      142717252   \n",
       "\n",
       "       End_position       Strand Variant_Classification Variant_Type  \\\n",
       "189786     12359347  __UNKNOWN__      Nonsense_Mutation          SNP   \n",
       "959813     12726322  __UNKNOWN__           In_Frame_Del          DEL   \n",
       "189787     24484172  __UNKNOWN__                 Silent          SNP   \n",
       "189788     33954141  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "189789     89657103  __UNKNOWN__                 Silent          SNP   \n",
       "...             ...          ...                    ...          ...   \n",
       "901963     83126495  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "901964    107400439  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "901965    107821607  __UNKNOWN__                 Silent          SNP   \n",
       "901966    135426613  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "901967    142717252  __UNKNOWN__      Missense_Mutation          SNP   \n",
       "\n",
       "       Reference_Allele  ... TCGAhsCnt isCOSMIChotspot COSMIChsCnt   ExAC_AF  \\\n",
       "189786                C  ...       NaN           False         0.0       NaN   \n",
       "959813  CTGGCGTGACGCCAT  ...       NaN           False         3.0       NaN   \n",
       "189787                G  ...       NaN           False         0.0       NaN   \n",
       "189788                T  ...       NaN           False         0.0       NaN   \n",
       "189789                G  ...       NaN           False         0.0  0.000016   \n",
       "...                 ...  ...       ...             ...         ...       ...   \n",
       "901963                C  ...       NaN           False         0.0       NaN   \n",
       "901964                A  ...       NaN           False         0.0       NaN   \n",
       "901965                A  ...       NaN           False         0.0       NaN   \n",
       "901966                A  ...       NaN           False         0.0       NaN   \n",
       "901967                A  ...       NaN           False         0.0       NaN   \n",
       "\n",
       "       CGA_WES_AC PERC_CGA_WES_AC MAX_AF PERC_PERC_CGA_WES_AC INCLUDE  \\\n",
       "189786     34:213        0.137652    0.5                  0.5    True   \n",
       "959813     57:141        0.287879    0.5                  0.5    True   \n",
       "189787      118:0        1.000000    1.0                  0.5    True   \n",
       "189788      28:62        0.311111    0.5                  0.5    True   \n",
       "189789       96:0        1.000000    1.0                  0.5    True   \n",
       "...           ...             ...    ...                  ...     ...   \n",
       "901963      34:38        0.472222    0.5                  0.5    True   \n",
       "901964       59:0        1.000000    1.0                  0.5    True   \n",
       "901965       95:0        1.000000    1.0                  0.5    True   \n",
       "901966       48:0        1.000000    1.0                  0.5    True   \n",
       "901967      105:0        1.000000    1.0                  0.5    True   \n",
       "\n",
       "                                 grouped  \n",
       "189786    ACH-000001_1:12359347-12359347  \n",
       "959813    ACH-000001_1:12726308-12726322  \n",
       "189787    ACH-000001_1:24484172-24484172  \n",
       "189788    ACH-000001_1:33954141-33954141  \n",
       "189789    ACH-000001_1:89657103-89657103  \n",
       "...                                  ...  \n",
       "901963    ACH-003000_X:83126495-83126495  \n",
       "901964  ACH-003000_X:107400439-107400439  \n",
       "901965  ACH-003000_X:107821607-107821607  \n",
       "901966  ACH-003000_X:135426613-135426613  \n",
       "901967  ACH-003000_X:142717252-142717252  \n",
       "\n",
       "[986724 rows x 31 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newmutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filterCoverage(maf, loc=['CGA_WES_AC'], sep=':',cov=4):\n",
    "    muts=np.zeroes((len(maf),2))\n",
    "    for val in loc:\n",
    "        muts+= np.array([[v[0],0] if 'NA' in v else v for v in mutations_20Q2_all[val].fillna('0'+sep+'0').astype(str).str.split(sep).tolist()]).astype(int)\n",
    "    return maf[muts[:,1]>=cov]\n",
    "\n",
    "def filterAllelicFraction(maf, loc=['CGA_WES_AC'], sep=':',frac=0.3):\n",
    "    muts=np.zeroes((len(maf),2))\n",
    "    for val in loc:\n",
    "        muts+= np.array([[v[0],0] if 'NA' in v else v for v in mutations_20Q2_all[val].fillna('0'+sep+'0').astype(str).str.split(sep).tolist()]).astype(int)\n",
    "    muts = muts[:,0]/muts[:,1]\n",
    "    return maf[muts>=frac]\n",
    "\n",
    "def mergeAnnotations(newmaf, additionalmaf):\n",
    "    \n",
    "\n",
    "def addAnnotation(maf):\n",
    "    \n",
    "\n",
    "def mafToMat(maf, col, boolify = False, samplesCol = \"DepMap_ID\", mutNameCol=\"Hugo_Symbol\"):\n",
    "    maf = maf.sort_values(by = mutNameCol)\n",
    "    mut = pd.DataFrame(data = np.zeroes((len(set(maf[mutNameCol])), 1)), columns=['fake'], index=set(maf[mutNameCol])).astype(float)\n",
    "    for val in set(maf[samplesCol]):\n",
    "        mut.append(maf[maf[samplesCol]==val], axis=)\n",
    "    return mut.nan_to_num(0).astype(bool if boolify else float).drop(columns=['fake'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_mutations = filterCoverage(mutations)\n",
    "filtered_mutations = filterAllelicFraction(filtered_mutations)\n",
    "\n",
    "merged_mutations = addAnnotation(mutations)\n",
    "\n",
    "mafToMat(filtered_mutations[filtered_mutations.damaging]).to_csv('.csv')\n",
    "mafToMat(filtered_mutations[filtered_mutations.other]).to_csv('.csv')\n",
    "mafToMat(filtered_mutations[filtered_mutations.hotspot]).to_csv('.csv')\n",
    "\n",
    "\n",
    "CCLE2othermutations = \n",
    "\n",
    "mutations = mergeAnnotations(filtered_mutations, CCLE2othermutations)\n",
    "\n",
    "#making \n",
    "for muttype in ['']:\n",
    "    mafToMat(CCLE2othermutations[CCLE2othermutations.damaging & CCLE2othermutations[muttype]]).to_csv(''+muttype+\".csv\")\n",
    "    mafToMat(CCLE2othermutations[CCLE2othermutations.other & CCLE2othermutations[muttype]]).to_csv(''+muttype+\".csv\")\n",
    "    mafToMat(CCLE2othermutations[CCLE2othermutations.hotspot & CCLE2othermutations[muttype]]).to_csv(''+muttype+\".csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cd ../depmap-release-readmes && git pull && mv release-'+releAse+'/dmc-'+releAse+'.txt ../ccle_processing/temp/README && cd -')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1301656\n",
      "1297418\n",
      "1757\n",
      "1744\n",
      "1758\n",
      "1745\n",
      "1725\n",
      "1712\n"
     ]
    }
   ],
   "source": [
    "print(len(mutations))\n",
    "mutations = mutations[~mutations.DepMap_ID.isin(wes_embargo)]\n",
    "print(len(mutations))\n",
    "mutations.to_csv('temp/depmap_'+release+'_mutation_calls.all', index=False)\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation = damaging_mutation[~damaging_mutation.index.isin(wes_embargo)]\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation.to_csv('temp/damaging_mutation.all')\n",
    "print(len(other_mutation))\n",
    "other_mutation = other_mutation[~other_mutation.index.isin(wes_embargo)]\n",
    "print(len(other_mutation))\n",
    "other_mutation.to_csv('temp/other_mutation.all',)\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation = hotspot_mutation[~hotspot_mutation.index.isin(wes_embargo)]\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation.to_csv('temp/hotspot_mutation.all',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[##################]100% |  40.4 MiB/s | 277.3 MiB / 277.3 MiB | Time:  0:00:06\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shoud be None\n",
      "set()\n",
      "new lines\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ACH-001533', 'ACH-001574', 'ACH-002021', 'ACH-002065'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevmut = tc.get(name='depmap-mutation-calls-dfce', version=15, file='depmap_'+prevname+'_mutation_calls')\n",
    "print('shoud be None')\n",
    "print(set(prevmut.DepMap_ID) - set(mutations.DepMap_ID))\n",
    "print(\"new lines\")\n",
    "newlines = set(mutations.DepMap_ID) - set(prevmut.DepMap_ID) \n",
    "newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc.update_dataset(dataset_permaname=\"depmap-mutation-calls-dfce\",\n",
    "                 upload_file_path_dict={'temp/depmap_'+release+'_mutation_calls.all': 'TableCSV',\n",
    "                                        'temp/damaging_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/other_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/hotspot_mutation.all': 'NumericMatrixCSV',\n",
    "                                       },#'temp/README': 'Raw'},\n",
    "                 dataset_description=\"\"\"\n",
    "# DMC Mutations\n",
    "\n",
    "* Version 1-5 DMC 19Q1*\n",
    "\n",
    "version 5 is a one-off portal thing because dmc wanted to be able to plot if a gene has any mutation as one-hot encoded value in the x/y axes of the data explorer It adds the any_mutation matrix, but does not change the others. Code used to generate:\n",
    "\n",
    "```\n",
    "from taigapy import TaigaClient\n",
    "\n",
    "c = TaigaClient()\n",
    "\n",
    "dmc_19q1_mutation_taiga_root = \"depmap-mutation-calls-dfce.3/\"\n",
    "other_matrix = c.get(dmc_19q1_mutation_taiga_root + \"other_mutation\")\n",
    "damaging_matrix = c.get(dmc_19q1_mutation_taiga_root + \"damaging_mutation\")\n",
    "hotspot_matrix = c.get(dmc_19q1_mutation_taiga_root + \"hotspot_mutation\")\n",
    "\n",
    "df = other_matrix.append(damaging_matrix)\n",
    "df = df.groupby(level=0).sum()\n",
    "\n",
    "df = df.append(hotspot_matrix)\n",
    "df = df.groupby(level=0).sum()\n",
    "\n",
    "df[df > 1] = 1\n",
    "\n",
    "df.to_csv('any_mutation.csv')\n",
    "```\n",
    "The code uses version 3 because the dmc portal was using version 3\n",
    "\n",
    "version 4 updates the column name from VA_WES_AC to CCLE_WES_AC\n",
    "\n",
    "version 3 has an updated definition for hotspot mutations\n",
    "\n",
    "version 2+ contains the correct data for 19Q1\n",
    "\n",
    "* Version 6 DMC 19Q2*\n",
    "\n",
    "* Version 7-8 DMC 19Q3*\n",
    "version 8 fixed entrez ids\n",
    "\n",
    "* Version 9 DMC 19Q4*\n",
    "adding 52 new cell lines.\n",
    "\n",
    "* Version 10 DMC 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 11 DMC 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 12 Internal 20Q1*\n",
    "uploading 8 new lines\n",
    "\n",
    "* Version 13 Internal 20Q1*\n",
    "removing unauthorized cl\n",
    "\n",
    "* Version 14 Internal 20Q2*\n",
    "uploading 8 new lines and adding .all to express the fact that this data is the aggregate of all different sequencing methods.\n",
    "\n",
    "* Version 15 Internal 20Q2*\n",
    "removing 2 lines\n",
    "\n",
    "* Version 15 Internal 20Q3*\n",
    "nothing different from 20Q2. no new cell lines\n",
    "\n",
    "* Version 15 Internal 20Q3*\n",
    "updating the blacklists\n",
    "\n",
    "\n",
    "MAF file, added column (Variant_annotation) classifying each variant as either silent, damaging, other conserving, or other non-conserving, based on this mapping (old annotation from Variant_Classification column - new annotation):\n",
    "\n",
    "Silent - silent\n",
    "Splice_Site - damaging\n",
    "Missense_Mutation - other non-conserving\n",
    "Nonsense_Mutation - damaging\n",
    "De_novo_Start_OutOfFrame - damaging\n",
    "Nonstop_Mutation - other non-conserving\n",
    "Frame_Shift_Del - damaging\n",
    "Frame_Shift_Ins - damaging\n",
    "In_Frame_Del - other non-conserving\n",
    "In_Frame_Ins - other non-conserving\n",
    "Stop_Codon_Del - other non-conserving\n",
    "Stop_Codon_Ins - other non-conserving\n",
    "Start_Codon_SNP - damaging\n",
    "Start_Codon_Del - damaging\n",
    "Start_Codon_Ins - damaging\n",
    "5'Flank - other conserving\n",
    "Intron - other conserving\n",
    "IGR - other conserving\n",
    "3'UTR - other conserving\n",
    "5'UTR - other conserving\n",
    "Binary matrices:\n",
    "- damaging: if damaging\n",
    "- other: if other conserving or other non-conserving\n",
    "- hotspot: if it is not a silent mutation and is either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, DepMap (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "NEW LINES:\n",
    "\"\"\"+newlines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To add to a virtual dataset\n",
    "AddToVirtual(virtual_dmc, 'depmap-mutation-calls-dfce', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls'),])#('README','README')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Public"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cd ../depmap-release-readmes && git pull && mv release-'+releAse+'/public-'+releAse+'.txt README && cd -')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#damaging_mutation\n",
    "mutations=depmap_20Q3_mutation_calls\n",
    "#hotspot_mutation\n",
    "#other_mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1297418\n",
      "1297418\n",
      "1744\n",
      "1744\n",
      "1745\n",
      "1745\n",
      "1712\n",
      "1712\n"
     ]
    }
   ],
   "source": [
    "print(len(mutations))\n",
    "mutations = mutations[mutations.DepMap_ID.isin(prevprev)]\n",
    "mutations = mutations[~mutations.DepMap_ID.isin(wes_dmc_embargo)]\n",
    "print(len(mutations))\n",
    "mutations.to_csv('temp/depmap_'+release+'_mutation_calls.all', index=False)\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation = damaging_mutation[damaging_mutation.index.isin(prevprev)]\n",
    "damaging_mutation = damaging_mutation[~damaging_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation.to_csv('temp/damaging_mutation.all')\n",
    "print(len(other_mutation))\n",
    "other_mutation = other_mutation[other_mutation.index.isin(prevprev)]\n",
    "other_mutation = other_mutation[~other_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(other_mutation))\n",
    "other_mutation.to_csv('temp/other_mutation.all')\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation = hotspot_mutation[hotspot_mutation.index.isin(prevprev)]\n",
    "hotspot_mutation = hotspot_mutation[~hotspot_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation.to_csv('temp/hotspot_mutation.all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shoud be None\n",
      "set()\n",
      "new lines\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ACH-001533', 'ACH-001574', 'ACH-002021', 'ACH-002065'}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevmut = tc.get(name='depmap-mutation-calls-9a1a', version=18, file='depmap_'+prevname+'_mutation_calls')\n",
    "print('shoud be None')\n",
    "ermgency_removed = set(prevmut.DepMap_ID) - set(mutations.DepMap_ID)\n",
    "print(ermgency_removed) \n",
    "print(\"new lines\")\n",
    "newlines = set(mutations.DepMap_ID) - set(prevmut.DepMap_ID) \n",
    "newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading depmap_20Q3_mutation_calls...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: depmap_20Q3_mutation_calls properly converted and uploaded\n",
      "Uploading damaging_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: damaging_mutation properly converted and uploaded\n",
      "Uploading other_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: other_mutation properly converted and uploaded\n",
      "Uploading hotspot_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: hotspot_mutation properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 8cc4900a50874d8593b0bfc591001360 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/8cc4900a50874d8593b0bfc591001360\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'8cc4900a50874d8593b0bfc591001360'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "description=\"\"\"\n",
    "# Public Mutations\n",
    "\n",
    "Mutation calls for Public DepMap data\n",
    "\n",
    "* Version 1 Public 18Q1*\n",
    "\n",
    "original source: CCLE data portal\n",
    "* Version 2 Public 18Q2*\n",
    "\n",
    "merged mutations and indels file (1,549 cell lines total, including data for 63 newly released cell lines)\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q2/CCLE_DepMap_18Q2_maf_20180502.txt`\n",
    "* Version 3-4 Public 18Q3*\n",
    "\n",
    "version 3 deprecated\n",
    "\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q3/CCLE_DepMap_18q3_maf_20180718.txt`\n",
    "\n",
    "Binary matrices:\n",
    "damaging: if isDeleterious is true\n",
    "missense: if isDeleterious is false\n",
    "hotspot: if missense and either TCGA or COSMIC hotspot\n",
    "Rows: cell line, Broad (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "MAF file\n",
    "\n",
    "* Version 5 Public 18Q4*\n",
    "\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q4/CCLE_DepMap_18q4_maf_20181029.txt`\n",
    "\n",
    "* Version 6-9 Public 19Q1*\n",
    "\n",
    "version 9 updates the column name from VA_WES_AC to CCLE_WES_AC\n",
    "\n",
    "version 8 uses an updated definition for hotspot mutations\n",
    "\n",
    "version 9 contains the correct data for 19Q1\n",
    "\n",
    "* Version 10 Public 19Q2*\n",
    "\n",
    "* Version 11-12 Public 19Q3*\n",
    "\n",
    "version 12 fixed entrez ids\n",
    "\n",
    "* Version 13 Public 19Q4*\n",
    "\n",
    "adding 52 new cell lines\n",
    "\n",
    "* Version 14 Public 19Q4*\n",
    "removing unauthorized lines and setting matrices\n",
    "\n",
    "* Version 15 Public 20Q1*\n",
    "adding 8 new lines \n",
    "\n",
    "* Version 16 Public 20Q1*\n",
    "removing an unauthorized line\n",
    "\n",
    "* Version 17 Internal 20Q2*\n",
    "uploading 8 new lines and adding .all to express the fact that this data is the aggregate of all different sequencing methods.\n",
    "\n",
    "* Version 18 Internal 20Q2*\n",
    "removing 2 lines\n",
    "\n",
    "* Version 19 Internal 20Q3*\n",
    "nothing different from 20Q2. no new cell lines\n",
    "\n",
    "* Version 20 Internal 20Q3*\n",
    "updating the blacklists\n",
    "\n",
    "* Version 21 Internal 20Q3*\n",
    "updating the dmc\n",
    "\n",
    "* Version 22 Internal 20Q3*\n",
    "readding two already released samples to the public list\n",
    "\n",
    "MAF file, added column (Variant_annotation) classifying each variant as either silent, damaging, other conserving, or other non-conserving, based on this mapping (old annotation from Variant_Classification column - new annotation):\n",
    "\n",
    "Silent - silent\n",
    "Splice_Site - damaging\n",
    "Missense_Mutation - other non-conserving\n",
    "Nonsense_Mutation - damaging\n",
    "De_novo_Start_OutOfFrame - damaging\n",
    "Nonstop_Mutation - other non-conserving\n",
    "Frame_Shift_Del - damaging\n",
    "Frame_Shift_Ins - damaging\n",
    "In_Frame_Del - other non-conserving\n",
    "In_Frame_Ins - other non-conserving\n",
    "Stop_Codon_Del - other non-conserving\n",
    "Stop_Codon_Ins - other non-conserving\n",
    "Start_Codon_SNP - damaging\n",
    "Start_Codon_Del - damaging\n",
    "Start_Codon_Ins - damaging\n",
    "5'Flank - other conserving\n",
    "Intron - other conserving\n",
    "IGR - other conserving\n",
    "3'UTR - other conserving\n",
    "5'UTR - other conserving\n",
    "Binary matrices:\n",
    "\n",
    "- damaging: if damaging\n",
    "- other: if other conserving or other non-conserving\n",
    "- hotspot: if it is not a silent mutation and is either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, DepMap (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "NEW LINES:\n",
    "\"\"\"+str(newlines)\n",
    "\n",
    "if len(ermgency_removed):\n",
    "    description+=\"\"\"\n",
    "    \n",
    "    !! WE REMOVED!!:\n",
    "    \"\"\"+str(ermgency_removed)\n",
    "\n",
    "tc.update_dataset(dataset_permaname=\"depmap-mutation-calls-9a1a\",\n",
    "                 upload_file_path_dict={'temp/depmap_'+release+'_mutation_calls.all': 'TableCSV',\n",
    "                                        'temp/damaging_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/other_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/hotspot_mutation.all': 'NumericMatrixCSV',\n",
    "                                       },#'temp/README': 'Raw'},\n",
    "                 dataset_description=description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_mutations', 'depmap-mutation-calls-9a1a.22/depmap_20Q3_mutation_calls'), ('CCLE_gene_cn', 'depmap-wes-cn-data-97cc.34/public_20Q3_gene_cn'), ('Achilles_gene_effect_unscaled', 'avana-public-tentative-20q3-3e73.5/gene_effect_unscaled'), ('Achilles_high_variance_genes', 'avana-public-tentative-20q3-3e73.5/high_variance_genes'), ('Achilles_guide_efficacy', 'avana-public-tentative-20q3-3e73.5/guide_efficacy'), ('CCLE_fusions_unfiltered', 'gene-fusions-6212.14/unfiltered_fusions_20Q3'), ('common_essentials', 'avana-public-tentative-20q3-3e73.5/essential_genes'), ('Achilles_logfold_change_failures', 'avana-public-tentative-20q3-3e73.5/logfold_change_failures'), ('CCLE_expression', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_proteincoding_tpm'), ('Achilles_raw_readcounts', 'avana-public-tentative-20q3-3e73.5/raw_readcounts'), ('Achilles_raw_readcounts_failures', 'avana-public-tentative-20q3-3e73.5/raw_readcounts_failing'), ('README', 'public-20q3-3d35.22/README'), ('CCLE_segment_cn', 'depmap-wes-cn-data-97cc.34/public_20Q3_segs_cn'), ('CCLE_RNAseq_transcripts', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_transcripts_tpm'), ('nonessentials', 'avana-public-tentative-20q3-3e73.5/nonessential_genes'), ('Achilles_guide_map', 'avana-public-tentative-20q3-3e73.5/guide_gene_map'), ('Achilles_dropped_guides', 'avana-public-tentative-20q3-3e73.5/dropped_guides'), ('CCLE_fusions', 'gene-fusions-6212.14/filtered_fusions_20Q3'), ('CCLE_RNAseq_reads', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_counts'), ('CCLE_expression_full', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_tpm'), ('Achilles_replicate_map', 'avana-public-tentative-20q3-3e73.5/replicate_map'), ('Achilles_gene_dependency', 'avana-public-tentative-20q3-3e73.5/gene_dependency'), ('Achilles_common_essentials', 'avana-public-tentative-20q3-3e73.5/pan_dependent_genes'), ('Achilles_logfold_change', 'avana-public-tentative-20q3-3e73.5/logfold_change'), ('Achilles_gene_effect', 'avana-public-tentative-20q3-3e73.5/gene_effect')]\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id f803fe38cdb04895b6285533dcc8b00b created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/f803fe38cdb04895b6285533dcc8b00b\n"
     ]
    }
   ],
   "source": [
    "# To add to a virtual dataset\n",
    "AddToVirtual(virtual_public, 'depmap-mutation-calls-9a1a', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls'),])#('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "notify_time": "5",
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "277px",
    "width": "375px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "198.8px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "position": {
    "height": "617.8px",
    "left": "1232.4px",
    "right": "20px",
    "top": "120px",
    "width": "262.8px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
