{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading the necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import depmapomics.patch_firecloud\n",
    "depmapomics.patch_firecloud.install_patches()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "from depmapomics import constants\n",
    "from depmapomics import env_config\n",
    "\n",
    "from depmapomics import fusions, expressions, dm_omics\n",
    "from mgenepy import terra\n",
    "from mgenepy.utils import helper as h\n",
    "\n",
    "import dalmatian as dm\n",
    "\n",
    "from bokeh.plotting import output_notebook\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isCCLE = True\n",
    "doCleanup = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading new data\n",
    "\n",
    "Currently, sequenced data for DepMap is generated by the Genomics Platform (GP) at the Broad who deposits them into several different Terra workspaces. Therefore, the first step of this pipeline is to look at these workspaces and:\n",
    "\n",
    " - identify new samples by looking at the bam files and compare them with bams we have already onboarded\n",
    " - remove duplicates and ones with broken file paths\n",
    " - map files to profiles in Gumbo, if possible\n",
    " - onboard new samples and new versions of old cell lines if we find any"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The following two cells scan the delivery workspaces and add new samples to gumbo. Currently under construction to be regularly run off-cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Currently working on running this step off-cycle\n",
    "# if isCCLE:\n",
    "#     from depmap_omics_upload import loading\n",
    "#     print(\"loading new RNAseq data\")\n",
    "#     rnasamples, unmapped = loading.loadFromMultipleWorkspaces(RNAWORKSPACES, EXTRACT_DEFAULTS[\"sm_id\"], \"pdo_sample\", \"rna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Currently working on running this step off-cycle\n",
    "# if isCCLE:\n",
    "#     # write samples to Sequencing table, copy bam files to internal storage bucket:\n",
    "#     rnasamples, cmds = loading.addSamplesToGumbo(rnasamples, 'rna', RNA_GCS_PATH)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All RNAseq sequencingIDs in gumbo that are not in the RNAseq terra workspace yet are considered \"new\" for the current release. Here we add them to the terra processing workspace as a sample set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    from depmap_omics_upload import loading\n",
    "    # load new rna samples from gumbo to RNAseq terra workspace\n",
    "    loading.addSamplesToDepMapWorkspace('rna', env_config.RNAWORKSPACE, samplesetname=constants.SAMPLESETNAME, add_to_samplesets=[\"all_stranded\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run pipeline on Terra\n",
    "\n",
    "We are using Dalmatian to send requests to Terra. See [our readme](https://github.com/broadinstitute/depmap_omics/blob/master/documentation/DepMap_processing_pipeline.md) for detailed breakdown of the subtasks in our RNAseq pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For non internal users, your Terra workspace needs to be correctly setup:\n",
    "\n",
    "Please follow instructions in the readme and make sure that you created your sampleset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"running Terra pipeline\")\n",
    "refwm = dm.WorkspaceManager(env_config.RNAWORKSPACE)\n",
    "submission_id = refwm.create_submission(\"RNA_pipeline\", constants.SAMPLESETNAME,'sample_set',expression='this.samples')\n",
    "await terra.waitForSubmission(env_config.RNAWORKSPACE, submission_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assuming all new samples are stranded\n",
    "submission_id = refwm.create_submission(\"RNA_stranded_rsem_rnaseqc2\", constants.SAMPLESETNAME,'sample_set',expression='this.samples')\n",
    "await terra.waitForSubmission(env_config.RNAWORKSPACE, submission_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"Aggregate_Fusion_Calls\", constants.SAMPLESETNAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"rsem_aggregate_results_with_length\", 'all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"rsem_aggregate_results_with_length_stranded\", 'all_stranded')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Postprocessing on local\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression post processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    await dm_omics.expressionPostProcessing(samplesetname=constants.SAMPLESETNAME, recompute_ssgsea=True, compute_enrichment=True)\n",
    "else:\n",
    "    await expressions.postProcessing(refworkspace=env_config.RNAWORKSPACE, samplesetname=constants.SAMPLESETNAME, recompute_ssgsea=False, compute_enrichment=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fusion post processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "old_fusion_fn = \"\"\n",
    "new_fusion_fn = \"\"\n",
    "\n",
    "new = pd.read_csv(new_fusion_fn, sep='\\t')\n",
    "old = pd.read_csv(old_fusion_fn, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([old, new], ignore_index=True).to_csv(\"fusion_stitched_\" + constants.SAMPLESETNAME + \".csv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    await dm_omics.fusionPostProcessing(samplesetname=constants.SAMPLESETNAME)\n",
    "else:\n",
    "    await fusions.postProcessing(env_config.RNAWORKSPACE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cleaning workspaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from depmap_omics_upload.mgenepy import terra as terra_cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if doCleanup:\n",
    "    print(\"cleaning workspaces\")\n",
    "    torm = await terra_cleanup.deleteHeavyFiles(env_config.RNAWORKSPACE)\n",
    "    h.parrun(['gsutil rm '+i for i in torm], cores=8)\n",
    "    terra_cleanup.removeFromFailedWorkflows(env_config.RNAWORKSPACE, dryrun=False, everythingFor=[])\n",
    "    print(\"done\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "229.391px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 371.4,
   "position": {
    "height": "393.4px",
    "left": "657px",
    "right": "20px",
    "top": "73px",
    "width": "351px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "1f170dacc8a54decc88aa6dec68bcf2a5d65c3893a5412f67865d27128dae76b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
