{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WGS PiPeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1002\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1002' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n",
       "  var css_urls = [];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1002\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1002' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n  var css_urls = [];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import os.path\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "sys.path.insert(0, '..')\n",
    "\n",
    "from src.CCLE_postp_function import *\n",
    "from JKBio import Datanalytics as da \n",
    "from JKBio import TerraFunction as terra\n",
    "from JKBio import Helper as h\n",
    "from JKBio.helper.google_sheet import GSheet\n",
    "from gsheets import Sheets\n",
    "from taigapy import TaigaClient\n",
    "import dalmatian as dm\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "from bokeh.plotting import *\n",
    "from bokeh.models import HoverTool\n",
    "from collections import OrderedDict\n",
    "from IPython.display import Image,display\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "tc = TaigaClient()\n",
    "output_notebook()\n",
    "\n",
    "my_id = '~/.client_secret.json'\n",
    "mystorage_id = \"~/.storage.json\"\n",
    "sheets = Sheets.from_files(my_id, mystorage_id)\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## boot up\n",
    "\n",
    "we are instanciating all the parameters needed for this pipeline to run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesetname = \"20Q4\"\n",
    "prevname=\"20Q3\"\n",
    "prevversion = 0\n",
    "prevprevname ='20Q2'\n",
    "prevprevversion= 0\n",
    "virtual_public='public-20q3-3d35'\n",
    "virtual_dmc='dmc-20q3-033d'\n",
    "virtual_internal='internal-20q3-00d0'\n",
    "\n",
    "workspace1=\"terra-broad-cancer-prod/DepMap_WGS\"\n",
    "workspace2=\"terra-broad-cancer-prod/Getz_IBM_CellLines_WGS\"\n",
    "\n",
    "\n",
    "refworkspace=\"broad-firecloud-ccle/DepMap_WGS_CN\"\n",
    "cgaworkspace=\"broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline-wgs\"\n",
    "\n",
    "source1=\"ccle\"\n",
    "source2=\"ibm\"\n",
    "\n",
    "refsheet_url = \"https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE\"\n",
    "sheeturl = \"https://docs.google.com/spreadsheets/d/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4\"\n",
    "potential_list_url = \"https://docs.google.com/spreadsheets/d/1BEgH03V4OmGhYeciLCZV00h6hp3WkO0basahS93akCE\"\n",
    "release = samplesetname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "wm1 = dm.WorkspaceManager(workspace1)\n",
    "wm2 = dm.WorkspaceManager(workspace2)\n",
    "\n",
    "refwm = dm.WorkspaceManager(refworkspace).disable_hound()\n",
    "cgawm = dm.WorkspaceManager(cgaworkspace).disable_hound()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_to_change = {'from_arxspan_id': 'participant'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [],
   "source": [
    "potential_list = sheets.get(potential_list_url).sheets[0].to_frame().values.T[0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n"
     ]
    }
   ],
   "source": [
    "ccle_refsamples = sheets.get(refsheet_url).sheets[0].to_frame().set_index('cds_sample_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding new data\n",
    "\n",
    "We are looking for new samples in a range of workspaces.\n",
    "\n",
    "They are quite messy and might contains duplicates, contain broken file paths...\n",
    "\n",
    "- We are thus looking at the bam files one by one and comparing them with our own bams. \n",
    "- We remove broken files, duplicates and add new version of a cell line's bam if we find some."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "refsamples is overrided by a refurl\n",
      "Getting sample infos...\n",
      "\n",
      "The shape of the sample tsv from <dalmatian.wmanager.WorkspaceManager terra-broad-cancer-prod/DepMap_WGS>: (25, 223)\n",
      "Identifying any true duplicates by checking file hashes (this runs for each data source)...\n",
      "This step can take a while as we need to use gsutil to check the size of each potential duplicate...\n",
      "listing files in gs\n",
      "These 0 bam file path do not exist: set()\n",
      "listing files in gs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:231: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  if extract['size'] not in samples.columns or recomputesize:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "listing files in gs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:233: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  for k, val in samples.iterrows():\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.75\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:242: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples[extract['release_date']] = list(h.datetoint(samples[extract['release_date']].values))\n",
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:243: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:246: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples[extract['patient_id']] = ['PT-' + h.randomString(stringLength=8, stype='all', withdigits=True) for _ in range(len(samples))]\n",
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:247: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples.reset_index(drop=True, inplace=True)\n",
      "/home/jeremie/ccle_processing/src/CCLE_postp_function.py:249: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "listing files in gs\n",
      "listing files in gs\n",
      "we found and removed 0 samples which did not match our id names: ['ACH-', 'CDS-']\n",
      "found 0 likely replicate\n",
      "listing files in gs\n",
      "Len of samples before removal: 21\n",
      "Dups from this workspace has len 0:\n",
      " []\n",
      "Len of samples after removal: 21\n",
      "\n",
      "The shape of the sample tsv from <dalmatian.wmanager.WorkspaceManager terra-broad-cancer-prod/Getz_IBM_CellLines_WGS>: (4, 233)\n",
      "Identifying any true duplicates by checking file hashes (this runs for each data source)...\n",
      "This step can take a while as we need to use gsutil to check the size of each potential duplicate...\n",
      "listing files in gs\n",
      "These 0 bam file path do not exist: set()\n",
      "listing files in gs\n",
      "listing files in gs\n",
      "listing files in gs\n",
      "listing files in gs\n",
      "we found and removed 0 samples which did not match our id names: ['ACH-', 'CDS-']\n",
      "found 0 likely replicate\n",
      "listing files in gs\n",
      "Len of samples before removal: 3\n",
      "Dups from this workspace has len 0:\n",
      " []\n",
      "Len of samples after removal: 3\n",
      "we had 0 duplicates in the release buckets\n",
      "found 0 matched normals\n"
     ]
    }
   ],
   "source": [
    "# we will be missing \"primary disease\",\"sm_id\", \"cellosaurus_id\", \"gender, \"age\", \"primary_site\", \"primary_disease\", \"subtype\", \"subsubtype\", \"origin\", \"comments\"\n",
    "#when SMid: match== \n",
    "samples, pairs, noarxspan = GetNewCellLinesFromWorkspaces(refworkspace, stype='wgs', refurl=refsheet_url, wmfroms = [workspace1, workspace2], sources=[source1, source2], match=['ACH-','CDS-'], participantslicepos=10, accept_unknowntypes=True, extract=extract_to_change, recomputedate=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>internal_bam_filepath</th>\n",
       "      <th>internal_bai_filepath</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>arxspan_id</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>crc32c_hash</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS-ymRIxH</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>HCC-1588</td>\n",
       "      <td>HCC-1588</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-7zhvxCOS</td>\n",
       "      <td>2ldWlQ==</td>\n",
       "      <td>108374791873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-5vtea4</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO83_1</td>\n",
       "      <td>MERO83</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-Vd7k8elh</td>\n",
       "      <td>nRYbAw==</td>\n",
       "      <td>116306111418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-X7YJtK</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO95_1</td>\n",
       "      <td>MERO95</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-FJCfUk9B</td>\n",
       "      <td>YkQgKw==</td>\n",
       "      <td>117193717024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-LvlS3j</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>TASK1_1</td>\n",
       "      <td>TASK1</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-ZxcFHknh</td>\n",
       "      <td>Xi90TA==</td>\n",
       "      <td>118399266655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-R22IHj</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>MONO-MAC-1</td>\n",
       "      <td>MONO-MAC-1</td>\n",
       "      <td>734878</td>\n",
       "      <td>PT-G7iG07aJ</td>\n",
       "      <td>y2QITw==</td>\n",
       "      <td>152230340144</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        internal_bam_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "                                        internal_bai_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "           stripped_cell_line_name  arxspan_id  sequencing_date  \\\n",
       "sample_id                                                         \n",
       "CDS-ymRIxH                HCC-1588    HCC-1588           737552   \n",
       "CDS-5vtea4                MERO83_1      MERO83           737547   \n",
       "CDS-X7YJtK                MERO95_1      MERO95           737547   \n",
       "CDS-LvlS3j                 TASK1_1       TASK1           737547   \n",
       "CDS-R22IHj              MONO-MAC-1  MONO-MAC-1           734878   \n",
       "\n",
       "           participant_id crc32c_hash          size  \n",
       "sample_id                                            \n",
       "CDS-ymRIxH    PT-7zhvxCOS    2ldWlQ==  108374791873  \n",
       "CDS-5vtea4    PT-Vd7k8elh    nRYbAw==  116306111418  \n",
       "CDS-X7YJtK    PT-FJCfUk9B    YkQgKw==  117193717024  \n",
       "CDS-LvlS3j    PT-ZxcFHknh    Xi90TA==  118399266655  \n",
       "CDS-R22IHj    PT-G7iG07aJ    y2QITw==  152230340144  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noarxspan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "noarxspan.to_csv('temp/noarxspan_wgs_'+release+'.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### finding back arxspan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "noarxspan['arxspan_id'] = ['ACH-001078', 'ACH-001560', 'ACH-001562', 'ACH-001670', 'ACH-001129']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "we found and removed 0 samples which did not match our id names: ['ACH', 'CDS']\n",
      "found 1 likely replicate\n",
      "listing files in gs\n",
      "Len of samples before removal: 5\n",
      "Dups from this workspace has len 0:\n",
      " []\n",
      "Len of samples after removal: 5\n"
     ]
    }
   ],
   "source": [
    "noarxspan = resolveFromWorkspace(noarxspan, refsamples = ccle_refsamples[ccle_refsamples['datatype'] == 'wgs'], match = ['ACH','CDS'], participantslicepos = 10, accept_unknowntypes = True, extract = extract_to_change)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>internal_bam_filepath</th>\n",
       "      <th>internal_bai_filepath</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>arxspan_id</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>crc32c_hash</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS-ymRIxH</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>HCC-1588</td>\n",
       "      <td>ACH-001078</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-7zhvxCOS</td>\n",
       "      <td>2ldWlQ==</td>\n",
       "      <td>108374791873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-5vtea4</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO83_1</td>\n",
       "      <td>ACH-001560</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-Vd7k8elh</td>\n",
       "      <td>nRYbAw==</td>\n",
       "      <td>116306111418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-X7YJtK</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO95_1</td>\n",
       "      <td>ACH-001562</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-FJCfUk9B</td>\n",
       "      <td>YkQgKw==</td>\n",
       "      <td>117193717024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-LvlS3j</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>TASK1_1</td>\n",
       "      <td>ACH-001670</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-ZxcFHknh</td>\n",
       "      <td>Xi90TA==</td>\n",
       "      <td>118399266655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-R22IHj</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>MONO-MAC-1</td>\n",
       "      <td>ACH-001129</td>\n",
       "      <td>734878</td>\n",
       "      <td>PT-G7iG07aJ</td>\n",
       "      <td>y2QITw==</td>\n",
       "      <td>152230340144</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        internal_bam_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "                                        internal_bai_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "           stripped_cell_line_name  arxspan_id  sequencing_date  \\\n",
       "sample_id                                                         \n",
       "CDS-ymRIxH                HCC-1588  ACH-001078           737552   \n",
       "CDS-5vtea4                MERO83_1  ACH-001560           737547   \n",
       "CDS-X7YJtK                MERO95_1  ACH-001562           737547   \n",
       "CDS-LvlS3j                 TASK1_1  ACH-001670           737547   \n",
       "CDS-R22IHj              MONO-MAC-1  ACH-001129           734878   \n",
       "\n",
       "           participant_id crc32c_hash          size  \n",
       "sample_id                                            \n",
       "CDS-ymRIxH    PT-7zhvxCOS    2ldWlQ==  108374791873  \n",
       "CDS-5vtea4    PT-Vd7k8elh    nRYbAw==  116306111418  \n",
       "CDS-X7YJtK    PT-FJCfUk9B    YkQgKw==  117193717024  \n",
       "CDS-LvlS3j    PT-ZxcFHknh    Xi90TA==  118399266655  \n",
       "CDS-R22IHj    PT-G7iG07aJ    y2QITw==  152230340144  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noarxspan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.concat([samples, noarxspan], sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n",
      "we had 0 duplicates in the release buckets\n"
     ]
    }
   ],
   "source": [
    "samples = assessAllSamples(samples, ccle_refsamples, stype='wgs', rename={}, extract={})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>internal_bam_filepath</th>\n",
       "      <th>internal_bai_filepath</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>arxspan_id</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>crc32c_hash</th>\n",
       "      <th>size</th>\n",
       "      <th>datatype</th>\n",
       "      <th>version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS-ttmLuC</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>C396</td>\n",
       "      <td>ACH-001818</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-oZH9BTUX</td>\n",
       "      <td>wP5Uig==</td>\n",
       "      <td>118951856040</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-VVWDLi</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>CHP134</td>\n",
       "      <td>ACH-001338</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-sqRdd0QF</td>\n",
       "      <td>TKjC7Q==</td>\n",
       "      <td>111261029620</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-cjMTSm</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>HCC1739BL</td>\n",
       "      <td>ACH-001828</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-OCJbBZ3S</td>\n",
       "      <td>qID1QA==</td>\n",
       "      <td>113315538347</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-AQpAIt</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>KARPAS1718</td>\n",
       "      <td>ACH-001533</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-ceOWs8HZ</td>\n",
       "      <td>vaScFg==</td>\n",
       "      <td>114662826746</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-cXQOkP</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>KOSC2CL343</td>\n",
       "      <td>ACH-001543</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-gCMWGzPt</td>\n",
       "      <td>gRp/bw==</td>\n",
       "      <td>96804199449</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-YLQJ19</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>M140325</td>\n",
       "      <td>ACH-002511</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-ELEMMjP8</td>\n",
       "      <td>87LwZQ==</td>\n",
       "      <td>125056253166</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-AqZLna</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>M1610113</td>\n",
       "      <td>ACH-002512</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-pUBtKkYR</td>\n",
       "      <td>7TbnGw==</td>\n",
       "      <td>110877810899</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-3PGQ84</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MEGA2</td>\n",
       "      <td>ACH-001553</td>\n",
       "      <td>737560</td>\n",
       "      <td>PT-SMrpbPt7</td>\n",
       "      <td>GGyO/Q==</td>\n",
       "      <td>100818382371</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-nKfczI</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MEL270</td>\n",
       "      <td>ACH-002014</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-RkMpD0qb</td>\n",
       "      <td>OHynTg==</td>\n",
       "      <td>109896913581</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-deAUlH</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MM28__</td>\n",
       "      <td>ACH-002013</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-ws3gMJ8I</td>\n",
       "      <td>BVyqdQ==</td>\n",
       "      <td>118332580889</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-exVsHa</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>RAMOS</td>\n",
       "      <td>ACH-001636</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-aFGn7OAh</td>\n",
       "      <td>YN5zyw==</td>\n",
       "      <td>133789949053</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-qyMeyg</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>RPMI2650</td>\n",
       "      <td>ACH-001385</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-g7W1qRmq</td>\n",
       "      <td>s6C72w==</td>\n",
       "      <td>103759426410</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-M1FWuz</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>SAT</td>\n",
       "      <td>ACH-001641</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-uOqhQJIC</td>\n",
       "      <td>w6hcug==</td>\n",
       "      <td>114090564340</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-KA0Ncu</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>SCC3</td>\n",
       "      <td>ACH-001642</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-CF1O2RQl</td>\n",
       "      <td>9wgf2g==</td>\n",
       "      <td>132865094686</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-FbQkW1</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>SNU2372_1</td>\n",
       "      <td>ACH-001662</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-bdAvWDQz</td>\n",
       "      <td>ysP94Q==</td>\n",
       "      <td>108771829505</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-allTcx</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>SW954</td>\n",
       "      <td>ACH-001400</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-bFHPciIP</td>\n",
       "      <td>ooDiUg==</td>\n",
       "      <td>123447299572</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-uYOFVp</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>TKKK</td>\n",
       "      <td>ACH-001863</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-2Cb1QaHZ</td>\n",
       "      <td>WpyzsA==</td>\n",
       "      <td>113987500107</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-JO3DcF</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>UPCISCC131</td>\n",
       "      <td>ACH-001699</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-79r1UuWc</td>\n",
       "      <td>EsEFRw==</td>\n",
       "      <td>119534283403</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-WgAbdT</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>VAL</td>\n",
       "      <td>ACH-001703</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-iFueiW0E</td>\n",
       "      <td>NRAWDA==</td>\n",
       "      <td>118478462240</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-vp6X0E</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>VP267</td>\n",
       "      <td>ACH-001705</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-gkuA6nS5</td>\n",
       "      <td>XIf9Ew==</td>\n",
       "      <td>114853944490</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-bTfRlF</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>WAOSEL</td>\n",
       "      <td>ACH-001707</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-heEmqvCI</td>\n",
       "      <td>AMuCew==</td>\n",
       "      <td>90677570187</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-OQITe6</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>CORL26</td>\n",
       "      <td>ACH-001046</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-Z2Fq3yte</td>\n",
       "      <td>7+vYIA==</td>\n",
       "      <td>100035318330</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-753lzg</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>Set2_</td>\n",
       "      <td>ACH-000195</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-Ftjkb7Jw</td>\n",
       "      <td>IuFwxQ==</td>\n",
       "      <td>109584388878</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-HoA5rp</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>UPCISCC090</td>\n",
       "      <td>ACH-001227</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-mkNL4JuM</td>\n",
       "      <td>IZjIhA==</td>\n",
       "      <td>117747270799</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-ymRIxH</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>HCC-1588</td>\n",
       "      <td>ACH-001078</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-epLMsCoH</td>\n",
       "      <td>2ldWlQ==</td>\n",
       "      <td>108374791873</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-5vtea4</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO83_1</td>\n",
       "      <td>ACH-001560</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-75blYymy</td>\n",
       "      <td>nRYbAw==</td>\n",
       "      <td>116306111418</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-X7YJtK</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MERO95_1</td>\n",
       "      <td>ACH-001562</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-qbPTkrGa</td>\n",
       "      <td>YkQgKw==</td>\n",
       "      <td>117193717024</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-LvlS3j</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>TASK1_1</td>\n",
       "      <td>ACH-001670</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-eRPKQiFq</td>\n",
       "      <td>Xi90TA==</td>\n",
       "      <td>118399266655</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-R22IHj</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>MONO-MAC-1</td>\n",
       "      <td>ACH-001129</td>\n",
       "      <td>734878</td>\n",
       "      <td>PT-EDkpMwcH</td>\n",
       "      <td>y2QITw==</td>\n",
       "      <td>152230340144</td>\n",
       "      <td>wgs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        internal_bam_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ttmLuC  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-VVWDLi  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-cjMTSm  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-AQpAIt  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-cXQOkP  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-YLQJ19  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-AqZLna  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-3PGQ84  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-nKfczI  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-deAUlH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-exVsHa  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-qyMeyg  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-M1FWuz  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-KA0Ncu  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-FbQkW1  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-allTcx  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-uYOFVp  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-JO3DcF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-WgAbdT  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-vp6X0E  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-bTfRlF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-OQITe6  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-753lzg  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-HoA5rp  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "                                        internal_bai_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-ttmLuC  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-VVWDLi  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-cjMTSm  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-AQpAIt  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-cXQOkP  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-YLQJ19  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-AqZLna  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-3PGQ84  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-nKfczI  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-deAUlH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-exVsHa  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-qyMeyg  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-M1FWuz  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-KA0Ncu  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-FbQkW1  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-allTcx  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-uYOFVp  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-JO3DcF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-WgAbdT  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-vp6X0E  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-bTfRlF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-OQITe6  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-753lzg  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-HoA5rp  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "CDS-ymRIxH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-5vtea4  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-X7YJtK  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-LvlS3j  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-R22IHj  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "           stripped_cell_line_name  arxspan_id  sequencing_date  \\\n",
       "sample_id                                                         \n",
       "CDS-ttmLuC                    C396  ACH-001818           737547   \n",
       "CDS-VVWDLi                  CHP134  ACH-001338           737547   \n",
       "CDS-cjMTSm               HCC1739BL  ACH-001828           737552   \n",
       "CDS-AQpAIt              KARPAS1718  ACH-001533           737552   \n",
       "CDS-cXQOkP              KOSC2CL343  ACH-001543           737547   \n",
       "CDS-YLQJ19                 M140325  ACH-002511           737552   \n",
       "CDS-AqZLna                M1610113  ACH-002512           737547   \n",
       "CDS-3PGQ84                   MEGA2  ACH-001553           737560   \n",
       "CDS-nKfczI                  MEL270  ACH-002014           737547   \n",
       "CDS-deAUlH                  MM28__  ACH-002013           737547   \n",
       "CDS-exVsHa                   RAMOS  ACH-001636           737552   \n",
       "CDS-qyMeyg                RPMI2650  ACH-001385           737547   \n",
       "CDS-M1FWuz                     SAT  ACH-001641           737547   \n",
       "CDS-KA0Ncu                    SCC3  ACH-001642           737547   \n",
       "CDS-FbQkW1               SNU2372_1  ACH-001662           737547   \n",
       "CDS-allTcx                   SW954  ACH-001400           737547   \n",
       "CDS-uYOFVp                    TKKK  ACH-001863           737547   \n",
       "CDS-JO3DcF              UPCISCC131  ACH-001699           737547   \n",
       "CDS-WgAbdT                     VAL  ACH-001703           737547   \n",
       "CDS-vp6X0E                   VP267  ACH-001705           737552   \n",
       "CDS-bTfRlF                  WAOSEL  ACH-001707           737552   \n",
       "CDS-OQITe6                  CORL26  ACH-001046           737546   \n",
       "CDS-753lzg                   Set2_  ACH-000195           737546   \n",
       "CDS-HoA5rp              UPCISCC090  ACH-001227           737546   \n",
       "CDS-ymRIxH                HCC-1588  ACH-001078           737552   \n",
       "CDS-5vtea4                MERO83_1  ACH-001560           737547   \n",
       "CDS-X7YJtK                MERO95_1  ACH-001562           737547   \n",
       "CDS-LvlS3j                 TASK1_1  ACH-001670           737547   \n",
       "CDS-R22IHj              MONO-MAC-1  ACH-001129           734878   \n",
       "\n",
       "           participant_id crc32c_hash          size datatype  version  \n",
       "sample_id                                                              \n",
       "CDS-ttmLuC    PT-oZH9BTUX    wP5Uig==  118951856040      wgs        1  \n",
       "CDS-VVWDLi    PT-sqRdd0QF    TKjC7Q==  111261029620      wgs        1  \n",
       "CDS-cjMTSm    PT-OCJbBZ3S    qID1QA==  113315538347      wgs        1  \n",
       "CDS-AQpAIt    PT-ceOWs8HZ    vaScFg==  114662826746      wgs        1  \n",
       "CDS-cXQOkP    PT-gCMWGzPt    gRp/bw==   96804199449      wgs        1  \n",
       "CDS-YLQJ19    PT-ELEMMjP8    87LwZQ==  125056253166      wgs        1  \n",
       "CDS-AqZLna    PT-pUBtKkYR    7TbnGw==  110877810899      wgs        1  \n",
       "CDS-3PGQ84    PT-SMrpbPt7    GGyO/Q==  100818382371      wgs        1  \n",
       "CDS-nKfczI    PT-RkMpD0qb    OHynTg==  109896913581      wgs        1  \n",
       "CDS-deAUlH    PT-ws3gMJ8I    BVyqdQ==  118332580889      wgs        1  \n",
       "CDS-exVsHa    PT-aFGn7OAh    YN5zyw==  133789949053      wgs        1  \n",
       "CDS-qyMeyg    PT-g7W1qRmq    s6C72w==  103759426410      wgs        1  \n",
       "CDS-M1FWuz    PT-uOqhQJIC    w6hcug==  114090564340      wgs        1  \n",
       "CDS-KA0Ncu    PT-CF1O2RQl    9wgf2g==  132865094686      wgs        1  \n",
       "CDS-FbQkW1    PT-bdAvWDQz    ysP94Q==  108771829505      wgs        1  \n",
       "CDS-allTcx    PT-bFHPciIP    ooDiUg==  123447299572      wgs        1  \n",
       "CDS-uYOFVp    PT-2Cb1QaHZ    WpyzsA==  113987500107      wgs        1  \n",
       "CDS-JO3DcF    PT-79r1UuWc    EsEFRw==  119534283403      wgs        1  \n",
       "CDS-WgAbdT    PT-iFueiW0E    NRAWDA==  118478462240      wgs        1  \n",
       "CDS-vp6X0E    PT-gkuA6nS5    XIf9Ew==  114853944490      wgs        1  \n",
       "CDS-bTfRlF    PT-heEmqvCI    AMuCew==   90677570187      wgs        1  \n",
       "CDS-OQITe6    PT-Z2Fq3yte    7+vYIA==  100035318330      wgs        1  \n",
       "CDS-753lzg    PT-Ftjkb7Jw    IuFwxQ==  109584388878      wgs        1  \n",
       "CDS-HoA5rp    PT-mkNL4JuM    IZjIhA==  117747270799      wgs        1  \n",
       "CDS-ymRIxH    PT-epLMsCoH    2ldWlQ==  108374791873      wgs        1  \n",
       "CDS-5vtea4    PT-75blYymy    nRYbAw==  116306111418      wgs        1  \n",
       "CDS-X7YJtK    PT-qbPTkrGa    YkQgKw==  117193717024      wgs        1  \n",
       "CDS-LvlS3j    PT-eRPKQiFq    Xi90TA==  118399266655      wgs        1  \n",
       "CDS-R22IHj    PT-EDkpMwcH    y2QITw==  152230340144      wgs        2  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## getting the addtional data and writing it here in the right order 'as shown above'\n",
    "- use the stripped_cell_line_name to find the samples on https://docs.google.com/spreadsheets/d/1uqCOos-T9EMQU7y2ZUw4Nm84opU5fIT1y7jet1vnScE/edit#gid=356471436. \n",
    "- Make sure that we don't have duplicate cell lines in there. Otherwise, use the duplicate renaming function\n",
    "- copy Primary Site, Primary Disease, Subtype, Comments, Disease Sub-subtype, if they exist. (sometimes subtype and subsubtype are the same.. don't use subsubtype then.\n",
    "- look for the cell line in cellosaurus, you might need to use one of the aliases given in master depmap pv..\n",
    "- copy  cellosaurus_id gender age info or write 'U' if they don't exist. 'can be a number or {Embryonic, Children, Adult, Fetus, U} \n",
    "- check that it does not say this cell line is not a duplicate from another cell line\n",
    "- check that if it says this cell line is derived/children/father/samepatient from other cell lines, and that if we have any of the other cell lines, that the patient id is changed to be the same one for all (be sure that you are updating everywhere these patient ids are used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If I have a previous samples I can update unknown data directly\n",
    "index=[]\n",
    "notfound=[]\n",
    "toupdate = {\"sex\":[],\n",
    "\"primary_disease\":[],\n",
    "\"cellosaurus_id\":[],\n",
    "\"age\":[],\n",
    "\"primary_site\":[],\n",
    "\"subtype\":[],\n",
    "\"subsubtype\":[],\n",
    "\"comments\":[],\n",
    "\"stripped_cell_line_name\":[],\n",
    "\"parent_cell_line\":[''],\n",
    "\"matched_normal\":[''],\n",
    "\"participant_id\":[]}\n",
    "for k, val in samples.iterrows():\n",
    "    dat = ccle_refsamples[ccle_refsamples['arxspan_id']==val['arxspan_id']]\n",
    "    if len(dat)>0:\n",
    "        index.append(k)\n",
    "        for k, v in toupdate.items():\n",
    "            toupdate[k].append(dat[k].tolist()[0])\n",
    "    else:\n",
    "        notfound.append(k)\n",
    "# doing so..\n",
    "for k, v in toupdate.items():\n",
    "    samples.loc[index,k] =v\n",
    "len(samples.loc[notfound].patient_id),samples.loc[notfound].patient_id.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>internal_bam_filepath</th>\n",
       "      <th>internal_bai_filepath</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>arxspan_id</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>crc32c_hash</th>\n",
       "      <th>size</th>\n",
       "      <th>datatype</th>\n",
       "      <th>version</th>\n",
       "      <th>sex</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>cellosaurus_id</th>\n",
       "      <th>age</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subsubtype</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS-cjMTSm</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>HCC1739BL</td>\n",
       "      <td>ACH-001828</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-OCJbBZ3S</td>\n",
       "      <td>qID1QA==</td>\n",
       "      <td>113315538347</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-3PGQ84</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MEGA2</td>\n",
       "      <td>ACH-001553</td>\n",
       "      <td>737560</td>\n",
       "      <td>PT-SMrpbPt7</td>\n",
       "      <td>GGyO/Q==</td>\n",
       "      <td>100818382371</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-deAUlH</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>MM28__</td>\n",
       "      <td>ACH-002013</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-ws3gMJ8I</td>\n",
       "      <td>BVyqdQ==</td>\n",
       "      <td>118332580889</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-vp6X0E</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>VP267</td>\n",
       "      <td>ACH-001705</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-gkuA6nS5</td>\n",
       "      <td>XIf9Ew==</td>\n",
       "      <td>114853944490</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-bTfRlF</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...</td>\n",
       "      <td>WAOSEL</td>\n",
       "      <td>ACH-001707</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-heEmqvCI</td>\n",
       "      <td>AMuCew==</td>\n",
       "      <td>90677570187</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-OQITe6</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...</td>\n",
       "      <td>CORL26</td>\n",
       "      <td>ACH-001046</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-Z2Fq3yte</td>\n",
       "      <td>7+vYIA==</td>\n",
       "      <td>100035318330</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        internal_bam_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-cjMTSm  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-3PGQ84  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-deAUlH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-vp6X0E  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-bTfRlF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-OQITe6  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "                                        internal_bai_filepath  \\\n",
       "sample_id                                                       \n",
       "CDS-cjMTSm  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-3PGQ84  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-deAUlH  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-vp6X0E  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-bTfRlF  gs://fc-d2a92869-a028-40cf-a400-6ce6f7a5e92b/D...   \n",
       "CDS-OQITe6  gs://fc-49b66034-d2f1-4d65-84db-891ae1be11d4/G...   \n",
       "\n",
       "           stripped_cell_line_name  arxspan_id  sequencing_date  \\\n",
       "sample_id                                                         \n",
       "CDS-cjMTSm               HCC1739BL  ACH-001828           737552   \n",
       "CDS-3PGQ84                   MEGA2  ACH-001553           737560   \n",
       "CDS-deAUlH                  MM28__  ACH-002013           737547   \n",
       "CDS-vp6X0E                   VP267  ACH-001705           737552   \n",
       "CDS-bTfRlF                  WAOSEL  ACH-001707           737552   \n",
       "CDS-OQITe6                  CORL26  ACH-001046           737546   \n",
       "\n",
       "           participant_id crc32c_hash          size datatype  version  sex  \\\n",
       "sample_id                                                                    \n",
       "CDS-cjMTSm    PT-OCJbBZ3S    qID1QA==  113315538347      wgs        1  NaN   \n",
       "CDS-3PGQ84    PT-SMrpbPt7    GGyO/Q==  100818382371      wgs        1  NaN   \n",
       "CDS-deAUlH    PT-ws3gMJ8I    BVyqdQ==  118332580889      wgs        1  NaN   \n",
       "CDS-vp6X0E    PT-gkuA6nS5    XIf9Ew==  114853944490      wgs        1  NaN   \n",
       "CDS-bTfRlF    PT-heEmqvCI    AMuCew==   90677570187      wgs        1  NaN   \n",
       "CDS-OQITe6    PT-Z2Fq3yte    7+vYIA==  100035318330      wgs        1  NaN   \n",
       "\n",
       "           primary_disease cellosaurus_id  age primary_site subtype  \\\n",
       "sample_id                                                             \n",
       "CDS-cjMTSm             NaN            NaN  NaN          NaN     NaN   \n",
       "CDS-3PGQ84             NaN            NaN  NaN          NaN     NaN   \n",
       "CDS-deAUlH             NaN            NaN  NaN          NaN     NaN   \n",
       "CDS-vp6X0E             NaN            NaN  NaN          NaN     NaN   \n",
       "CDS-bTfRlF             NaN            NaN  NaN          NaN     NaN   \n",
       "CDS-OQITe6             NaN            NaN  NaN          NaN     NaN   \n",
       "\n",
       "           subsubtype comments  \n",
       "sample_id                       \n",
       "CDS-cjMTSm        NaN      NaN  \n",
       "CDS-3PGQ84        NaN      NaN  \n",
       "CDS-deAUlH        NaN      NaN  \n",
       "CDS-vp6X0E        NaN      NaN  \n",
       "CDS-bTfRlF        NaN      NaN  \n",
       "CDS-OQITe6        NaN      NaN  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for these samples I will need to check and manually add the data in the list \n",
    "samples.loc[notfound]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = changeCellLineNameInNewSet(new = samples, ref=ccle_refsamples, datatype=\"wgs\", dupdict=dup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "toupdate = {\"sex\":[\"Female\",\"Male\",\"Male\",\"Female\",\"Male\",\"Female\"],\n",
    "\"primary_disease\":[\"Breast Cancer\",\"Leukemia\",\"Eye Cancer\",\"Breast Cancer\",\"Leukemia\",\"Lung Cancer\"],\n",
    "\"cellosaurus_id\":[\"CVCL_3280\",\"CVCL_1832\",\"CVCL_4D15\",\"CVCL_2755\",\"CVCL_Y549\",\"CVCL_2410\"],\n",
    "\"age\":[51,24,72,48,\"Adult\",\"U\"],\n",
    "\"primary_site\":[\"pleural_effusion\",\"haematopoietic_and_lymphoid_tissue\",\"liver\",\"breast\",\"haematopoietic_and_lymphoid_tissue\",\"bone_marrow\"],\n",
    "\"subtype\":[\"Carcinoma\",\"CML\",\"Uveal Melanoma\",\"Carcinoma\",\"CLL\",\"Small Cell Lung Cancer (SCLC)\"],\n",
    "\"subsubtype\":[\"breast_ductal_carcinoma\",\"\",\"\",\"breast_ductal_carcinoma\",\"b_cell\",\"\"],\n",
    "\"comments\":[\"Transformant: NCBI_TaxID; 10376; Epstein-Barr virus (EBV). TNBC\", \"\",\"\",\"Derived from a breast cancer biopsy removed from a 48 year-old female. The histology of the breast cancer biopsy was defined as Ductal Grade III, tumour cells did not stain positive for oestrogen receptors. The VP267 cell line was derived from the same in\",\"\",\"\"],\n",
    "\"parent_cell_line\":['','','','','',''],\n",
    "\"matched_normal\":['','','','','',''],\n",
    "\"participant_id\":['PT-OCJbBZ3S','PT-SMrpbPt7',\"PT-ws3gMJ8I\",\"PT-n3CSJEGT\",'PT-y3RbI7uD',\"PT-Z2Fq3yte\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>cellosaurus_id</th>\n",
       "      <th>age</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subsubtype</th>\n",
       "      <th>comments</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>CVCL_3280</td>\n",
       "      <td>51</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>Carcinoma</td>\n",
       "      <td>breast_ductal_carcinoma</td>\n",
       "      <td>Transformant: NCBI_TaxID; 10376; Epstein-Barr ...</td>\n",
       "      <td>PT-OCJbBZ3S</td>\n",
       "      <td>HCC1739BL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_1832</td>\n",
       "      <td>24</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>CML</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PT-SMrpbPt7</td>\n",
       "      <td>MEGA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Male</td>\n",
       "      <td>Eye Cancer</td>\n",
       "      <td>CVCL_4D15</td>\n",
       "      <td>72</td>\n",
       "      <td>liver</td>\n",
       "      <td>Uveal Melanoma</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PT-ws3gMJ8I</td>\n",
       "      <td>MM28__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Female</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>CVCL_2755</td>\n",
       "      <td>48</td>\n",
       "      <td>breast</td>\n",
       "      <td>Carcinoma</td>\n",
       "      <td>breast_ductal_carcinoma</td>\n",
       "      <td>Derived from a breast cancer biopsy removed fr...</td>\n",
       "      <td>PT-n3CSJEGT</td>\n",
       "      <td>VP267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_Y549</td>\n",
       "      <td>Adult</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>CLL</td>\n",
       "      <td>b_cell</td>\n",
       "      <td></td>\n",
       "      <td>PT-y3RbI7uD</td>\n",
       "      <td>WAOSEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Female</td>\n",
       "      <td>Lung Cancer</td>\n",
       "      <td>CVCL_2410</td>\n",
       "      <td>U</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>Small Cell Lung Cancer (SCLC)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PT-Z2Fq3yte</td>\n",
       "      <td>CORL26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sex primary_disease cellosaurus_id    age  \\\n",
       "0  Female   Breast Cancer      CVCL_3280     51   \n",
       "1    Male        Leukemia      CVCL_1832     24   \n",
       "2    Male      Eye Cancer      CVCL_4D15     72   \n",
       "3  Female   Breast Cancer      CVCL_2755     48   \n",
       "4    Male        Leukemia      CVCL_Y549  Adult   \n",
       "5  Female     Lung Cancer      CVCL_2410      U   \n",
       "\n",
       "                         primary_site                        subtype  \\\n",
       "0                    pleural_effusion                      Carcinoma   \n",
       "1  haematopoietic_and_lymphoid_tissue                            CML   \n",
       "2                               liver                 Uveal Melanoma   \n",
       "3                              breast                      Carcinoma   \n",
       "4  haematopoietic_and_lymphoid_tissue                            CLL   \n",
       "5                         bone_marrow  Small Cell Lung Cancer (SCLC)   \n",
       "\n",
       "                subsubtype                                           comments  \\\n",
       "0  breast_ductal_carcinoma  Transformant: NCBI_TaxID; 10376; Epstein-Barr ...   \n",
       "1                                                                               \n",
       "2                                                                               \n",
       "3  breast_ductal_carcinoma  Derived from a breast cancer biopsy removed fr...   \n",
       "4                   b_cell                                                      \n",
       "5                                                                               \n",
       "\n",
       "  participant_id       name  \n",
       "0    PT-OCJbBZ3S  HCC1739BL  \n",
       "1    PT-SMrpbPt7      MEGA2  \n",
       "2    PT-ws3gMJ8I     MM28__  \n",
       "3    PT-n3CSJEGT      VP267  \n",
       "4    PT-y3RbI7uD     WAOSEL  \n",
       "5    PT-Z2Fq3yte     CORL26  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a  = pd.DataFrame(toupdate)\n",
    "a['name'] = samples.loc[notfound,\"stripped_cell_line_name\"].tolist()\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# updating..\n",
    "for k, v in toupdate.items():\n",
    "    samples.loc[notfound,k] =v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>internal_bam_filepath</th>\n",
       "      <th>internal_bai_filepath</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>arxspan_id</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>crc32c_hash</th>\n",
       "      <th>size</th>\n",
       "      <th>datatype</th>\n",
       "      <th>version</th>\n",
       "      <th>sex</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>cellosaurus_id</th>\n",
       "      <th>age</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subsubtype</th>\n",
       "      <th>comments</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS-ttmLuC</td>\n",
       "      <td>gs://cclebams/wgs/CDS-ttmLuC.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-ttmLuC.bai</td>\n",
       "      <td>C396</td>\n",
       "      <td>ACH-001818</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-oZH9BTUX</td>\n",
       "      <td>wP5Uig==</td>\n",
       "      <td>118951856040</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>U</td>\n",
       "      <td>Bone Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>bone</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Received from Academic lab (Gorlick, Montefior...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-VVWDLi</td>\n",
       "      <td>gs://cclebams/wgs/CDS-VVWDLi.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-VVWDLi.bai</td>\n",
       "      <td>CHP134</td>\n",
       "      <td>ACH-001338</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-sqRdd0QF</td>\n",
       "      <td>TKjC7Q==</td>\n",
       "      <td>111261029620</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Neuroblastoma</td>\n",
       "      <td>CVCL_1124</td>\n",
       "      <td>43952</td>\n",
       "      <td>autonomic_ganglia</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYCN amp; adrenergic ; 1p - (literature)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-cjMTSm</td>\n",
       "      <td>gs://cclebams/wgs/CDS-cjMTSm.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-cjMTSm.bai</td>\n",
       "      <td>HCC1739BL</td>\n",
       "      <td>ACH-001828</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-OCJbBZ3S</td>\n",
       "      <td>qID1QA==</td>\n",
       "      <td>113315538347</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>CVCL_3280</td>\n",
       "      <td>51</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>Carcinoma</td>\n",
       "      <td>breast_ductal_carcinoma</td>\n",
       "      <td>Transformant: NCBI_TaxID; 10376; Epstein-Barr ...</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-AQpAIt</td>\n",
       "      <td>gs://cclebams/wgs/CDS-AQpAIt.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-AQpAIt.bai</td>\n",
       "      <td>KARPAS1718</td>\n",
       "      <td>ACH-001533</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-ceOWs8HZ</td>\n",
       "      <td>vaScFg==</td>\n",
       "      <td>114662826746</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Lymphoma</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>non_hodgkin_lymphoma</td>\n",
       "      <td>splenic_lymphoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-cXQOkP</td>\n",
       "      <td>gs://cclebams/wgs/CDS-cXQOkP.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-cXQOkP.bai</td>\n",
       "      <td>KOSC2CL343</td>\n",
       "      <td>ACH-001543</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-gCMWGzPt</td>\n",
       "      <td>gRp/bw==</td>\n",
       "      <td>96804199449</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>51.0</td>\n",
       "      <td>upper_aerodigestive_tract</td>\n",
       "      <td>upper_aerodigestive_squamous</td>\n",
       "      <td>oral</td>\n",
       "      <td>Human Tumor cells derived from oral cancer at ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-YLQJ19</td>\n",
       "      <td>gs://cclebams/wgs/CDS-YLQJ19.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-YLQJ19.bai</td>\n",
       "      <td>M140325</td>\n",
       "      <td>ACH-002511</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-ELEMMjP8</td>\n",
       "      <td>87LwZQ==</td>\n",
       "      <td>125056253166</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Skin Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>64.0</td>\n",
       "      <td>skin</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>mucosal</td>\n",
       "      <td>Biopsy site - vagina; Received from Academic l...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-AqZLna</td>\n",
       "      <td>gs://cclebams/wgs/CDS-AqZLna.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-AqZLna.bai</td>\n",
       "      <td>M1610113</td>\n",
       "      <td>ACH-002512</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-pUBtKkYR</td>\n",
       "      <td>7TbnGw==</td>\n",
       "      <td>110877810899</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Skin Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U</td>\n",
       "      <td>skin</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>acral</td>\n",
       "      <td>M1610113 (p9); Mutation: NRASG12R; biopsy site...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-3PGQ84</td>\n",
       "      <td>gs://cclebams/wgs/CDS-3PGQ84.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-3PGQ84.bai</td>\n",
       "      <td>MEGA2</td>\n",
       "      <td>ACH-001553</td>\n",
       "      <td>737560</td>\n",
       "      <td>PT-SMrpbPt7</td>\n",
       "      <td>GGyO/Q==</td>\n",
       "      <td>100818382371</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_1832</td>\n",
       "      <td>24</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>CML</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-nKfczI</td>\n",
       "      <td>gs://cclebams/wgs/CDS-nKfczI.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-nKfczI.bai</td>\n",
       "      <td>MEL270</td>\n",
       "      <td>ACH-002014</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-RkMpD0qb</td>\n",
       "      <td>OHynTg==</td>\n",
       "      <td>109896913581</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Eye Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>62.0</td>\n",
       "      <td>uvea</td>\n",
       "      <td>uveal_melanoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mel270, OMM23, and OMM25 are from the same pat...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-deAUlH</td>\n",
       "      <td>gs://cclebams/wgs/CDS-deAUlH.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-deAUlH.bai</td>\n",
       "      <td>MM28__</td>\n",
       "      <td>ACH-002013</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-ws3gMJ8I</td>\n",
       "      <td>BVyqdQ==</td>\n",
       "      <td>118332580889</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eye Cancer</td>\n",
       "      <td>CVCL_4D15</td>\n",
       "      <td>72</td>\n",
       "      <td>liver</td>\n",
       "      <td>Uveal Melanoma</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-exVsHa</td>\n",
       "      <td>gs://cclebams/wgs/CDS-exVsHa.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-exVsHa.bai</td>\n",
       "      <td>RAMOS</td>\n",
       "      <td>ACH-001636</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-aFGn7OAh</td>\n",
       "      <td>YN5zyw==</td>\n",
       "      <td>133789949053</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Lymphoma</td>\n",
       "      <td>U</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ascites</td>\n",
       "      <td>non_hodgkin_lymphoma</td>\n",
       "      <td>b_cell_burkitt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-qyMeyg</td>\n",
       "      <td>gs://cclebams/wgs/CDS-qyMeyg.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-qyMeyg.bai</td>\n",
       "      <td>RPMI2650</td>\n",
       "      <td>ACH-001385</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-g7W1qRmq</td>\n",
       "      <td>s6C72w==</td>\n",
       "      <td>103759426410</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "      <td>CVCL_1664</td>\n",
       "      <td>52.0</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>upper_aerodigestive_squamous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-M1FWuz</td>\n",
       "      <td>gs://cclebams/wgs/CDS-M1FWuz.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-M1FWuz.bai</td>\n",
       "      <td>SAT</td>\n",
       "      <td>ACH-001641</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-uOqhQJIC</td>\n",
       "      <td>w6hcug==</td>\n",
       "      <td>114090564340</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>U</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "      <td>CVCL_3160</td>\n",
       "      <td>U</td>\n",
       "      <td>upper_aerodigestive_tract</td>\n",
       "      <td>upper_aerodigestive_squamous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-KA0Ncu</td>\n",
       "      <td>gs://cclebams/wgs/CDS-KA0Ncu.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-KA0Ncu.bai</td>\n",
       "      <td>SCC3</td>\n",
       "      <td>ACH-001642</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-CF1O2RQl</td>\n",
       "      <td>9wgf2g==</td>\n",
       "      <td>132865094686</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Lymphoma</td>\n",
       "      <td>CVCL_1683</td>\n",
       "      <td>20.0</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>non_hodgkin_lymphoma</td>\n",
       "      <td>b_cell</td>\n",
       "      <td>Human monocytic cell line expressing Interleuk...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-FbQkW1</td>\n",
       "      <td>gs://cclebams/wgs/CDS-FbQkW1.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-FbQkW1.bai</td>\n",
       "      <td>SNU2372</td>\n",
       "      <td>ACH-001662</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-bdAvWDQz</td>\n",
       "      <td>ysP94Q==</td>\n",
       "      <td>108771829505</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>CVCL_T041</td>\n",
       "      <td>55</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>Breast Ductal Carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>breast cancer, Infiltrating ductal carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-allTcx</td>\n",
       "      <td>gs://cclebams/wgs/CDS-allTcx.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-allTcx.bai</td>\n",
       "      <td>SW954</td>\n",
       "      <td>ACH-001400</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-bFHPciIP</td>\n",
       "      <td>ooDiUg==</td>\n",
       "      <td>123447299572</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Cervical Cancer</td>\n",
       "      <td>CVCL_1732</td>\n",
       "      <td>86.0</td>\n",
       "      <td>cervix</td>\n",
       "      <td>cervical_squamous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Vulvar squamous cell carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-uYOFVp</td>\n",
       "      <td>gs://cclebams/wgs/CDS-uYOFVp.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-uYOFVp.bai</td>\n",
       "      <td>TKKK</td>\n",
       "      <td>ACH-001863</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-2Cb1QaHZ</td>\n",
       "      <td>WpyzsA==</td>\n",
       "      <td>113987500107</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Bile Duct Cancer</td>\n",
       "      <td>CVCL_5599</td>\n",
       "      <td>U</td>\n",
       "      <td>biliary_tract</td>\n",
       "      <td>cholangiocarcinoma</td>\n",
       "      <td>intrahepatic</td>\n",
       "      <td>Received from Academic lab (Bardeesy, MGH) - e...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-JO3DcF</td>\n",
       "      <td>gs://cclebams/wgs/CDS-JO3DcF.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-JO3DcF.bai</td>\n",
       "      <td>UPCISCC131</td>\n",
       "      <td>ACH-001699</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-79r1UuWc</td>\n",
       "      <td>EsEFRw==</td>\n",
       "      <td>119534283403</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>73.0</td>\n",
       "      <td>upper_aerodigestive_tract</td>\n",
       "      <td>upper_aerodigestive_squamous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-WgAbdT</td>\n",
       "      <td>gs://cclebams/wgs/CDS-WgAbdT.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-WgAbdT.bai</td>\n",
       "      <td>VAL</td>\n",
       "      <td>ACH-001703</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-iFueiW0E</td>\n",
       "      <td>NRAWDA==</td>\n",
       "      <td>118478462240</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Lymphoma</td>\n",
       "      <td>CVCL_1819</td>\n",
       "      <td>50.0</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>non_hodgkin_lymphoma</td>\n",
       "      <td>DLBCL</td>\n",
       "      <td>Bone Marrow metastatic site</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-vp6X0E</td>\n",
       "      <td>gs://cclebams/wgs/CDS-vp6X0E.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-vp6X0E.bai</td>\n",
       "      <td>VP267</td>\n",
       "      <td>ACH-001705</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-n3CSJEGT</td>\n",
       "      <td>XIf9Ew==</td>\n",
       "      <td>114853944490</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>CVCL_2755</td>\n",
       "      <td>48</td>\n",
       "      <td>breast</td>\n",
       "      <td>Carcinoma</td>\n",
       "      <td>breast_ductal_carcinoma</td>\n",
       "      <td>Derived from a breast cancer biopsy removed fr...</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-bTfRlF</td>\n",
       "      <td>gs://cclebams/wgs/CDS-bTfRlF.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-bTfRlF.bai</td>\n",
       "      <td>WAOSEL</td>\n",
       "      <td>ACH-001707</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-y3RbI7uD</td>\n",
       "      <td>AMuCew==</td>\n",
       "      <td>90677570187</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_Y549</td>\n",
       "      <td>Adult</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>CLL</td>\n",
       "      <td>b_cell</td>\n",
       "      <td></td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-OQITe6</td>\n",
       "      <td>gs://cclebams/wgs/CDS-OQITe6.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-OQITe6.bai</td>\n",
       "      <td>CORL26</td>\n",
       "      <td>ACH-001046</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-Z2Fq3yte</td>\n",
       "      <td>7+vYIA==</td>\n",
       "      <td>100035318330</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lung Cancer</td>\n",
       "      <td>CVCL_2410</td>\n",
       "      <td>U</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>Small Cell Lung Cancer (SCLC)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-753lzg</td>\n",
       "      <td>gs://cclebams/wgs/CDS-753lzg.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-753lzg.bai</td>\n",
       "      <td>SET2</td>\n",
       "      <td>ACH-000195</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-Ftjkb7Jw</td>\n",
       "      <td>IuFwxQ==</td>\n",
       "      <td>109584388878</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_2187</td>\n",
       "      <td>71.0</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCLE line</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-HoA5rp</td>\n",
       "      <td>gs://cclebams/wgs/CDS-HoA5rp.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-HoA5rp.bai</td>\n",
       "      <td>UPCISCC090</td>\n",
       "      <td>ACH-001227</td>\n",
       "      <td>737546</td>\n",
       "      <td>PT-mkNL4JuM</td>\n",
       "      <td>IZjIhA==</td>\n",
       "      <td>117747270799</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "      <td>U</td>\n",
       "      <td>44.0</td>\n",
       "      <td>upper_aerodigestive_tract</td>\n",
       "      <td>upper_aerodigestive_squamous</td>\n",
       "      <td>oral</td>\n",
       "      <td>Received from academic lab</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-ymRIxH</td>\n",
       "      <td>gs://cclebams/wgs/CDS-ymRIxH.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-ymRIxH.bai</td>\n",
       "      <td>HCC1588</td>\n",
       "      <td>ACH-001078</td>\n",
       "      <td>737552</td>\n",
       "      <td>PT-epLMsCoH</td>\n",
       "      <td>2ldWlQ==</td>\n",
       "      <td>108374791873</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Lung Cancer</td>\n",
       "      <td>CVCL_A351</td>\n",
       "      <td>63.0</td>\n",
       "      <td>lung</td>\n",
       "      <td>NSCLC</td>\n",
       "      <td>NSCLC_squamous</td>\n",
       "      <td>CCLE line; Identical lines: HCC-1588 and LS513...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-5vtea4</td>\n",
       "      <td>gs://cclebams/wgs/CDS-5vtea4.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-5vtea4.bai</td>\n",
       "      <td>MERO83</td>\n",
       "      <td>ACH-001560</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-75blYymy</td>\n",
       "      <td>nRYbAw==</td>\n",
       "      <td>116306111418</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Lung Cancer</td>\n",
       "      <td>CVCL_2595</td>\n",
       "      <td>U</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>mesothelioma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A human mesothelioma cell line derived from a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-X7YJtK</td>\n",
       "      <td>gs://cclebams/wgs/CDS-X7YJtK.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-X7YJtK.bai</td>\n",
       "      <td>MERO95</td>\n",
       "      <td>ACH-001562</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-qbPTkrGa</td>\n",
       "      <td>YkQgKw==</td>\n",
       "      <td>117193717024</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Lung Cancer</td>\n",
       "      <td>CVCL_2597</td>\n",
       "      <td>U</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>mesothelioma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A human mesothelioma cell line derived from a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-LvlS3j</td>\n",
       "      <td>gs://cclebams/wgs/CDS-LvlS3j.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-LvlS3j.bai</td>\n",
       "      <td>TASK1</td>\n",
       "      <td>ACH-001670</td>\n",
       "      <td>737547</td>\n",
       "      <td>PT-eRPKQiFq</td>\n",
       "      <td>Xi90TA==</td>\n",
       "      <td>118399266655</td>\n",
       "      <td>wgs</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>Brain Cancer</td>\n",
       "      <td>CVCL_3177</td>\n",
       "      <td>21.0</td>\n",
       "      <td>central_nervous_system</td>\n",
       "      <td>PNET</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human cell line with primitive neuroectodermal...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS-R22IHj</td>\n",
       "      <td>gs://cclebams/wgs/CDS-R22IHj.bam</td>\n",
       "      <td>gs://cclebams/wgs/CDS-R22IHj.bai</td>\n",
       "      <td>MONOMAC1</td>\n",
       "      <td>ACH-001129</td>\n",
       "      <td>734878</td>\n",
       "      <td>PT-EDkpMwcH</td>\n",
       "      <td>y2QITw==</td>\n",
       "      <td>152230340144</td>\n",
       "      <td>wgs</td>\n",
       "      <td>2</td>\n",
       "      <td>Male</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>CVCL_1425</td>\n",
       "      <td>64.0</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCLE line; MLL-AF4; p53 mut; RAS WT; FLT3-ITD;...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       internal_bam_filepath  \\\n",
       "sample_id                                      \n",
       "CDS-ttmLuC  gs://cclebams/wgs/CDS-ttmLuC.bam   \n",
       "CDS-VVWDLi  gs://cclebams/wgs/CDS-VVWDLi.bam   \n",
       "CDS-cjMTSm  gs://cclebams/wgs/CDS-cjMTSm.bam   \n",
       "CDS-AQpAIt  gs://cclebams/wgs/CDS-AQpAIt.bam   \n",
       "CDS-cXQOkP  gs://cclebams/wgs/CDS-cXQOkP.bam   \n",
       "CDS-YLQJ19  gs://cclebams/wgs/CDS-YLQJ19.bam   \n",
       "CDS-AqZLna  gs://cclebams/wgs/CDS-AqZLna.bam   \n",
       "CDS-3PGQ84  gs://cclebams/wgs/CDS-3PGQ84.bam   \n",
       "CDS-nKfczI  gs://cclebams/wgs/CDS-nKfczI.bam   \n",
       "CDS-deAUlH  gs://cclebams/wgs/CDS-deAUlH.bam   \n",
       "CDS-exVsHa  gs://cclebams/wgs/CDS-exVsHa.bam   \n",
       "CDS-qyMeyg  gs://cclebams/wgs/CDS-qyMeyg.bam   \n",
       "CDS-M1FWuz  gs://cclebams/wgs/CDS-M1FWuz.bam   \n",
       "CDS-KA0Ncu  gs://cclebams/wgs/CDS-KA0Ncu.bam   \n",
       "CDS-FbQkW1  gs://cclebams/wgs/CDS-FbQkW1.bam   \n",
       "CDS-allTcx  gs://cclebams/wgs/CDS-allTcx.bam   \n",
       "CDS-uYOFVp  gs://cclebams/wgs/CDS-uYOFVp.bam   \n",
       "CDS-JO3DcF  gs://cclebams/wgs/CDS-JO3DcF.bam   \n",
       "CDS-WgAbdT  gs://cclebams/wgs/CDS-WgAbdT.bam   \n",
       "CDS-vp6X0E  gs://cclebams/wgs/CDS-vp6X0E.bam   \n",
       "CDS-bTfRlF  gs://cclebams/wgs/CDS-bTfRlF.bam   \n",
       "CDS-OQITe6  gs://cclebams/wgs/CDS-OQITe6.bam   \n",
       "CDS-753lzg  gs://cclebams/wgs/CDS-753lzg.bam   \n",
       "CDS-HoA5rp  gs://cclebams/wgs/CDS-HoA5rp.bam   \n",
       "CDS-ymRIxH  gs://cclebams/wgs/CDS-ymRIxH.bam   \n",
       "CDS-5vtea4  gs://cclebams/wgs/CDS-5vtea4.bam   \n",
       "CDS-X7YJtK  gs://cclebams/wgs/CDS-X7YJtK.bam   \n",
       "CDS-LvlS3j  gs://cclebams/wgs/CDS-LvlS3j.bam   \n",
       "CDS-R22IHj  gs://cclebams/wgs/CDS-R22IHj.bam   \n",
       "\n",
       "                       internal_bai_filepath stripped_cell_line_name  \\\n",
       "sample_id                                                              \n",
       "CDS-ttmLuC  gs://cclebams/wgs/CDS-ttmLuC.bai                    C396   \n",
       "CDS-VVWDLi  gs://cclebams/wgs/CDS-VVWDLi.bai                  CHP134   \n",
       "CDS-cjMTSm  gs://cclebams/wgs/CDS-cjMTSm.bai               HCC1739BL   \n",
       "CDS-AQpAIt  gs://cclebams/wgs/CDS-AQpAIt.bai              KARPAS1718   \n",
       "CDS-cXQOkP  gs://cclebams/wgs/CDS-cXQOkP.bai              KOSC2CL343   \n",
       "CDS-YLQJ19  gs://cclebams/wgs/CDS-YLQJ19.bai                 M140325   \n",
       "CDS-AqZLna  gs://cclebams/wgs/CDS-AqZLna.bai                M1610113   \n",
       "CDS-3PGQ84  gs://cclebams/wgs/CDS-3PGQ84.bai                   MEGA2   \n",
       "CDS-nKfczI  gs://cclebams/wgs/CDS-nKfczI.bai                  MEL270   \n",
       "CDS-deAUlH  gs://cclebams/wgs/CDS-deAUlH.bai                  MM28__   \n",
       "CDS-exVsHa  gs://cclebams/wgs/CDS-exVsHa.bai                   RAMOS   \n",
       "CDS-qyMeyg  gs://cclebams/wgs/CDS-qyMeyg.bai                RPMI2650   \n",
       "CDS-M1FWuz  gs://cclebams/wgs/CDS-M1FWuz.bai                     SAT   \n",
       "CDS-KA0Ncu  gs://cclebams/wgs/CDS-KA0Ncu.bai                    SCC3   \n",
       "CDS-FbQkW1  gs://cclebams/wgs/CDS-FbQkW1.bai                 SNU2372   \n",
       "CDS-allTcx  gs://cclebams/wgs/CDS-allTcx.bai                   SW954   \n",
       "CDS-uYOFVp  gs://cclebams/wgs/CDS-uYOFVp.bai                    TKKK   \n",
       "CDS-JO3DcF  gs://cclebams/wgs/CDS-JO3DcF.bai              UPCISCC131   \n",
       "CDS-WgAbdT  gs://cclebams/wgs/CDS-WgAbdT.bai                     VAL   \n",
       "CDS-vp6X0E  gs://cclebams/wgs/CDS-vp6X0E.bai                   VP267   \n",
       "CDS-bTfRlF  gs://cclebams/wgs/CDS-bTfRlF.bai                  WAOSEL   \n",
       "CDS-OQITe6  gs://cclebams/wgs/CDS-OQITe6.bai                  CORL26   \n",
       "CDS-753lzg  gs://cclebams/wgs/CDS-753lzg.bai                    SET2   \n",
       "CDS-HoA5rp  gs://cclebams/wgs/CDS-HoA5rp.bai              UPCISCC090   \n",
       "CDS-ymRIxH  gs://cclebams/wgs/CDS-ymRIxH.bai                 HCC1588   \n",
       "CDS-5vtea4  gs://cclebams/wgs/CDS-5vtea4.bai                  MERO83   \n",
       "CDS-X7YJtK  gs://cclebams/wgs/CDS-X7YJtK.bai                  MERO95   \n",
       "CDS-LvlS3j  gs://cclebams/wgs/CDS-LvlS3j.bai                   TASK1   \n",
       "CDS-R22IHj  gs://cclebams/wgs/CDS-R22IHj.bai                MONOMAC1   \n",
       "\n",
       "            arxspan_id  sequencing_date participant_id crc32c_hash  \\\n",
       "sample_id                                                            \n",
       "CDS-ttmLuC  ACH-001818           737547    PT-oZH9BTUX    wP5Uig==   \n",
       "CDS-VVWDLi  ACH-001338           737547    PT-sqRdd0QF    TKjC7Q==   \n",
       "CDS-cjMTSm  ACH-001828           737552    PT-OCJbBZ3S    qID1QA==   \n",
       "CDS-AQpAIt  ACH-001533           737552    PT-ceOWs8HZ    vaScFg==   \n",
       "CDS-cXQOkP  ACH-001543           737547    PT-gCMWGzPt    gRp/bw==   \n",
       "CDS-YLQJ19  ACH-002511           737552    PT-ELEMMjP8    87LwZQ==   \n",
       "CDS-AqZLna  ACH-002512           737547    PT-pUBtKkYR    7TbnGw==   \n",
       "CDS-3PGQ84  ACH-001553           737560    PT-SMrpbPt7    GGyO/Q==   \n",
       "CDS-nKfczI  ACH-002014           737547    PT-RkMpD0qb    OHynTg==   \n",
       "CDS-deAUlH  ACH-002013           737547    PT-ws3gMJ8I    BVyqdQ==   \n",
       "CDS-exVsHa  ACH-001636           737552    PT-aFGn7OAh    YN5zyw==   \n",
       "CDS-qyMeyg  ACH-001385           737547    PT-g7W1qRmq    s6C72w==   \n",
       "CDS-M1FWuz  ACH-001641           737547    PT-uOqhQJIC    w6hcug==   \n",
       "CDS-KA0Ncu  ACH-001642           737547    PT-CF1O2RQl    9wgf2g==   \n",
       "CDS-FbQkW1  ACH-001662           737547    PT-bdAvWDQz    ysP94Q==   \n",
       "CDS-allTcx  ACH-001400           737547    PT-bFHPciIP    ooDiUg==   \n",
       "CDS-uYOFVp  ACH-001863           737547    PT-2Cb1QaHZ    WpyzsA==   \n",
       "CDS-JO3DcF  ACH-001699           737547    PT-79r1UuWc    EsEFRw==   \n",
       "CDS-WgAbdT  ACH-001703           737547    PT-iFueiW0E    NRAWDA==   \n",
       "CDS-vp6X0E  ACH-001705           737552    PT-n3CSJEGT    XIf9Ew==   \n",
       "CDS-bTfRlF  ACH-001707           737552    PT-y3RbI7uD    AMuCew==   \n",
       "CDS-OQITe6  ACH-001046           737546    PT-Z2Fq3yte    7+vYIA==   \n",
       "CDS-753lzg  ACH-000195           737546    PT-Ftjkb7Jw    IuFwxQ==   \n",
       "CDS-HoA5rp  ACH-001227           737546    PT-mkNL4JuM    IZjIhA==   \n",
       "CDS-ymRIxH  ACH-001078           737552    PT-epLMsCoH    2ldWlQ==   \n",
       "CDS-5vtea4  ACH-001560           737547    PT-75blYymy    nRYbAw==   \n",
       "CDS-X7YJtK  ACH-001562           737547    PT-qbPTkrGa    YkQgKw==   \n",
       "CDS-LvlS3j  ACH-001670           737547    PT-eRPKQiFq    Xi90TA==   \n",
       "CDS-R22IHj  ACH-001129           734878    PT-EDkpMwcH    y2QITw==   \n",
       "\n",
       "                    size datatype  version     sex       primary_disease  \\\n",
       "sample_id                                                                  \n",
       "CDS-ttmLuC  118951856040      wgs        1       U           Bone Cancer   \n",
       "CDS-VVWDLi  111261029620      wgs        1    Male         Neuroblastoma   \n",
       "CDS-cjMTSm  113315538347      wgs        1     NaN         Breast Cancer   \n",
       "CDS-AQpAIt  114662826746      wgs        1    Male              Lymphoma   \n",
       "CDS-cXQOkP   96804199449      wgs        1    Male  Head and Neck Cancer   \n",
       "CDS-YLQJ19  125056253166      wgs        1  Female           Skin Cancer   \n",
       "CDS-AqZLna  110877810899      wgs        1    Male           Skin Cancer   \n",
       "CDS-3PGQ84  100818382371      wgs        1     NaN              Leukemia   \n",
       "CDS-nKfczI  109896913581      wgs        1    Male            Eye Cancer   \n",
       "CDS-deAUlH  118332580889      wgs        1     NaN            Eye Cancer   \n",
       "CDS-exVsHa  133789949053      wgs        1    Male              Lymphoma   \n",
       "CDS-qyMeyg  103759426410      wgs        1    Male  Head and Neck Cancer   \n",
       "CDS-M1FWuz  114090564340      wgs        1       U  Head and Neck Cancer   \n",
       "CDS-KA0Ncu  132865094686      wgs        1  Female              Lymphoma   \n",
       "CDS-FbQkW1  108771829505      wgs        1  Female         Breast Cancer   \n",
       "CDS-allTcx  123447299572      wgs        1  Female       Cervical Cancer   \n",
       "CDS-uYOFVp  113987500107      wgs        1    Male      Bile Duct Cancer   \n",
       "CDS-JO3DcF  119534283403      wgs        1    Male  Head and Neck Cancer   \n",
       "CDS-WgAbdT  118478462240      wgs        1  Female              Lymphoma   \n",
       "CDS-vp6X0E  114853944490      wgs        1     NaN         Breast Cancer   \n",
       "CDS-bTfRlF   90677570187      wgs        1     NaN              Leukemia   \n",
       "CDS-OQITe6  100035318330      wgs        1     NaN           Lung Cancer   \n",
       "CDS-753lzg  109584388878      wgs        1  Female              Leukemia   \n",
       "CDS-HoA5rp  117747270799      wgs        1    Male  Head and Neck Cancer   \n",
       "CDS-ymRIxH  108374791873      wgs        1  Female           Lung Cancer   \n",
       "CDS-5vtea4  116306111418      wgs        1    Male           Lung Cancer   \n",
       "CDS-X7YJtK  117193717024      wgs        1    Male           Lung Cancer   \n",
       "CDS-LvlS3j  118399266655      wgs        1  Female          Brain Cancer   \n",
       "CDS-R22IHj  152230340144      wgs        2    Male              Leukemia   \n",
       "\n",
       "           cellosaurus_id    age                        primary_site  \\\n",
       "sample_id                                                              \n",
       "CDS-ttmLuC              U      U                                bone   \n",
       "CDS-VVWDLi      CVCL_1124  43952                   autonomic_ganglia   \n",
       "CDS-cjMTSm      CVCL_3280     51                    pleural_effusion   \n",
       "CDS-AQpAIt              U      U  haematopoietic_and_lymphoid_tissue   \n",
       "CDS-cXQOkP              U   51.0           upper_aerodigestive_tract   \n",
       "CDS-YLQJ19              U   64.0                                skin   \n",
       "CDS-AqZLna            NaN      U                                skin   \n",
       "CDS-3PGQ84      CVCL_1832     24  haematopoietic_and_lymphoid_tissue   \n",
       "CDS-nKfczI              U   62.0                                uvea   \n",
       "CDS-deAUlH      CVCL_4D15     72                               liver   \n",
       "CDS-exVsHa              U    3.0                             ascites   \n",
       "CDS-qyMeyg      CVCL_1664   52.0                    pleural_effusion   \n",
       "CDS-M1FWuz      CVCL_3160      U           upper_aerodigestive_tract   \n",
       "CDS-KA0Ncu      CVCL_1683   20.0                    pleural_effusion   \n",
       "CDS-FbQkW1      CVCL_T041     55                    pleural_effusion   \n",
       "CDS-allTcx      CVCL_1732   86.0                              cervix   \n",
       "CDS-uYOFVp      CVCL_5599      U                       biliary_tract   \n",
       "CDS-JO3DcF              U   73.0           upper_aerodigestive_tract   \n",
       "CDS-WgAbdT      CVCL_1819   50.0                         bone_marrow   \n",
       "CDS-vp6X0E      CVCL_2755     48                              breast   \n",
       "CDS-bTfRlF      CVCL_Y549  Adult  haematopoietic_and_lymphoid_tissue   \n",
       "CDS-OQITe6      CVCL_2410      U                         bone_marrow   \n",
       "CDS-753lzg      CVCL_2187   71.0  haematopoietic_and_lymphoid_tissue   \n",
       "CDS-HoA5rp              U   44.0           upper_aerodigestive_tract   \n",
       "CDS-ymRIxH      CVCL_A351   63.0                                lung   \n",
       "CDS-5vtea4      CVCL_2595      U                    pleural_effusion   \n",
       "CDS-X7YJtK      CVCL_2597      U                    pleural_effusion   \n",
       "CDS-LvlS3j      CVCL_3177   21.0              central_nervous_system   \n",
       "CDS-R22IHj      CVCL_1425   64.0  haematopoietic_and_lymphoid_tissue   \n",
       "\n",
       "                                  subtype               subsubtype  \\\n",
       "sample_id                                                            \n",
       "CDS-ttmLuC                   osteosarcoma                      NaN   \n",
       "CDS-VVWDLi                  neuroblastoma                      NaN   \n",
       "CDS-cjMTSm                      Carcinoma  breast_ductal_carcinoma   \n",
       "CDS-AQpAIt           non_hodgkin_lymphoma         splenic_lymphoma   \n",
       "CDS-cXQOkP   upper_aerodigestive_squamous                     oral   \n",
       "CDS-YLQJ19                       melanoma                  mucosal   \n",
       "CDS-AqZLna                       Melanoma                    acral   \n",
       "CDS-3PGQ84                            CML                            \n",
       "CDS-nKfczI                 uveal_melanoma                      NaN   \n",
       "CDS-deAUlH                 Uveal Melanoma                            \n",
       "CDS-exVsHa           non_hodgkin_lymphoma           b_cell_burkitt   \n",
       "CDS-qyMeyg   upper_aerodigestive_squamous                      NaN   \n",
       "CDS-M1FWuz   upper_aerodigestive_squamous                      NaN   \n",
       "CDS-KA0Ncu           non_hodgkin_lymphoma                   b_cell   \n",
       "CDS-FbQkW1        Breast Ductal Carcinoma                      NaN   \n",
       "CDS-allTcx              cervical_squamous                      NaN   \n",
       "CDS-uYOFVp             cholangiocarcinoma             intrahepatic   \n",
       "CDS-JO3DcF   upper_aerodigestive_squamous                      NaN   \n",
       "CDS-WgAbdT           non_hodgkin_lymphoma                    DLBCL   \n",
       "CDS-vp6X0E                      Carcinoma  breast_ductal_carcinoma   \n",
       "CDS-bTfRlF                            CLL                   b_cell   \n",
       "CDS-OQITe6  Small Cell Lung Cancer (SCLC)                            \n",
       "CDS-753lzg                            AML                      NaN   \n",
       "CDS-HoA5rp   upper_aerodigestive_squamous                     oral   \n",
       "CDS-ymRIxH                          NSCLC           NSCLC_squamous   \n",
       "CDS-5vtea4                   mesothelioma                      NaN   \n",
       "CDS-X7YJtK                   mesothelioma                      NaN   \n",
       "CDS-LvlS3j                           PNET                      NaN   \n",
       "CDS-R22IHj                            AML                      NaN   \n",
       "\n",
       "                                                     comments  gender  \n",
       "sample_id                                                              \n",
       "CDS-ttmLuC  Received from Academic lab (Gorlick, Montefior...     NaN  \n",
       "CDS-VVWDLi           MYCN amp; adrenergic ; 1p - (literature)     NaN  \n",
       "CDS-cjMTSm  Transformant: NCBI_TaxID; 10376; Epstein-Barr ...  Female  \n",
       "CDS-AQpAIt                                                NaN     NaN  \n",
       "CDS-cXQOkP  Human Tumor cells derived from oral cancer at ...     NaN  \n",
       "CDS-YLQJ19  Biopsy site - vagina; Received from Academic l...     NaN  \n",
       "CDS-AqZLna  M1610113 (p9); Mutation: NRASG12R; biopsy site...     NaN  \n",
       "CDS-3PGQ84                                                       Male  \n",
       "CDS-nKfczI  Mel270, OMM23, and OMM25 are from the same pat...     NaN  \n",
       "CDS-deAUlH                                                       Male  \n",
       "CDS-exVsHa                                                NaN     NaN  \n",
       "CDS-qyMeyg                                                NaN     NaN  \n",
       "CDS-M1FWuz                                                NaN     NaN  \n",
       "CDS-KA0Ncu  Human monocytic cell line expressing Interleuk...     NaN  \n",
       "CDS-FbQkW1       breast cancer, Infiltrating ductal carcinoma     NaN  \n",
       "CDS-allTcx                     Vulvar squamous cell carcinoma     NaN  \n",
       "CDS-uYOFVp  Received from Academic lab (Bardeesy, MGH) - e...     NaN  \n",
       "CDS-JO3DcF                                                NaN     NaN  \n",
       "CDS-WgAbdT                        Bone Marrow metastatic site     NaN  \n",
       "CDS-vp6X0E  Derived from a breast cancer biopsy removed fr...  Female  \n",
       "CDS-bTfRlF                                                       Male  \n",
       "CDS-OQITe6                                                     Female  \n",
       "CDS-753lzg                                          CCLE line     NaN  \n",
       "CDS-HoA5rp                         Received from academic lab     NaN  \n",
       "CDS-ymRIxH  CCLE line; Identical lines: HCC-1588 and LS513...     NaN  \n",
       "CDS-5vtea4  A human mesothelioma cell line derived from a ...     NaN  \n",
       "CDS-X7YJtK  A human mesothelioma cell line derived from a ...     NaN  \n",
       "CDS-LvlS3j  Human cell line with primitive neuroectodermal...     NaN  \n",
       "CDS-R22IHj  CCLE line; MLL-AF4; p53 mut; RAS WT; FLT3-ITD;...     NaN  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# uploading to our bucket (now a new function)\n",
    "h.changeToBucket(samples,'gs://cclebams/wgs/', name_col= \"index\" , values=['internal_bam_filepath','internal_bai_filepath'], filetypes=['bam', 'bai'], catchdup=True, test=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names=[]\n",
    "subccle_refsamples = ccle_refsamples[ccle_refsamples['datatype'] == \"wgs\"]\n",
    "for k, val in samples.iterrows():\n",
    "    val = val[\"arxspan_id\"]\n",
    "    names.append(val)\n",
    "    samples.loc[k, 'version'] = len(subccle_refsamples[subccle_refsamples['arxspan_id'] == val]) + names.count(val)\n",
    "samples['version'] = samples['version'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples = pd.read_csv('temp/updated_ref_samples.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subccle_refsamples.sequencing_date = h.datetoint(subccle_refsamples.sequencing_date.values, split='/', order = \"asc\")\n",
    "for k, val in samples.iterrows():\n",
    "    loc = subccle_refsamples[subccle_refsamples.arxspan_id==val.arxspan_id]\n",
    "    if len(loc)>0:\n",
    "        if val.sequencing_date > 0:\n",
    "            for i, v in loc.iterrows():\n",
    "                if v.sequencing_date > val.sequencing_date:\n",
    "                    ccle_refsamples.loc[i,'version']+=1\n",
    "                    samples.loc[k, 'version']-=1\n",
    "        else:\n",
    "            if max(loc['size']) > val['size']:\n",
    "                samples.loc[k, 'version'] = 1\n",
    "                ccle_refsamples.loc[loc.index,'version'] = ccle_refsamples.loc[loc.index,'version'].values+1     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples = ccle_refsamples.append(samples, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "for val in samepatient:\n",
    "    sub = ccle_refsamples[ccle_refsamples.arxspan_id.isin(val)]\n",
    "    if len(set(sub.participant_id))>2:\n",
    "        print('we found a missig participant relationship')\n",
    "        # ccle_refsamples.loc[ccle_refsamples.index, \"participant_id\"]=sub.participant_id[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples.to_csv('temp/updated_ref_samples.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'subccle_refsamples' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m--------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                    Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-116-c619ef0d2942>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpairs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msetupPairsFromSamples\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msamples\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubccle_refsamples\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextract\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'patient_id'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m'participant_id'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'subccle_refsamples' is not defined"
     ]
    }
   ],
   "source": [
    "#pairs = setupPairsFromSamples(samples, subccle_refsamples, extract={'patient_id':'participant_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully imported 29 participants.\n",
      "Successfully imported 29 samples.\n",
      "Successfully imported 1 sample sets:\n",
      "  * 20Q4 (29 samples)\n",
      "Successfully imported 1 sample sets:\n",
      "  * all (359 samples)\n"
     ]
    }
   ],
   "source": [
    "#uploading new samples to mut\n",
    "refwm.upload_samples(samples)\n",
    "#refwm.upload_entities('pairs', pairs)\n",
    "#refwm.update_pair_set(pair_set_id=samplesetname,pair_ids=pairs.index)\n",
    "sam = refwm.get_samples()\n",
    "\n",
    "#pair = refwm.get_pairs()\n",
    "#refwm.update_pair_set(pair_set_id='all',pair_ids=pair.index)\n",
    "#creating a sample set\n",
    "refwm.update_sample_set(sample_set_id=samplesetname, sample_ids=samples.index)\n",
    "refwm.update_sample_set(sample_set_id='all', sample_ids=[i for i in sam.index.tolist() if i!='nan'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully imported 1 sample sets:\n",
      "  * all (359 samples)\n"
     ]
    }
   ],
   "source": [
    "#and CGA\n",
    "cgawm.upload_samples(samples)\n",
    "sam = cgawm.get_samples()\n",
    "#creating a sample set\n",
    "cgawm.update_sample_set(sample_set_id=samplesetname, sample_ids=samples.index)\n",
    "cgawm.update_sample_set(sample_set_id='all', sample_ids=[i for i in sam.index.tolist() if i!='nan'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check that we have all the cell lines we expect for this release\n",
    "\n",
    "This involves comparing to the list in the Google sheet \"Cell Line Profiling Status.\"\n",
    "\n",
    "_As the list cannot be parsed, we are not comparing it for now_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function may not work - it hasn't been tested\n",
    "url = 'https://docs.google.com/spreadsheets/d/1qus-9TKzqzwUMNWp8S1QP4s4-3SsMo2vuQRZrNXf7ag/edit?ts=5db85e27#gid=0&fvid=1627883727'\n",
    "\n",
    "compareToCuratedGS(url, sample = newsample[0], samplesetname = samplesetname, colname = 'CN New to internal')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run the pipeline\n",
    "\n",
    "We are using Dalmatian to send request to Terra, we are running a set of 5 functions To generate the mutation dataset:\n",
    "\n",
    "*   For new samples in DepMap, run the ICE version of this task. CCLE2 samples used Agilent targets, so this pipeline should be used instead. The pipelines are identical in terms of their outputs, but the proper targets, baits, and pseudo normal should be used based on how the samples were sequenced.\n",
    "\n",
    "    **ICE_CGA_Production_Analysis_Pipeline_Cell_Lines_copy** (cclf/CGA_Production_Analysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22) OR\n",
    "\n",
    "\n",
    "    **AGILENT_CGA_Production_Analysis_Pipeline_Cell_Lines** (cclf/CGA_Production_Anablysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22)\n",
    "\n",
    "*   **common_variant_filter** (breardon/common_variant_filterSnapshot ID: 3)\n",
    "*   **filterMAF_on_CGA_pipeline** (gkugener/filterMAF_on_CGA_pipelineSnapshot ID: 8)\n",
    "*   **aggregateMAFs_selectFields** (ccle_mg/aggregateMAFs_selectFieldsSnapshot ID: 1)\n",
    "\n",
    "This outputs to be downloaded will be saved in the sample set that was run. The output we use for the release is:\n",
    "\n",
    "\n",
    "*   **passedCGA_filteredMAF_aggregated** \n",
    "\n",
    "There are several other tasks in this workspace. In brief:\n",
    "\n",
    "\n",
    "\n",
    "*   **CGA_Production_Analysis_Pipeline_Cell_Lines** (lelagina/CGA_Production_Analysis_Pipeline_Cell_LinesSnapshot ID: 12). This task is the same as the ICE and AGILENT prefixed version above, except that it relied on pulling the baits and targets to use from the metadata stored for the samples. Having AGILENT and ICE versions specified made the uploading and running process easier.\n",
    "*   **SANGER_CGA_Production_Analysis_Pipeline_Cell_Lines** (cclf/CGA_Production_Analysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22). This task was trying to run the CGA pipeline on the Sanger WES data, using a Sanger pseudo normal. In its current implementation, this task fails to complete for the samples.\n",
    "*   **UNFILTERED_aggregateMAFs_selectFields** (ccle_mg/aggregateMAFs_selectFieldsSnapshot ID: 1). Aggregates the MAF outputted by the CGA cell line pipeline prior to the common variant filter and germline filtering tasks. This can give us insight to which mutations are getting filtered out when. We may want to potentially include this MAF in the release so people can see why certain mutations of interest may be getting filtered out.\n",
    "*   WES_DM_Mutation_Calling_Pipeline_(standard |expensive) (gkugener/WES_DM_Mutation_Calling_PipelineSnapshot ID: 2). This was a previous mutation calling pipeline implemented for CCLE. We do not use this pipeline any more as the CGA pipeline looks better.\n",
    "*   aggregate_filterMAF_CGA (CCLE/aggregate_filterMAF_CGASnapshot ID: 1). An aggregation MAF task that we used in the past. We do not use this task anymore.\n",
    "*   calculate_mutational_burden (breardon/calculate_mutational_burdenSnapshot ID: 21). This task can be used to calculate the mutational rate of the samples. We do not make use of this data in the release although it could be of interest.\n",
    "*   summarizeWigFile (breardon/summarizeWigFileSnapshot ID: 5). CCLF ran this task (might be necessary for the mutational burden task). For our workflow, we do not run it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## On Terra"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### realigning to hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmap = \"BamToUnmappedRGBams\"\n",
    "groupfiles= \"ArrayOfFilesToTxt\"\n",
    "mapping = \"PreProcessingForVariantDiscovery_GATK4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission 5e61d3db-3cdd-4f20-9608-d67fb06a4905.\n"
     ]
    }
   ],
   "source": [
    "submission_id2 = refwm.create_submission(unmap, samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = cgawm.get_samples().loc[cgawm.get_sample_sets().loc[samplesetname].samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n",
      "Successfully updated attributes '['ubam_file_list']' for 1 samples.\n"
     ]
    }
   ],
   "source": [
    "for k, i in a.iterrows():\n",
    "    refwm.update_sample_attributes({'ubam_file_list': i['file_list_name']}, sample_id=k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calling mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission fb0ba43c-fb90-4c92-a0e1-f92d9c35c51e.\n"
     ]
    }
   ],
   "source": [
    "submission_id1 = cgawm.create_submission(\"CGA_WES_CCLE_Characterization_Pipeline_v0.1_Jul2019_copy\", samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission c9d45fdd-9578-417f-a1bf-b406622e542c.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(cgaworkspace, submission_id2)\n",
    "submission_id2 = cgawm.create_submission(groupfiles, samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission c79fd133-58ac-4ac5-9742-414bf3f9e48e.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(refworkspace, submission_id2)\n",
    "submission_id2 = refwm.create_submission(mapping, samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission 37ea7504-48da-4169-b9a9-ca65ae53abad.\n",
      "Successfully created submission 55dd3446-214f-4421-87d9-e385c53e99fb.\n",
      "Successfully created submission 601a257a-2704-4180-93c5-4cb706d64154.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(refworkspace, submission_id2)\n",
    "submission_id2 = refwm.create_submission(\"cnn-variant-filter\", samplesetname,'sample_set',expression='this.samples')\n",
    "submission_id3 = refwm.create_submission(\"CNV_Somatic_Workflow_on_Sample\", samplesetname,'sample_set',expression='this.samples')\n",
    "submission_id4 = refwm.create_submission(\"Manta_SomaticSV\", samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission db715982-826a-45ee-a2f8-43705ba746a4.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(cgaworkspace, submission_id1)\n",
    "submission_id1 = cgawm.create_submission(\"common_variant_filter\", samplesetname, 'sample_set', expression='this.samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0 of jobs Succeeded in submission 0.sion 0. 8 mn elapsed..\n",
      "Successfully created submission 0257e1e2-f84d-4478-942a-62c5790da1bc.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(cgaworkspace, submission_id1)\n",
    "submission_id1 = cgawm.create_submission(\"filterMAF_on_CGA_pipeline\", samplesetname,'sample_set',expression='this.samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aggregations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission e7387274-2f53-4ab3-943c-967107094a60.\n"
     ]
    }
   ],
   "source": [
    "#filtered\n",
    "terra.waitForSubmission(cgaworkspace, submission_id1)\n",
    "submission_id1 = cgawm.create_submission(\"aggregateMAFs_selectFields_copy\", samplesetname)\n",
    "#unfiltered \n",
    "submission_id5 = cgawm.create_submission(\"aggregateMAFs_selectFields_unfiltered\", samplesetname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created submission 58e2c1d1-47a6-4540-8e53-1a2fcfa2dc7e.\n",
      "Successfully created submission 21c66ea6-14e0-4fb9-a1b5-924b4c566b4d.\n"
     ]
    }
   ],
   "source": [
    "terra.waitForSubmission(refworkspace, [submission_id2,submission_id3,submission_id4])\n",
    "submission_id2 = refwm.create_submission(\"aggregate_vcfs_sv\", \"all\")\n",
    "submission_id3 = refwm.create_submission(\"aggregate_vcfs\", \"all\")\n",
    "submission_id4 = refwm.create_submission(\"aggregate_vcfs_snp\", \"all\")\n",
    "submission_id6 = refwm.create_submission(\"aggregate_vcfs_sv_somatic\", \"all\")\n",
    "submission_id7 = refwm.create_submission(\"Aggregate_CN_seg_files\", \"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terra.waitForSubmission(refworkspace, [submission_id2, submission_id3, submission_id4, submission_id6, submission_id7])\n",
    "terra.waitForSubmission(cgaworkspace, [submission_id1, submission_id5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy pairs data to sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs = refwm.get_pairs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs = pairs[pairs.index.isin(tokeep)]\n",
    "pairs = pairs[~pairs['mutation_validator_validated_maf'].isna()]\n",
    "pairs = pairs.drop(columns=['case_sample','control_sample','participant_id'])\n",
    "pairs.index = [i.split('_')[0] for i in pairs.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refwm.update_sample_attributes(pairs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the workflow configurations used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terra.saveConfigs(refworkspace,'./data/'+samplesetname+'/Mutconfig')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## On local\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove some datafile to save money¶"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "val = \"\"\n",
    "! gsutil -m rm gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/$val/**/call-tumorMM_Task/*.cleaned.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'readgroup_ubams'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2896\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2897\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2898\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'readgroup_ubams'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-301-3966d9c191ec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrefwm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_samples\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0me\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"readgroup_ubams\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnan\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2978\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2979\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2980\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2981\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2982\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2897\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2898\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2899\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2900\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2901\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'readgroup_ubams'"
     ]
    }
   ],
   "source": [
    "# sometimes it does not work so better check again\n",
    "a = refwm.get_samples()\n",
    "e = []\n",
    "for i in a[\"readgroup_ubams\"].values.tolist():\n",
    "    if i is not np.nan:\n",
    "        e.extend(i)\n",
    "gcp.rmFiles(e)\n",
    "toremove = [\"readgroup_ubams\",]\n",
    "for val in toremove:\n",
    "    refwm.disable_hound().delete_entity_attributes('sample', res[val])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples = pd.read_csv('temp/updated_ref_samples.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "using the data from broad-firecloud-ccle/DepMap_WGS_CN sample list\n",
      "this should only contains gs:// paths otherwise precise columns using \"onlycol\"\n",
      "Successfully updated attributes '['analysis_ready_bam', 'analysis_ready_bam_index', 'analysis_ready_bam_md5']' for 29 samples.\n"
     ]
    }
   ],
   "source": [
    "#move HG38 to our bucket\n",
    "onlycol = ['analysis_ready_bam', 'analysis_ready_bam_index', 'analysis_ready_bam_md5']\n",
    "samplesinset= [i['entityName'] for i in refwm.get_entities('sample_set').loc[samplesetname].samples]\n",
    "wgs_newgs = 'gs://cclebams/wgs_hg38/'\n",
    "wgs_res, _ = terra.changeGSlocation(refworkspace, newgs=wgs_newgs, onlysamples=samplesinset, onlycol=onlycol, entity='sample', keeppath=False, dry_run = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "listing files in gs\n",
      "listing files in gs\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "ccle_refsamples.loc[wgs_res.index.tolist()][['legacy_bam_filepath','legacy_bai_filepath','legacy_size',\t'legacy_crc32c_hash']] = ccle_refsamples.loc[wgs_res.index.tolist()][['internal_bam_filepath','internal_bai_filepath','size','crc32c_hash']]\n",
    "ccle_refsamples.loc[wgs_res.index.tolist()][['internal_bam_filepath','internal_bai_filepath']] = wgs_res[['analysis_ready_bam','analysis_ready_bam_index']]\n",
    "ccle_refsamples.loc[wgs_res.index.tolist(),'size'] = [gcp.extractSize(i)[1] for i in gcp.lsFiles(wgs_res['analysis_ready_bam'].tolist(),'-l')]\n",
    "ccle_refsamples.loc[wgs_res.index.tolist(),'crc32c_hash'] = [gcp.extractHash(i) for i in gcp.lsFiles(wgs_res['analysis_ready_bam'].tolist(),'-L')]\n",
    "ccle_refsamples.loc[wgs_res.index.tolist(),'md5_hash'] = gcp.catFiles(wgs_res['analysis_ready_bam_md5'].tolist(), cut=32)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get QC files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMut = getQC(workspace=cgaworkspace ,only=[], qcname=[\"gatk_cnv_all_plots\", \"lego_plotter_pngs\", \"copy_number_qc_report\", \"ffpe_OBF_figures\", \"mut_legos_html\", \"oxoG_OBF_figures\", \"tumor_bam_base_distribution_by_cycle_metrics\", \"tumor_bam_converted_oxog_metrics\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataBamCGA = getQC(workspace=cgaworkspace ,only=[], qcname=[\"tumor_bam_alignment_summary_metrics\", \"tumor_bam_bait_bias_summary_metrics\", \"tumor_bam_gc_bias_summary_metrics\", \"tumor_bam_hybrid_selection_metrics\", \"tumor_bam_insert_size_histogram\", \"tumor_bam_insert_size_metrics\", \"tumor_bam_pre_adapter_summary_metrics\", \"tumor_bam_quality_by_cycle_metrics\", \"tumor_bam_quality_distribution_metrics\", \"tumor_bam_quality_yield_metrics\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataBamGATK = getQC(workspace=refworkspace, only=[], qcname=[\"duplication_metrics\", \"bqsr_report\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataCN = getQC(workspace=refworkspace,only=[], qcname=[\"allelic_counts_tumor\",\"delta_MAD_tumor\",\"denoised_MAD_tumor\",\"scaled_delta_MAD_tumor\",\"denoised_copy_ratios_lim_4_plot_tumor\",\"denoised_copy_ratios_plot_tumor\",\"modeled_segments_plot_tumor\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k,v in dataMut.items():\n",
    "    if k =='nan':\n",
    "        continue\n",
    "    ccle_refsamples.loc[k,'processing_qc'] = str(v) + ',' #+ ccle_refsamples.loc[k,'processing_qc']\n",
    "for k,v in dataCN.items():\n",
    "    if k =='nan':\n",
    "        continue\n",
    "    ccle_refsamples.loc[k,'processing_qc'] = str(v) + ',' + ccle_refsamples.loc[k,'processing_qc']\n",
    "for k,v in dataBamGATK.items():\n",
    "    if k =='nan':\n",
    "        continue\n",
    "    ccle_refsamples.loc[k,'bam_qc'] = str(v) + ',' #+ ccle_refsamples.loc[k,'bam_qc']\n",
    "for k,v in dataBamCGA.items():\n",
    "    if k =='nan':\n",
    "        continue\n",
    "    ccle_refsamples.loc[k,'bam_qc'] = str(v) + ',' + ccle_refsamples.loc[k,'bam_qc']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### downloading from terra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "samples              [CDS-00Nrci, CDS-0sdrSB, CDS-0Z9yIY, CDS-17moN...\n",
       "merged_vcf           gs://fc-secure-bd7b8bc9-f665-4269-997e-5a40208...\n",
       "combined_seg_file    gs://fc-secure-bd7b8bc9-f665-4269-997e-5a40208...\n",
       "Name: all, dtype: object"
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = refwm.get_sample_sets().loc[\"all\"]\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://fc-secure-bd7b8bc9-f665-4269-997e-5a402088a369/2493f472-63da-4d2d-83c0-a6217e06d3a4/aggregate_CN_segments_wrkflw/cd7cd515-1710-4818-98fa-02b8a5e1afd0/call-aggregate_CN_segments/all.called.seg...\n",
      "\\ [1 files][138.8 MiB/138.8 MiB]                                                \n",
      "Operation completed over 1 objects/138.8 MiB.                                    \n"
     ]
    }
   ],
   "source": [
    "snps = res['merged_vcf']\n",
    "segments = res['combined_seg_file']\n",
    "! gsutil cp $segments \"temp/segments_terra_merged_wgs.seg\"\n",
    "! gsutil cp $snps gs://cclebams/germline_data/\n",
    "! gsutil -m acl ch -ru taiga-892@cds-logging.iam.gserviceaccount.com:R gs://cclebams/germline_data/all.called.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = cgawm.get_sample_sets().loc[\"all\"]\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered = res['filtered_CGA_MAF_aggregated']\n",
    "! gsutil cp $filtered \"temp/mutation_filtered_terra_merged_wgs.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unfiltered = res['unfiltered_CGA_MAF_aggregated']\n",
    "! gsutil cp $unfiltered \"temp/mutation_unfiltered_terra_merged.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = pd.read_csv('temp/segments_terra_merged_wgs.seg', sep='\\t')\n",
    "#new_refsamples = pd.read_csv('temp/newrefCN.csv',index_col=\"cds_sample_id\")\n",
    "#new_refsamples = pd.read_csv('temp/newrefCN.csv',index_col=\"cds_sample_id\")\n",
    "#new_refsamples = pd.read_csv('temp/newrefCN.csv',index_col=\"cds_sample_id\")\n",
    "#new_refsamples = pd.read_csv('temp/newrefCN.csv',index_col=\"cds_sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "359"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(segments.Sample))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = segments.rename(columns={'CONTIG':'Chromosome',\n",
    "'START':'Start',\n",
    "'END':'End',\n",
    "'Sample':\"DepMap_ID\",\n",
    "'NUM_POINTS_COPY_RATIO':'Num_Probes',\n",
    "'MEAN_LOG2_COPY_RATIO':'Segment_Mean',\n",
    "'CALL':'Status'})\n",
    "segments['Source'] = \"Broad WGS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "removed 2 duplicate samples\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'file' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-380-3048b86870a2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mrenaming\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mremoveOlderVersions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msegments\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'DepMap_ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrefsamples\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrefwm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_samples\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marxspan_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"arxspan_id\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mversion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"version\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Chromosome'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'0'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'file' is not defined"
     ]
    }
   ],
   "source": [
    "renaming = removeOlderVersions(names = set(segments['DepMap_ID']), refsamples = refwm.get_samples(), arxspan_id = \"arxspan_id\", version=\"version\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = segments[segments['DepMap_ID'].isin(renaming.keys())].replace({'DepMap_ID':renaming})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### saving samples used for 20Q2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples.loc[renaming.keys(),samplesetname]=1\n",
    "ccle_refsamples.to_csv('temp/updated_ref_samples.csv', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### knowing the missing samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "wgs  = set(refwm.get_samples().arxspan_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACH-001714', 'ACH-002709', 'ACH-002874', 'ACH-002875'}"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%store -r notWESnotlegacy \n",
    "notWESnotlegacy - wgs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### postprocessing\n",
    "\n",
    "\n",
    "Here, rather than rerunning the entire analysis, because we know we are adding only WES samples, we can download the previous release's MAF, add the samples, update any annotations, and perform any global filters at the end.\n",
    "\n",
    "First we need to do an additional step of filtering on coverage and number \n",
    "\n",
    "- readMutations\n",
    "- createSNPs\n",
    "- addToMainMutation\n",
    "- filterAllelicFraction\n",
    "- filterMinCoverage\n",
    "- mergeAnnotations\n",
    "- addAnnotation\n",
    "- maf_add_variant_annotations\n",
    "- mutation_maf_to_binary_matrix (x3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_mapping = tc.get(name='depmap-wes-cn-data--08f3', file='WES_CN_hg38_gene_DM19Q1')\n",
    "gene_mapping = gene_mapping[gene_mapping.columns[:5]]\n",
    "gene_mapping = gene_mapping.rename(columns={'EGID':'ensembl_id','SYMBOL':'symbol','CHR':'Chromosome','CHRLOC':'start','CHRLOCEND':'end'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments['Chromosome'] = [i[3:] for i in segments['Chromosome'].tolist()]\n",
    "segments = segments.sort_values(by=['DepMap_ID','Chromosome','Start','End'])\n",
    "gene_mapping = gene_mapping.sort_values(by=['Chromosome','start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments.to_csv('temp/WGS_segments_all_'+samplesetname+'.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gapmergedsegs = manageGapsInSegments(segments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn = toGeneMatrix(gapmergedsegs, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecn = np.log2((2**genecn)+1)\n",
    "genecn.mean(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "newly_merged_maf <- readMutations('temp/mutation_filtered_terra_merged.txt')\n",
    "new_release <- createSNPs(newly_merged_maf)\n",
    "names(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "previous.release.maf <- load.from.taiga(data.name='depmap-mutations-maf-35fe', data.file=paste0('mutations.',prevname),data.version=prevversion)\n",
    "if (colnames(previous.release.maf)[1] == 'X1' || colnames(previous.release.maf)[1] == \"\") {\n",
    " previous.release.maf[,1] <- NULL \n",
    "}\n",
    "prevnames <- names(previous.release.maf)\n",
    "prevnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "print(nrow(previous.release.maf))\n",
    "merged <- addToMainMutation(previous.release.maf, new_release)\n",
    "print(nrow(merged))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "## Adding more\n",
    "newly_merged_maf <- readMutations('temp/mutation_filtered_terra_merged.txt')\n",
    "new_release <- createSNPs(newly_merged_maf)\n",
    "print(names(new_release))\n",
    "merged <- addToMainMutation(merged, new_release)\n",
    "nrow(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "setdiff(names(merged), names(previous.release.maf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "## check if some rows have nans\n",
    "length(which(is.na(merged$Hugo_Symbol)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "dim(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "filtered <- filterAllelicFraction(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "head(new_release[which(new_release$Tumor_Sample_Barcode=='ACH-002466'),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "filtered <- filterMinCoverage(filtered$merged, filtered$removed_from_maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "head(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "clean_annotations <- mergeAnnotations(merged,previous.release.maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Guillaume's version\n",
    "new_release <- addAnnotation(filtered$merged, clean_annotations, colnames(previous.release.maf))\n",
    "# Allie's version\n",
    "new_release <- maf_add_variant_annotations(new_release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# some matric files that does get used internaly and might be useful\n",
    "damaging_mutation <- mutation_maf_to_binary_matrix(new_release, damaging =  TRUE)\n",
    "other_mutation <- mutation_maf_to_binary_matrix(new_release, other = TRUE)\n",
    "hotspot_mutation <- mutation_maf_to_binary_matrix(new_release, hotspot = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  new_release, \n",
    "  paste0('temp/mutations.', release, '.all.csv'), sep = ',', quote = F, row.names = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  damaging_mutation, \n",
    "  paste0('temp/damaging_mutation.', release, '.all.csv'), sep = ',', quote = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  other_mutation, \n",
    "  paste0('temp/other_mutation.', release, '.all.csv'), sep = ',', quote = F)\n",
    "# Save the ready to upload file to upload to taiga\n",
    "write.table(\n",
    "  hotspot_mutation, \n",
    "  paste0('temp/hotspot_mutation.', release, '.all.csv'), sep = ',', quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare to previous release\n",
    "\n",
    "I would run some checks here comparing the results to the previous releases MAF. Namely:\n",
    "\n",
    "- Count the total number of mutations per cell line, split by type (SNP, INS, DEL)\n",
    "- Count the total number of mutations observed by position (group by chromosome, start position, end position and count the number of mutations)\n",
    "- Look at specific differences between the two MAFs (join on DepMap_ID, Chromosome, Start position, End position, Variant_Type). I would do this for WES only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "mutations = pd.read_csv('temp/mutations.'+release+'.all.csv')\n",
    "damaging_mutation = pd.read_csv('temp/damaging_mutation.'+release+'.all.csv')\n",
    "print(len(damaging_mutation))\n",
    "other_mutation = pd.read_csv('temp/other_mutation.'+release+'.all.csv')\n",
    "print(len(other_mutation))\n",
    "hotspot_mutation = pd.read_csv('temp/hotspot_mutation.'+release+'.all.csv')\n",
    "print(len(hotspot_mutation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(mutations.DepMap_ID) - set(mutations[~(mutations['CGA_WES_AC'].isna() & mutations['SangerWES_AC'].isna() & mutations['WGS_AC'].isna() & mutations['SangerRecalibWES_AC'].isna())].DepMap_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[mutations.DepMap_ID==\"ACH-000458\"].sum(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[mutations[\"Hugo_Symbol\"]==\"ACOT4\"][mutations['Start_position']==74058831]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_data = mutations[[val for val in mutations.columns.values if '_AC' in val]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_names = ac_data.columns.values\n",
    "ac_data = ac_data.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_data.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do some checks and manual rescuing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[mutations.DepMap_ID==\"ACH-003000\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check important mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check MOLM13, MV411 cell lines- The well known mutation status of FLT3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check TP53 mutation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "toofew = 0\n",
    "allnan = 0\n",
    "for pos, val in enumerate(ac_data):\n",
    "    i = 0\n",
    "    print(str(100*pos/ac_data.shape[0]),end='\\r')\n",
    "    for p, v in enumerate(val):\n",
    "        if v is np.nan:\n",
    "            i+=1\n",
    "    if i==7:\n",
    "        mutations = mutations.drop[pos]\n",
    "        allnan+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allnan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### basic counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Count the total number of mutations per cell line, split by type (SNP, INS, DEL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the total number of mutations observed by position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are mutation consistent?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  to check this, if you group all the mutations in the mutations table by Chromosome, Start_position, End_position, Reference_Allele, Tumor_Seq_Allele1 columns, they should all have the same annotation for the other columns (protein change, exac_af, etc...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "QC mutations, for a known dependency, check if it matches mutation of this gene. (if P53 is mutated, cannot have dependency on P53 or MDM2 MDM4/ inverse fir BRAF and KRAF to themselves)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevprevname,prevprevversion "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[mutations.DepMap_ID==\"ACH-001546\"][mutations.columns[-17:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevprev= set(tc.get(name='depmap-mutation-calls-9be3', file= \"depmap_\"+prevprevname+\"_mutation_calls\", version = prevprevversion).DepMap_ID.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# uploading on taiga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsheets = sheets.get(sheeturl).sheets[6].to_frame()\n",
    "wes_dmc_embargo = [i for i in gsheets['WES_DMC_embargo'].values.tolist() if str(i) != \"nan\"]\n",
    "wes_embargo = [i for i in gsheets['WES_embargo'].values.tolist() if str(i) != \"nan\"]\n",
    "blacklist = [i for i in gsheets['blacklist'].values.tolist() if str(i) != \"nan\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wes_embargo, wes_dmc_embargo, blacklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "local_metadata": {},
    "remote_metadata": {
     "scrolled": true
    }
   },
   "outputs": [],
   "source": [
    "! cd .. && git clone https://github.com/broadinstitute/depmap-release-readmes.git && cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cd ../depmap-release-readmes && git pull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cd ../depmap-release-readmes/ && python3 make_new_release.py $release && git add . && git commit -m $release && git push "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cd ../depmap-release-readmes && git pull && mv release-'+release+'/internal-'+release+'.txt ../ccle_processing/temp/README && cd -')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc.update_dataset(dataset_permaname=\"depmap-mutations-maf-35fe\",\n",
    "                 upload_file_path_dict={'temp/mutations.'+release+'.all.csv': 'TableCSV'}, \n",
    "                 dataset_description=\"\"\"\n",
    "# Mutations\n",
    "\n",
    "filtered and unfiltered mutation files from Broad WES and Sanger WES data mapped to hg19\n",
    "The MAF file for DepMap that includes all of the latest WES samples. This MAF is generated by merging CCLE (WGS, RNAseq, RD, HC) and Sanger (WES) data.\n",
    "\n",
    "PORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal. Must use subsetted dataset instead. These data will not make it on the portal starting 19Q1. With the DMC portal, there is new cell line release prioritization as to which lines can be included, so a new taiga dataset will be created containing CN for the portal.\n",
    "\n",
    "version 1:  In 19Q1 the WES_AC column has been replaced by two columns, VA_WES_AC and CGA_WES_AC. We are currently using the Van Allen and CGA based pipeline to generate mutation calls. The CGA pipeline includes more filtering on the MAFs than VA and has a better INDEL caller. However, some of these filters may be removing some variants of interest that are still capture by the VA pipeline, which is why both a retained for now. DEPRECATED:  Missing the VA_WES_AC, CGA_WES_AC columns\n",
    "version 2: 19Q1 data\n",
    "version 3: 19Q2 data. We are no longer using the CCLE_WES_AC column. We are only using the CGA pipeline for mutation calls.\n",
    "version 4: Updating to 19Q3interim DEPRECATED\n",
    "version 5: Updating to 19Q3interim DEPRECATED\n",
    "version 6: Updating to 19Q3interim\n",
    "version 7: Updating to 19Q3 DEPRECATED\n",
    "version 8: reparing the missing mutation problem DEPRECATED\n",
    "version 9: reparing the missing column problem\n",
    "\n",
    "\n",
    "version10:\n",
    "Adding 52 new cell lines. \n",
    "Some cells lines have been flagged as:\n",
    "\n",
    "version11:\n",
    "adding missing cell lines\n",
    "\n",
    "Adding 52 new cell lines. \n",
    "Some cells lines have been flagged as:\n",
    "\n",
    " - having bad looking copy ration plots = \n",
    " - Genes having a similar CN value accross all []\n",
    "\n",
    "version 12:\n",
    "\n",
    "adding 8 new cell lines\n",
    "\n",
    "version 13:\n",
    "\n",
    "removing a wrong column\n",
    "\n",
    "version 14:\n",
    "\n",
    "adding 8 new cell lines. Adding .all. since we are soon going to release a restricted set of mutations. this one contains everything which is not necessarily what we want\n",
    "\n",
    "\n",
    "genes (gene rpkm):\n",
    "__Rows__:\n",
    "__Columns__:\n",
    "Counts (gene counts):\n",
    "__Rows__:\n",
    "__Columns__:\n",
    "Gene level CN data:\n",
    "__Rows__:\n",
    "__Columns__:\n",
    " DepMap cell line IDs\n",
    " gene names in the format HGNC\\_symbol (Entrez\\_ID)\n",
    "DepMap\\_ID, Chromosome, Start, End, Num\\_Probes, Segment\\_Mean\n",
    " \"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Internal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspot_mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevmut = tc.get(name='depmap-mutation-calls-9be3', version=24, file='depmap_'+prevname+'_mutation_calls.all')\n",
    "print('shoud be None')\n",
    "print(set(prevmut.DepMap_ID) - set(mutations.DepMap_ID))\n",
    "print(\"new lines\")\n",
    "newlines = set(mutations.DepMap_ID) - set(prevmut.DepMap_ID) \n",
    "newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc.update_dataset(dataset_permaname=\"depmap-mutation-calls-9be3\",\n",
    "                 upload_file_path_dict={'temp/depmap_'+release+'_mutation_calls.all': 'TableCSV',\n",
    "                                        'temp/damaging_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/other_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/hotspot_mutation.all': 'NumericMatrixCSV',\n",
    "                                       },#'temp/README': 'Raw'},\n",
    "                 dataset_description=\"\"\"\n",
    "# Internal Mutations\n",
    "\n",
    "Mutation calls for Internal DepMap data\n",
    "\n",
    "* Version 1 Internal 18Q1*\n",
    "\n",
    "original source: `/xchip/ccle_dist/broad_only/CMAG/mutations/CCLE_depMap_18Q1_maf_20180202.txt`\n",
    "* Version 2-4 Internal 18Q2*\n",
    "\n",
    "merged mutations and indels file (1,606 cell lines, including CCLE and Sanger WES reanalysis)\n",
    "original source: \n",
    "`/xchip/ccle_dist/broad_only/CMAG/mutations/CCLE_depMap_18q2_maf_20180502.txt`\n",
    "Binary matrices:\n",
    "- damaging: if isDeleterious is true\n",
    "- missense: if isDeleterious is false\n",
    "- hotspot: if missense and either TCGA or COSMIC hotspot\n",
    "Version 2 contains the MAF file\n",
    "* Version 5-6 Internal 18Q3*\n",
    "\n",
    "version 5 deprecated\n",
    "\n",
    "original source: `/xchip/ccle_dist/broad_only/CMAG/mutations/CCLE_depMap_18q3_maf_20180716.txt`\n",
    "\n",
    "Binary matrices:\n",
    "- damaging: if isDeleterious is true\n",
    "- missense: if isDeleterious is false\n",
    "- hotspot: if missense and either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, Broad (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "MAF file\n",
    "\n",
    "* Version 7-8 Internal 18Q4*\n",
    "\n",
    "version 8 just changes a column name in the MAF file from Broad_ID to DepMap_ID\n",
    "\n",
    "original source: `/xchip/ccle_dist/broad_only/CMAG/mutations/CCLE_DepMap_18Q4_maf_20181028.txt`\n",
    "\n",
    "* Version 9-12 Internal 19Q1*\n",
    "\n",
    "version 12 updates the column name from VA_WES_AC to CCLE_WES_AC\n",
    "\n",
    "version 11+ uses an updated definition for hotspot mutations\n",
    "\n",
    "version 12 contains the correct data for 19Q1\n",
    "\n",
    "* Version 13 Internal 19Q2*\n",
    "\n",
    "* Version 14-15 Internal 19Q3*\n",
    "\n",
    "version 15 fixed entrez ids\n",
    "\n",
    "* Version 16 Internal 19Q4*\n",
    "\n",
    "adding 35 new cell lines.\n",
    "\n",
    "* Version 16 Internal 19Q4*\n",
    "uploading as matrices\n",
    "\n",
    "* Version 17 Internal 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 18 Internal 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 19 Internal 20Q1*\n",
    "uploading 8 new lines\n",
    "\n",
    "* Version 20 Internal 20Q1*\n",
    "removing unauthorized cl\n",
    "\n",
    "* Version 21 Internal 20Q2*\n",
    "uploading 8 new lines and adding .all to express the fact that this data is the aggregate of all different sequencing methods.\n",
    "\n",
    "* Version 22 Internal 20Q2*\n",
    "removing 2 cell lines\n",
    "\n",
    "* Version 23 Internal 20Q3*\n",
    "nothing different from 20Q2. no new cell lines\n",
    "\n",
    "* Version 24 Internal 20Q2*\n",
    "updating the blacklists\n",
    "\n",
    "*** Variant annotation column ***\n",
    "\n",
    "MAF file, added column (Variant_annotation) classifying each variant as either silent, damaging, other conserving, or other non-conserving, based on this mapping (old annotation from Variant_Classification column - new annotation):\n",
    "\n",
    "Silent - silent\n",
    "Splice_Site - damaging\n",
    "Missense_Mutation - other non-conserving\n",
    "Nonsense_Mutation - damaging\n",
    "De_novo_Start_OutOfFrame - damaging\n",
    "Nonstop_Mutation - other non-conserving\n",
    "Frame_Shift_Del - damaging\n",
    "Frame_Shift_Ins - damaging\n",
    "In_Frame_Del - other non-conserving\n",
    "In_Frame_Ins - other non-conserving\n",
    "Stop_Codon_Del - other non-conserving\n",
    "Stop_Codon_Ins - other non-conserving\n",
    "Start_Codon_SNP - damaging\n",
    "Start_Codon_Del - damaging\n",
    "Start_Codon_Ins - damaging\n",
    "5'Flank - other conserving\n",
    "Intron - other conserving\n",
    "IGR - other conserving\n",
    "3'UTR - other conserving\n",
    "5'UTR - other conserving\n",
    "Binary matrices:\n",
    "\n",
    "- damaging: if damaging\n",
    "- other: if other conserving or other non-conserving\n",
    "- hotspot: if it is not a silent mutation and is either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, DepMap (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "NEW LINES:\n",
    "\"\"\"+newlines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To add to a virtual dataset\n",
    "AddToVirtual(virtual_internal, 'depmap-mutation-calls-9be3', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls'),])#('README','README')])\n",
    "# To add to a eternal dataset\n",
    "AddToVirtual('depmap-a0ab', 'depmap-mutation-calls-9be3', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cd ../depmap-release-readmes && git pull && mv release-'+releAse+'/dmc-'+releAse+'.txt ../ccle_processing/temp/README && cd -')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1301656\n",
      "1297418\n",
      "1757\n",
      "1744\n",
      "1758\n",
      "1745\n",
      "1725\n",
      "1712\n"
     ]
    }
   ],
   "source": [
    "print(len(mutations))\n",
    "mutations = mutations[~mutations.DepMap_ID.isin(wes_embargo)]\n",
    "print(len(mutations))\n",
    "mutations.to_csv('temp/depmap_'+release+'_mutation_calls.all', index=False)\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation = damaging_mutation[~damaging_mutation.index.isin(wes_embargo)]\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation.to_csv('temp/damaging_mutation.all')\n",
    "print(len(other_mutation))\n",
    "other_mutation = other_mutation[~other_mutation.index.isin(wes_embargo)]\n",
    "print(len(other_mutation))\n",
    "other_mutation.to_csv('temp/other_mutation.all',)\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation = hotspot_mutation[~hotspot_mutation.index.isin(wes_embargo)]\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation.to_csv('temp/hotspot_mutation.all',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[##################]100% |  40.4 MiB/s | 277.3 MiB / 277.3 MiB | Time:  0:00:06\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shoud be None\n",
      "set()\n",
      "new lines\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ACH-001533', 'ACH-001574', 'ACH-002021', 'ACH-002065'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevmut = tc.get(name='depmap-mutation-calls-dfce', version=15, file='depmap_'+prevname+'_mutation_calls')\n",
    "print('shoud be None')\n",
    "print(set(prevmut.DepMap_ID) - set(mutations.DepMap_ID))\n",
    "print(\"new lines\")\n",
    "newlines = set(mutations.DepMap_ID) - set(prevmut.DepMap_ID) \n",
    "newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc.update_dataset(dataset_permaname=\"depmap-mutation-calls-dfce\",\n",
    "                 upload_file_path_dict={'temp/depmap_'+release+'_mutation_calls.all': 'TableCSV',\n",
    "                                        'temp/damaging_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/other_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/hotspot_mutation.all': 'NumericMatrixCSV',\n",
    "                                       },#'temp/README': 'Raw'},\n",
    "                 dataset_description=\"\"\"\n",
    "# DMC Mutations\n",
    "\n",
    "* Version 1-5 DMC 19Q1*\n",
    "\n",
    "version 5 is a one-off portal thing because dmc wanted to be able to plot if a gene has any mutation as one-hot encoded value in the x/y axes of the data explorer It adds the any_mutation matrix, but does not change the others. Code used to generate:\n",
    "\n",
    "```\n",
    "from taigapy import TaigaClient\n",
    "\n",
    "c = TaigaClient()\n",
    "\n",
    "dmc_19q1_mutation_taiga_root = \"depmap-mutation-calls-dfce.3/\"\n",
    "other_matrix = c.get(dmc_19q1_mutation_taiga_root + \"other_mutation\")\n",
    "damaging_matrix = c.get(dmc_19q1_mutation_taiga_root + \"damaging_mutation\")\n",
    "hotspot_matrix = c.get(dmc_19q1_mutation_taiga_root + \"hotspot_mutation\")\n",
    "\n",
    "df = other_matrix.append(damaging_matrix)\n",
    "df = df.groupby(level=0).sum()\n",
    "\n",
    "df = df.append(hotspot_matrix)\n",
    "df = df.groupby(level=0).sum()\n",
    "\n",
    "df[df > 1] = 1\n",
    "\n",
    "df.to_csv('any_mutation.csv')\n",
    "```\n",
    "The code uses version 3 because the dmc portal was using version 3\n",
    "\n",
    "version 4 updates the column name from VA_WES_AC to CCLE_WES_AC\n",
    "\n",
    "version 3 has an updated definition for hotspot mutations\n",
    "\n",
    "version 2+ contains the correct data for 19Q1\n",
    "\n",
    "* Version 6 DMC 19Q2*\n",
    "\n",
    "* Version 7-8 DMC 19Q3*\n",
    "version 8 fixed entrez ids\n",
    "\n",
    "* Version 9 DMC 19Q4*\n",
    "adding 52 new cell lines.\n",
    "\n",
    "* Version 10 DMC 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 11 DMC 19Q4*\n",
    "removing unauthorized lines and setting as matrices\n",
    "\n",
    "* Version 12 Internal 20Q1*\n",
    "uploading 8 new lines\n",
    "\n",
    "* Version 13 Internal 20Q1*\n",
    "removing unauthorized cl\n",
    "\n",
    "* Version 14 Internal 20Q2*\n",
    "uploading 8 new lines and adding .all to express the fact that this data is the aggregate of all different sequencing methods.\n",
    "\n",
    "* Version 15 Internal 20Q2*\n",
    "removing 2 lines\n",
    "\n",
    "* Version 15 Internal 20Q3*\n",
    "nothing different from 20Q2. no new cell lines\n",
    "\n",
    "* Version 15 Internal 20Q3*\n",
    "updating the blacklists\n",
    "\n",
    "\n",
    "MAF file, added column (Variant_annotation) classifying each variant as either silent, damaging, other conserving, or other non-conserving, based on this mapping (old annotation from Variant_Classification column - new annotation):\n",
    "\n",
    "Silent - silent\n",
    "Splice_Site - damaging\n",
    "Missense_Mutation - other non-conserving\n",
    "Nonsense_Mutation - damaging\n",
    "De_novo_Start_OutOfFrame - damaging\n",
    "Nonstop_Mutation - other non-conserving\n",
    "Frame_Shift_Del - damaging\n",
    "Frame_Shift_Ins - damaging\n",
    "In_Frame_Del - other non-conserving\n",
    "In_Frame_Ins - other non-conserving\n",
    "Stop_Codon_Del - other non-conserving\n",
    "Stop_Codon_Ins - other non-conserving\n",
    "Start_Codon_SNP - damaging\n",
    "Start_Codon_Del - damaging\n",
    "Start_Codon_Ins - damaging\n",
    "5'Flank - other conserving\n",
    "Intron - other conserving\n",
    "IGR - other conserving\n",
    "3'UTR - other conserving\n",
    "5'UTR - other conserving\n",
    "Binary matrices:\n",
    "- damaging: if damaging\n",
    "- other: if other conserving or other non-conserving\n",
    "- hotspot: if it is not a silent mutation and is either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, DepMap (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "NEW LINES:\n",
    "\"\"\"+newlines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To add to a virtual dataset\n",
    "AddToVirtual(virtual_dmc, 'depmap-mutation-calls-dfce', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls'),])#('README','README')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Public"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cd ../depmap-release-readmes && git pull && mv release-'+releAse+'/public-'+releAse+'.txt README && cd -')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#damaging_mutation\n",
    "mutations=depmap_20Q3_mutation_calls\n",
    "#hotspot_mutation\n",
    "#other_mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1297418\n",
      "1297418\n",
      "1744\n",
      "1744\n",
      "1745\n",
      "1745\n",
      "1712\n",
      "1712\n"
     ]
    }
   ],
   "source": [
    "print(len(mutations))\n",
    "mutations = mutations[mutations.DepMap_ID.isin(prevprev)]\n",
    "mutations = mutations[~mutations.DepMap_ID.isin(wes_dmc_embargo)]\n",
    "print(len(mutations))\n",
    "mutations.to_csv('temp/depmap_'+release+'_mutation_calls.all', index=False)\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation = damaging_mutation[damaging_mutation.index.isin(prevprev)]\n",
    "damaging_mutation = damaging_mutation[~damaging_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(damaging_mutation))\n",
    "damaging_mutation.to_csv('temp/damaging_mutation.all')\n",
    "print(len(other_mutation))\n",
    "other_mutation = other_mutation[other_mutation.index.isin(prevprev)]\n",
    "other_mutation = other_mutation[~other_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(other_mutation))\n",
    "other_mutation.to_csv('temp/other_mutation.all')\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation = hotspot_mutation[hotspot_mutation.index.isin(prevprev)]\n",
    "hotspot_mutation = hotspot_mutation[~hotspot_mutation.index.isin(wes_dmc_embargo)]\n",
    "print(len(hotspot_mutation))\n",
    "hotspot_mutation.to_csv('temp/hotspot_mutation.all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shoud be None\n",
      "set()\n",
      "new lines\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ACH-001533', 'ACH-001574', 'ACH-002021', 'ACH-002065'}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevmut = tc.get(name='depmap-mutation-calls-9a1a', version=18, file='depmap_'+prevname+'_mutation_calls')\n",
    "print('shoud be None')\n",
    "ermgency_removed = set(prevmut.DepMap_ID) - set(mutations.DepMap_ID)\n",
    "print(ermgency_removed) \n",
    "print(\"new lines\")\n",
    "newlines = set(mutations.DepMap_ID) - set(prevmut.DepMap_ID) \n",
    "newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading depmap_20Q3_mutation_calls...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: depmap_20Q3_mutation_calls properly converted and uploaded\n",
      "Uploading damaging_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: damaging_mutation properly converted and uploaded\n",
      "Uploading other_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 500\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1000\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: other_mutation properly converted and uploaded\n",
      "Uploading hotspot_mutation...\n",
      "hitting https://cds.team/taiga/api/datafile/00b151ca35a14f6d9f1be95ef24ea368\n",
      "Conversion and upload...:\n",
      "\t Downloading the file from S3\n",
      "\t Downloading the file from S3\n",
      "\t Scanning through file to determine size (line 1001)\n",
      "\t Conversion in progress, line 250\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 750\n",
      "\t Conversion in progress, line 1250\n",
      "\t Conversion in progress, line 1500\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\t Uploading to S3\n",
      "\n",
      "\t Done: hotspot_mutation properly converted and uploaded\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id 8cc4900a50874d8593b0bfc591001360 created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/8cc4900a50874d8593b0bfc591001360\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'8cc4900a50874d8593b0bfc591001360'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "description=\"\"\"\n",
    "# Public Mutations\n",
    "\n",
    "Mutation calls for Public DepMap data\n",
    "\n",
    "* Version 1 Public 18Q1*\n",
    "\n",
    "original source: CCLE data portal\n",
    "* Version 2 Public 18Q2*\n",
    "\n",
    "merged mutations and indels file (1,549 cell lines total, including data for 63 newly released cell lines)\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q2/CCLE_DepMap_18Q2_maf_20180502.txt`\n",
    "* Version 3-4 Public 18Q3*\n",
    "\n",
    "version 3 deprecated\n",
    "\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q3/CCLE_DepMap_18q3_maf_20180718.txt`\n",
    "\n",
    "Binary matrices:\n",
    "damaging: if isDeleterious is true\n",
    "missense: if isDeleterious is false\n",
    "hotspot: if missense and either TCGA or COSMIC hotspot\n",
    "Rows: cell line, Broad (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "MAF file\n",
    "\n",
    "* Version 5 Public 18Q4*\n",
    "\n",
    "original source: `/xchip/ccle_dist/public/DepMap_18Q4/CCLE_DepMap_18q4_maf_20181029.txt`\n",
    "\n",
    "* Version 6-9 Public 19Q1*\n",
    "\n",
    "version 9 updates the column name from VA_WES_AC to CCLE_WES_AC\n",
    "\n",
    "version 8 uses an updated definition for hotspot mutations\n",
    "\n",
    "version 9 contains the correct data for 19Q1\n",
    "\n",
    "* Version 10 Public 19Q2*\n",
    "\n",
    "* Version 11-12 Public 19Q3*\n",
    "\n",
    "version 12 fixed entrez ids\n",
    "\n",
    "* Version 13 Public 19Q4*\n",
    "\n",
    "adding 52 new cell lines\n",
    "\n",
    "* Version 14 Public 19Q4*\n",
    "removing unauthorized lines and setting matrices\n",
    "\n",
    "* Version 15 Public 20Q1*\n",
    "adding 8 new lines \n",
    "\n",
    "* Version 16 Public 20Q1*\n",
    "removing an unauthorized line\n",
    "\n",
    "* Version 17 Internal 20Q2*\n",
    "uploading 8 new lines and adding .all to express the fact that this data is the aggregate of all different sequencing methods.\n",
    "\n",
    "* Version 18 Internal 20Q2*\n",
    "removing 2 lines\n",
    "\n",
    "* Version 19 Internal 20Q3*\n",
    "nothing different from 20Q2. no new cell lines\n",
    "\n",
    "* Version 20 Internal 20Q3*\n",
    "updating the blacklists\n",
    "\n",
    "* Version 21 Internal 20Q3*\n",
    "updating the dmc\n",
    "\n",
    "* Version 22 Internal 20Q3*\n",
    "readding two already released samples to the public list\n",
    "\n",
    "MAF file, added column (Variant_annotation) classifying each variant as either silent, damaging, other conserving, or other non-conserving, based on this mapping (old annotation from Variant_Classification column - new annotation):\n",
    "\n",
    "Silent - silent\n",
    "Splice_Site - damaging\n",
    "Missense_Mutation - other non-conserving\n",
    "Nonsense_Mutation - damaging\n",
    "De_novo_Start_OutOfFrame - damaging\n",
    "Nonstop_Mutation - other non-conserving\n",
    "Frame_Shift_Del - damaging\n",
    "Frame_Shift_Ins - damaging\n",
    "In_Frame_Del - other non-conserving\n",
    "In_Frame_Ins - other non-conserving\n",
    "Stop_Codon_Del - other non-conserving\n",
    "Stop_Codon_Ins - other non-conserving\n",
    "Start_Codon_SNP - damaging\n",
    "Start_Codon_Del - damaging\n",
    "Start_Codon_Ins - damaging\n",
    "5'Flank - other conserving\n",
    "Intron - other conserving\n",
    "IGR - other conserving\n",
    "3'UTR - other conserving\n",
    "5'UTR - other conserving\n",
    "Binary matrices:\n",
    "\n",
    "- damaging: if damaging\n",
    "- other: if other conserving or other non-conserving\n",
    "- hotspot: if it is not a silent mutation and is either TCGA or COSMIC hotspot\n",
    "- Rows: cell line, DepMap (arxspan) IDs\n",
    "\n",
    "Columns: Gene, HGNC symbol (Entrez ID)\n",
    "\n",
    "NEW LINES:\n",
    "\"\"\"+str(newlines)\n",
    "\n",
    "if len(ermgency_removed):\n",
    "    description+=\"\"\"\n",
    "    \n",
    "    !! WE REMOVED!!:\n",
    "    \"\"\"+str(ermgency_removed)\n",
    "\n",
    "tc.update_dataset(dataset_permaname=\"depmap-mutation-calls-9a1a\",\n",
    "                 upload_file_path_dict={'temp/depmap_'+release+'_mutation_calls.all': 'TableCSV',\n",
    "                                        'temp/damaging_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/other_mutation.all': 'NumericMatrixCSV',\n",
    "                                        'temp/hotspot_mutation.all': 'NumericMatrixCSV',\n",
    "                                       },#'temp/README': 'Raw'},\n",
    "                 dataset_description=description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CCLE_mutations', 'depmap-mutation-calls-9a1a.22/depmap_20Q3_mutation_calls'), ('CCLE_gene_cn', 'depmap-wes-cn-data-97cc.34/public_20Q3_gene_cn'), ('Achilles_gene_effect_unscaled', 'avana-public-tentative-20q3-3e73.5/gene_effect_unscaled'), ('Achilles_high_variance_genes', 'avana-public-tentative-20q3-3e73.5/high_variance_genes'), ('Achilles_guide_efficacy', 'avana-public-tentative-20q3-3e73.5/guide_efficacy'), ('CCLE_fusions_unfiltered', 'gene-fusions-6212.14/unfiltered_fusions_20Q3'), ('common_essentials', 'avana-public-tentative-20q3-3e73.5/essential_genes'), ('Achilles_logfold_change_failures', 'avana-public-tentative-20q3-3e73.5/logfold_change_failures'), ('CCLE_expression', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_proteincoding_tpm'), ('Achilles_raw_readcounts', 'avana-public-tentative-20q3-3e73.5/raw_readcounts'), ('Achilles_raw_readcounts_failures', 'avana-public-tentative-20q3-3e73.5/raw_readcounts_failing'), ('README', 'public-20q3-3d35.22/README'), ('CCLE_segment_cn', 'depmap-wes-cn-data-97cc.34/public_20Q3_segs_cn'), ('CCLE_RNAseq_transcripts', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_transcripts_tpm'), ('nonessentials', 'avana-public-tentative-20q3-3e73.5/nonessential_genes'), ('Achilles_guide_map', 'avana-public-tentative-20q3-3e73.5/guide_gene_map'), ('Achilles_dropped_guides', 'avana-public-tentative-20q3-3e73.5/dropped_guides'), ('CCLE_fusions', 'gene-fusions-6212.14/filtered_fusions_20Q3'), ('CCLE_RNAseq_reads', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_counts'), ('CCLE_expression_full', 'depmap-rnaseq-expression-data-ccd0.25/public_20Q3_tpm'), ('Achilles_replicate_map', 'avana-public-tentative-20q3-3e73.5/replicate_map'), ('Achilles_gene_dependency', 'avana-public-tentative-20q3-3e73.5/gene_dependency'), ('Achilles_common_essentials', 'avana-public-tentative-20q3-3e73.5/pan_dependent_genes'), ('Achilles_logfold_change', 'avana-public-tentative-20q3-3e73.5/logfold_change'), ('Achilles_gene_effect', 'avana-public-tentative-20q3-3e73.5/gene_effect')]\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datafile/ad76ca2f1c2f43c1b57c7f84a1adaafe\n",
      "hitting https://cds.team/taiga/api/datasetVersion\n",
      "\n",
      "Dataset version with id f803fe38cdb04895b6285533dcc8b00b created. You can access to this dataset version directly with this url: https://cds.team/taiga/dataset_version/f803fe38cdb04895b6285533dcc8b00b\n"
     ]
    }
   ],
   "source": [
    "# To add to a virtual dataset\n",
    "AddToVirtual(virtual_public, 'depmap-mutation-calls-9a1a', [('CCLE_mutations', 'depmap_'+release+'_mutation_calls'),])#('README','README')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "notify_time": "5",
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "277px",
    "width": "375px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "198.8px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "position": {
    "height": "617.8px",
    "left": "1232.4px",
    "right": "20px",
    "top": "120px",
    "width": "262.8px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
