{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading the necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import depmapomics.patch_firecloud\n",
    "depmapomics.patch_firecloud.install_patches()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "from depmapomics import constants\n",
    "from depmapomics import env_config\n",
    "\n",
    "from depmapomics import dm_omics\n",
    "from depmapomics import mutations as omics_mut\n",
    "from depmapomics import copynumbers as omics_cn\n",
    "from depmapomics import fingerprinting as fp\n",
    "\n",
    "from mgenepy import terra\n",
    "import dalmatian as dm\n",
    "from bokeh.plotting import output_notebook\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "isCCLE = True\n",
    "doCleanup = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading new data\n",
    "\n",
    "Currently, sequenced data for DepMap is generated by the Genomics Platform (GP) at the Broad who deposits them into several different Terra workspaces. Therefore, the first step of this pipeline is to look at these workspaces and:\n",
    "\n",
    " - identify new samples by looking at the bam files and compare them with bams we have already onboarded\n",
    " - remove duplicates and ones with broken file paths\n",
    " - map files to profiles in Gumbo, if possible\n",
    " - onboard new samples and new versions of old cell lines if we find any"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The following two cells scan the delivery workspaces and add new samples to gumbo. Currently under construction to be regularly run off-cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Currently working on running this step off-cycle\n",
    "# if isCCLE:\n",
    "#     print(\"loading new WGS data\")\n",
    "#     from depmap_omics_upload import loading\n",
    "#     wgssamples, unmapped = loading.loadFromMultipleWorkspaces(WGSWORKSPACES, EXTRACT_DEFAULTS[\"sm_id\"], \"SMIDOrdered\", \"wgs\", bamcol=\"cram_path\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Currently working on running this step off-cycle\n",
    "# if isCCLE:\n",
    "#     from depmap_omics_upload import loading\n",
    "#     # write samples to Sequencing table, copy bam files to internal storage bucket:\n",
    "#     wgssamples, cmds = loading.addSamplesToGumbo(wgssamples, 'wgs', WGS_GCS_PATH, filetypes=[\"cram\", \"crai\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All WGS sequencingIDs in gumbo that are not in the WGS terra workspace yet are considered \"new\" for the current release. Here we add them to the terra processing workspace as a sample set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    from depmap_omics_upload import loading\n",
    "    # load new rna samples from gumbo to WGS terra workspace:\n",
    "    loading.addSamplesToDepMapWorkspace('wgs', env_config.WGSWORKSPACE, samplesetname=constants.SAMPLESETNAME, add_to_samplesets=['allcurrent'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run SNP fingerprinting, new (rna + wgs) vs all existing samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wgs_wm = dm.WorkspaceManager(env_config.WGSWORKSPACE)\n",
    "rna_wm = dm.WorkspaceManager(env_config.RNAWORKSPACE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wgs_all = wgs_wm.get_sample_sets().loc[constants.SAMPLESETNAME, \"samples\"]\n",
    "rna_all = rna_wm.get_sample_sets().loc[constants.SAMPLESETNAME, \"samples\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    updated_lod_mat, mismatches, matches = await fp._CCLEFingerPrint(rna_all, wgs_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run pipeline on Terra\n",
    "\n",
    "We are using Dalmatian to send requests to Terra. See [our readme](https://github.com/broadinstitute/depmap_omics/blob/master/documentation/DepMap_processing_pipeline.md) for detailed breakdown of the subtasks in our WGS pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For non internal users, please make sure that your workspace is correctly setup\n",
    "\n",
    "To set up your workspace, follow the instructions in the README page."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"running Terra pipeline\")\n",
    "refwm = dm.WorkspaceManager(env_config.WGSWORKSPACE)\n",
    "submission_id = refwm.create_submission(\"WGS_pipeline\", constants.SAMPLESETNAME, 'sample_set', expression='this.samples')\n",
    "await terra.waitForSubmission(env_config.WGSWORKSPACE, submission_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_id = refwm.create_submission(\"Aggregate_CN_seg_files\", 'all')\n",
    "await terra.waitForSubmission(env_config.WGSWORKSPACE, submission_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the workflow configurations used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terra.saveWorkspace(env_config.WGSWORKSPACE,'data/'+constants.SAMPLESETNAME+'/WGSconfig/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Postprocessing on local\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wgs_wm = dm.WorkspaceManager(env_config.WGSWORKSPACE)\n",
    "wgs_samples = wgs_wm.get_samples()\n",
    "wgs_purecn = wgs_samples[(~wgs_samples.PureCN_loh.isna()) & (wgs_samples.PureCN_loh != \"NA\")].index.tolist()\n",
    "wgs_wm.update_sample_set(sample_set_id=\"PureCN\", sample_ids=wgs_purecn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if isCCLE:\n",
    "    wespriosegs, wgspriosegs = await dm_omics.cnPostProcessing(samplesetname=constants.SAMPLESETNAME, wesrefworkspace=env_config.WESCNWORKSPACE, wgsrefworkspace=env_config.WGSWORKSPACE, dryrun=False, useCache=False)\n",
    "else:\n",
    "    segments, genecn, failed, purecn_segments, purecn_genecn, loh_status, feature_table = await omics_cn.postProcess(env_config.WGSWORKSPACE, sampleset=constants.SAMPLESETNAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Somatic Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from depmapomics.config_prod import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p output/23Q4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DOING WES\n",
      "loading from Terra\n",
      "MUTECT1_CS_SNV\n",
      "MUTECT1_VEP_annotated_vcf\n",
      "MUTECT2_VCF_ALL\n",
      "MUTECT2_VCF_INDELS\n",
      "MUTECT2_VEP_annotated_vcf\n",
      "MuTect1_merged_coverage_wig\n",
      "MuTect1_merged_power_wig\n",
      "PDO\n",
      "ProfileID\n",
      "PureCN_chromosomes_pdf\n",
      "PureCN_cin\n",
      "PureCN_cin_allele_specific\n",
      "PureCN_cin_allele_specific_ploidy_robust\n",
      "PureCN_cin_ploidy_robust\n",
      "PureCN_comment\n",
      "PureCN_contamination\n",
      "PureCN_curated\n",
      "PureCN_curated_solution\n",
      "PureCN_dnacopy\n",
      "PureCN_failed\n",
      "PureCN_flagged\n",
      "PureCN_genes\n",
      "PureCN_local_optima_pdf\n",
      "PureCN_log\n",
      "PureCN_loh\n",
      "PureCN_loh_fraction\n",
      "PureCN_ploidy\n",
      "PureCN_purity\n",
      "PureCN_rds\n",
      "PureCN_segmentation\n",
      "PureCN_selected_solution\n",
      "PureCN_solutions_pdf\n",
      "PureCN_variants\n",
      "PureCN_wgd\n",
      "STRELKA_VEP_annotated_vcf\n",
      "SmId\n",
      "absolute_highres_plot\n",
      "absolute_rdata\n",
      "age\n",
      "allele_fraction_legacy_segments_normal\n",
      "allele_fraction_legacy_segments_tumor\n",
      "allele_fraction_parameters_begin_normal\n",
      "allele_fraction_parameters_begin_tumor\n",
      "allele_fraction_parameters_normal\n",
      "allele_fraction_parameters_tumor\n",
      "allelic_counts_entity_id_normal\n",
      "allelic_counts_entity_id_tumor\n",
      "allelic_counts_normal\n",
      "allelic_counts_tumor\n",
      "alleliccapseg_plot\n",
      "alleliccapseg_skew\n",
      "alleliccapseg_tsv\n",
      "analysis_ready_bam_md5\n",
      "analysis_ready_bam_md5_alt\n",
      "annotatedMAF_cga_cf\n",
      "arxspan_id\n",
      "baits\n",
      "blat_all_maf\n",
      "blat_number_mutations_passed\n",
      "blat_number_mutations_rejected\n",
      "blat_passed_mutations\n",
      "bmutect2_with_germline_amout_index\n",
      "bqsr_report\n",
      "bqsr_report_alt\n",
      "called_copy_ratio_segments_normal\n",
      "called_copy_ratio_segments_tumor\n",
      "candidate_indel_vcf\n",
      "candidate_indel_vcf_index\n",
      "candidate_sv_vcf\n",
      "candidate_sv_vcf_index\n",
      "cellosaurus_id\n",
      "cnn_filtered_vcf\n",
      "cnn_filtered_vcf_h19_rejected\n",
      "cnn_filtered_vcf_hg19\n",
      "cnn_filtered_vcf_index\n",
      "comments\n",
      "consideredCount_cga_cf\n",
      "contamination_data\n",
      "contestAFFile\n",
      "contest_base_report\n",
      "contest_validation\n",
      "copy_number_qc_mix_ups\n",
      "copy_number_qc_report\n",
      "copy_number_qc_report_png\n",
      "copy_ratio_legacy_segments_normal\n",
      "copy_ratio_legacy_segments_tumor\n",
      "copy_ratio_only_segments_normal\n",
      "copy_ratio_only_segments_tumor\n",
      "copy_ratio_parameters_begin_normal\n",
      "copy_ratio_parameters_begin_tumor\n",
      "copy_ratio_parameters_normal\n",
      "copy_ratio_parameters_tumor\n",
      "coverage_file\n",
      "crc32c_hash\n",
      "cross_check_fingprt_metrics\n",
      "cross_check_fingprt_min_lod_lanes\n",
      "cross_check_fingprt_min_lod_value\n",
      "cross_check_fingprt_report\n",
      "datatype\n",
      "delta_MAD_normal\n",
      "delta_MAD_tumor\n",
      "denoised_MAD_normal\n",
      "denoised_MAD_tumor\n",
      "denoised_copy_ratios_lim_4_plot_normal\n",
      "denoised_copy_ratios_lim_4_plot_tumor\n",
      "denoised_copy_ratios_normal\n",
      "denoised_copy_ratios_plot_normal\n",
      "denoised_copy_ratios_plot_tumor\n",
      "denoised_copy_ratios_tumor\n",
      "depmap_maf\n",
      "depmap_maf_23q4\n",
      "depmap_maf_whitelisted\n",
      "dna_pipeline_main_parquet\n",
      "duplication_metrics\n",
      "duplication_metrics_alt\n",
      "expected_type\n",
      "ffpe_OBF_figures\n",
      "ffpe_OBF_number_mutations_passed\n",
      "ffpe_OBF_number_mutations_rejected\n",
      "ffpe_OBF_passed_and_rejected_mutations\n",
      "ffpe_OBF_passed_mutations\n",
      "ffpe_OBF_q_val\n",
      "filter_passed_and_rejected_mutations\n",
      "filter_passed_mutations\n",
      "filtered_annotated_sv\n",
      "filters_passed_merged_intersection_maf\n",
      "filters_passed_merged_union_maf\n",
      "fixed_vcf\n",
      "fracContam\n",
      "full_annotated_parquet\n",
      "full_file\n",
      "full_file_whitelisted\n",
      "full_vcf\n",
      "full_vcf_idx\n",
      "gatk_cnv_all_plots\n",
      "gatk_cnv_coverage_file\n",
      "gatk_cnv_pre_tn_coverage\n",
      "gatk_cnv_seg_file\n",
      "gatk_cnv_tn_coverage\n",
      "gatk_het_ad_normal\n",
      "gatk_het_ad_tumor\n",
      "gender\n",
      "germline_sv_vcf\n",
      "germline_sv_vcf_index\n",
      "het_allelic_counts_normal\n",
      "het_allelic_counts_tumor\n",
      "hg19_bai_filepath\n",
      "hg19_bam_filepath\n",
      "hg38_analysis_ready_bam\n",
      "hg38_analysis_ready_bam_index\n",
      "hg38_analysis_ready_bam_md5\n",
      "hg38_bqsr_report\n",
      "hg38_duplication_metrics\n",
      "hgvs_maf_masked\n",
      "hgvs_oc_vcf_masked\n",
      "internal_bai_filepath\n",
      "internal_bai_filepath_alt\n",
      "internal_bam_filepath\n",
      "internal_bam_filepath_alt\n",
      "legacy_bai_filepath\n",
      "legacy_bam_filepath\n",
      "legacy_crc32c_hash\n",
      "legacy_size\n",
      "lego_plotter_ais\n",
      "lego_plotter_figs\n",
      "lego_plotter_pngs\n",
      "lego_plotter_pss\n",
      "maf\n",
      "matched_normal\n",
      "md5_hash\n",
      "mean_corrected_auto_depth\n",
      "mean_haploid_depth\n",
      "mean_mito_depth\n",
      "mediatype\n",
      "mito_ratio\n",
      "modeled_segments_begin_normal\n",
      "modeled_segments_begin_tumor\n",
      "modeled_segments_normal\n",
      "modeled_segments_plot_normal\n",
      "modeled_segments_plot_tumor\n",
      "modeled_segments_tumor\n",
      "msisensor2_output\n",
      "msisensor2_output_dis\n",
      "msisensor2_output_somatic\n",
      "msisensor2_score\n",
      "mut_legos_html\n",
      "mutation_validator_validated_maf\n",
      "mutect1_snv_mutect2_indel_strelka_indel_annotated_maf\n",
      "mutect2_bamout\n",
      "mutect2_bamout_index\n",
      "mutect2_contamination_table\n",
      "mutect2_filtered_vcf\n",
      "mutect2_filtered_vcf_idx\n",
      "mutect2_filtering_stats\n",
      "mutect2_fixed_vcf\n",
      "mutect2_funcotated_file\n",
      "mutect2_funcotated_file_index\n",
      "mutect2_maf_segments\n",
      "mutect2_mutect_stats\n",
      "mutect2_out_vcf\n",
      "mutect2_out_vcf_index\n",
      "mutect2_read_orientation_model_params\n",
      "mutect2_with_germline_bamout\n",
      "mutect2_with_germline_bamout_index\n",
      "mutect2_with_germline_contamination_table\n",
      "mutect2_with_germline_filtered_vcf\n",
      "mutect2_with_germline_filtered_vcf_idx\n",
      "mutect2_with_germline_filtering_stats\n",
      "mutect2_with_germline_funcotated_file\n",
      "mutect2_with_germline_funcotated_file_index\n",
      "mutect2_with_germline_maf_segments\n",
      "mutect2_with_germline_mutect_stats\n",
      "mutect2_with_germline_read_orientation_model_params\n",
      "mutect_force_call_cs\n",
      "mutect_force_call_cw\n",
      "mutect_force_call_pw\n",
      "normal_bam_alignment_summary_metrics\n",
      "normal_bam_bait_bias_detail_metrics\n",
      "normal_bam_bait_bias_summary_metrics\n",
      "normal_bam_bam_validation\n",
      "normal_bam_base_distribution_by_cycle\n",
      "normal_bam_base_distribution_by_cycle_metrics\n",
      "normal_bam_converted_oxog_metrics\n",
      "normal_bam_gc_bias\n",
      "normal_bam_gc_bias_detail_metrics\n",
      "normal_bam_gc_bias_summary_metrics\n",
      "normal_bam_hybrid_selection_metrics\n",
      "normal_bam_insert_size_histogram\n",
      "normal_bam_insert_size_metrics\n",
      "normal_bam_lane_list\n",
      "normal_bam_pre_adapter_detail_metrics\n",
      "normal_bam_pre_adapter_summary_metrics\n",
      "normal_bam_quality_by_cycle\n",
      "normal_bam_quality_by_cycle_metrics\n",
      "normal_bam_quality_distribution\n",
      "normal_bam_quality_distribution_metrics\n",
      "normal_bam_quality_yield_metrics\n",
      "normal_bam_read_coverage_lane\n",
      "normal_het_allelic_counts_normal\n",
      "normal_het_allelic_counts_tumor\n",
      "number_mutations_passed\n",
      "number_mutations_rejected\n",
      "oc_err\n",
      "oc_error_file\n",
      "oc_error_files\n",
      "oc_log\n",
      "oc_log_file\n",
      "oc_log_files\n",
      "oc_main_file\n",
      "oc_main_files\n",
      "oc_sql_files\n",
      "oc_sqlite_file\n",
      "oc_tsv_file\n",
      "oncotated_called_file_tumor\n",
      "oncotated_called_gene_list_file_tumor\n",
      "origin\n",
      "outMAFannotatedFN\n",
      "outMAFfn\n",
      "output_vcf\n",
      "output_vcf_index\n",
      "oxoG_OBF_figures\n",
      "oxoG_OBF_number_mutations_passed\n",
      "oxoG_OBF_number_mutations_rejected\n",
      "oxoG_OBF_passed_and_rejected_mutations\n",
      "oxoG_OBF_passed_mutations\n",
      "oxoG_OBF_q_val\n",
      "parent_cell_line\n",
      "participant\n",
      "participant_id\n",
      "passCount_cga_cf\n",
      "passedMAF_cga_cf\n",
      "patient_id\n",
      "preprocessed_intervals\n",
      "primary_disease\n",
      "primary_site\n",
      "read_counts_entity_id_normal\n",
      "read_counts_entity_id_tumor\n",
      "read_counts_normal\n",
      "read_counts_tumor\n",
      "rejectCount_cga_cf\n",
      "rejectedMAF_cga_cf\n",
      "scaled_delta_MAD_normal\n",
      "scaled_delta_MAD_tumor\n",
      "sequencing_date\n",
      "sex\n",
      "size\n",
      "somatic_annotated_sv\n",
      "somatic_sv_vcf\n",
      "somatic_sv_vcf_index\n",
      "source\n",
      "standardized_MAD_normal\n",
      "standardized_MAD_tumor\n",
      "standardized_copy_ratios_normal\n",
      "standardized_copy_ratios_tumor\n",
      "strelka_all_indels\n",
      "strelka_all_snvs\n",
      "strelka_passed_indels\n",
      "strelka_passed_snvs\n",
      "stripped_cell_line_name\n",
      "subsubtype\n",
      "subtype\n",
      "sv_dropped\n",
      "tumor_bam_alignment_summary_metrics\n",
      "tumor_bam_bait_bias_detail_metrics\n",
      "tumor_bam_bait_bias_summary_metrics\n",
      "tumor_bam_bam_validation\n",
      "tumor_bam_base_distribution_by_cycle\n",
      "tumor_bam_base_distribution_by_cycle_metrics\n",
      "tumor_bam_converted_oxog_metrics\n",
      "tumor_bam_gc_bias\n",
      "tumor_bam_gc_bias_detail_metrics\n",
      "tumor_bam_gc_bias_summary_metrics\n",
      "tumor_bam_hybrid_selection_metrics\n",
      "tumor_bam_insert_size_histogram\n",
      "tumor_bam_insert_size_metrics\n",
      "tumor_bam_lane_list\n",
      "tumor_bam_pre_adapter_detail_metrics\n",
      "tumor_bam_pre_adapter_summary_metrics\n",
      "tumor_bam_quality_by_cycle\n",
      "tumor_bam_quality_by_cycle_metrics\n",
      "tumor_bam_quality_distribution\n",
      "tumor_bam_quality_distribution_metrics\n",
      "tumor_bam_quality_yield_metrics\n",
      "tumor_bam_read_coverage_lane\n",
      "ubam_validation_reports\n",
      "unmapped_bams_list\n",
      "update_time\n",
      "vcf_fixedclust\n",
      "vcf_fixedcol\n",
      "vcf_fixedploid\n",
      "vcf_fully_annotated\n",
      "vep\n",
      "version\n",
      "depmap_maf_23q4\n",
      "0 samples don't have corresponding maf:  set()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  9%|████████████▍                                                                                                                       | 197/2095 [00:54<07:37,  4.15it/s]/Users/qqin/Workspace/depmap_omics_upstream/depmapomics/mutations.py:201: DtypeWarning: Columns (21,54,59,67,70,72,78) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  maf = pd.read_csv(row[mafcol])\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2095/2095 [07:49<00:00,  4.46it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "further filtering and standardizing maf\n",
      "saving somatic mutations (all)\n",
      "done\n",
      "connecting to gumbo@localhost:5432/gumbo\n",
      "Warning: Connect fail, but will retry: connection to server at \"localhost\" (::1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "connection to server at \"localhost\" (127.0.0.1), port 5432 failed: server closed the connection unexpectedly\n",
      "\tThis probably means the server terminated abnormally\n",
      "\tbefore or while processing the request.\n",
      "\n",
      "Warning: Connect fail, but will retry: connection to server at \"localhost\" (::1), port 5432 failed: Connection refused\n",
      "\tIs the server running on that host and accepting TCP/IP connections?\n",
      "connection to server at \"localhost\" (127.0.0.1), port 5432 failed: server closed the connection unexpectedly\n",
      "\tThis probably means the server terminated abnormally\n",
      "\tbefore or while processing the request.\n",
      "\n",
      "setting username to szhang\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/qqin/miniconda3/envs/cds-omics/lib/python3.9/site-packages/gumbo_client/client.py:273: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df = pd.read_sql(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clearing username\n",
      "DOING WGS\n",
      "loading from Terra\n",
      "MUTECT1_CS_SNV\n",
      "MUTECT1_VEP_annotated_vcf\n",
      "MUTECT2_VCF_ALL\n",
      "MUTECT2_VCF_INDELS\n",
      "MUTECT2_VEP_annotated_vcf\n",
      "MuTect1_merged_coverage_wig\n",
      "MuTect1_merged_power_wig\n",
      "PDO\n",
      "PDO-ID\n",
      "PdoId\n",
      "ProfileID\n",
      "PureCN_chromosomes_pdf\n",
      "PureCN_cin\n",
      "PureCN_cin_allele_specific\n",
      "PureCN_cin_allele_specific_ploidy_robust\n",
      "PureCN_cin_ploidy_robust\n",
      "PureCN_comment\n",
      "PureCN_contamination\n",
      "PureCN_curated\n",
      "PureCN_curated_solution\n",
      "PureCN_dnacopy\n",
      "PureCN_failed\n",
      "PureCN_flagged\n",
      "PureCN_genes\n",
      "PureCN_local_optima_pdf\n",
      "PureCN_log\n",
      "PureCN_loh\n",
      "PureCN_loh_fraction\n",
      "PureCN_ploidy\n",
      "PureCN_purity\n",
      "PureCN_rds\n",
      "PureCN_segmentation\n",
      "PureCN_selected_solution\n",
      "PureCN_solutions_pdf\n",
      "PureCN_variants\n",
      "PureCN_wgd\n",
      "SM-ID\n",
      "STRELKA_VEP_annotated_vcf\n",
      "SmId\n",
      "absolute_highres_plot\n",
      "absolute_rdata\n",
      "age\n",
      "allele_fraction_legacy_segments_normal\n",
      "allele_fraction_legacy_segments_tumor\n",
      "allele_fraction_parameters_begin_normal\n",
      "allele_fraction_parameters_begin_tumor\n",
      "allele_fraction_parameters_normal\n",
      "allele_fraction_parameters_tumor\n",
      "allelic_counts_entity_id_normal\n",
      "allelic_counts_entity_id_tumor\n",
      "allelic_counts_normal\n",
      "allelic_counts_tumor\n",
      "alleliccapseg_plot\n",
      "alleliccapseg_skew\n",
      "alleliccapseg_tsv\n",
      "analysis_ready_bam\n",
      "analysis_ready_bam_index\n",
      "analysis_ready_bam_md5\n",
      "annotatedMAF\n",
      "annotatedMAF_cga_cf\n",
      "arxspan_id\n",
      "bai_from_cram\n",
      "baits\n",
      "bam_from_cram\n",
      "bamout\n",
      "bamout_index\n",
      "blacklist\n",
      "blat_all_maf\n",
      "blat_number_mutations_passed\n",
      "blat_number_mutations_rejected\n",
      "blat_passed_mutations\n",
      "bqsr_report\n",
      "called_copy_ratio_legacy_segments_normal\n",
      "called_copy_ratio_legacy_segments_tumor\n",
      "called_copy_ratio_segments_normal\n",
      "called_copy_ratio_segments_tumor\n",
      "candidate_indel_vcf\n",
      "candidate_indel_vcf_index\n",
      "candidate_sv_vcf\n",
      "candidate_sv_vcf_index\n",
      "cellosaurus_id\n",
      "cnn_filtered_vcf\n",
      "cnn_filtered_vcf_index\n",
      "collection_site\n",
      "comments\n",
      "consideredCount\n",
      "consideredCount_cga_cf\n",
      "contamination_data\n",
      "contestAFFile\n",
      "contest_base_report\n",
      "contest_validation\n",
      "copy_number_qc_mix_ups\n",
      "copy_number_qc_report\n",
      "copy_number_qc_report_png\n",
      "copy_ratio_legacy_segments_normal\n",
      "copy_ratio_legacy_segments_tumor\n",
      "copy_ratio_only_segments_normal\n",
      "copy_ratio_only_segments_tumor\n",
      "copy_ratio_parameters_begin_normal\n",
      "copy_ratio_parameters_begin_tumor\n",
      "copy_ratio_parameters_normal\n",
      "copy_ratio_parameters_tumor\n",
      "crc32c_hash\n",
      "cross_check_fingprt_metrics\n",
      "cross_check_fingprt_min_lod_lanes\n",
      "cross_check_fingprt_min_lod_value\n",
      "cross_check_fingprt_report\n",
      "datatype\n",
      "delta_MAD_normal\n",
      "delta_MAD_tumor\n",
      "delta_MAD_value_normal\n",
      "delta_MAD_value_tumor\n",
      "denoised_MAD_normal\n",
      "denoised_MAD_tumor\n",
      "denoised_MAD_value_normal\n",
      "denoised_MAD_value_tumor\n",
      "denoised_copy_ratios_lim_4_plot_normal\n",
      "denoised_copy_ratios_lim_4_plot_tumor\n",
      "denoised_copy_ratios_normal\n",
      "denoised_copy_ratios_plot_normal\n",
      "denoised_copy_ratios_plot_tumor\n",
      "denoised_copy_ratios_tumor\n",
      "depmap_maf\n",
      "depmap_maf_23q4\n",
      "depmap_maf_whitelisted\n",
      "dna_pipeline_main_parquet\n",
      "duplication_metrics\n",
      "expected_type\n",
      "ffpe_OBF_figures\n",
      "ffpe_OBF_number_mutations_passed\n",
      "ffpe_OBF_number_mutations_rejected\n",
      "ffpe_OBF_passed_and_rejected_mutations\n",
      "ffpe_OBF_passed_mutations\n",
      "ffpe_OBF_q_val\n",
      "file_list_name\n",
      "filter_passed_and_rejected_mutations\n",
      "filter_passed_mutations\n",
      "filtered_annotated_sv\n",
      "filters_passed_merged_intersection_maf\n",
      "filters_passed_merged_union_maf\n",
      "final_vcf\n",
      "fracContam\n",
      "full_annotated_parquet\n",
      "full_file\n",
      "full_file_whitelisted\n",
      "full_vcf_idx\n",
      "gatk_cnv_all_plots\n",
      "gatk_cnv_coverage_file\n",
      "gatk_cnv_pre_tn_coverage\n",
      "gatk_cnv_seg_file\n",
      "gatk_cnv_tn_coverage\n",
      "gatk_het_ad_normal\n",
      "gatk_het_ad_tumor\n",
      "germline_sv_vcf\n",
      "germline_sv_vcf_index\n",
      "het_allelic_counts_normal\n",
      "het_allelic_counts_tumor\n",
      "hg19_bai_filepath\n",
      "hg19_bam_filepath\n",
      "hg38_crai_filepath\n",
      "hg38_cram_filepath\n",
      "hgvs_maf_masked\n",
      "hgvs_oc_vcf_masked\n",
      "internal_bai_filepath\n",
      "internal_bam_filepath\n",
      "issue\n",
      "legacy_crc32c_hash\n",
      "legacy_size\n",
      "lego_plotter_ais\n",
      "lego_plotter_figs\n",
      "lego_plotter_pngs\n",
      "lego_plotter_pss\n",
      "lifted_vcf\n",
      "lineage\n",
      "md5_hash\n",
      "mediatype\n",
      "modeled_segments_begin_normal\n",
      "modeled_segments_begin_tumor\n",
      "modeled_segments_normal\n",
      "modeled_segments_plot_normal\n",
      "modeled_segments_plot_tumor\n",
      "modeled_segments_tumor\n",
      "msisensor2_output\n",
      "msisensor2_output_dis\n",
      "msisensor2_output_somatic\n",
      "msisensor2_score\n",
      "mut_legos_html\n",
      "mutation_validator_validated_maf\n",
      "mutect1_snv_mutect2_indel_strelka_indel_annotated_maf\n",
      "mutect2_bamout\n",
      "mutect2_bamout_index\n",
      "mutect2_contamination_table\n",
      "mutect2_filtered_vcf\n",
      "mutect2_filtered_vcf_idx\n",
      "mutect2_filtering_stats\n",
      "mutect2_fixed_vcf\n",
      "mutect2_funcotated_file\n",
      "mutect2_funcotated_file_index\n",
      "mutect2_maf_segments\n",
      "mutect2_mutect_stats\n",
      "mutect2_read_orientation_model_params\n",
      "mutect_force_call_cs\n",
      "mutect_force_call_cw\n",
      "mutect_force_call_pw\n",
      "normal_bam_alignment_summary_metrics\n",
      "normal_bam_bait_bias_detail_metrics\n",
      "normal_bam_bait_bias_summary_metrics\n",
      "normal_bam_bam_validation\n",
      "normal_bam_base_distribution_by_cycle\n",
      "normal_bam_base_distribution_by_cycle_metrics\n",
      "normal_bam_converted_oxog_metrics\n",
      "normal_bam_gc_bias\n",
      "normal_bam_gc_bias_detail_metrics\n",
      "normal_bam_gc_bias_summary_metrics\n",
      "normal_bam_hybrid_selection_metrics\n",
      "normal_bam_insert_size_histogram\n",
      "normal_bam_insert_size_metrics\n",
      "normal_bam_lane_list\n",
      "normal_bam_pre_adapter_detail_metrics\n",
      "normal_bam_pre_adapter_summary_metrics\n",
      "normal_bam_quality_by_cycle\n",
      "normal_bam_quality_by_cycle_metrics\n",
      "normal_bam_quality_distribution\n",
      "normal_bam_quality_distribution_metrics\n",
      "normal_bam_quality_yield_metrics\n",
      "normal_bam_read_coverage_lane\n",
      "normal_het_allelic_counts_normal\n",
      "normal_het_allelic_counts_tumor\n",
      "number_mutations_passed\n",
      "number_mutations_rejected\n",
      "oc_error_file\n",
      "oc_error_files\n",
      "oc_log_file\n",
      "oc_log_files\n",
      "oc_main_files\n",
      "oc_sql_files\n",
      "oncotated_called_file_tumor\n",
      "oncotated_called_gene_list_file_tumor\n",
      "origin\n",
      "outMAFannotatedFN\n",
      "outMAFfn\n",
      "output_sv_vcf\n",
      "output_sv_vcf_index\n",
      "oxoG_OBF_figures\n",
      "oxoG_OBF_number_mutations_passed\n",
      "oxoG_OBF_number_mutations_rejected\n",
      "oxoG_OBF_passed_and_rejected_mutations\n",
      "oxoG_OBF_passed_mutations\n",
      "oxoG_OBF_q_val\n",
      "parent_cell_line\n",
      "participant\n",
      "participant_id\n",
      "passCount\n",
      "passCount_cga_cf\n",
      "passedMAF\n",
      "passedMAF_cga_cf\n",
      "patient_id\n",
      "preprocessed_intervals\n",
      "primary_disease\n",
      "primary_site\n",
      "read_counts_entity_id_normal\n",
      "read_counts_entity_id_tumor\n",
      "read_counts_normal\n",
      "read_counts_tumor\n",
      "rejectCount\n",
      "rejectCount_cga_cf\n",
      "rejectedMAF\n",
      "rejectedMAF_cga_cf\n",
      "rejected_mutations_fromliftover\n",
      "scaled_delta_MAD_normal\n",
      "scaled_delta_MAD_tumor\n",
      "scaled_delta_MAD_value_normal\n",
      "scaled_delta_MAD_value_tumor\n",
      "sequencing_date\n",
      "sex\n",
      "size\n",
      "somatic_annotated_sv\n",
      "somatic_sv_vcf\n",
      "somatic_sv_vcf_index\n",
      "source\n",
      "standardized_MAD_normal\n",
      "standardized_MAD_tumor\n",
      "standardized_MAD_value_normal\n",
      "standardized_MAD_value_tumor\n",
      "standardized_copy_ratios_normal\n",
      "standardized_copy_ratios_tumor\n",
      "str_hipstr\n",
      "strelka_all_indels\n",
      "strelka_all_snvs\n",
      "strelka_passed_indels\n",
      "strelka_passed_snvs\n",
      "stripped_cell_line_name\n",
      "subsubtype\n",
      "subtype\n",
      "sv_dropped\n",
      "tumor_bam_alignment_summary_metrics\n",
      "tumor_bam_bait_bias_detail_metrics\n",
      "tumor_bam_bait_bias_summary_metrics\n",
      "tumor_bam_bam_validation\n",
      "tumor_bam_base_distribution_by_cycle\n",
      "tumor_bam_base_distribution_by_cycle_metrics\n",
      "tumor_bam_converted_oxog_metrics\n",
      "tumor_bam_gc_bias\n",
      "tumor_bam_gc_bias_detail_metrics\n",
      "tumor_bam_gc_bias_summary_metrics\n",
      "tumor_bam_hybrid_selection_metrics\n",
      "tumor_bam_insert_size_histogram\n",
      "tumor_bam_insert_size_metrics\n",
      "tumor_bam_lane_list\n",
      "tumor_bam_pre_adapter_detail_metrics\n",
      "tumor_bam_pre_adapter_summary_metrics\n",
      "tumor_bam_quality_by_cycle\n",
      "tumor_bam_quality_by_cycle_metrics\n",
      "tumor_bam_quality_distribution\n",
      "tumor_bam_quality_distribution_metrics\n",
      "tumor_bam_quality_yield_metrics\n",
      "tumor_bam_read_coverage_lane\n",
      "ubam_file_list\n",
      "ubam_validation_reports\n",
      "update_time\n",
      "version\n",
      "depmap_maf_23q4\n",
      "1 samples don't have corresponding maf:  {'CDS-fUSLRB'}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 83%|████████████████████████████████████████████████████████████████████████████████████████████████████████████                       | 1039/1259 [03:56<00:46,  4.72it/s]/Users/qqin/Workspace/depmap_omics_upstream/depmapomics/mutations.py:201: DtypeWarning: Columns (53,62,67,72,75,78) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  maf = pd.read_csv(row[mafcol])\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1259/1259 [04:46<00:00,  4.40it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "further filtering and standardizing maf\n",
      "saving somatic mutations (all)\n",
      "done\n",
      "merging WES and WGS\n",
      "downloading gene names from biomart\n",
      "adding entrez id column\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/qqin/Workspace/depmap_omics_upstream/depmapomics/dm_omics.py:822: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  if \"Y\" in mergedmutations[col].values:\n",
      "/Users/qqin/Workspace/depmap_omics_upstream/depmapomics/dm_omics.py:845: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  if \"Y\" in merged[col].values:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading gene names from biomart\n",
      "adding entrez id column\n",
      "creating mutation matrices\n",
      "generating genotyped driver and damaging mutation matrix\n",
      "No dataset version provided. Using version 25.\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'output/23Q4/merged_binary_germline_avana.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m isCCLE:\n\u001b[0;32m----> 2\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m dm_omics\u001b[38;5;241m.\u001b[39mmutationPostProcessing(wesrefworkspace\u001b[38;5;241m=\u001b[39mWESCNWORKSPACE, wgsrefworkspace\u001b[38;5;241m=\u001b[39mWGSWORKSPACE, run_guidemat\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, run_sv\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, mafcol\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdepmap_maf_23q4\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m omics_mut\u001b[38;5;241m.\u001b[39mpostProcess(env_config\u001b[38;5;241m.\u001b[39mWGSWORKSPACE, samplesetname\u001b[38;5;241m=\u001b[39mconstants\u001b[38;5;241m.\u001b[39mSAMPLESETNAME)\n",
      "File \u001b[0;32m~/Workspace/depmap_omics_upstream/depmapomics/dm_omics.py:913\u001b[0m, in \u001b[0;36mmutationPostProcessing\u001b[0;34m(wesrefworkspace, wgsrefworkspace, vcfdir, vcf_colname, samplesetname, AllSamplesetName, taiga_description, taiga_dataset, bed_locations, sv_col, sv_filename, mutcol, standardmafcol, mafcol, doCleanup, run_sv, run_guidemat, upload_taiga, **kwargs)\u001b[0m\n\u001b[1;32m    911\u001b[0m \u001b[38;5;66;03m# uploading to taiga\u001b[39;00m\n\u001b[1;32m    912\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m upload_taiga:\n\u001b[0;32m--> 913\u001b[0m     \u001b[43mtc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdate_dataset\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    914\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchanges_description\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mnew \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msamplesetname\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m release!\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    915\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset_permaname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtaiga_dataset\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    916\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupload_files\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\n\u001b[1;32m    917\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    918\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations_genotyped_hotspot_profile.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    919\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_genotypedMatrix_hotspot_profile\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    920\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mNumericMatrixCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    921\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    922\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    923\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    924\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations_genotyped_damaging_profile.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    925\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_genotypedMatrix_damaging_profile\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    926\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mNumericMatrixCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    927\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    928\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    929\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    930\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations_profile.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    931\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_profile\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    932\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    933\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    934\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    935\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    936\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations_profile.maf.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    937\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_profile_maf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    938\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    939\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    940\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    941\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    942\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations_all_cols_profile.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    943\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_profile_maf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    944\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    945\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    946\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    947\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    948\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomatic_mutations.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    949\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msomaticMutations_withReplicates\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    950\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    951\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    952\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    953\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    954\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_germline_avana.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    955\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_mutation_avana\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    956\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    957\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    958\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    959\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    960\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_germline_ky.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    961\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_mutation_ky\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    962\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    963\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    964\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    965\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    966\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_germline_humagne.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    967\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinary_mutation_humagne\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    968\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    969\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    970\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    971\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    972\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msvs.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    973\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstructuralVariants_withReplicates\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    974\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    975\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    976\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    977\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    978\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msvs_profile.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    979\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstructuralVariants_profile\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    980\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTableCSV\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    981\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mutf-8\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    982\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    983\u001b[0m \u001b[43m        \u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    984\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupload_async\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    985\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset_description\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtaiga_description\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    986\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/cds-omics/lib/python3.9/site-packages/taigapy/client.py:773\u001b[0m, in \u001b[0;36mTaigaClient.update_dataset\u001b[0;34m(self, dataset_id, dataset_permaname, dataset_version, dataset_description, changes_description, upload_files, add_taiga_ids, add_gcs_files, add_all_existing_files, upload_async)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_set_token_and_initialized_api()\n\u001b[1;32m    769\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    770\u001b[0m     (\n\u001b[1;32m    771\u001b[0m         all_uploads,\n\u001b[1;32m    772\u001b[0m         dataset_version_metadata,\n\u001b[0;32m--> 773\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_update_dataset_arguments\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    774\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    775\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset_permaname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    776\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset_version\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    777\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchanges_description\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    778\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupload_files\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    779\u001b[0m \u001b[43m        \u001b[49m\u001b[43madd_taiga_ids\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    780\u001b[0m \u001b[43m        \u001b[49m\u001b[43madd_gcs_files\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    781\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    782\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mValueError\u001b[39;00m, Taiga404Exception) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    783\u001b[0m     \u001b[38;5;28mprint\u001b[39m(cf\u001b[38;5;241m.\u001b[39mred(\u001b[38;5;28mstr\u001b[39m(e)))\n",
      "File \u001b[0;32m~/miniconda3/envs/cds-omics/lib/python3.9/site-packages/taigapy/client.py:446\u001b[0m, in \u001b[0;36mTaigaClient._validate_update_dataset_arguments\u001b[0;34m(self, dataset_id, dataset_permaname, dataset_version, changes_description, upload_files, add_taiga_ids, add_gcs_files)\u001b[0m\n\u001b[1;32m    434\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\n\u001b[1;32m    435\u001b[0m         cf\u001b[38;5;241m.\u001b[39morange(\n\u001b[1;32m    436\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo dataset version provided. Using version \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    439\u001b[0m         )\n\u001b[1;32m    440\u001b[0m     )\n\u001b[1;32m    442\u001b[0m dataset_version_metadata: DatasetVersionMetadataDict \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    443\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_dataset_metadata(dataset_permaname, dataset_version)\n\u001b[1;32m    444\u001b[0m )\n\u001b[0;32m--> 446\u001b[0m all_uploads \u001b[38;5;241m=\u001b[39m \u001b[43mtransform_upload_args_to_upload_list\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupload_files\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m    \u001b[49m\u001b[43madd_taiga_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m    \u001b[49m\u001b[43madd_gcs_files\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdataset_version_metadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    453\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m all_uploads, dataset_version_metadata\n",
      "File \u001b[0;32m~/miniconda3/envs/cds-omics/lib/python3.9/site-packages/taigapy/utils.py:170\u001b[0m, in \u001b[0;36mtransform_upload_args_to_upload_list\u001b[0;34m(upload_files, add_taiga_ids, add_gcs_files, dataset_version_metadata)\u001b[0m\n\u001b[1;32m    168\u001b[0m add_as_virtual \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    169\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m upload_file_dict \u001b[38;5;129;01min\u001b[39;00m upload_files:\n\u001b[0;32m--> 170\u001b[0m     sha256, md5 \u001b[38;5;241m=\u001b[39m \u001b[43mget_file_hashes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mupload_file_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    171\u001b[0m     \u001b[38;5;66;03m# check to see if the previous version had a datafile with the same hashes\u001b[39;00m\n\u001b[1;32m    172\u001b[0m     matching_file: Optional[DatasetVersionFiles] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(\n\u001b[1;32m    173\u001b[0m         (\n\u001b[1;32m    174\u001b[0m             f\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    187\u001b[0m         \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    188\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/cds-omics/lib/python3.9/site-packages/taigapy/utils.py:247\u001b[0m, in \u001b[0;36mget_file_hashes\u001b[0;34m(file_name)\u001b[0m\n\u001b[1;32m    245\u001b[0m sha256 \u001b[38;5;241m=\u001b[39m hashlib\u001b[38;5;241m.\u001b[39msha256()\n\u001b[1;32m    246\u001b[0m md5 \u001b[38;5;241m=\u001b[39m hashlib\u001b[38;5;241m.\u001b[39mmd5()\n\u001b[0;32m--> 247\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfile_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m fd:\n\u001b[1;32m    248\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    249\u001b[0m         buffer \u001b[38;5;241m=\u001b[39m fd\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;241m1024\u001b[39m \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m1024\u001b[39m)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'output/23Q4/merged_binary_germline_avana.csv'"
     ]
    }
   ],
   "source": [
    "if isCCLE:\n",
    "    await dm_omics.mutationPostProcessing(wesrefworkspace=WESCNWORKSPACE, wgsrefworkspace=WGSWORKSPACE, run_guidemat=False, run_sv=False, mafcol=\"depmap_maf_23q4\")\n",
    "else:\n",
    "    await omics_mut.postProcess(env_config.WGSWORKSPACE, samplesetname=constants.SAMPLESETNAME)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset and upload\n",
    "\n",
    "Based on release dates and embargo status in gumbo, subset and upload datasets for each release audience, and hand off to the portal team."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from depmap_omics_upload import tracker\n",
    "from depmap_omics_upload import upload\n",
    "from mgenepy.utils import helper as h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import date\n",
    "import datetime\n",
    "release_date = datetime.date(2023, 5, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "virtual = upload.initVirtualDatasets(samplesetname=constants.SAMPLESETNAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upload.checkDataPermission()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upload.uploadAuxTables(taiga_ids=virtual, today=release_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upload.makeModelLvMatrices(virtual_ids=virtual, today=release_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upload.makePRLvMatrices(virtual_ids=virtual, files_nummat={}, files_table={}, files_raw={\"mutations-latest-ed72\": {\"somaticMutations_profile_maf\": \"OmicsSomaticMutationsMAFProfile.maf\"}}, today=release_date)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Managing release readmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! cd .. && git clone https://github.com/broadinstitute/depmap-release-readmes.git && cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cd ../depmap-release-readmes && git pull --no-commit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cd ../depmap-release-readmes/ && python3 make_new_release.py $constants.RELEASE  && git add . && git commit -m $constants.RELEASE && git push "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cleaning workspaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from depmap_omics_upload.mgenepy import terra as terra_cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if doCleanup:\n",
    "    print(\"cleaning workspaces\")\n",
    "    torm = await terra_cleanup.deleteHeavyFiles(env_config.WGSWORKSPACE)\n",
    "    h.parrun(['gsutil rm '+i for i in torm], cores=8)\n",
    "    terra_cleanup.removeFromFailedWorkflows(env_config.WGSWORKSPACE, dryrun=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving workspace configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! terra-sync export broad-firecloud-ccle/DepMap_WGS_CN data/$constants.SAMPLESETNAME/WGSconfig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! terra-sync export broad-firecloud-ccle/DepMap_hg38_RNAseq data/$constants.SAMPLESETNAME/RNAconfig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cd data/$constants.SAMPLESETNAME/WGSconfig && mv */*/* . && rm -r configs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cd ../RNAconfig && mv */*/* . && rm -r configs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "notify_time": "5",
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "277px",
    "width": "375px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "386.4px",
    "left": "762px",
    "top": "202px",
    "width": "198.8px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 577.8,
   "position": {
    "height": "40px",
    "left": "1285.4px",
    "right": "20px",
    "top": "106px",
    "width": "254.8px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "none",
   "window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "1f170dacc8a54decc88aa6dec68bcf2a5d65c3893a5412f67865d27128dae76b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
