
## RNA_pipeline

### Inputs

#### Required

  * `ctat_genome_lib_build_dir_files` (Array[File], **required**)
  * `genes_gtf` (File, **required**)
  * `input_bam_cram` (File, **required**)
  * `ref_genome_fa_star_idx_files` (Array[File], **required**)
  * `reference_fasta` (File, **required**)
  * `rsem_reference` (File, **required**)
  * `sample_id` (String, **required**)
  * `star_index` (File, **required**)

#### Optional

  * `rnaseqc2.flags` (String?)
  * `rnaseqc2.intervals_bed` (File?)
  * `rnaseqc2.strandedness` (String?)
  * `rsem.calc_ci` (String?)
  * `rsem.ci_memory` (Int?)
  * `star.sjdbFileChrStartEnd` (File?)
  * `star.varVCFfile` (File?)
  * `star.waspOutputMode` (String?)

#### Defaults

  * `StarFusion.disk_space` (Int, default=500)
  * `StarFusion.docker` (String, default="trinityctat/starfusion:1.7.0")
  * `StarFusion.memory` (Int, default=64)
  * `StarFusion.num_preempt` (Int, default=2)
  * `StarFusion.num_threads` (Int, default=8)
  * `rnaseqc2.disk_space` (Int, default=75)
  * `rnaseqc2.memory` (Int, default=8)
  * `rnaseqc2.num_preempt` (Int, default=1)
  * `rnaseqc2.num_threads` (Int, default=4)
  * `rsem.disk_space` (Int, default=500)
  * `rsem.estimate_rspd` (String, default="true")
  * `rsem.max_frag_len` (Int, default=1000)
  * `rsem.memory` (Int, default=128)
  * `rsem.num_preempt` (Int, default=1)
  * `rsem.num_threads` (Int, default=32)
  * `samtofastq.disk_space` (Int, default=400)
  * `samtofastq.java_memory` (Int, default=floor((memory - 0.5)))
  * `samtofastq.memory` (Float, default=64)
  * `samtofastq.num_preempt` (Int, default=5)
  * `samtofastq.num_threads` (Int, default=8)
  * `star.alignIntronMax` (Int, default=1000000)
  * `star.alignIntronMin` (Int, default=20)
  * `star.alignMatesGapMax` (Int, default=1000000)
  * `star.alignSJDBoverhangMin` (Int, default=1)
  * `star.alignSJoverhangMin` (Int, default=8)
  * `star.alignSoftClipAtReferenceEnds` (String, default="Yes")
  * `star.chimJunctionOverhangMin` (Int, default=15)
  * `star.chimMainSegmentMultNmax` (Int, default=1)
  * `star.chimOutJunctionFormat` (Int, default=1)
  * `star.chimOutType` (String, default="Junctions WithinBAM SoftClip")
  * `star.chimSegmentMin` (Int, default=15)
  * `star.disk_space` (Int, default=500)
  * `star.limitSjdbInsertNsj` (Int, default=1200000)
  * `star.memory` (Int, default=64)
  * `star.num_preempt` (Int, default=1)
  * `star.num_threads` (Int, default=16)
  * `star.outFilterIntronMotifs` (String, default="None")
  * `star.outFilterMatchNminOverLread` (Float, default=0.33)
  * `star.outFilterMismatchNmax` (Int, default=999)
  * `star.outFilterMismatchNoverLmax` (Float, default=0.1)
  * `star.outFilterMultimapNmax` (Int, default=20)
  * `star.outFilterScoreMinOverLread` (Float, default=0.33)
  * `star.outFilterType` (String, default="BySJout")
  * `star.outSAMattrRGline` (String, default="ID:rg1 SM:sm1")
  * `star.outSAMattributes` (String, default="NH HI AS nM NM ch")
  * `star.outSAMstrandField` (String, default="intronMotif")
  * `star.quantMode` (String, default="TranscriptomeSAM GeneCounts")

### Outputs

  * `bam_file` (File)
  * `bam_index` (File)
  * `transcriptome_bam` (File)
  * `chimeric_junctions` (File)
  * `chimeric_bam_file` (File)
  * `read_counts` (File)
  * `junctions` (File)
  * `junctions_pass1` (File)
  * `logs` (Array[File])
  * `gene_tpm` (File)
  * `gene_counts` (File)
  * `exon_counts` (File)
  * `metrics` (File)
  * `insertsize_distr` (File)
  * `genes` (File)
  * `isoforms` (File)
  * `fusion_predictions` (File)
  * `fusion_predictions_abridged` (File)
