{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you need to have installed JKBio in the same folder as ccle_processing\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n",
       "  var css_urls = [];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.3.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.3.4.min.js\"];\n  var css_urls = [];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import os.path\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "sys.path.insert(0, '../..')\n",
    "\n",
    "from src.CCLE_postp_function import *\n",
    "from JKBio import Datanalytics as da \n",
    "from JKBio import TerraFunction as terra\n",
    "from JKBio import Helper as h\n",
    "from gsheets import Sheets\n",
    "from taigapy import TaigaClient\n",
    "import dalmatian as dm\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "from bokeh.plotting import *\n",
    "from bokeh.models import HoverTool\n",
    "from collections import OrderedDict\n",
    "from IPython.display import Image,display\n",
    "\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "tc = TaigaClient()\n",
    "output_notebook()\n",
    "sheets = Sheets.from_files('~/.client_secret.json', '~/.storage.json')\n",
    "replace = {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesetname = \"20Q2\"\n",
    "prevname=\"20Q1\"\n",
    "prevprevname ='19Q4'\n",
    "prevprevversion= 18\n",
    "virtual_public='public-20q2-075d'\n",
    "virtual_dmc='dmc-20q2-2db6'\n",
    "virtual_internal='internal-20q2-7f46'\n",
    "\n",
    "workspace1=\"broad-genomics-delivery/Getz_IBM_CellLines_Exomes\"\n",
    "workspace2=\"broad-firecloud-ccle/CCLE_DepMap_WES\"\n",
    "workspace3=\"broad-genomics-delivery/CCLE_DepMap_WES\"\n",
    "\n",
    "workspace6=\"terra-broad-cancer-prod/CCLE_DepMap_WES\"\n",
    "\n",
    "refworkspace=\"broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline\"\n",
    "\n",
    "rnaworkspace=\"broad-firecloud-ccle/DepMap_hg38_RNAseq\"\n",
    "\n",
    "source1=\"ibm\"\n",
    "source2=\"ccle\"\n",
    "source3=\"ccle\"\n",
    "source6=\"ccle\"\n",
    "source7=\"ibm\"\n",
    "\n",
    "refsheet_url = \"https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE\"\n",
    "sheeturl = \"https://docs.google.com/spreadsheets/d/115TUgA1t_mD32SnWAGpW9OKmJ2W5WYAOs3SuSdedpX4\"\n",
    "\n",
    "release = samplesetname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wm1 = dm.WorkspaceManager(workspace1)\n",
    "wm2 = dm.WorkspaceManager(workspace2)\n",
    "wm3 = dm.WorkspaceManager(workspace3)\n",
    "\n",
    "wm6 = dm.WorkspaceManager(workspace6)\n",
    "\n",
    "refwm = dm.WorkspaceManager(refworkspace)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_refsamples = sheets.get(refsheet_url).sheets[0].to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! wget https://github.com/mskcc/vcf2maf/archive/v1.6.18.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3249: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'AC': '\"Allele count in genotypes, for each ALT allele, in the same order as listed\"', 'AF': '\"Allele Frequency, for each ALT allele, in the same order as listed\"', 'AN': '\"Total number of alleles in called genotypes\"', 'BaseQRankSum': '\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\"', 'DB': '\"dbSNP Membership\"', 'ExcessHet': '\"Phred-scaled p-value for exact test of excess heterozygosity\"', 'FS': '\"Phred-scaled p-value using Fisher\\'s exact test to detect strand bias\"', 'InbreedingCoeff': '\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\"', 'MLEAC': '\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\"', 'MLEAF': '\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\"', 'MQ': '\"RMS Mapping Quality\"', 'MQRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\"', 'QD': '\"Variant Confidence/Quality by Depth\"', 'ReadPosRankSum': '\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\"', 'SOR': '\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\"'}\n"
     ]
    }
   ],
   "source": [
    "# for rnaseq\n",
    "#! gsutil cp gs://fc-secure-639c94ba-2b0d-4960-92fc-9cd50046a968/8af15313-7faf-4344-a7b7-97dd52353dc8/RNAseq/0de1da92-8821-4cbd-a9ce-96ed8b06f395/call-MergeVCFs/G41677.MV-4-11.5.g.vcf.gz ../temp/\n",
    "path= \"../temp/G41677.MV-4-11.5.g.vcf.gz\"\n",
    "rnaseq, description = h.vcf_to_df(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'AC': '\"Allele count in genotypes, for each ALT allele, in the same order as listed\"', 'AF': '\"Allele Frequency, for each ALT allele, in the same order as listed\"', 'AN': '\"Total number of alleles in called genotypes\"', 'BaseQRankSum': '\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\"', 'DB': '\"dbSNP Membership\"', 'ExcessHet': '\"Phred-scaled p-value for exact test of excess heterozygosity\"', 'FS': '\"Phred-scaled p-value using Fisher\\'s exact test to detect strand bias\"', 'InbreedingCoeff': '\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\"', 'MLEAC': '\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\"', 'MLEAF': '\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\"', 'MQ': '\"RMS Mapping Quality\"', 'MQRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\"', 'QD': '\"Variant Confidence/Quality by Depth\"', 'ReadPosRankSum': '\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\"', 'SOR': '\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\"'}\n"
     ]
    }
   ],
   "source": [
    "# for rnaseq filtered\n",
    "#!gsutil cp gs://fc-secure-639c94ba-2b0d-4960-92fc-9cd50046a968/8af15313-7faf-4344-a7b7-97dd52353dc8/RNAseq/0de1da92-8821-4cbd-a9ce-96ed8b06f395/call-VariantFiltration/G41677.MV-4-11.5.variant_filtered.vcf.gz ../temp/\n",
    "rnaseqfilt, description = h.vcf_to_df(\"../temp/G41677.MV-4-11.5.variant_filtered.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'MIN_DP': '\"Minimum DP observed within the GVCF block\"', 'PGT': '\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another\"', 'PID': '\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'SB': '\"Per-sample component statistics which comprise the Fisher\\'s Exact Test to detect strand bias.\"', 'BaseQRankSum': '\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\"', 'ClippingRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref number of hard clipped bases\"', 'DS': '\"Were any of the samples downsampled?\"', 'END': '\"Stop position of the interval\"', 'ExcessHet': '\"Phred-scaled p-value for exact test of excess heterozygosity\"', 'InbreedingCoeff': '\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\"', 'MLEAC': '\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\"', 'MLEAF': '\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\"', 'MQ': '\"RMS Mapping Quality\"', 'MQRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\"', 'RAW_MQ': '\"Raw data for RMS Mapping Quality\"', 'ReadPosRankSum': '\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\"'}\n"
     ]
    }
   ],
   "source": [
    "# for unfiltered wes\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/51fcb10c-9a67-4ff9-b13c-d9667e2d49d0/HaplotypeCallerGvcf_GATK4/6611b28a-12c2-484e-a4a3-901708785f95/call-MergeGVCFs/SANGER_MV411_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE.g.vcf.gz ../temp/\n",
    "wes, description = h.vcf_to_df(\"../temp/SANGER_MV411_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE.g.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'AC': '\"Allele count in genotypes, for each ALT allele, in the same order as listed\"', 'AF': '\"Allele Frequency, for each ALT allele, in the same order as listed\"', 'AN': '\"Total number of alleles in called genotypes\"', 'BaseQRankSum': '\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\"', 'CNN_1D': '\"Log odds of being a true variant versus being false under the trained 1D Convolutional Neural Network\"', 'DS': '\"Were any of the samples downsampled?\"', 'ExcessHet': '\"Phred-scaled p-value for exact test of excess heterozygosity\"', 'FS': '\"Phred-scaled p-value using Fisher\\'s exact test to detect strand bias\"', 'InbreedingCoeff': '\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\"', 'MLEAC': '\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\"', 'MLEAF': '\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\"', 'MQ': '\"RMS Mapping Quality\"', 'MQRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\"', 'QD': '\"Variant Confidence/Quality by Depth\"', 'ReadPosRankSum': '\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\"', 'SOR': '\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\"'}\n"
     ]
    }
   ],
   "source": [
    "# for filtered wes\n",
    "#! gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/d953cc3f-7941-4426-bb44-a41992d57606/Cram2FilteredVcf/dc943e9d-aa73-4314-b43d-208a80319997/call-FilterVariantTranches/CDS-FInR9b_cnn_filtered.vcf.gz ../temp/\n",
    "wesfilt, description = h.vcf_to_df(\"../temp/CDS-FInR9b_cnn_filtered.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (19,29,30) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# for somatic filtered wes\n",
    "somwesfilt = pd.read_csv('../temp/mutations.'+release+'.all.tsv', sep='\\t')\n",
    "somwesfilt = somwesfilt[somwesfilt.DepMap_ID==\"ACH-000045\"]\n",
    "somwesfilt = somwesfilt.rename(columns=h.rename_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for somatic unfiltered wes\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/9e3cc501-3f08-47fb-87a5-0359febb833c/CGA_Production_Analysis_Workflow/72efd16d-fa2d-4d78-b7c7-81ec2a6d08e9/call-Oncotate_Task/CDS-FInR9b.MuTect1_SNV.MuTect2_INDEL.Strelka_INDEL.annotated.maf ../temp/\n",
    "somwes = pd.read_csv(\"../temp/CDS-FInR9b.MuTect1_SNV.MuTect2_INDEL.Strelka_INDEL.annotated.maf\", sep='\\t')\n",
    "somwes = somwes.rename(columns=h.rename_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Hugo_Symbol',\n",
       " 'Entrez_Gene_Id',\n",
       " 'Center',\n",
       " 'NCBI_Build',\n",
       " 'chr',\n",
       " 'pos',\n",
       " 'End_position',\n",
       " 'Strand',\n",
       " 'Variant_Classification',\n",
       " 'Variant_Type',\n",
       " 'ref',\n",
       " 'alt',\n",
       " 'Tumor_Seq_Allele2',\n",
       " 'dbSNP_RS',\n",
       " 'dbSNP_Val_Status',\n",
       " 'Tumor_Sample_Barcode',\n",
       " 'Matched_Norm_Sample_Barcode',\n",
       " 'Match_Norm_Seq_Allele1',\n",
       " 'Match_Norm_Seq_Allele2',\n",
       " 'Tumor_Validation_Allele1',\n",
       " 'Tumor_Validation_Allele2',\n",
       " 'Match_Norm_Validation_Allele1',\n",
       " 'Match_Norm_Validation_Allele2',\n",
       " 'Verification_Status',\n",
       " 'Validation_Status',\n",
       " 'Mutation_Status',\n",
       " 'Sequencing_Phase',\n",
       " 'Sequence_Source',\n",
       " 'Validation_Method',\n",
       " 'Score',\n",
       " 'BAM_file',\n",
       " 'Sequencer',\n",
       " 'dbNSFP_aapos_SIFT',\n",
       " 'n_ref_count',\n",
       " 'HGNC_Ensembl Gene ID',\n",
       " 'dbNSFP_phyloP46way_placental_rankscore',\n",
       " 'HGVS_protein_change',\n",
       " 'dbNSFP_MutationTaster_score',\n",
       " 'TAR',\n",
       " 'ESP_AA_GTC',\n",
       " 'dbNSFP_Ancestral_allele',\n",
       " 'ESP_EA_AGE',\n",
       " '1000gp3_SVTYPE',\n",
       " 'gencode_transcript_name',\n",
       " 'UniProt_Site',\n",
       " 'ClinVar_SYM',\n",
       " 'ESP_Chromosome',\n",
       " '1000gp3_SVLEN',\n",
       " 'dbNSFP_phastCons46way_primate',\n",
       " 'Codon_Change',\n",
       " 'HGNC_HGNC ID',\n",
       " 'HGNC_OMIM ID(supplied by NCBI)',\n",
       " 'alt_allele_seen',\n",
       " 'dbNSFP_phastCons100way_vertebrate',\n",
       " 't_q20_count',\n",
       " 'dbNSFP_cds_strand',\n",
       " 'HGNC_Previous Symbols',\n",
       " 't_ins_count',\n",
       " 'dbNSFP_MutationAssessor_pred',\n",
       " 't_lod_fstar',\n",
       " 'allelic_depth',\n",
       " 'ESP_AvgAAsampleReadDepth',\n",
       " 'dbNSFP_1000Gp1_AFR_AC',\n",
       " 'dbNSFP_1000Gp1_AFR_AF',\n",
       " 'dbNSFP_phastCons46way_placental_rankscore',\n",
       " '1000gp3_CS',\n",
       " 'ClinVar_TYPE',\n",
       " 'dbNSFP_LRT_converted_rankscore',\n",
       " 'dbNSFP_aaalt',\n",
       " 'dbNSFP_GERP++_RS_rankscore',\n",
       " 'dbNSFP_1000Gp1_ASN_AC',\n",
       " 'dbNSFP_SIFT_converted_rankscore',\n",
       " 'dbNSFP_1000Gp1_ASN_AF',\n",
       " 'TQSI_NT',\n",
       " 'normal_f',\n",
       " 'HGNC_Date Symbol Changed',\n",
       " 'FDP50',\n",
       " 'HGNC_Record Type',\n",
       " 'ESP_MAF',\n",
       " 'dbNSFP_SiPhy_29way_pi',\n",
       " 'HGNC_Accession Numbers',\n",
       " 'ESP_DP',\n",
       " 'dbNSFP_Uniprot_aapos',\n",
       " 't_lod_fstar_forward',\n",
       " '1000gp3_EUR_AF',\n",
       " 'CGC_Tissue Type',\n",
       " 'HGNC_UCSC ID(supplied by UCSC)',\n",
       " 'normal_best_gt',\n",
       " 'dbNSFP_Interpro_domain',\n",
       " 'TQSI',\n",
       " '1000gp3_AFR_AF',\n",
       " 'gnomADg',\n",
       " 'dbNSFP_aapos_FATHMM',\n",
       " 'CGC_Name',\n",
       " 'TOR',\n",
       " 'ESP_EA_GTC',\n",
       " 'dbNSFP_aapos',\n",
       " 'CCLE_ONCOMAP_overlapping_mutations',\n",
       " 'TCGAscape_Amplification_Peaks',\n",
       " 'dbNSFP_SLR_test_statistic',\n",
       " 'dbNSFP_LR_rankscore',\n",
       " 'Matched_Norm_Sample_UUID',\n",
       " 'dbNSFP_Uniprot_id',\n",
       " 'dbNSFP_FATHMM_rankscore',\n",
       " 'dbNSFP_SIFT_pred',\n",
       " 'TCGAscape_Deletion_Peaks',\n",
       " 'dbNSFP_LRT_pred',\n",
       " '1000gp3_NS',\n",
       " 'tumor_alt_rpir_median',\n",
       " 'ESP_CP',\n",
       " '1000gp3_TSD',\n",
       " 'DP2',\n",
       " 'ESP_Position',\n",
       " 'gnomADg_AF_EAS',\n",
       " 'Transcript_Position',\n",
       " 'dbNSFP_phyloP100way_vertebrate_rankscore',\n",
       " 'gnomADg_AF_OTH',\n",
       " '1000gp3_IMPRECISE',\n",
       " 'strand_bias_counts',\n",
       " 'RU',\n",
       " 'dbNSFP_SiPhy_29way_logOdds',\n",
       " 'RC',\n",
       " 'covered',\n",
       " 'observed_in_normals_count',\n",
       " 'dbNSFP_phyloP100way_vertebrate',\n",
       " 'dbNSFP_ESP6500_AA_AF',\n",
       " 'Transcript_Exon',\n",
       " 'contaminant_lod',\n",
       " 'normal_power_wsp',\n",
       " 'UniProt_alt_uniprot_accessions',\n",
       " 'dbNSFP_SIFT_score',\n",
       " 'ESP_TotalAAsamplesCovered',\n",
       " 'n_alt_count',\n",
       " 'dbNSFP_phyloP46way_primate_rankscore',\n",
       " 'dbNSFP_Polyphen2_HVAR_pred',\n",
       " 'iHpol',\n",
       " 'failure_reasons',\n",
       " '1000gp3_DP',\n",
       " 'dbNSFP_genename',\n",
       " 'HGNC_Locus Group',\n",
       " 'DNARepairGenes_Role',\n",
       " 'dbNSFP_phastCons100way_vertebrate_rankscore',\n",
       " 'map_Q0_reads',\n",
       " 'HGNC_Chromosome',\n",
       " 'CGC_Cancer Somatic Mut',\n",
       " 't_alt_max_mapq',\n",
       " 'Tumor_Sample_UUID',\n",
       " 'UniProt_Natural_Variations',\n",
       " 'CSQ',\n",
       " 'Other_Variant_Caller',\n",
       " 'ESP_CA',\n",
       " 'CGC_Cancer Molecular Genetics',\n",
       " 'ESP_CG',\n",
       " 'HGNC_Enzyme IDs',\n",
       " 'n_q20_count',\n",
       " 'ESP_GTC',\n",
       " 'HGNC_UniProt ID(supplied by UniProt)',\n",
       " 'ESP_GTS',\n",
       " 'Repeat',\n",
       " 'genotype',\n",
       " 'gnomADg_AN',\n",
       " 'CGC_Cancer Syndrome',\n",
       " 'QSI_NT',\n",
       " 'ESP_TotalSamplesCovered',\n",
       " 'gnomADg_AF',\n",
       " 'gnomADg_AC',\n",
       " 'HGNC_Synonyms',\n",
       " 'contaminant_fraction',\n",
       " 'dbNSFP_Polyphen2_HDIV_score',\n",
       " 'qual',\n",
       " 'tumor_alt_rpir_mad',\n",
       " 'dbNSFP_Uniprot_acc',\n",
       " 'Transcript_Strand',\n",
       " 'ClinVar_HGMD_ID',\n",
       " 'SVTYPE',\n",
       " 'dbNSFP_LR_pred',\n",
       " 'CGC_Other Germline Mut',\n",
       " 'COSMIC_tissue_types_affected',\n",
       " 'ClinVar_rs',\n",
       " 'build',\n",
       " 'SwissProt_entry_Id',\n",
       " '1000gp3_MLEN',\n",
       " 'cDNA_Change',\n",
       " 'tumor_f',\n",
       " 'UniProt_Experimental_Info',\n",
       " 'Annotation_Transcript',\n",
       " 'dbNSFP_RadialSVM_pred',\n",
       " '1000gp3_EAS_AF',\n",
       " 'power_to_detect_negative_strand_artifact',\n",
       " 'HGNC_Date Modified',\n",
       " 'total_reads',\n",
       " 't_ref_max_mapq',\n",
       " 'dbNSFP_FATHMM_score',\n",
       " 'QSI',\n",
       " 'ESP_GS',\n",
       " 'ESP_GL',\n",
       " 'ESP_GM',\n",
       " 'annotation_transcript',\n",
       " 'dbNSFP_MutationTaster_converted_rankscore',\n",
       " 'dbNSFP_phastCons46way_placental',\n",
       " 'QSI_ref',\n",
       " 'dbNSFP_Polyphen2_HDIV_pred',\n",
       " 'dbNSFP_refcodon',\n",
       " 'HGNC_VEGA IDs',\n",
       " 'HGNC_Locus Type',\n",
       " 'dbNSFP_fold-degenerate',\n",
       " 'COSMIC_overlapping_mutation_descriptions',\n",
       " 'BCNoise',\n",
       " 'ESP_AvgEAsampleReadDepth',\n",
       " 'secondary_variant_classification',\n",
       " 'dbNSFP_Polyphen2_HVAR_score',\n",
       " 'dbNSFP_Polyphen2_HVAR_rankscore',\n",
       " 'dbNSFP_1000Gp1_AMR_AF',\n",
       " 'IHP',\n",
       " 'dbNSFP_1000Gp1_AMR_AC',\n",
       " 'ESP_PP',\n",
       " 'Ensembl_so_accession',\n",
       " 'init_n_lod',\n",
       " 'ESP_PH',\n",
       " '1000gp3_AN',\n",
       " '1000gp3_AA',\n",
       " '1000gp3_AC',\n",
       " 'gene_type',\n",
       " '1000gp3_AF',\n",
       " 'dbNSFP_RadialSVM_rankscore',\n",
       " 'ORegAnno_bin',\n",
       " 'COSMIC_fusion_genes',\n",
       " 'UniProt_Region',\n",
       " 'gencode_transcript_status',\n",
       " 'dbNSFP_Ensembl_transcriptid',\n",
       " 'CGC_Tumor_Types_Germline',\n",
       " 'Refseq_mRNA_Id',\n",
       " 'Familial_Cancer_Genes_Synonym',\n",
       " 'CGC_Translocation_Partner',\n",
       " 'Familial_Cancer_Genes_Reference',\n",
       " 'dbNSFP_phyloP46way_primate',\n",
       " 't_alt_sum',\n",
       " 'score',\n",
       " 'SUBDP50',\n",
       " 'dbNSFP_CADD_raw_rankscore',\n",
       " 'ESP_EXOME_CHIP',\n",
       " 'ESP_TAC',\n",
       " 't_lod_fstar_reverse',\n",
       " 'Genome_Change',\n",
       " 'gencode_transcript_type',\n",
       " 'context',\n",
       " 'read_depth',\n",
       " 'ESP_EA_AC',\n",
       " 'CGC_Tumor_Types_Somatic',\n",
       " 'dbNSFP_Polyphen2_HDIV_rankscore',\n",
       " 'ClinVar_ASSEMBLY',\n",
       " 't_ref_sum',\n",
       " 'dbNSFP_hg18_pos(1-coor)',\n",
       " 'HGVS_genomic_change',\n",
       " '1000gp3_MSTART',\n",
       " 'GO_Cellular_Component',\n",
       " 'OVERLAP',\n",
       " 'ESP_AA_AC',\n",
       " 'Tumorscape_Deletion_Peaks',\n",
       " 'CGC_Chr Band',\n",
       " 'dbNSFP_LR_score',\n",
       " '1000gp3_CIPOS',\n",
       " 'GO_Molecular_Function',\n",
       " 'n_ref_sum',\n",
       " 'gc_content',\n",
       " '1000gp3_AMR_AF',\n",
       " 'dbNSFP_LRT_Omega',\n",
       " 'dbNSFP_phyloP46way_placental',\n",
       " 'somatic',\n",
       " 'id',\n",
       " 'gnomADg_AF_FIN',\n",
       " 'Other_Transcripts',\n",
       " 'dbNSFP_LRT_score',\n",
       " 't_alt_count',\n",
       " 'ESP_FG',\n",
       " 'gnomADg_GT',\n",
       " 'dbNSFP_RadialSVM_score',\n",
       " 'refseq_mrna_id',\n",
       " 'dbNSFP_Reliability_index',\n",
       " 'Ensembl_so_term',\n",
       " 'HGNC_Status',\n",
       " 'power_to_detect_positive_strand_artifact',\n",
       " 'dbNSFP_ESP6500_EA_AF',\n",
       " 'havana_transcript',\n",
       " 'IC',\n",
       " 'CGC_Other_Diseases',\n",
       " 'normal_power',\n",
       " 'Refseq_prot_Id',\n",
       " 'dbsnp_site',\n",
       " 'COSMIC_overlapping_mutations',\n",
       " 'gnomADg_AF_AMR',\n",
       " 'HGNC_Ensembl ID(supplied by Ensembl)',\n",
       " 'TIR',\n",
       " 'CGC_Mutation_Type',\n",
       " '1000gp3_CIEND',\n",
       " '1000gp3_MEINFO',\n",
       " 'ESP_GWAS_PUBMED',\n",
       " 'OREGANNO_ID',\n",
       " 'HGNC_Secondary IDs',\n",
       " 'UniProt_AApos',\n",
       " 'dbNSFP_codonpos',\n",
       " 'HGNC_Pubmed IDs',\n",
       " 'judgement',\n",
       " 'dbNSFP_phastCons46way_primate_rankscore',\n",
       " 'HGNC_Date Name Changed',\n",
       " 'ESP_TotalEAsamplesCovered',\n",
       " 'SwissProt_acc_Id',\n",
       " 'ESP_AA',\n",
       " 'COSMIC_total_alterations_in_gene',\n",
       " 'COSMIC_overlapping_primary_sites',\n",
       " 'dbNSFP_MutationTaster_pred',\n",
       " 'HGNC_Gene family description',\n",
       " 'HGNC_Previous Names',\n",
       " 'HGNC_CCDS IDs',\n",
       " 'CGC_Cancer Germline Mut',\n",
       " 'normal_power_nsp',\n",
       " 'ACHILLES_Lineage_Results_Top_Genes',\n",
       " 'FamilialCancerDatabase_Syndromes',\n",
       " 'ref_context',\n",
       " 'CCLE_ONCOMAP_total_mutations_in_gene',\n",
       " 'HGVS_coding_DNA_change',\n",
       " 'dbNSFP_Ensembl_geneid',\n",
       " 'gnomADg_AF_NFE',\n",
       " 'HGNC_Name Synonyms',\n",
       " 'init_t_lod',\n",
       " 'power',\n",
       " 'ESP_AA_AGE',\n",
       " 'tumor_alt_fpir_median',\n",
       " '1000gp3_END',\n",
       " 'dbNSFP_GERP++_NR',\n",
       " 'MUTSIG_Published_Results',\n",
       " 'CGC_GeneID',\n",
       " 'ESP_AvgSampleReadDepth',\n",
       " '1000gp3_SAS_AF',\n",
       " 'dbNSFP_MutationAssessor_score',\n",
       " 'tumor_power',\n",
       " 'dbNSFP_GERP++_RS',\n",
       " 'dbNSFP_CADD_raw',\n",
       " 'tumor_alt_fpir_mad',\n",
       " '1000gp3_MC',\n",
       " 'dbNSFP_MutationAssessor_rankscore',\n",
       " 'DrugBank',\n",
       " 'dbNSFP_CADD_phred',\n",
       " 'dbNSFP_SiPhy_29way_logOdds_rankscore',\n",
       " 'NT',\n",
       " 'ESP_AAC',\n",
       " 'gencode_transcript_tags',\n",
       " 't_del_count',\n",
       " 'entrez_gene_id',\n",
       " 'gnomADg_AF_AFR',\n",
       " 'ESP_DBSNP',\n",
       " 'ESP_CDP',\n",
       " 'n_alt_sum',\n",
       " 'GO_Biological_Process',\n",
       " 'COSMIC_n_overlapping_mutations',\n",
       " 'dbNSFP_UniSNP_ids',\n",
       " 't_ref_count',\n",
       " 'dbNSFP_1000Gp1_EUR_AC',\n",
       " 'Tumorscape_Amplification_Peaks',\n",
       " 'dbNSFP_1000Gp1_AC',\n",
       " 'dbNSFP_1000Gp1_EUR_AF',\n",
       " 'dbNSFP_1000Gp1_AF',\n",
       " 'CGC_Chr',\n",
       " '1000gp3_MEND',\n",
       " 'Description',\n",
       " 'SGT',\n",
       " 'Protein_Change',\n",
       " 'ccds_id',\n",
       " 'HGNC_RefSeq(supplied by NCBI)',\n",
       " 'dbNSFP_FATHMM_pred',\n",
       " 'DP50',\n",
       " 'OREGANNO_Values',\n",
       " 'matched_norm_sample_barcode',\n",
       " 'gnomADg_AF_ASJ',\n",
       " 'dbNSFP_aaref',\n",
       " 'HGNC_Primary IDs']"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "somwes.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'QSS': '\"Quality score for any somatic snv, ie. for the ALT allele to be present at a significantly different frequency in the tumor and normal\"', 'TQSS': '\"Data tier used to compute QSS\"', 'NT': '\"Genotype of the normal in all data tiers, as used to classify somatic variants. One of {ref,het,hom,conflict}.\"', 'QSS_NT': '\"Quality score reflecting the joint probability of a somatic variant and NT\"', 'TQSS_NT': '\"Data tier used to compute QSS_NT\"', 'SGT': '\"Most likely somatic genotype excluding normal noise states\"', 'SOMATIC': '\"Somatic mutation\"', 'DP': '\"Read depth for tier1 (used+filtered)\"', 'FDP': '\"Number of basecalls filtered from original read depth for tier1\"', 'SDP': '\"Number of reads with deletions spanning this site at tier1\"', 'SUBDP': '\"Number of reads below tier1 mapping quality threshold aligned across this site\"', 'AU': '\"Number of \\'A\\' alleles used in tiers 1,2\"', 'CU': '\"Number of \\'C\\' alleles used in tiers 1,2\"', 'GU': '\"Number of \\'G\\' alleles used in tiers 1,2\"', 'TU': '\"Number of \\'T\\' alleles used in tiers 1,2\"'}\n",
      "{'QSI': '\"Quality score for any somatic variant, ie. for the ALT haplotype to be present at a significantly different frequency in the tumor and normal\"', 'TQSI': '\"Data tier used to compute QSI\"', 'NT': '\"Genotype of the normal in all data tiers, as used to classify somatic variants. One of {ref,het,hom,conflict}.\"', 'QSI_NT': '\"Quality score reflecting the joint probability of a somatic variant and NT\"', 'TQSI_NT': '\"Data tier used to compute QSI_NT\"', 'SGT': '\"Most likely somatic genotype excluding normal noise states\"', 'RU': '\"Smallest repeating sequence unit in inserted or deleted sequence\"', 'RC': '\"Number of times RU repeats in the reference allele\"', 'IC': '\"Number of times RU repeats in the indel allele\"', 'IHP': '\"Largest reference interupted homopolymer length intersecting with the indel\"', 'SVTYPE': '\"Type of structural variant\"', 'SOMATIC': '\"Somatic mutation\"', 'OVERLAP': '\"Somatic indel possibly overlaps a second indel.\"', 'DP': '\"Read depth for tier1\"', 'DP2': '\"Read depth for tier2\"', 'TAR': '\"Reads strongly supporting alternate allele for tiers 1,2\"', 'TIR': '\"Reads strongly supporting indel allele for tiers 1,2\"', 'TOR': '\"Other reads (weak support or insufficient indel breakpoint overlap) for tiers 1,2\"', 'DP50': '\"Average tier1 read depth within 50 bases\"', 'FDP50': '\"Average tier1 number of basecalls filtered from original read depth within 50 bases\"', 'SUBDP50': '\"Average number of reads below tier1 mapping quality threshold aligned across sites within 50 bases\"'}\n"
     ]
    }
   ],
   "source": [
    "# for strelka wes\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/9e3cc501-3f08-47fb-87a5-0359febb833c/CGA_Production_Analysis_Workflow/72efd16d-fa2d-4d78-b7c7-81ec2a6d08e9/call-Strelka/CDS-FInR9b.all.somatic.snvs.vcf ../temp/\n",
    "strelka, description = h.vcf_to_df(\"../temp/CDS-FInR9b.all.somatic.snvs.vcf\", hasfilter=True,samples=['sample1_A','sample1_B'])\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/9e3cc501-3f08-47fb-87a5-0359febb833c/CGA_Production_Analysis_Workflow/72efd16d-fa2d-4d78-b7c7-81ec2a6d08e9/call-Strelka/CDS-FInR9b.all.somatic.indels.vcf ../temp/\n",
    "strelka2, description = h.vcf_to_df(\"../temp/CDS-FInR9b.all.somatic.indels.vcf\",samples=['sample1_A','sample1_B'],hasfilter=True)\n",
    "strelka = pd.concat([strelka,strelka2], ignore_index=True,sort=False).sort_values(by=['chr','pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jeremie/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# for mutect1 wes\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/9e3cc501-3f08-47fb-87a5-0359febb833c/CGA_Production_Analysis_Workflow/72efd16d-fa2d-4d78-b7c7-81ec2a6d08e9/call-Gather_Task/CDS-FInR9b.MuTect1.call_stats.txt ../temp/\n",
    "mutect1 = pd.read_csv('../temp/CDS-FInR9b.MuTect1.call_stats.txt', sep='\\t', header=1, index_col=None)\n",
    "mutect1 = mutect1.rename(columns=h.rename_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'AF': '\"Allele fractions of alternate alleles in the tumor\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'F1R2': '\"Count of reads in F1R2 pair orientation supporting each allele\"', 'F2R1': '\"Count of reads in F2R1 pair orientation supporting each allele\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'MBQ': '\"median base quality\"', 'MFRL': '\"median fragment length\"', 'MMQ': '\"median mapping quality\"', 'MPOS': '\"median distance from end of read\"', 'PGT': '\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another\"', 'PID': '\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'SA_MAP_AF': '\"MAP estimates of allele fraction given z\"', 'SA_POST_PROB': '\"posterior probabilities of the presence of strand artifact\"', 'ECNT': '\"Number of events in this haplotype\"', 'IN_PON': '\"site found in panel of normals\"', 'NLOD': '\"Normal LOD score\"', 'N_ART_LOD': '\"log odds of artifact in normal with same allele fraction as tumor\"', 'POP_AF': '\"population allele frequencies of alt alleles\"', 'P_CONTAM': '\"Posterior probability for alt allele to be due to contamination\"', 'P_GERMLINE': '\"Posterior probability for alt allele to be germline variants\"', 'RPA': '\"Number of times tandem repeat unit is repeated, for each allele (including reference)\"', 'RU': '\"Tandem repeat unit (bases)\"', 'STR': '\"Variant is a short tandem repeat\"', 'TLOD': '\"Tumor LOD score\"'}\n"
     ]
    }
   ],
   "source": [
    "# for mutect 2 wes\n",
    "#!gsutil cp gs://fc-secure-012d088c-f039-4d36-bde5-ee9b1b76b912/9e3cc501-3f08-47fb-87a5-0359febb833c/CGA_Production_Analysis_Workflow/72efd16d-fa2d-4d78-b7c7-81ec2a6d08e9/call-Gather_Task/CDS-FInR9b.MuTect2.call_stats.vcf ../temp/\n",
    "mutect2, description = h.vcf_to_df('../temp/CDS-FInR9b.MuTect2.call_stats.vcf', hasfilter=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AD': '\"Allelic depths for the ref and alt alleles in the order listed\"', 'DP': '\"Approximate read depth; some reads may have been filtered\"', 'GQ': '\"Genotype Quality\"', 'GT': '\"Genotype\"', 'PL': '\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\"', 'AC': '\"Allele count in genotypes, for each ALT allele, in the same order as listed\"', 'AF': '\"Allele Frequency, for each ALT allele, in the same order as listed\"', 'AN': '\"Total number of alleles in called genotypes\"', 'BaseQRankSum': '\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\"', 'CNN_1D': '\"Log odds of being a true variant versus being false under the trained 1D Convolutional Neural Network\"', 'DS': '\"Were any of the samples downsampled?\"', 'END': '\"Stop position of the interval\"', 'ExcessHet': '\"Phred-scaled p-value for exact test of excess heterozygosity\"', 'FS': '\"Phred-scaled p-value using Fisher\\'s exact test to detect strand bias\"', 'InbreedingCoeff': '\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\"', 'MLEAC': '\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\"', 'MLEAF': '\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\"', 'MQ': '\"RMS Mapping Quality\"', 'MQRankSum': '\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\"', 'NEGATIVE_TRAIN_SITE': '\"This variant was used to build the negative training set of bad variants\"', 'POSITIVE_TRAIN_SITE': '\"This variant was used to build the positive training set of good variants\"', 'QD': '\"Variant Confidence/Quality by Depth\"', 'RAW_MQ': '\"Raw data for RMS Mapping Quality\"', 'ReadPosRankSum': '\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\"', 'ReverseComplementedAlleles': '\"The REF and the ALT alleles have been reverse complemented in liftover since the mapping from the previous reference to the current one was on the negative strand.\"', 'SOR': '\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\"', 'SwappedAlleles': '\"The REF and the ALT alleles have been swapped in liftover due to changes in the reference. It is possible that not all INFO annotations reflect this swap, and in the genotypes, only the GT, PL, and AD fields have been modified. You should check the TAGS_TO_REVERSE parameter that was used during the LiftOver to be sure.\"', 'VQSLOD': '\"Log odds of being a true variant versus being false under the trained gaussian mixture model\"', 'culprit': '\"The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out\"'}\n"
     ]
    }
   ],
   "source": [
    "# for unfiltered wgs\n",
    "# for filtered wgs\n",
    "#! gsutil cp gs://fc-secure-bd7b8bc9-f665-4269-997e-5a402088a369/77dbe69a-509c-4618-a2bf-f809c1aea47a/liftover_workflows/2950a03d-c092-4355-b261-2e6eb50fca41/call-liftover/ccle_ACH-000045_liftover.vcf ../temp/\n",
    "wgsfilt, description = h.vcf_to_df('../temp/ccle_ACH-000045_liftover.vcf')\n",
    "wgsfilt.chr = [i[3:] for i in wgsfilt.chr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "data= {'wgsfilt': wgsfilt, 'mutect2': mutect2, 'mutect1': mutect1, 'strelka': strelka, 'somwes': somwes, 'somwesfilt': somwesfilt, 'wes': wes, 'rnaseq': rnaseq, 'rnaseqfilt': rnaseqfilt, 'wesfilt': wesfilt}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutset = {}\n",
    "for i,v in data.items():\n",
    "    chrom = v.chr.tolist()\n",
    "    alt = v.alt.tolist()\n",
    "    pos = v.pos.tolist()\n",
    "    mutset[i] = set([str(chrom[i])+'_'+str(pos[i]) for i, a in enumerate(alt)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['wgsfilt'] & mutset['wesfilt']) / len(mutset['wesfilt']), len(mutset['wgsfilt']), len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['rnaseqfilt']  & mutset['wesfilt']) / len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['rnaseq']  & mutset['wesfilt']) / len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['rnaseqfilt']  & mutset['rnaseq']) / len(mutset['rnaseq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['wes']  & mutset['wesfilt']) / len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['wes']  & mutset['wesfilt']) / len(mutset['wes'])\n",
    "# how many mutations is the cnn filtering.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['wesfilt']), len(mutset['wes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['strelka']), len(mutset['mutect1']), len(mutset['mutect2'])\n",
    "# how many mutations each contain and their similarity between each caller in cga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['strelka'] & mutset['mutect1'] & mutset['mutect2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes']), len(mutset['somwesfilt']), len(mutset['somwesfilt'] & mutset['somwes'])\n",
    "# how many mutations each contain and their similarity between cga middle and cga end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['mutect1']) / len(mutset['mutect1']), len(mutset['somwesfilt'] & mutset['mutect1']) / len(mutset['mutect1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['mutect2']) / len(mutset['mutect2']), len(mutset['somwesfilt'] & mutset['mutect2']) / len(mutset['mutect2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['strelka']) / len(mutset['strelka']), len(mutset['somwesfilt'] & mutset['strelka']) / len(mutset['strelka'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['wes']) / len(mutset['somwes']), len(mutset['wes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['wesfilt']) / len(mutset['somwes']), len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwesfilt'] & mutset['wesfilt']) / len(mutset['somwesfilt']), len(mutset['wesfilt'])\n",
    "# good agreement between wes hapcall and wes cga and we can see how much we recover once bad mutations get filtered!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['wesfilt']), len(mutset['somwesfilt'] & mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwes'] & mutset['wgsfilt']) / len(mutset['somwes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8333333333333334"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mutset['somwesfilt'] & mutset['wgsfilt']) / len(mutset['somwesfilt'])\n",
    "# good agreement between wgs hapcall and wes cga"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "17% of the mutations in WES could not be found in WGS. this is explained by read depth issues?\n",
    "--> artifacts? low coverage wgs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc= [176551, 1170265, 825909, 1098957, 826976, 278207, 391815] # 7 that we would have liked to find and likely 3 False positives from the previous somatic mutation pipeline. The rest could not have been caught on this experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ref</th>\n",
       "      <th>...</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>338957</td>\n",
       "      <td>EIF5B</td>\n",
       "      <td>9669</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>100015269</td>\n",
       "      <td>100015270</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69:314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>678687</td>\n",
       "      <td>MT-ATP6</td>\n",
       "      <td>4508</td>\n",
       "      <td>37</td>\n",
       "      <td>M</td>\n",
       "      <td>9150</td>\n",
       "      <td>9150</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1018133</td>\n",
       "      <td>SLC22A16</td>\n",
       "      <td>85413</td>\n",
       "      <td>37</td>\n",
       "      <td>6</td>\n",
       "      <td>110763879</td>\n",
       "      <td>110763880</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>444857</td>\n",
       "      <td>GOLGA8I</td>\n",
       "      <td>283796</td>\n",
       "      <td>37</td>\n",
       "      <td>15</td>\n",
       "      <td>23261942</td>\n",
       "      <td>23261942</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:73</td>\n",
       "      <td>16:90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>622626</td>\n",
       "      <td>LTN1</td>\n",
       "      <td>26046</td>\n",
       "      <td>37</td>\n",
       "      <td>21</td>\n",
       "      <td>30303518</td>\n",
       "      <td>30303519</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:76</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>296745</td>\n",
       "      <td>DLGAP2</td>\n",
       "      <td>9228</td>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>1628713</td>\n",
       "      <td>1628713</td>\n",
       "      <td>+</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:35</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445004</td>\n",
       "      <td>GOLGA8R</td>\n",
       "      <td>101059918</td>\n",
       "      <td>37</td>\n",
       "      <td>15</td>\n",
       "      <td>30702129</td>\n",
       "      <td>30702129</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>883766</td>\n",
       "      <td>PRDX6</td>\n",
       "      <td>9588</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>173456921</td>\n",
       "      <td>173456922</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37:264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>930038</td>\n",
       "      <td>RBM42</td>\n",
       "      <td>79171</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>36120079</td>\n",
       "      <td>36120079</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>871718</td>\n",
       "      <td>POU6F2</td>\n",
       "      <td>11281</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>39379285</td>\n",
       "      <td>39379285</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47:55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>176551</td>\n",
       "      <td>CCDC65</td>\n",
       "      <td>85478</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>49312156</td>\n",
       "      <td>49312157</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CA</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40:8</td>\n",
       "      <td>49:74</td>\n",
       "      <td>12:10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1170265</td>\n",
       "      <td>TROVE2</td>\n",
       "      <td>6738</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>193038192</td>\n",
       "      <td>193038192</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:7</td>\n",
       "      <td>21:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:20</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>825909</td>\n",
       "      <td>PDPR</td>\n",
       "      <td>55066</td>\n",
       "      <td>37</td>\n",
       "      <td>16</td>\n",
       "      <td>70154418</td>\n",
       "      <td>70154418</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32:145</td>\n",
       "      <td>30:130</td>\n",
       "      <td>8:29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1168917</td>\n",
       "      <td>TRIP12</td>\n",
       "      <td>9320</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>230724210</td>\n",
       "      <td>230724211</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9:51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>343760</td>\n",
       "      <td>ENAH</td>\n",
       "      <td>55740</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>225707079</td>\n",
       "      <td>225707079</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1098957</td>\n",
       "      <td>TAF3</td>\n",
       "      <td>83860</td>\n",
       "      <td>37</td>\n",
       "      <td>10</td>\n",
       "      <td>8005934</td>\n",
       "      <td>8005934</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25:7</td>\n",
       "      <td>32:43</td>\n",
       "      <td>7:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:25</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1282118</td>\n",
       "      <td>ZNF513</td>\n",
       "      <td>130557</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>27601679</td>\n",
       "      <td>27601681</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>ATG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8:64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>62244</td>\n",
       "      <td>ANO7</td>\n",
       "      <td>50636</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>242135234</td>\n",
       "      <td>242135234</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13:42</td>\n",
       "      <td>11:39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>826976</td>\n",
       "      <td>PDZD2</td>\n",
       "      <td>23037</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>32089889</td>\n",
       "      <td>32089897</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GCGACAGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:0</td>\n",
       "      <td>9:24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15:20</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>836283</td>\n",
       "      <td>PHKA2</td>\n",
       "      <td>5256</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>18969282</td>\n",
       "      <td>18969283</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:2</td>\n",
       "      <td>18:6</td>\n",
       "      <td>11:25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:7</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1199549</td>\n",
       "      <td>UFL1</td>\n",
       "      <td>23376</td>\n",
       "      <td>37</td>\n",
       "      <td>6</td>\n",
       "      <td>96985372</td>\n",
       "      <td>96985372</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>115368</td>\n",
       "      <td>BDP1</td>\n",
       "      <td>55814</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>70858261</td>\n",
       "      <td>70858262</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1105410</td>\n",
       "      <td>TBC1D4</td>\n",
       "      <td>9882</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>76055659</td>\n",
       "      <td>76055659</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>465238</td>\n",
       "      <td>GTF2E1</td>\n",
       "      <td>2960</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>120489633</td>\n",
       "      <td>120489634</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10:55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71924</td>\n",
       "      <td>ARAP2</td>\n",
       "      <td>116984</td>\n",
       "      <td>37</td>\n",
       "      <td>4</td>\n",
       "      <td>36152653</td>\n",
       "      <td>36152653</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:25</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>881304</td>\n",
       "      <td>PRAMEF21</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>13522203</td>\n",
       "      <td>13522203</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31:0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8:0</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>360682</td>\n",
       "      <td>EXD3</td>\n",
       "      <td>54932</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>140218237</td>\n",
       "      <td>140218237</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8:40</td>\n",
       "      <td>8:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>526281</td>\n",
       "      <td>IRS2</td>\n",
       "      <td>8660</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>110434578</td>\n",
       "      <td>110434583</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GCGGCG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:0</td>\n",
       "      <td>4:8</td>\n",
       "      <td>9:0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>391815</td>\n",
       "      <td>FCGBP</td>\n",
       "      <td>8857</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>40366263</td>\n",
       "      <td>40366263</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:4</td>\n",
       "      <td>25:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>897992</td>\n",
       "      <td>PSMD1</td>\n",
       "      <td>5707</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>231931679</td>\n",
       "      <td>231931680</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29:124</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>924206</td>\n",
       "      <td>RASA4B</td>\n",
       "      <td>100271927</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>102137315</td>\n",
       "      <td>102137315</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1264242</td>\n",
       "      <td>ZMYM6</td>\n",
       "      <td>9204</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>35452789</td>\n",
       "      <td>35452790</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3313</td>\n",
       "      <td>ABCA13</td>\n",
       "      <td>154664</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>48559709</td>\n",
       "      <td>48559709</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34:148</td>\n",
       "      <td>33:148</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>778104</td>\n",
       "      <td>OR2T8</td>\n",
       "      <td>343172</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>248085025</td>\n",
       "      <td>248085025</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:2</td>\n",
       "      <td>8:19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>278207</td>\n",
       "      <td>DCHS2</td>\n",
       "      <td>54798</td>\n",
       "      <td>37</td>\n",
       "      <td>4</td>\n",
       "      <td>155253735</td>\n",
       "      <td>155253735</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27:11</td>\n",
       "      <td>38:63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:11</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>909946</td>\n",
       "      <td>PUS7</td>\n",
       "      <td>54517</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>105148863</td>\n",
       "      <td>105148864</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1267443</td>\n",
       "      <td>ZNF184</td>\n",
       "      <td>7738</td>\n",
       "      <td>37</td>\n",
       "      <td>6</td>\n",
       "      <td>27420172</td>\n",
       "      <td>27420173</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id  NCBI_Build chr        pos  End_position  \\\n",
       "338957        EIF5B            9669          37   2  100015269     100015270   \n",
       "678687      MT-ATP6            4508          37   M       9150          9150   \n",
       "1018133    SLC22A16           85413          37   6  110763879     110763880   \n",
       "444857      GOLGA8I          283796          37  15   23261942      23261942   \n",
       "622626         LTN1           26046          37  21   30303518      30303519   \n",
       "296745       DLGAP2            9228          37   8    1628713       1628713   \n",
       "445004      GOLGA8R       101059918          37  15   30702129      30702129   \n",
       "883766        PRDX6            9588          37   1  173456921     173456922   \n",
       "930038        RBM42           79171          37  19   36120079      36120079   \n",
       "871718       POU6F2           11281          37   7   39379285      39379285   \n",
       "176551       CCDC65           85478          37  12   49312156      49312157   \n",
       "1170265      TROVE2            6738          37   1  193038192     193038192   \n",
       "825909         PDPR           55066          37  16   70154418      70154418   \n",
       "1168917      TRIP12            9320          37   2  230724210     230724211   \n",
       "343760         ENAH           55740          37   1  225707079     225707079   \n",
       "1098957        TAF3           83860          37  10    8005934       8005934   \n",
       "1282118      ZNF513          130557          37   2   27601679      27601681   \n",
       "62244          ANO7           50636          37   2  242135234     242135234   \n",
       "826976        PDZD2           23037          37   5   32089889      32089897   \n",
       "836283        PHKA2            5256          37   X   18969282      18969283   \n",
       "1199549        UFL1           23376          37   6   96985372      96985372   \n",
       "115368         BDP1           55814          37   5   70858261      70858262   \n",
       "1105410      TBC1D4            9882          37  13   76055659      76055659   \n",
       "465238       GTF2E1            2960          37   3  120489633     120489634   \n",
       "71924         ARAP2          116984          37   4   36152653      36152653   \n",
       "881304     PRAMEF21               0          37   1   13522203      13522203   \n",
       "360682         EXD3           54932          37   9  140218237     140218237   \n",
       "526281         IRS2            8660          37  13  110434578     110434583   \n",
       "391815        FCGBP            8857          37  19   40366263      40366263   \n",
       "897992        PSMD1            5707          37   2  231931679     231931680   \n",
       "924206       RASA4B       100271927          37   7  102137315     102137315   \n",
       "1264242       ZMYM6            9204          37   1   35452789      35452790   \n",
       "3313         ABCA13          154664          37   7   48559709      48559709   \n",
       "778104        OR2T8          343172          37   1  248085025     248085025   \n",
       "278207        DCHS2           54798          37   4  155253735     155253735   \n",
       "909946         PUS7           54517          37   7  105148863     105148864   \n",
       "1267443      ZNF184            7738          37   6   27420172      27420173   \n",
       "\n",
       "        Strand Variant_Classification Variant_Type        ref  ...   ExAC_AF  \\\n",
       "338957       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "678687       +                 Silent          SNP          A  ...       NaN   \n",
       "1018133      +        Frame_Shift_Ins          INS          -  ...  0.000016   \n",
       "444857       +      Missense_Mutation          SNP          A  ...  0.000034   \n",
       "622626       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "296745       +            Splice_Site          SNP          G  ...       NaN   \n",
       "445004       +      Missense_Mutation          SNP          A  ...       NaN   \n",
       "883766       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "930038       +                 Silent          SNP          G  ...       NaN   \n",
       "871718       +      Missense_Mutation          SNP          C  ...  0.000042   \n",
       "176551       +        Frame_Shift_Del          DEL         CA  ...  0.000016   \n",
       "1170265      +        Frame_Shift_Del          DEL          A  ...       NaN   \n",
       "825909       +      Missense_Mutation          SNP          C  ...  0.000008   \n",
       "1168917      +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "343760       +      Missense_Mutation          SNP          T  ...       NaN   \n",
       "1098957      +        Frame_Shift_Del          DEL          T  ...       NaN   \n",
       "1282118      +           In_Frame_Del          DEL        ATG  ...       NaN   \n",
       "62244        +      Missense_Mutation          SNP          C  ...  0.000016   \n",
       "826976       +           In_Frame_Del          DEL  GCGACAGAC  ...       NaN   \n",
       "836283       +        Frame_Shift_Del          DEL         GG  ...       NaN   \n",
       "1199549      +      Missense_Mutation          SNP          G  ...       NaN   \n",
       "115368       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "1105410      +      Missense_Mutation          SNP          A  ...       NaN   \n",
       "465238       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "71924        +        Frame_Shift_Del          DEL          T  ...       NaN   \n",
       "881304       +                 Silent          SNP          C  ...       NaN   \n",
       "360682       +                 Silent          SNP          C  ...  0.000017   \n",
       "526281       +           In_Frame_Del          DEL     GCGGCG  ...       NaN   \n",
       "391815       +        Frame_Shift_Del          DEL          C  ...  0.000017   \n",
       "897992       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "924206       +                 Silent          SNP          G  ...       NaN   \n",
       "1264242      +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "3313         +                 Silent          SNP          C  ...  0.000008   \n",
       "778104       +        Frame_Shift_Del          DEL          T  ...       NaN   \n",
       "278207       +        Frame_Shift_Del          DEL          C  ...       NaN   \n",
       "909946       +        Frame_Shift_Ins          INS          -  ...       NaN   \n",
       "1267443      +        Frame_Shift_Ins          INS          -  ...  0.000008   \n",
       "\n",
       "        CGA_WES_AC SangerWES_AC SangerRecalibWES_AC RNAseq_AC HC_AC RD_AC  \\\n",
       "338957         NaN          NaN                 NaN    69:314   NaN   NaN   \n",
       "678687         NaN          NaN                45:1       NaN   NaN   NaN   \n",
       "1018133        NaN          NaN                 NaN     11:60   NaN   NaN   \n",
       "444857         NaN        16:73               16:90       NaN   NaN   NaN   \n",
       "622626         NaN          NaN                 NaN     16:76   NaN   NaN   \n",
       "296745         NaN          NaN                 NaN       NaN   NaN   NaN   \n",
       "445004         NaN          NaN                 NaN      16:1   NaN   NaN   \n",
       "883766         NaN          NaN                 NaN    37:264   NaN   NaN   \n",
       "930038         NaN          NaN                 4:8       NaN   NaN   NaN   \n",
       "871718         NaN          NaN               47:55       NaN   NaN   NaN   \n",
       "176551         NaN         40:8               49:74     12:10   NaN   NaN   \n",
       "1170265        NaN         12:7               21:30       NaN   NaN   NaN   \n",
       "825909         NaN       32:145              30:130      8:29   NaN   NaN   \n",
       "1168917        NaN          NaN                 NaN      9:51   NaN   NaN   \n",
       "343760         NaN          NaN                 6:4       NaN   NaN   NaN   \n",
       "1098957        NaN         25:7               32:43      7:16   NaN   NaN   \n",
       "1282118        NaN          NaN                 NaN      8:64   NaN   NaN   \n",
       "62244          NaN        13:42               11:39       NaN   NaN   NaN   \n",
       "826976         NaN          7:0                9:24       NaN   NaN   NaN   \n",
       "836283         NaN         17:2                18:6     11:25   NaN   NaN   \n",
       "1199549        NaN          NaN                 NaN      4:25   NaN   NaN   \n",
       "115368         NaN          NaN                 NaN      9:33   NaN   NaN   \n",
       "1105410        NaN          NaN                4:16       NaN   NaN   NaN   \n",
       "465238         NaN          NaN                 NaN     10:55   NaN   NaN   \n",
       "71924          NaN          NaN                 NaN       NaN   NaN   NaN   \n",
       "881304         NaN          NaN                31:0       NaN   NaN   NaN   \n",
       "360682         NaN         8:40                8:41       NaN   NaN   NaN   \n",
       "526281         NaN          5:0                 4:8       9:0   NaN   NaN   \n",
       "391815         NaN         20:4               25:16       NaN   NaN   NaN   \n",
       "897992         NaN          NaN                 NaN    29:124   NaN   NaN   \n",
       "924206         NaN          NaN                 NaN       4:9   NaN   NaN   \n",
       "1264242        NaN          NaN                 NaN    17:102   NaN   NaN   \n",
       "3313           NaN       34:148              33:148       NaN   NaN   NaN   \n",
       "778104         NaN          6:2                8:19       NaN   NaN   NaN   \n",
       "278207         NaN        27:11               38:63       NaN   NaN   NaN   \n",
       "909946         NaN          NaN                 NaN     18:84   NaN   NaN   \n",
       "1267443        NaN          NaN                 NaN      6:28   NaN   NaN   \n",
       "\n",
       "        WGS_AC    Variant_annotation   DepMap_ID  \n",
       "338957     NaN              damaging  ACH-000045  \n",
       "678687     NaN                silent  ACH-000045  \n",
       "1018133    NaN              damaging  ACH-000045  \n",
       "444857     NaN  other non-conserving  ACH-000045  \n",
       "622626     NaN              damaging  ACH-000045  \n",
       "296745   12:35              damaging  ACH-000045  \n",
       "445004     NaN  other non-conserving  ACH-000045  \n",
       "883766     NaN              damaging  ACH-000045  \n",
       "930038     NaN                silent  ACH-000045  \n",
       "871718     NaN  other non-conserving  ACH-000045  \n",
       "176551   11:17              damaging  ACH-000045  \n",
       "1170265   6:20              damaging  ACH-000045  \n",
       "825909     NaN  other non-conserving  ACH-000045  \n",
       "1168917    NaN              damaging  ACH-000045  \n",
       "343760     NaN  other non-conserving  ACH-000045  \n",
       "1098957   5:25              damaging  ACH-000045  \n",
       "1282118    NaN  other non-conserving  ACH-000045  \n",
       "62244      NaN  other non-conserving  ACH-000045  \n",
       "826976   15:20  other non-conserving  ACH-000045  \n",
       "836283     3:7              damaging  ACH-000045  \n",
       "1199549    NaN  other non-conserving  ACH-000045  \n",
       "115368     NaN              damaging  ACH-000045  \n",
       "1105410    NaN  other non-conserving  ACH-000045  \n",
       "465238     NaN              damaging  ACH-000045  \n",
       "71924     7:25              damaging  ACH-000045  \n",
       "881304     8:0                silent  ACH-000045  \n",
       "360682     NaN                silent  ACH-000045  \n",
       "526281     NaN  other non-conserving  ACH-000045  \n",
       "391815   26:17              damaging  ACH-000045  \n",
       "897992     NaN              damaging  ACH-000045  \n",
       "924206     NaN                silent  ACH-000045  \n",
       "1264242    NaN              damaging  ACH-000045  \n",
       "3313       NaN                silent  ACH-000045  \n",
       "778104     NaN              damaging  ACH-000045  \n",
       "278207   16:11              damaging  ACH-000045  \n",
       "909946     NaN              damaging  ACH-000045  \n",
       "1267443    NaN              damaging  ACH-000045  \n",
       "\n",
       "[37 rows x 34 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = subset(data['somwesfilt'],mutset['somwesfilt'] - mutset['wgsfilt'])\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "locify(mutset['somwesfilt'] - mutset['wgsfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutset['somwesfilt'] & mutset['wes']) / len(mutset['somwesfilt']), len(mutset['wes'])\n",
    "# good agreement between wes hapcall and wes cga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we don't have a good similarity wes/wgs, can it be because of slight differences in the base distances? can it be because of the mutation caller?\n",
    "len(set([i for i in mutset['wesfilt']]) & set([i for i in mutset['wgsfilt']]))/len(mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locify(dat):\n",
    "    return ['chr'+d.split('_')[0]+':'+str(int(d.split('_')[1])-40)+'-'+str(int(d.split('_')[1])+40) for d in dat]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'chr6:27420132-27420212':'rna',\n",
    "'chr5:32089849-32089929':'del',\n",
    "'chr4:100574667-100574747':'wgs',\n",
    "'chr7:100643254-100643334':'wgs',\n",
    "'chr15:89070794-89070874':'wgs',\n",
    "'chr19:36120039-36120119':'snp',\n",
    "'chr20:31196288-31196368':'snp',\n",
    "'chr15:42343066-42343146':False,\n",
    "'chr11:6568654-6568734':'snp',\n",
    "\n",
    "'chr9:24543236-24543316':'snp',\n",
    "'chr15:23261902-23261982':'snp',\n",
    "'chr4:84255808-84255888':True,\n",
    "'chr13:76055619-76055699':'snp',\n",
    "'chrX:18969242-18969322':,\n",
    "'chr4:36152613-36152693':,\n",
    "'chr19:50354695-50354775':,\n",
    "'chr4:100574668-100574748':,\n",
    "'chr6:110763839-110763919':,\n",
    "'chr7:100648111-100648191':,\n",
    "'chr5:70858221-70858301':,\n",
    "'chr3:120489593-120489673':,\n",
    "'chr2:27601639-27601719':,\n",
    "'chr4:77184865-77184945':,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'locify' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-bc209fe04a92>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mlocify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutset\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'somwesfilt'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mmutset\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'wesfilt'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'locify' is not defined"
     ]
    }
   ],
   "source": [
    "locify(mutset['somwesfilt'] - mutset['wesfilt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subset(df,mut):\n",
    "    return pd.concat([df[(df.chr==val.split('_')[0]) & (df.pos==int(val.split('_')[1]))] for val in mut])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ref</th>\n",
       "      <th>...</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>678687</td>\n",
       "      <td>MT-ATP6</td>\n",
       "      <td>4508</td>\n",
       "      <td>37</td>\n",
       "      <td>M</td>\n",
       "      <td>9150</td>\n",
       "      <td>9150</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>444857</td>\n",
       "      <td>GOLGA8I</td>\n",
       "      <td>283796</td>\n",
       "      <td>37</td>\n",
       "      <td>15</td>\n",
       "      <td>23261942</td>\n",
       "      <td>23261942</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:73</td>\n",
       "      <td>16:90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>930038</td>\n",
       "      <td>RBM42</td>\n",
       "      <td>79171</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>36120079</td>\n",
       "      <td>36120079</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>176551</td>\n",
       "      <td>CCDC65</td>\n",
       "      <td>85478</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>49312156</td>\n",
       "      <td>49312157</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CA</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40:8</td>\n",
       "      <td>49:74</td>\n",
       "      <td>12:10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>309536</td>\n",
       "      <td>DNHD1</td>\n",
       "      <td>144132</td>\n",
       "      <td>37</td>\n",
       "      <td>11</td>\n",
       "      <td>6568694</td>\n",
       "      <td>6568694</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29:29</td>\n",
       "      <td>30:29</td>\n",
       "      <td>22:24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26:17</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1170265</td>\n",
       "      <td>TROVE2</td>\n",
       "      <td>6738</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>193038192</td>\n",
       "      <td>193038192</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:7</td>\n",
       "      <td>21:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:20</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78368</td>\n",
       "      <td>ARHGEF16</td>\n",
       "      <td>27237</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>3380076</td>\n",
       "      <td>3380076</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:4</td>\n",
       "      <td>5:4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26:18</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>956348</td>\n",
       "      <td>RP11-410N8.4</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>20</td>\n",
       "      <td>31196328</td>\n",
       "      <td>31196328</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:6</td>\n",
       "      <td>4:7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:23</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>343760</td>\n",
       "      <td>ENAH</td>\n",
       "      <td>55740</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>225707079</td>\n",
       "      <td>225707079</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1098957</td>\n",
       "      <td>TAF3</td>\n",
       "      <td>83860</td>\n",
       "      <td>37</td>\n",
       "      <td>10</td>\n",
       "      <td>8005934</td>\n",
       "      <td>8005934</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25:7</td>\n",
       "      <td>32:43</td>\n",
       "      <td>7:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:25</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>826976</td>\n",
       "      <td>PDZD2</td>\n",
       "      <td>23037</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>32089889</td>\n",
       "      <td>32089897</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GCGACAGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:0</td>\n",
       "      <td>9:24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15:20</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>505498</td>\n",
       "      <td>IFFO2</td>\n",
       "      <td>126917</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>19244904</td>\n",
       "      <td>19244904</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:24</td>\n",
       "      <td>18:22</td>\n",
       "      <td>5:3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:11</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>836283</td>\n",
       "      <td>PHKA2</td>\n",
       "      <td>5256</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>18969282</td>\n",
       "      <td>18969283</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:2</td>\n",
       "      <td>18:6</td>\n",
       "      <td>11:25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:7</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1105410</td>\n",
       "      <td>TBC1D4</td>\n",
       "      <td>9882</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>76055659</td>\n",
       "      <td>76055659</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>526281</td>\n",
       "      <td>IRS2</td>\n",
       "      <td>8660</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>110434578</td>\n",
       "      <td>110434583</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GCGGCG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5:0</td>\n",
       "      <td>4:8</td>\n",
       "      <td>9:0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>690021</td>\n",
       "      <td>MUC12</td>\n",
       "      <td>10071</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>100648151</td>\n",
       "      <td>100648151</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56:71</td>\n",
       "      <td>54:70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13:16</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>391815</td>\n",
       "      <td>FCGBP</td>\n",
       "      <td>8857</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>40366263</td>\n",
       "      <td>40366263</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:4</td>\n",
       "      <td>25:16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>778104</td>\n",
       "      <td>OR2T8</td>\n",
       "      <td>343172</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>248085025</td>\n",
       "      <td>248085025</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:2</td>\n",
       "      <td>8:19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>377082</td>\n",
       "      <td>FAM47E</td>\n",
       "      <td>100129583</td>\n",
       "      <td>37</td>\n",
       "      <td>4</td>\n",
       "      <td>77184905</td>\n",
       "      <td>77184905</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50:70</td>\n",
       "      <td>49:68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13:13</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>278207</td>\n",
       "      <td>DCHS2</td>\n",
       "      <td>54798</td>\n",
       "      <td>37</td>\n",
       "      <td>4</td>\n",
       "      <td>155253735</td>\n",
       "      <td>155253735</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27:11</td>\n",
       "      <td>38:63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16:11</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Hugo_Symbol  Entrez_Gene_Id  NCBI_Build chr        pos  \\\n",
       "678687        MT-ATP6            4508          37   M       9150   \n",
       "444857        GOLGA8I          283796          37  15   23261942   \n",
       "930038          RBM42           79171          37  19   36120079   \n",
       "176551         CCDC65           85478          37  12   49312156   \n",
       "309536          DNHD1          144132          37  11    6568694   \n",
       "1170265        TROVE2            6738          37   1  193038192   \n",
       "78368        ARHGEF16           27237          37   1    3380076   \n",
       "956348   RP11-410N8.4               0          37  20   31196328   \n",
       "343760           ENAH           55740          37   1  225707079   \n",
       "1098957          TAF3           83860          37  10    8005934   \n",
       "826976          PDZD2           23037          37   5   32089889   \n",
       "505498          IFFO2          126917          37   1   19244904   \n",
       "836283          PHKA2            5256          37   X   18969282   \n",
       "1105410        TBC1D4            9882          37  13   76055659   \n",
       "526281           IRS2            8660          37  13  110434578   \n",
       "690021          MUC12           10071          37   7  100648151   \n",
       "391815          FCGBP            8857          37  19   40366263   \n",
       "778104          OR2T8          343172          37   1  248085025   \n",
       "377082         FAM47E       100129583          37   4   77184905   \n",
       "278207          DCHS2           54798          37   4  155253735   \n",
       "\n",
       "         End_position Strand Variant_Classification Variant_Type        ref  \\\n",
       "678687           9150      +                 Silent          SNP          A   \n",
       "444857       23261942      +      Missense_Mutation          SNP          A   \n",
       "930038       36120079      +                 Silent          SNP          G   \n",
       "176551       49312157      +        Frame_Shift_Del          DEL         CA   \n",
       "309536        6568694      +                 Silent          SNP          C   \n",
       "1170265     193038192      +        Frame_Shift_Del          DEL          A   \n",
       "78368         3380076      +      Missense_Mutation          SNP          G   \n",
       "956348       31196328      +      Missense_Mutation          SNP          A   \n",
       "343760      225707079      +      Missense_Mutation          SNP          T   \n",
       "1098957       8005934      +        Frame_Shift_Del          DEL          T   \n",
       "826976       32089897      +           In_Frame_Del          DEL  GCGACAGAC   \n",
       "505498       19244904      +                 Silent          SNP          G   \n",
       "836283       18969283      +        Frame_Shift_Del          DEL         GG   \n",
       "1105410      76055659      +      Missense_Mutation          SNP          A   \n",
       "526281      110434583      +           In_Frame_Del          DEL     GCGGCG   \n",
       "690021      100648151      +                 Silent          SNP          G   \n",
       "391815       40366263      +        Frame_Shift_Del          DEL          C   \n",
       "778104      248085025      +        Frame_Shift_Del          DEL          T   \n",
       "377082       77184905      +      Missense_Mutation          SNP          G   \n",
       "278207      155253735      +        Frame_Shift_Del          DEL          C   \n",
       "\n",
       "         ...   ExAC_AF CGA_WES_AC SangerWES_AC SangerRecalibWES_AC RNAseq_AC  \\\n",
       "678687   ...       NaN        NaN          NaN                45:1       NaN   \n",
       "444857   ...  0.000034        NaN        16:73               16:90       NaN   \n",
       "930038   ...       NaN        NaN          NaN                 4:8       NaN   \n",
       "176551   ...  0.000016        NaN         40:8               49:74     12:10   \n",
       "309536   ...       NaN        NaN        29:29               30:29     22:24   \n",
       "1170265  ...       NaN        NaN         12:7               21:30       NaN   \n",
       "78368    ...       NaN        NaN          5:4                 5:4       NaN   \n",
       "956348   ...       NaN        NaN          4:6                 4:7       NaN   \n",
       "343760   ...       NaN        NaN          NaN                 6:4       NaN   \n",
       "1098957  ...       NaN        NaN         25:7               32:43      7:16   \n",
       "826976   ...       NaN        NaN          7:0                9:24       NaN   \n",
       "505498   ...       NaN        NaN        20:24               18:22       5:3   \n",
       "836283   ...       NaN        NaN         17:2                18:6     11:25   \n",
       "1105410  ...       NaN        NaN          NaN                4:16       NaN   \n",
       "526281   ...       NaN        NaN          5:0                 4:8       9:0   \n",
       "690021   ...       NaN        NaN        56:71               54:70       NaN   \n",
       "391815   ...  0.000017        NaN         20:4               25:16       NaN   \n",
       "778104   ...       NaN        NaN          6:2                8:19       NaN   \n",
       "377082   ...  0.000039        NaN        50:70               49:68       NaN   \n",
       "278207   ...       NaN        NaN        27:11               38:63       NaN   \n",
       "\n",
       "        HC_AC RD_AC WGS_AC    Variant_annotation   DepMap_ID  \n",
       "678687    NaN   NaN    NaN                silent  ACH-000045  \n",
       "444857    NaN   NaN    NaN  other non-conserving  ACH-000045  \n",
       "930038    NaN   NaN    NaN                silent  ACH-000045  \n",
       "176551    NaN   NaN  11:17              damaging  ACH-000045  \n",
       "309536    NaN   NaN  26:17                silent  ACH-000045  \n",
       "1170265   NaN   NaN   6:20              damaging  ACH-000045  \n",
       "78368     NaN   NaN  26:18  other non-conserving  ACH-000045  \n",
       "956348    NaN   NaN  16:23  other non-conserving  ACH-000045  \n",
       "343760    NaN   NaN    NaN  other non-conserving  ACH-000045  \n",
       "1098957   NaN   NaN   5:25              damaging  ACH-000045  \n",
       "826976    NaN   NaN  15:20  other non-conserving  ACH-000045  \n",
       "505498    NaN   NaN  18:11                silent  ACH-000045  \n",
       "836283    NaN   NaN    3:7              damaging  ACH-000045  \n",
       "1105410   NaN   NaN    NaN  other non-conserving  ACH-000045  \n",
       "526281    NaN   NaN    NaN  other non-conserving  ACH-000045  \n",
       "690021    NaN   NaN  13:16                silent  ACH-000045  \n",
       "391815    NaN   NaN  26:17              damaging  ACH-000045  \n",
       "778104    NaN   NaN    NaN              damaging  ACH-000045  \n",
       "377082    NaN   NaN  13:13  other non-conserving  ACH-000045  \n",
       "278207    NaN   NaN  16:11              damaging  ACH-000045  \n",
       "\n",
       "[20 rows x 34 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = subset(data['somwesfilt'],mutset['somwesfilt'] - mutset['wesfilt'])\n",
    "print(60-a.SangerRecalibWES_AC.isna().sum())\n",
    "a[~a.SangerRecalibWES_AC.isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  we have 60-40=20 that the old pipeline has found but not us. of these 20; 7 were suspicious as they were not confirmed by any other sequencing methods, had low depth etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ref</th>\n",
       "      <th>...</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>SangerRecalibWES_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>DepMap_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>527683</td>\n",
       "      <td>ISOC1</td>\n",
       "      <td>51015</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>128430479</td>\n",
       "      <td>128430479</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:7</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>729754</td>\n",
       "      <td>NEURL</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>10</td>\n",
       "      <td>105344322</td>\n",
       "      <td>105344322</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:10</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>335550</td>\n",
       "      <td>EHMT1</td>\n",
       "      <td>79813</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>140707977</td>\n",
       "      <td>140707977</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147:142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:17</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>138731</td>\n",
       "      <td>C1orf122</td>\n",
       "      <td>127687</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>38274362</td>\n",
       "      <td>38274362</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102:85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14:10</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>247398</td>\n",
       "      <td>CPVL</td>\n",
       "      <td>54504</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>29035415</td>\n",
       "      <td>29035415</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:12</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>385824</td>\n",
       "      <td>FBLN2</td>\n",
       "      <td>2199</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>13655514</td>\n",
       "      <td>13655514</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14:22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:16</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1172171</td>\n",
       "      <td>TRPM5</td>\n",
       "      <td>29850</td>\n",
       "      <td>37</td>\n",
       "      <td>11</td>\n",
       "      <td>2434364</td>\n",
       "      <td>2434364</td>\n",
       "      <td>+</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:17</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1242740</td>\n",
       "      <td>XRCC3</td>\n",
       "      <td>7517</td>\n",
       "      <td>37</td>\n",
       "      <td>14</td>\n",
       "      <td>104169511</td>\n",
       "      <td>104169511</td>\n",
       "      <td>+</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>112:106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9:16</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>335548</td>\n",
       "      <td>EHMT1</td>\n",
       "      <td>79813</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>140707976</td>\n",
       "      <td>140707976</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>149:139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:17</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>291218</td>\n",
       "      <td>DHX36</td>\n",
       "      <td>170506</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>154018843</td>\n",
       "      <td>154018843</td>\n",
       "      <td>+</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46:45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8:24</td>\n",
       "      <td>damaging</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>756847</td>\n",
       "      <td>NSMF</td>\n",
       "      <td>26012</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>140352100</td>\n",
       "      <td>140352100</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>170:182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24:22</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>513218</td>\n",
       "      <td>IL17RA</td>\n",
       "      <td>23765</td>\n",
       "      <td>37</td>\n",
       "      <td>22</td>\n",
       "      <td>17590137</td>\n",
       "      <td>17590137</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>161:149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15:16</td>\n",
       "      <td>silent</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>247396</td>\n",
       "      <td>CPVL</td>\n",
       "      <td>54504</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>29035414</td>\n",
       "      <td>29035414</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21:29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:12</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>527682</td>\n",
       "      <td>ISOC1</td>\n",
       "      <td>51015</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>128430478</td>\n",
       "      <td>128430478</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11:7</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>291221</td>\n",
       "      <td>DHX36</td>\n",
       "      <td>170506</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>154018844</td>\n",
       "      <td>154018844</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46:43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9:24</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>385827</td>\n",
       "      <td>FBLN2</td>\n",
       "      <td>2199</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>13655515</td>\n",
       "      <td>13655515</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15:23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10:18</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>ACH-000045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id  NCBI_Build chr        pos  End_position  \\\n",
       "527683        ISOC1           51015          37   5  128430479     128430479   \n",
       "729754        NEURL               0          37  10  105344322     105344322   \n",
       "335550        EHMT1           79813          37   9  140707977     140707977   \n",
       "138731     C1orf122          127687          37   1   38274362      38274362   \n",
       "247398         CPVL           54504          37   7   29035415      29035415   \n",
       "385824        FBLN2            2199          37   3   13655514      13655514   \n",
       "1172171       TRPM5           29850          37  11    2434364       2434364   \n",
       "1242740       XRCC3            7517          37  14  104169511     104169511   \n",
       "335548        EHMT1           79813          37   9  140707976     140707976   \n",
       "291218        DHX36          170506          37   3  154018843     154018843   \n",
       "756847         NSMF           26012          37   9  140352100     140352100   \n",
       "513218       IL17RA           23765          37  22   17590137      17590137   \n",
       "247396         CPVL           54504          37   7   29035414      29035414   \n",
       "527682        ISOC1           51015          37   5  128430478     128430478   \n",
       "291221        DHX36          170506          37   3  154018844     154018844   \n",
       "385827        FBLN2            2199          37   3   13655515      13655515   \n",
       "\n",
       "        Strand Variant_Classification Variant_Type ref  ...   ExAC_AF  \\\n",
       "527683       +      Missense_Mutation          SNP   C  ...       NaN   \n",
       "729754       +      Missense_Mutation          SNP   C  ...       NaN   \n",
       "335550       +      Missense_Mutation          SNP   C  ...       NaN   \n",
       "138731       +                 Silent          SNP   A  ...  0.000010   \n",
       "247398       +                 Silent          SNP   T  ...       NaN   \n",
       "385824       +      Missense_Mutation          SNP   C  ...       NaN   \n",
       "1172171      +            Splice_Site          SNP   C  ...       NaN   \n",
       "1242740      +            Splice_Site          SNP   A  ...       NaN   \n",
       "335548       +                 Silent          SNP   T  ...  0.000016   \n",
       "291218       +      Nonsense_Mutation          SNP   C  ...       NaN   \n",
       "756847       +                 Silent          SNP   C  ...       NaN   \n",
       "513218       +                 Silent          SNP   C  ...       NaN   \n",
       "247396       +      Missense_Mutation          SNP   T  ...       NaN   \n",
       "527682       +      Missense_Mutation          SNP   G  ...       NaN   \n",
       "291221       +      Missense_Mutation          SNP   T  ...       NaN   \n",
       "385827       +      Missense_Mutation          SNP   G  ...       NaN   \n",
       "\n",
       "        CGA_WES_AC SangerWES_AC SangerRecalibWES_AC RNAseq_AC HC_AC RD_AC  \\\n",
       "527683         NaN          NaN                 NaN     47:47   NaN   NaN   \n",
       "729754         NaN          NaN                 NaN       NaN   NaN   NaN   \n",
       "335550         NaN        12:12                 NaN   147:142   NaN   NaN   \n",
       "138731         NaN          NaN                 NaN    102:85   NaN   NaN   \n",
       "247398         NaN        20:29                 NaN       NaN   NaN   NaN   \n",
       "385824         NaN        14:22                 NaN       NaN   NaN   NaN   \n",
       "1172171        NaN          NaN                 NaN       NaN   NaN   NaN   \n",
       "1242740        NaN          NaN                 NaN   112:106   NaN   NaN   \n",
       "335548         NaN        11:12                 NaN   149:139   NaN   NaN   \n",
       "291218         NaN        46:45                 NaN       NaN   NaN   NaN   \n",
       "756847         NaN          NaN                 NaN   170:182   NaN   NaN   \n",
       "513218         NaN          NaN                 NaN   161:149   NaN   NaN   \n",
       "247396         NaN        21:29                 NaN       NaN   NaN   NaN   \n",
       "527682         NaN          NaN                 NaN     47:47   NaN   NaN   \n",
       "291221         NaN        46:43                 NaN       NaN   NaN   NaN   \n",
       "385827         NaN        15:23                 NaN       NaN   NaN   NaN   \n",
       "\n",
       "        WGS_AC    Variant_annotation   DepMap_ID  \n",
       "527683    11:7  other non-conserving  ACH-000045  \n",
       "729754    7:10  other non-conserving  ACH-000045  \n",
       "335550   18:17  other non-conserving  ACH-000045  \n",
       "138731   14:10                silent  ACH-000045  \n",
       "247398   17:12                silent  ACH-000045  \n",
       "385824   11:16  other non-conserving  ACH-000045  \n",
       "1172171  20:17              damaging  ACH-000045  \n",
       "1242740   9:16              damaging  ACH-000045  \n",
       "335548   18:17                silent  ACH-000045  \n",
       "291218    8:24              damaging  ACH-000045  \n",
       "756847   24:22                silent  ACH-000045  \n",
       "513218   15:16                silent  ACH-000045  \n",
       "247396   18:12  other non-conserving  ACH-000045  \n",
       "527682    11:7  other non-conserving  ACH-000045  \n",
       "291221    9:24  other non-conserving  ACH-000045  \n",
       "385827   10:18  other non-conserving  ACH-000045  \n",
       "\n",
       "[16 rows x 34 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = subset(data['somwesfilt'],mutset['somwesfilt'] & mutset['wesfilt'])\n",
    "print(res.SangerRecalibWES_AC.isna().sum())\n",
    "res[res.SangerRecalibWES_AC.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have found 16 mutations that the old pipeline did not fouund, of these, only one was only confirmed by one sequencing method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset(data['somwesfilt'],mutset['rnaseqfilt'] & mutset['somwesfilt']).RNAseq_AC.isna().sum() #113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset(data['somwesfilt'],mutset['somwesfilt'] - mutset['rnaseqfilt']).RNAseq_AC.isna().sum() #113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have found 12 more rnaseq mutations and lost 9 rnaseq mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'WGS_AC'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-35adca389dbe>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msubset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'wgsfilt'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutset\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'somwesfilt'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mmutset\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'wgsfilt'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m37\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mWGS_AC\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5177\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5178\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5179\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5181\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'WGS_AC'"
     ]
    }
   ],
   "source": [
    "res = subset(data['somwesfilt'],mutset['somwesfilt'] - mutset['wgsfilt'])\n",
    "print(37-res.WGS_AC.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### we have high allele freq mut which seems it might be that they get filtered by the cnn. two seemed quite shady"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
