{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f426c1f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20e60ab5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on built-in function exec in module builtins:\n",
      "\n",
      "exec(source, globals=None, locals=None, /, *, closure=None)\n",
      "    Execute the given source in the context of globals and locals.\n",
      "    \n",
      "    The source may be a string representing one or more Python statements\n",
      "    or a code object as returned by compile().\n",
      "    The globals must be a dictionary and locals can be any mapping,\n",
      "    defaulting to the current globals and locals.\n",
      "    If only globals is given, locals defaults to it.\n",
      "    The closure must be a tuple of cellvars, and can only be used\n",
      "    when source is a code object requiring exactly that many cellvars.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(exec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5fb4d131",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "733c330d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'__builtins__': {'__name__': 'builtins',\n",
       "  '__doc__': \"Built-in functions, exceptions, and other objects.\\n\\nNoteworthy: None is the `nil' object; Ellipsis represents `...' in slices.\",\n",
       "  '__package__': '',\n",
       "  '__loader__': _frozen_importlib.BuiltinImporter,\n",
       "  '__spec__': ModuleSpec(name='builtins', loader=<class '_frozen_importlib.BuiltinImporter'>, origin='built-in'),\n",
       "  '__build_class__': <function __build_class__>,\n",
       "  '__import__': <function __import__(name, globals=None, locals=None, fromlist=(), level=0)>,\n",
       "  'abs': <function abs(x, /)>,\n",
       "  'all': <function all(iterable, /)>,\n",
       "  'any': <function any(iterable, /)>,\n",
       "  'ascii': <function ascii(obj, /)>,\n",
       "  'bin': <function bin(number, /)>,\n",
       "  'breakpoint': <function breakpoint>,\n",
       "  'callable': <function callable(obj, /)>,\n",
       "  'chr': <function chr(i, /)>,\n",
       "  'compile': <function compile(source, filename, mode, flags=0, dont_inherit=False, optimize=-1, *, _feature_version=-1)>,\n",
       "  'delattr': <function delattr(obj, name, /)>,\n",
       "  'dir': <function dir>,\n",
       "  'divmod': <function divmod(x, y, /)>,\n",
       "  'eval': <function eval(source, globals=None, locals=None, /)>,\n",
       "  'exec': <function exec(source, globals=None, locals=None, /, *, closure=None)>,\n",
       "  'format': <function format(value, format_spec='', /)>,\n",
       "  'getattr': <function getattr>,\n",
       "  'globals': <function globals()>,\n",
       "  'hasattr': <function hasattr(obj, name, /)>,\n",
       "  'hash': <function hash(obj, /)>,\n",
       "  'hex': <function hex(number, /)>,\n",
       "  'id': <function id(obj, /)>,\n",
       "  'input': <bound method Kernel.raw_input of <ipykernel.ipkernel.IPythonKernel object at 0x105751550>>,\n",
       "  'isinstance': <function isinstance(obj, class_or_tuple, /)>,\n",
       "  'issubclass': <function issubclass(cls, class_or_tuple, /)>,\n",
       "  'iter': <function iter>,\n",
       "  'aiter': <function aiter(async_iterable, /)>,\n",
       "  'len': <function len(obj, /)>,\n",
       "  'locals': <function locals()>,\n",
       "  'max': <function max>,\n",
       "  'min': <function min>,\n",
       "  'next': <function next>,\n",
       "  'anext': <function anext>,\n",
       "  'oct': <function oct(number, /)>,\n",
       "  'ord': <function ord(c, /)>,\n",
       "  'pow': <function pow(base, exp, mod=None)>,\n",
       "  'print': <function print(*args, sep=' ', end='\\n', file=None, flush=False)>,\n",
       "  'repr': <function repr(obj, /)>,\n",
       "  'round': <function round(number, ndigits=None)>,\n",
       "  'setattr': <function setattr(obj, name, value, /)>,\n",
       "  'sorted': <function sorted(iterable, /, *, key=None, reverse=False)>,\n",
       "  'sum': <function sum(iterable, /, start=0)>,\n",
       "  'vars': <function vars>,\n",
       "  'None': None,\n",
       "  'Ellipsis': Ellipsis,\n",
       "  'NotImplemented': NotImplemented,\n",
       "  'False': False,\n",
       "  'True': True,\n",
       "  'bool': bool,\n",
       "  'memoryview': memoryview,\n",
       "  'bytearray': bytearray,\n",
       "  'bytes': bytes,\n",
       "  'classmethod': classmethod,\n",
       "  'complex': complex,\n",
       "  'dict': dict,\n",
       "  'enumerate': enumerate,\n",
       "  'filter': filter,\n",
       "  'float': float,\n",
       "  'frozenset': frozenset,\n",
       "  'property': property,\n",
       "  'int': int,\n",
       "  'list': list,\n",
       "  'map': map,\n",
       "  'object': object,\n",
       "  'range': range,\n",
       "  'reversed': reversed,\n",
       "  'set': set,\n",
       "  'slice': slice,\n",
       "  'staticmethod': staticmethod,\n",
       "  'str': str,\n",
       "  'super': super,\n",
       "  'tuple': tuple,\n",
       "  'type': type,\n",
       "  'zip': zip,\n",
       "  '__debug__': True,\n",
       "  'BaseException': BaseException,\n",
       "  'BaseExceptionGroup': BaseExceptionGroup,\n",
       "  'Exception': Exception,\n",
       "  'GeneratorExit': GeneratorExit,\n",
       "  'KeyboardInterrupt': KeyboardInterrupt,\n",
       "  'SystemExit': SystemExit,\n",
       "  'ArithmeticError': ArithmeticError,\n",
       "  'AssertionError': AssertionError,\n",
       "  'AttributeError': AttributeError,\n",
       "  'BufferError': BufferError,\n",
       "  'EOFError': EOFError,\n",
       "  'ImportError': ImportError,\n",
       "  'LookupError': LookupError,\n",
       "  'MemoryError': MemoryError,\n",
       "  'NameError': NameError,\n",
       "  'OSError': OSError,\n",
       "  'ReferenceError': ReferenceError,\n",
       "  'RuntimeError': RuntimeError,\n",
       "  'StopAsyncIteration': StopAsyncIteration,\n",
       "  'StopIteration': StopIteration,\n",
       "  'SyntaxError': SyntaxError,\n",
       "  'SystemError': SystemError,\n",
       "  'TypeError': TypeError,\n",
       "  'ValueError': ValueError,\n",
       "  'Warning': Warning,\n",
       "  'FloatingPointError': FloatingPointError,\n",
       "  'OverflowError': OverflowError,\n",
       "  'ZeroDivisionError': ZeroDivisionError,\n",
       "  'BytesWarning': BytesWarning,\n",
       "  'DeprecationWarning': DeprecationWarning,\n",
       "  'EncodingWarning': EncodingWarning,\n",
       "  'FutureWarning': FutureWarning,\n",
       "  'ImportWarning': ImportWarning,\n",
       "  'PendingDeprecationWarning': PendingDeprecationWarning,\n",
       "  'ResourceWarning': ResourceWarning,\n",
       "  'RuntimeWarning': RuntimeWarning,\n",
       "  'SyntaxWarning': SyntaxWarning,\n",
       "  'UnicodeWarning': UnicodeWarning,\n",
       "  'UserWarning': UserWarning,\n",
       "  'BlockingIOError': BlockingIOError,\n",
       "  'ChildProcessError': ChildProcessError,\n",
       "  'ConnectionError': ConnectionError,\n",
       "  'FileExistsError': FileExistsError,\n",
       "  'FileNotFoundError': FileNotFoundError,\n",
       "  'InterruptedError': InterruptedError,\n",
       "  'IsADirectoryError': IsADirectoryError,\n",
       "  'NotADirectoryError': NotADirectoryError,\n",
       "  'PermissionError': PermissionError,\n",
       "  'ProcessLookupError': ProcessLookupError,\n",
       "  'TimeoutError': TimeoutError,\n",
       "  'IndentationError': IndentationError,\n",
       "  'IndexError': IndexError,\n",
       "  'KeyError': KeyError,\n",
       "  'ModuleNotFoundError': ModuleNotFoundError,\n",
       "  'NotImplementedError': NotImplementedError,\n",
       "  'RecursionError': RecursionError,\n",
       "  'UnboundLocalError': UnboundLocalError,\n",
       "  'UnicodeError': UnicodeError,\n",
       "  'BrokenPipeError': BrokenPipeError,\n",
       "  'ConnectionAbortedError': ConnectionAbortedError,\n",
       "  'ConnectionRefusedError': ConnectionRefusedError,\n",
       "  'ConnectionResetError': ConnectionResetError,\n",
       "  'TabError': TabError,\n",
       "  'UnicodeDecodeError': UnicodeDecodeError,\n",
       "  'UnicodeEncodeError': UnicodeEncodeError,\n",
       "  'UnicodeTranslateError': UnicodeTranslateError,\n",
       "  'ExceptionGroup': ExceptionGroup,\n",
       "  'EnvironmentError': OSError,\n",
       "  'IOError': OSError,\n",
       "  'open': <function io.open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)>,\n",
       "  'copyright': Copyright (c) 2001-2022 Python Software Foundation.\n",
       "  All Rights Reserved.\n",
       "  \n",
       "  Copyright (c) 2000 BeOpen.com.\n",
       "  All Rights Reserved.\n",
       "  \n",
       "  Copyright (c) 1995-2001 Corporation for National Research Initiatives.\n",
       "  All Rights Reserved.\n",
       "  \n",
       "  Copyright (c) 1991-1995 Stichting Mathematisch Centrum, Amsterdam.\n",
       "  All Rights Reserved.,\n",
       "  'credits':     Thanks to CWI, CNRI, BeOpen.com, Zope Corporation and a cast of thousands\n",
       "      for supporting Python development.  See www.python.org for more information.,\n",
       "  'license': Type license() to see the full license text,\n",
       "  'help': Type help() for interactive help, or help(object) for help about object.,\n",
       "  'execfile': <function _pydev_bundle._pydev_execfile.execfile(file, glob=None, loc=None)>,\n",
       "  'runfile': <function _pydev_bundle.pydev_umd.runfile(filename, args=None, wdir=None, namespace=None)>,\n",
       "  '__IPYTHON__': True,\n",
       "  'display': <function IPython.core.display_functions.display(*objs, include=None, exclude=None, metadata=None, transient=None, display_id=None, raw=False, clear=False, **kwargs)>,\n",
       "  'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.zmqshell.ZMQInteractiveShell object at 0x105765a50>>},\n",
       " 'np': <module 'numpy' from '/Users/pmontgom/Library/Caches/pypoetry/virtualenvs/depmap-omics-jrgI6E8X-py3.11/lib/python3.11/site-packages/numpy/__init__.py'>,\n",
       " 'SAMPLESETNAME': '22Q4',\n",
       " 'RELEASE': '22q4',\n",
       " 'SHEETCREDS': '../../.credentials.json',\n",
       " 'MY_ID': '~/.client_secret.json',\n",
       " 'MYSTORAGE_ID': '~/.storage.json',\n",
       " 'GCS_PAYER_PROJECT': 'broad-firecloud-ccle',\n",
       " 'SAMPLEID': 'DepMap_ID',\n",
       " 'SOURCE_PRIORITY': ['BROAD',\n",
       "  'DEPMAP',\n",
       "  'IBM',\n",
       "  'CCLE2',\n",
       "  'SANGER',\n",
       "  'CHORDOMA',\n",
       "  'PRISM'],\n",
       " 'ENSEMBL_SERVER_V': 'http://nov2020.archive.ensembl.org/biomart',\n",
       " 'HG38BAMCOL': ['bam_filepath', 'bai_filepath'],\n",
       " 'LEGACY_BAM_COLNAMES': ['hg19_bam_filepath', 'hg19_bai_filepath'],\n",
       " 'HG38_CRAM_COLNAMES': ['hg38_cram_filepath', 'hg38_crai_filepath'],\n",
       " 'TORAISE': (['ACH-001195'],),\n",
       " 'TO_UPDATE': {'primary_disease': ['Primary Disease'],\n",
       "  'sex': ['CCLF Gender'],\n",
       "  'collection_site': ['Sample Collection Site'],\n",
       "  'subtype': ['lineage_subtype'],\n",
       "  'subsubtype': ['lineage_sub_subtype'],\n",
       "  'lineage': ['lineage'],\n",
       "  'parent_cell_line': ['Parental ID'],\n",
       "  'comments': ['Comments'],\n",
       "  'stripped_cell_line_name': ['Stripped Cell Line Name'],\n",
       "  'cellosaurus_id': ['RRID'],\n",
       "  'age': ['CCLF Age']},\n",
       " 'MAXAGE': '2021-01-01',\n",
       " 'EXTRACT_TO_CHANGE': {'from_arxspan_id': 'participant'},\n",
       " 'REPLACE': {'T': 'Tumor', 'N': 'Normal', 'm': 'Unknown', 'L': 'Unknown'},\n",
       " 'values': (['legacy_bam_filepath', 'legacy_bai_filepath'],),\n",
       " 'filetypes': ['bam', 'bai'],\n",
       " 'MATCH': ['ACH-', 'CDS-'],\n",
       " 'EXTRACT_DEFAULTS': {'name': 'sample_alias',\n",
       "  'bai': 'crai_or_bai_path',\n",
       "  'bam': 'cram_or_bam_path',\n",
       "  'cram': 'cram_path',\n",
       "  'crai': 'crai_path',\n",
       "  'ref_bam': 'hg19_bam_filepath',\n",
       "  'ref_type': 'Datatype',\n",
       "  'ref_bai': 'hg19_bai_filepath',\n",
       "  'ref_cram': 'hg38_cram_filepath',\n",
       "  'ref_crai': 'hg38_crai_filepath',\n",
       "  'version': 'version',\n",
       "  'primary_disease': 'primary_disease',\n",
       "  'ref_arxspan_id': 'arxspan_id',\n",
       "  'ref_name': 'stripped_cell_line_name',\n",
       "  'source': 'source',\n",
       "  'size': 'size',\n",
       "  'legacy_size': 'legacy_size',\n",
       "  'from_arxspan_id': 'individual_alias',\n",
       "  'ref_id': 'sample_id',\n",
       "  'PDO_id_terra': 'PDO',\n",
       "  'PDO_id_gumbo': 'PdoId',\n",
       "  'update_time': 'update_time',\n",
       "  'from_patient_id': 'individual_alias',\n",
       "  'patient_id': 'participant_id',\n",
       "  'ref_date': 'date_sequenced',\n",
       "  'hs_hs_library_size': 'hs_hs_library_size',\n",
       "  'hs_het_snp_sensitivity': 'hs_het_snp_sensitivity',\n",
       "  'hs_mean_bait_coverage': 'hs_mean_bait_coverage',\n",
       "  'hs_mean_target_coverage': 'hs_mean_target_coverage',\n",
       "  'hs_on_target_bases': 'hs_on_target_bases',\n",
       "  'total_reads': 'total_reads',\n",
       "  'release_date': 'sequencing_date',\n",
       "  'hash': 'crc32c_hash',\n",
       "  'legacy_hash': 'legacy_crc32c_hash',\n",
       "  'mean_depth': 'mean_depth',\n",
       "  'root_sample_id': 'root_sample_id',\n",
       "  'sm_id': 'SmId',\n",
       "  'profile_id': 'ProfileID',\n",
       "  'expected_type': 'expected_type'},\n",
       " 'MINSIZES_BAM': {'rna': 2000000000, 'wes': 3000000000, 'wgs': 50000000000},\n",
       " 'MINSIZES_CRAM': {'wgs': 10000000000},\n",
       " 'DUP_ARXSPANS': {'ACH-001620': 'ACH-001605', 'ACH-001621': 'ACH-001606'},\n",
       " 'rnaworkspace6': 'terra-broad-cancer-prod/CCLE_DepMap_RNAseq',\n",
       " 'rnaworkspace7': 'terra-broad-cancer-prod/Getz_IBM_CellLines_RNASeqData',\n",
       " 'rnasource6': 'DEPMAP',\n",
       " 'rnasource7': 'IBM',\n",
       " 'RNAWORKSPACES': [('DEPMAP',\n",
       "   'terra-broad-cancer-prod/CCLE_DepMap_RNAseq',\n",
       "   'bam'),\n",
       "  ('IBM', 'terra-broad-cancer-prod/Getz_IBM_CellLines_RNASeqData', 'bam')],\n",
       " 'wgsworkspace1': 'terra-broad-cancer-prod/DepMap_WGS',\n",
       " 'wgsworkspace2': 'terra-broad-cancer-prod/Getz_IBM_CellLines_WGS',\n",
       " 'wgssource1': 'DEPMAP',\n",
       " 'wgssource2': 'IBM',\n",
       " 'WGSWORKSPACES': [('DEPMAP', 'terra-broad-cancer-prod/DepMap_WGS', 'bam'),\n",
       "  ('IBM', 'terra-broad-cancer-prod/Getz_IBM_CellLines_WGS', 'bam'),\n",
       "  ('DEPMAP', 'broad-genomics-data/DepMap_WGS', 'cram')],\n",
       " 'WGSSETENTITY': 'sample_set',\n",
       " 'WESSETENTITY': 'pair_set',\n",
       " 'FPALLBATCHPAIRSETS': 'all',\n",
       " 'WORKING_DIR': 'output/',\n",
       " 'PREV_VIRTUAL': {'public': 'public-22q1-305b',\n",
       "  'ibm': 'ibm-22q1-cce1',\n",
       "  'dmc': 'dmc-22q1-d00a',\n",
       "  'internal': 'internal-22q1-1778'},\n",
       " 'DATASETS': ['internal', 'dmc', 'public'],\n",
       " 'RUN_NOTEBOOKS': ['WGS_CCLE.ipynb', 'RNA_CCLE.ipynb'],\n",
       " 'PROCQC': ['allelic_counts_tumor',\n",
       "  'delta_MAD_tumor',\n",
       "  'denoised_MAD_tumor',\n",
       "  'scaled_delta_MAD_tumor',\n",
       "  'denoised_copy_ratios_lim_4_plot_tumor',\n",
       "  'denoised_copy_ratios_plot_tumor',\n",
       "  'modeled_segments_plot_tumor',\n",
       "  'gatk_cnv_all_plots',\n",
       "  'lego_plotter_pngs',\n",
       "  'copy_number_qc_report',\n",
       "  'ffpe_OBF_figures',\n",
       "  'mut_legos_html',\n",
       "  'oxoG_OBF_figures',\n",
       "  'tumor_bam_base_distribution_by_cycle_metrics',\n",
       "  'tumor_bam_converted_oxog_metrics'],\n",
       " 'BAMQC': ['duplication_metrics',\n",
       "  'bqsr_report',\n",
       "  'tumor_bam_alignment_summary_metrics',\n",
       "  'tumor_bam_bait_bias_summary_metrics',\n",
       "  'tumor_bam_gc_bias_summary_metrics',\n",
       "  'tumor_bam_hybrid_selection_metrics',\n",
       "  'tumor_bam_insert_size_histogram',\n",
       "  'tumor_bam_insert_size_metrics',\n",
       "  'tumor_bam_pre_adapter_summary_metrics',\n",
       "  'tumor_bam_quality_by_cycle_metrics',\n",
       "  'tumor_bam_quality_distribution_metrics',\n",
       "  'tumor_bam_quality_yield_metrics'],\n",
       " 'VCFDIR': '/tmp/vcfs/',\n",
       " 'GUIDESBED': {'avana': 'data/avana_guides.bed',\n",
       "  'humagne': 'data/humagne_guides.bed',\n",
       "  'ky': 'data/ky_score_guides.bed'},\n",
       " 'VCFCOLNAME': 'mutect2_filtered_vcf',\n",
       " 'GENECHANGETHR': 0.025,\n",
       " 'SEGMENTSTHR': 1500,\n",
       " 'MAXYCHROM': 150,\n",
       " 'COLRENAMING': {'CONTIG': 'Chromosome',\n",
       "  'START': 'Start',\n",
       "  'END': 'End',\n",
       "  'end': 'End',\n",
       "  'seqnames': 'Chromosome',\n",
       "  'start': 'Start',\n",
       "  'Sample': 'DepMap_ID',\n",
       "  'NUM_POINTS_COPY_RATIO': 'NumProbes',\n",
       "  'MEAN_LOG2_COPY_RATIO': 'SegmentMean',\n",
       "  'CALL': 'Status'},\n",
       " 'PURECN_MIN_GOF': 70,\n",
       " 'PURECN_MAX_PLOIDY': 5,\n",
       " 'PURECN_COLRENAMING': {'start': 'Start',\n",
       "  'end': 'End',\n",
       "  'chr': 'Chromosome',\n",
       "  'Sampleid': 'DepMap_ID',\n",
       "  'type': 'LoHStatus',\n",
       "  'C': 'MajorAlleleAbsoluteCN',\n",
       "  'M': 'MinorAlleleAbsoluteCN'},\n",
       " 'PURECN_LOH_COLNAME': 'PureCN_loh_merged',\n",
       " 'PURECN_FAILED_COLNAME': 'PureCN_failed',\n",
       " 'PURECN_SAMPLESET': 'PureCN',\n",
       " 'PURECN_LOHVALUES': ['LOH',\n",
       "  'COPY-NEUTRAL LOH',\n",
       "  'WHOLE ARM COPY-NEUTRAL LOH',\n",
       "  'WHOLE ARM LOH'],\n",
       " 'SIGTABLE_TERRACOLS': {'PureCN_cin_allele_specific_ploidy_robust',\n",
       "  'PureCN_loh_fraction',\n",
       "  'PureCN_ploidy',\n",
       "  'PureCN_wgd'},\n",
       " 'MISC_SIG_TERRACOLS': {'msisensor2_score'},\n",
       " 'SIGTABLE_BINARYCOLS': ['PureCN_wgd'],\n",
       " 'SIGTABLE_RENAMING': {'PureCN_loh_fraction': 'LoHFraction',\n",
       "  'PureCN_wgd': 'WGD',\n",
       "  'PureCN_cin_allele_specific_ploidy_robust': 'CIN',\n",
       "  'PureCN_ploidy': 'Ploidy',\n",
       "  'msisensor2_score': 'MSIScore'},\n",
       " 'SOURCE_RENAME': {'CCLF': 'Broad',\n",
       "  'CHORDOMA': 'Chordoma',\n",
       "  'SANGER': 'Sanger',\n",
       "  'IBM': 'Broad',\n",
       "  'CCLE2': 'Broad',\n",
       "  nan: 'Broad',\n",
       "  'DEPMAP': 'Broad',\n",
       "  'IBM WES': 'Broad WES',\n",
       "  'Broad CCLF': 'Broad WES'},\n",
       " 'MINFREQTOCALL': 0.25,\n",
       " 'SV_COLNAME': 'somatic_annotated_sv',\n",
       " 'SV_FILENAME': 'all_sv.csv',\n",
       " 'SV_COLRENAME': {'CHROM_A': 'ChromA',\n",
       "  'START_A': 'StartA',\n",
       "  'END_A': 'EndA',\n",
       "  'CHROM_B': 'ChromB',\n",
       "  'START_B': 'StartB',\n",
       "  'END_B': 'EndB',\n",
       "  'ID': 'ID',\n",
       "  'QUAL': 'Qual',\n",
       "  'STRAND_A': 'StrandA',\n",
       "  'STRAND_B': 'StrandB',\n",
       "  'TYPE': 'Type',\n",
       "  'FILTER': 'Filter',\n",
       "  'NAME_A': 'NameA',\n",
       "  'REF_A': 'RefA',\n",
       "  'ALT_A': 'AltA',\n",
       "  'NAME_B': 'NameB',\n",
       "  'REF_B': 'RefB',\n",
       "  'ALT_B': 'AltB',\n",
       "  'INFO_A': 'InfoA',\n",
       "  'INFO_B': 'InfoB',\n",
       "  'FORMAT': 'Format',\n",
       "  'SPAN': 'Span',\n",
       "  'HOMSEQ': 'HomSeq',\n",
       "  'HOMLEN': 'HomLen',\n",
       "  'BREAK_A_Ensembl_Gene': 'BreakAEnsemblGene',\n",
       "  'BREAK_A_Gene_Name': 'BreakAGeneName',\n",
       "  'BREAK_B_Ensembl_Gene': 'BreakBEnsemblGene',\n",
       "  'BREAK_B_Gene_Name': 'BreakBGeneName',\n",
       "  'BREAK_A_Ensembl_Exon': 'BreakAEnsemblExon',\n",
       "  'BREAK_B_Ensembl_Exon': 'BreakBEnsemblExon',\n",
       "  'Filter': 'Filter',\n",
       "  'SAMPLE': 'Sample',\n",
       "  'OCILY12': 'Sample'},\n",
       " 'MUTATION_GROUPS': {'other conserving': [\"5'Flank\",\n",
       "   'Intron',\n",
       "   'IGR',\n",
       "   \"3'UTR\",\n",
       "   \"5'UTR\"],\n",
       "  'other non-conserving': ['In_Frame_Del',\n",
       "   'In_Frame_Ins',\n",
       "   'Stop_Codon_Del',\n",
       "   'Stop_Codon_Ins',\n",
       "   'Missense_Mutation',\n",
       "   'Nonstop_Mutation'],\n",
       "  'silent': ['Silent'],\n",
       "  'damaging': ['De_novo_Start_OutOfFrame',\n",
       "   'Frame_Shift_Del',\n",
       "   'Frame_Shift_Ins',\n",
       "   'Splice_Site',\n",
       "   'Start_Codon_Del',\n",
       "   'Start_Codon_Ins',\n",
       "   'Start_Codon_SNP',\n",
       "   'Nonsense_Mutation']},\n",
       " 'MAF_COL': 'depmap_maf',\n",
       " 'HGNC_MAPPING': 'data/new_and_old_hgnc_symbols.txt',\n",
       " 'DNA_CHANGE_COL': 'DNAChange',\n",
       " 'CHROM_COL': 'Chrom',\n",
       " 'POS_COL': 'Pos',\n",
       " 'HOTSPOT_COL': 'CosmicHotspot',\n",
       " 'IMMORTALIZED_THR': 0.05,\n",
       " 'HUGO_COL': 'HugoSymbol',\n",
       " 'LIKELY_LOF_COL': 'LikelyLoF',\n",
       " 'CCLE_DELETERIOUS_COL': 'CCLEDeleterious',\n",
       " 'CIVIC_SCORE_COL': 'CivicScore',\n",
       " 'HESS_COL': 'HessDriver',\n",
       " 'MUTCOL_DEPMAP': {'chrom': 'Chrom',\n",
       "  'pos': 'Pos',\n",
       "  'ref': 'Ref',\n",
       "  'alt': 'Alt',\n",
       "  'af': 'AF',\n",
       "  'ref_count': 'RefCount',\n",
       "  'alt_count': 'AltCount',\n",
       "  'gt': 'GT',\n",
       "  'ps': 'PS',\n",
       "  'variant_type': 'VariantType',\n",
       "  'variant_info': 'VariantInfo',\n",
       "  'dna_change': 'DNAChange',\n",
       "  'protein_change': 'ProteinChange',\n",
       "  'hugo_symbol': 'HugoSymbol',\n",
       "  'hgnc_name': 'HgncName',\n",
       "  'hgnc_family': 'HgncFamily',\n",
       "  'transcript': 'Transcript',\n",
       "  'transcript_exon': 'TranscriptExon',\n",
       "  'transcript_strand': 'TranscriptStrand',\n",
       "  'uniprot_id': 'UniprotID',\n",
       "  'str': 'Str',\n",
       "  'dbsnp_id': 'DbsnpID',\n",
       "  'dbsnp_filter': 'DbsnpFilter',\n",
       "  'issues': 'Issues',\n",
       "  'gc_content': 'GcContent',\n",
       "  'lineage_association': 'LineageAssociation',\n",
       "  'cancer_molecular_genetics': 'CancerMolecularGenetics',\n",
       "  'ccle_deleterious': 'CCLEDeleterious',\n",
       "  'structural_relation': 'StructuralRelation',\n",
       "  'cosmic_hotspot': 'CosmicHotspot',\n",
       "  'cosmic_overlapping_mutations': 'CosmicOverlappingMutations',\n",
       "  'associated_with': 'AssociatedWith',\n",
       "  'lof': 'LoF',\n",
       "  'driver': 'Driver',\n",
       "  'likely_driver': 'LikelyDriver',\n",
       "  'transcript_likely_lof': 'TranscriptLikelyLoF',\n",
       "  'civic_id': 'CivicID',\n",
       "  'civic_description': 'CivicDescription',\n",
       "  'civic_score': 'CivicScore',\n",
       "  'popaf': 'Popaf',\n",
       "  'likely_gof': 'LikelyGoF',\n",
       "  'likely_lof': 'LikelyLoF',\n",
       "  'hess_driver': 'HessDriver',\n",
       "  'hess_signture': 'HessSignature',\n",
       "  'hess_signature': 'HessSignature',\n",
       "  'cscape_score': 'CscapeScore',\n",
       "  'dann_score': 'DannScore',\n",
       "  'revel_score': 'RevelScore',\n",
       "  'funseq2_score': 'Funseq2Score',\n",
       "  'pharmgkb_id': 'PharmgkbID',\n",
       "  'dida_id': 'DidaID',\n",
       "  'dida_name': 'DidaName',\n",
       "  'gwas_disease': 'GwasDisease',\n",
       "  'gwas_pmid': 'GwasPmID',\n",
       "  'gtex_gene': 'GTexGene',\n",
       "  'DepMap_ID': 'DepMap_ID'},\n",
       " 'FUSION_COLNAME': ['FusionName',\n",
       "  'JunctionReadCount',\n",
       "  'SpanningFragCount',\n",
       "  'SpliceType',\n",
       "  'LeftGene',\n",
       "  'LeftBreakpoint',\n",
       "  'RightGene',\n",
       "  'RightBreakpoint',\n",
       "  'LargeAnchorSupport',\n",
       "  'FFPM',\n",
       "  'LeftBreakDinuc',\n",
       "  'LeftBreakEntropy',\n",
       "  'RightBreakDinuc',\n",
       "  'RightBreakEntropy',\n",
       "  'annots'],\n",
       " 'FUSION_RED_HERRING': ['GTEx_recurrent',\n",
       "  'DGD_PARALOGS',\n",
       "  'HGNC_GENEFAM',\n",
       "  'Greger_Normal',\n",
       "  'Babiceanu_Normal',\n",
       "  'ConjoinG',\n",
       "  'NEIGHBORS'],\n",
       " 'FUSION_MAXFREQ': 0.1,\n",
       " 'FUSION_MINFFPM': 0.05,\n",
       " 'FUSION_MAXFFPM': 0.1,\n",
       " 'STARBAMCOLTERRA': ['internal_bam_filepath', 'internal_bai_filepath'],\n",
       " 'RSEM_TRANSCRIPTS': ['rsem_transcripts_expected_count',\n",
       "  'rsem_transcripts_tpm'],\n",
       " 'RSEMFILENAME_GENE': ['genes_tpm', 'genes_expected_count'],\n",
       " 'PROTEINEFILENAMES': ['proteincoding_genes_tpm',\n",
       "  'proteincoding_genes_expected_count'],\n",
       " 'RSEMFILENAME_TRANSCRIPTS': ['transcripts_tpm', 'transcripts_expected_count'],\n",
       " 'RSEMFILENAME': ['genes_tpm',\n",
       "  'genes_expected_count',\n",
       "  'transcripts_tpm',\n",
       "  'transcripts_expected_count'],\n",
       " 'SSGSEAFILEPATH': 'data/genesets/msigdb.v7.2.symbols.gmt',\n",
       " 'RNAMINSIMI': 0.95,\n",
       " 'RNASEQC_THRESHOLDS_LOWQUAL': {'minmapping': 0.85,\n",
       "  'minendmapping': 0.75,\n",
       "  'minefficiency': 0.75,\n",
       "  'maxendmismatch': 0.02,\n",
       "  'maxmismatch': 0.02,\n",
       "  'minhighqual': 0.8,\n",
       "  'minexon': 0.7,\n",
       "  'maxambiguous': 0.05,\n",
       "  'maxsplits': 0.1,\n",
       "  'maxalt': 0.2,\n",
       "  'maxchim': 0.05,\n",
       "  'minreads': 20000000,\n",
       "  'minlength': 80,\n",
       "  'maxgenes': 35000,\n",
       "  'mingenes': 12000},\n",
       " 'RNASEQC_THRESHOLDS_FAILED': {'minmapping': 0.7,\n",
       "  'minendmapping': 0.66,\n",
       "  'minefficiency': 0.6,\n",
       "  'maxendmismatch': 0.02,\n",
       "  'maxmismatch': 0.02,\n",
       "  'minhighqual': 0.7,\n",
       "  'minexon': 0.66,\n",
       "  'maxambiguous': 0.1,\n",
       "  'maxsplits': 0.1,\n",
       "  'maxalt': 0.5,\n",
       "  'maxchim': 0.2,\n",
       "  'minreads': 20000000,\n",
       "  'minlength': 80,\n",
       "  'maxgenes': 35000,\n",
       "  'mingenes': 10000},\n",
       " 'README_folder': '../depmap-release-readmes/',\n",
       " 'README_currentfolder': '../depmap-release-readmes/release-22q4/',\n",
       " 'README_changes': '\\n\\n\\n',\n",
       " 'SKIP_UPLOADING_README': True,\n",
       " 'FUSIONreadme': \"\\n# Fusions\\n\\nPORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal.\\n\\n/!\\\\ This is the most up to date version of the CCLE CN data.\\n\\n## Annotations\\n\\nDescription: Gene fusions derived from RNAseq data.\\n\\nRows: cell lines, IDs contained in the column DepMap_ID\\n\\nUnfiltered data contains all output fusions, while the filtered data uses the filters suggested by the star fusion docs. These filters are:\\n- FFPM > 0.1 -  a cutoff of 0.1 means&nbsp;at least 1 fusion-supporting RNAseq fragment per 10M total reads\\n- Remove known false positives, such as GTEx recurrent fusions and certain paralogs\\n- Genes that are next to each other\\n- Fusions with mitochondrial breakpoints\\n- Removing fusion involving mitochondrial chromosomes or HLA genes\\n- Removed common false positive fusions (red herring annotations as described in the STAR-Fusion docs)\\n- Recurrent fusions observed in CCLE across cell lines (in more than 10% of our samples)\\n- Removed fusions where SpliceType='INCL_NON_REF_SPLICE' and LargeAnchorSupport='NO_LDAS' and FFPM < 0.1\\n- FFPM < 0.05\\n\",\n",
       " 'RNAseqreadme': '\\n# RNAseq\\n\\nPORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal.\\n\\n/!\\\\ This is the most up to date version of the CCLE RNA data.\\n\\n## Annotations:\\n\\ntranscriptions (Transcripts rpkm):\\n\\ngenes (gene rpkm):\\n__Rows__:\\n__Columns__:\\nCounts (gene counts):\\n__Rows__:\\n__Columns__:\\nGene level CN data:\\n__Rows__:\\n__Columns__:\\n DepMap cell line IDs\\n gene names in the format HGNC\\\\_symbol (Entrez\\\\_ID)\\nDepMap\\\\_ID, Chromosome, Start, End, Num\\\\_Probes, Segment\\\\_Mean\\n',\n",
       " 'CNreadme': '\\n# Copy Number\\n\\nPORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal.\\n\\n/!\\\\ This is the most up to date version of the CCLE CN data.\\n\\n# Notations:\\n\\nall: everything\\n\\nallWES: all data comes from the WExomeS samples we posses\\n\\nallWGS: all data comes from the WGenomeS samples we posses\\n\\nwithreplicates: if we have two different sequencing from a sample, we kept both, see the depmap sample tracker for annotations [https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE](https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE). this dataset is more geared toward QC or in-depth analysis of a particular cell line.\\n\\nmerged: everything from both WGS and WES\\n\\nlatest: only the latest sequencing versions of the samples were kept\\n\\n\\nGene level CN data:\\n\\n__Rows__: cell line IDs\\n\\n__Columns__: gene names in the format HGNC\\\\_symbol (Entrez\\\\_ID)\\n\\nSegment level data:\\n\\n__Columns__: DepMap\\\\_ID, Chromosome, Start, End, Segment\\\\_Mean, Num\\\\_Probes, Calls',\n",
       " 'Achillesreadme': '\\n# Copy Number\\n\\nCombined segment and gene-level CN calls from Broad WES, Sanger WES, and Broad SNP. Relative CN, log2(x+1) transformed.\\n\\nPORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal. Must use subsetted dataset instead. These data will not make it on the portal starting 19Q1. With the DMC portal, there is new cell line release prioritization as to which lines can be included, so a new taiga dataset will be created containing CN for the portal.\\n\\nThese data are generated for Achilles to pull from to run CERES.\\n\\n\\nGene level CN data:\\n\\n__Rows__: DepMap cell line IDs\\n\\n__Columns__: gene names in the format HGNC\\\\_symbol (Entrez\\\\_ID)\\n\\nSegment level data:\\n\\n__Columns__: DepMap\\\\_ID, Chromosome, Start, End, Num\\\\_Probes, Segment\\\\_Mean',\n",
       " 'Mutationsreadme': '\\n# Mutations\\n\\nPORTAL TEAM SHOULD NOT USE THIS: There are lines here that should not make it even to internal.\\n\\n/!\\\\ This is the most up to date version of the CCLE Mutatios data.\\nThe data is most likely of a better quality that what is on other folder. It is however in beta version as not all changes have either been confirmed or accepted by the DepMap Ops and the DepMap Portal Team.\\n\\n# Notations:\\n\\nall: every cell lines we have\\n\\nWES: all data comes from the WExomeS samples we posses\\n\\nWGS: all data comes from the WGenomeS samples we posses\\n\\nwithreplicates: if we have two different sequencing from a sample, we kept both, see the depmap sample tracker for annotations [https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE](https://docs.google.com/spreadsheets/d/1XkZypRuOEXzNLxVk9EOHeWRE98Z8_DBvL4PovyM01FE). this dataset is more geared toward QC or in-depth analysis of a particular cell line.\\n\\nmerged: everything from both WGS and WES\\n\\nlatest: only the latest sequencing versions of the samples were kept\\n\\ngenes (gene rpkm):\\n__Rows__:\\n__Columns__:\\nCounts (gene counts):\\n__Rows__:\\n__Columns__:\\nGene level CN data:\\n__Rows__:\\n__Columns__:\\n DepMap cell line IDs\\n gene names in the format HGNC\\\\_symbol (Entrez\\\\_ID)\\nDepMap\\\\_ID, Chromosome, Start, End, Num\\\\_Probes, Segment\\\\_Mean\\n ',\n",
       " 'GUMBO_CLIENT_USERNAME': 'szhang',\n",
       " 'DATE_COL_DICT': {'internal': 'InternalReleaseDate',\n",
       "  'dmc': 'ConsortiumReleaseDate',\n",
       "  'public': 'PublicReleaseDate'},\n",
       " 'MODEL_TABLE_NAME': 'model',\n",
       " 'MODEL_TABLE_INDEX': 'ModelID',\n",
       " 'MC_TABLE_NAME': 'model_condition',\n",
       " 'MC_TABLE_INDEX': 'ModelConditionID',\n",
       " 'PR_TABLE_NAME': 'omics_profile',\n",
       " 'PR_TABLE_INDEX': 'ProfileID',\n",
       " 'SEQ_TABLE_NAME': 'omics_sequencing',\n",
       " 'SEQ_TABLE_INDEX': 'SequencingID',\n",
       " 'SAMPLE_TABLE_NAME': 'sample',\n",
       " 'ACH_CHOICE_TABLE_COLS': ['ModelConditionID', 'ProfileID', 'ProfileType'],\n",
       " 'ACH_CHOICE_TABLE_NAME': 'achilles_choice_table',\n",
       " 'DEFAULT_TABLE_COLS': ['ModelID', 'ProfileID', 'ProfileType'],\n",
       " 'DEFAULT_TABLE_NAME': 'default_model_table',\n",
       " 'PROFILE_TABLE_COLS': ['ModelCondition', 'ModelID', 'Datatype'],\n",
       " 'RELEASE_PR_TABLE_NAME': 'OmicsProfiles',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_MODEL': {'proteinCoding_genes_tpm_logp1_profile': 'OmicsExpressionProteinCodingGenesTPMLogp1',\n",
       "  'gene_set_enrichment_profile': 'OmicsExpressionGeneSetEnrichment'},\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_MODEL': {'merged_gene_cn_profile': 'OmicsCNGene',\n",
       "  'merged_absolute_gene_cn_profile': 'OmicsAbsoluteCNGene',\n",
       "  'merged_loh_profile': 'OmicsLoH',\n",
       "  'globalGenomicFeatures_profile': 'OmicsSignatures'},\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_MODEL': {'somaticMutations_genotypedMatrix_hotspot_profile': 'OmicsSomaticMutationsMatrixHotspot',\n",
       "  'somaticMutations_genotypedMatrix_damaging_profile': 'OmicsSomaticMutationsMatrixDamaging',\n",
       "  'somaticMutations_genotypedMatrix_driver_profile': 'OmicsSomaticMutationsMatrixDriver'},\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_MODEL': {'binary_germline_mutation': 'OmicsGuideMutationsBinaryAvana'},\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_MODEL': {'fusions_filtered_profile': 'OmicsFusionFiltered'},\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_MODEL': {},\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_MODEL': {'somaticMutations_profile': 'OmicsSomaticMutations',\n",
       "  'structuralVariants_profile': 'OmicsStructuralVariants'},\n",
       " 'VIRTUAL_FILENAMES_GUIDEMUT': {'binary_mutation_avana': 'OmicsGuideMutationsBinaryAvana',\n",
       "  'binary_mutation_ky': 'OmicsGuideMutationsBinaryKY',\n",
       "  'binary_mutation_humagne': 'OmicsGuideMutationsBinaryHumagne'},\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_PR': {'genes_expectedCount_profile': 'OmicsExpressionGenesExpectedCountProfile',\n",
       "  'transcripts_expectedCount_profile': 'OmicsExpressionTranscriptsExpectedCountProfile',\n",
       "  'gene_set_enrichment_profile': 'OmicsExpressionGeneSetEnrichmentProfile'},\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_PR': {'globalGenomicFeatures_profile': 'OmicsSignaturesProfile'},\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_PR': {},\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_PR': {},\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_PR': {'fusions_unfiltered_profile': 'OmicsFusionUnfilteredProfile'},\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_PR': {'merged_segments_profile': 'OmicsCNSegmentsProfile',\n",
       "  'merged_absolute_segments_profile': 'OmicsAbsoluteCNSegmentsProfile'},\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_PR': {'somaticMutations_profile': 'OmicsSomaticMutationsProfile',\n",
       "  'structuralVariants_profile': 'OmicsStructuralVariantsProfile'},\n",
       " 'SHEETNAME': 'ccle sample tracker sandbox',\n",
       " 'isCCLE': True,\n",
       " 'doCleanup': True,\n",
       " 'BAM_GCS_BUCKET': 'gs://cclebams-sandbox',\n",
       " 'RNA_GCS_PATH_HG38': 'gs://cclebams-sandbox/rnasq_hg38/',\n",
       " 'RNA_GCS_PATH': 'gs://cclebams-sandbox/rna/',\n",
       " 'WGS_GCS_PATH': 'gs://cclebams-sandbox/wgs/',\n",
       " 'WGS_GCS_PATH_HG38': 'gs://cclebams-sandbox/wgs_hg38/',\n",
       " 'WES_GCS_PATH': 'gs://cclebams-sandbox/wes/',\n",
       " 'REFSHEET_URL': 'https://docs.google.com/spreadsheets/d/161mmmHF5nc3nFhgpp_7zub5Erod8xDbXnNZ-u-7BWKI',\n",
       " 'DEPMAP_PV': 'https://docs.google.com/spreadsheets/d/1uqCOos-T9EMQU7y2ZUw4Nm84opU5fIT1y7jet1vnScE',\n",
       " 'POTENTIAL_LIST': 'https://docs.google.com/spreadsheets/d/1YuKEgZ1pFKRYzydvncQt9Y_BKToPlHP-oDB-0CAv3gE',\n",
       " 'SAMPLES_FOUND_NAME': 'depmap ALL samples found',\n",
       " 'SAMPLES_NOT_FOUND_NAME': 'depmap samples not found',\n",
       " 'SAMPLES_NOT_FOUND_URL': 'https://docs.google.com/spreadsheets/d/1yC3brpov3JELvzNoQe3eh0W196tfXzvpa0jUezMAxIg',\n",
       " 'SAMPLES_MISSING_ARXSPAN_NAME': 'depmap samples missing arxspan',\n",
       " 'SAMPLES_MISSING_ARXSPAN_URL': 'https://docs.google.com/spreadsheets/d/1htfgpXrMvXDlqbcZltpq6vOE_Vo2YZ3-3mdaXV-Irzk',\n",
       " 'TAIGA_ETERNAL': 'depmap-a0ab',\n",
       " 'TAIGA_ETERNAL_UPLOAD': 'eternal-74b2',\n",
       " 'DEPMAP_TAIGA': 'arxspan-cell-line-export-f808',\n",
       " 'TAIGA_MUTATION': 'mutations-latest-f263',\n",
       " 'TAIGA_CN': 'cn-latest-8bea',\n",
       " 'TAIGA_CN_ACHILLES': 'cn-achilles-version-43ea',\n",
       " 'TAIGA_EXPRESSION': 'expression-869e',\n",
       " 'TAIGA_FUSION': 'fusions-64c4',\n",
       " 'TAIGA_LEGACY_CN': 'copy-number-5f61',\n",
       " 'VIRTUAL_FOLDER': 'aee7ec053d434091a670bc64a9d7a3c1',\n",
       " 'VIRTUAL': {'internal': '', 'ibm': '', 'dmc': '', 'public': ''},\n",
       " 'RNAWORKSPACE': 'broad-firecloud-ccle/DEV_DepMap_hg38_RNAseq',\n",
       " 'WGSWORKSPACE': 'broad-firecloud-ccle/DEV_DepMap_WGS_CN',\n",
       " 'WESCNWORKSPACE': 'broad-firecloud-ccle/DepMap_WES_CN_hg38-sandbox',\n",
       " 'WESMUTWORKSPACE': 'broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline-sandbox',\n",
       " 'FPWORKSPACE': 'broad-firecloud-ccle/CCLE_SNP_QC-copy',\n",
       " 'TAIGA_FP': 'ccle-bam-fingerprints-4f4a',\n",
       " 'TAIGA_FP_FILENAME': 'fingerprint_lod_matrix',\n",
       " 'GUMBO_SHEET': 'https://docs.google.com/spreadsheets/d/10Lg0xkT5OHLYgJ9VKpkh8VR64TXfxPVJXRVAckU8uBg',\n",
       " 'GUMBO_SHEETNAME': 'Backfilled profile IDs',\n",
       " 'LATEST2FN_NUMMAT_MODEL': {'cn-latest-8bea': {'merged_gene_cn_profile': 'OmicsCNGene',\n",
       "   'merged_absolute_gene_cn_profile': 'OmicsAbsoluteCNGene',\n",
       "   'merged_loh_profile': 'OmicsLoH',\n",
       "   'globalGenomicFeatures_profile': 'OmicsSignatures'},\n",
       "  'expression-869e': {'proteinCoding_genes_tpm_logp1_profile': 'OmicsExpressionProteinCodingGenesTPMLogp1',\n",
       "   'gene_set_enrichment_profile': 'OmicsExpressionGeneSetEnrichment'},\n",
       "  'mutations-latest-f263': {'somaticMutations_genotypedMatrix_hotspot_profile': 'OmicsSomaticMutationsMatrixHotspot',\n",
       "   'somaticMutations_genotypedMatrix_damaging_profile': 'OmicsSomaticMutationsMatrixDamaging',\n",
       "   'somaticMutations_genotypedMatrix_driver_profile': 'OmicsSomaticMutationsMatrixDriver'}},\n",
       " 'LATEST2FN_TABLE_MODEL': {'cn-latest-8bea': {},\n",
       "  'fusions-64c4': {'fusions_filtered_profile': 'OmicsFusionFiltered'},\n",
       "  'mutations-latest-f263': {'somaticMutations_profile': 'OmicsSomaticMutations',\n",
       "   'structuralVariants_profile': 'OmicsStructuralVariants'}},\n",
       " 'LATEST2FN_NUMMAT_PR': {'cn-latest-8bea': {'globalGenomicFeatures_profile': 'OmicsSignaturesProfile'},\n",
       "  'expression-869e': {'genes_expectedCount_profile': 'OmicsExpressionGenesExpectedCountProfile',\n",
       "   'transcripts_expectedCount_profile': 'OmicsExpressionTranscriptsExpectedCountProfile',\n",
       "   'gene_set_enrichment_profile': 'OmicsExpressionGeneSetEnrichmentProfile'},\n",
       "  'mutations-latest-f263': {}},\n",
       " 'LATEST2FN_TABLE_PR': {'cn-latest-8bea': {'merged_segments_profile': 'OmicsCNSegmentsProfile',\n",
       "   'merged_absolute_segments_profile': 'OmicsAbsoluteCNSegmentsProfile'},\n",
       "  'fusions-64c4': {'fusions_unfiltered_profile': 'OmicsFusionUnfilteredProfile'},\n",
       "  'mutations-latest-f263': {'somaticMutations_profile': 'OmicsSomaticMutationsProfile',\n",
       "   'structuralVariants_profile': 'OmicsStructuralVariantsProfile'}}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "950c5b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def read_settings(fn):\n",
    "    content = open(fn).read()\n",
    "    globals = {}\n",
    "    locals = {}\n",
    "    exec(content, globals, locals)\n",
    "    return { k:v for k,v in locals.items() if re.match(\"[A-Z]+.*\", k) }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8c75e892",
   "metadata": {},
   "outputs": [],
   "source": [
    "dev = read_settings(\"config_dev.py\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6e7a2c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "shared = read_settings(\"config_global.py\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "380f81f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prod = read_settings(\"config_prod.py\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ed0e263a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GUMBO_SHEET',\n",
       " 'GUMBO_SHEETNAME',\n",
       " 'SAMPLES_FOUND_NAME',\n",
       " 'SAMPLES_MISSING_ARXSPAN_NAME',\n",
       " 'SAMPLES_MISSING_ARXSPAN_URL',\n",
       " 'SAMPLES_NOT_FOUND_NAME',\n",
       " 'SAMPLES_NOT_FOUND_URL'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(dev.keys()).difference(prod.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8cb217c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(prod.keys()).difference(dev.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4519a6a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACH_CHOICE_TABLE_COLS',\n",
       " 'ACH_CHOICE_TABLE_NAME',\n",
       " 'Achillesreadme',\n",
       " 'BAMQC',\n",
       " 'CCLE_DELETERIOUS_COL',\n",
       " 'CHROM_COL',\n",
       " 'CIVIC_SCORE_COL',\n",
       " 'CNreadme',\n",
       " 'COLRENAMING',\n",
       " 'DATASETS',\n",
       " 'DATE_COL_DICT',\n",
       " 'DEFAULT_TABLE_COLS',\n",
       " 'DEFAULT_TABLE_NAME',\n",
       " 'DNA_CHANGE_COL',\n",
       " 'DUP_ARXSPANS',\n",
       " 'ENSEMBL_SERVER_V',\n",
       " 'EXTRACT_DEFAULTS',\n",
       " 'EXTRACT_TO_CHANGE',\n",
       " 'FPALLBATCHPAIRSETS',\n",
       " 'FUSION_COLNAME',\n",
       " 'FUSION_MAXFFPM',\n",
       " 'FUSION_MAXFREQ',\n",
       " 'FUSION_MINFFPM',\n",
       " 'FUSION_RED_HERRING',\n",
       " 'FUSIONreadme',\n",
       " 'GCS_PAYER_PROJECT',\n",
       " 'GENECHANGETHR',\n",
       " 'GUIDESBED',\n",
       " 'GUMBO_CLIENT_USERNAME',\n",
       " 'HESS_COL',\n",
       " 'HG38BAMCOL',\n",
       " 'HG38_CRAM_COLNAMES',\n",
       " 'HGNC_MAPPING',\n",
       " 'HOTSPOT_COL',\n",
       " 'HUGO_COL',\n",
       " 'IMMORTALIZED_THR',\n",
       " 'LEGACY_BAM_COLNAMES',\n",
       " 'LIKELY_LOF_COL',\n",
       " 'MAF_COL',\n",
       " 'MATCH',\n",
       " 'MAXAGE',\n",
       " 'MAXYCHROM',\n",
       " 'MC_TABLE_INDEX',\n",
       " 'MC_TABLE_NAME',\n",
       " 'MINFREQTOCALL',\n",
       " 'MINSIZES_BAM',\n",
       " 'MINSIZES_CRAM',\n",
       " 'MISC_SIG_TERRACOLS',\n",
       " 'MODEL_TABLE_INDEX',\n",
       " 'MODEL_TABLE_NAME',\n",
       " 'MUTATION_GROUPS',\n",
       " 'MUTCOL_DEPMAP',\n",
       " 'MYSTORAGE_ID',\n",
       " 'MY_ID',\n",
       " 'Mutationsreadme',\n",
       " 'POS_COL',\n",
       " 'PREV_VIRTUAL',\n",
       " 'PROCQC',\n",
       " 'PROFILE_TABLE_COLS',\n",
       " 'PROTEINEFILENAMES',\n",
       " 'PR_TABLE_INDEX',\n",
       " 'PR_TABLE_NAME',\n",
       " 'PURECN_COLRENAMING',\n",
       " 'PURECN_FAILED_COLNAME',\n",
       " 'PURECN_LOHVALUES',\n",
       " 'PURECN_LOH_COLNAME',\n",
       " 'PURECN_MAX_PLOIDY',\n",
       " 'PURECN_MIN_GOF',\n",
       " 'PURECN_SAMPLESET',\n",
       " 'README_changes',\n",
       " 'README_currentfolder',\n",
       " 'README_folder',\n",
       " 'RELEASE',\n",
       " 'RELEASE_PR_TABLE_NAME',\n",
       " 'REPLACE',\n",
       " 'RNAMINSIMI',\n",
       " 'RNASEQC_THRESHOLDS_FAILED',\n",
       " 'RNASEQC_THRESHOLDS_LOWQUAL',\n",
       " 'RNAWORKSPACES',\n",
       " 'RNAseqreadme',\n",
       " 'RSEMFILENAME',\n",
       " 'RSEMFILENAME_GENE',\n",
       " 'RSEMFILENAME_TRANSCRIPTS',\n",
       " 'RSEM_TRANSCRIPTS',\n",
       " 'RUN_NOTEBOOKS',\n",
       " 'SAMPLEID',\n",
       " 'SAMPLESETNAME',\n",
       " 'SAMPLE_TABLE_NAME',\n",
       " 'SEGMENTSTHR',\n",
       " 'SEQ_TABLE_INDEX',\n",
       " 'SEQ_TABLE_NAME',\n",
       " 'SHEETCREDS',\n",
       " 'SIGTABLE_BINARYCOLS',\n",
       " 'SIGTABLE_RENAMING',\n",
       " 'SIGTABLE_TERRACOLS',\n",
       " 'SKIP_UPLOADING_README',\n",
       " 'SOURCE_PRIORITY',\n",
       " 'SOURCE_RENAME',\n",
       " 'SSGSEAFILEPATH',\n",
       " 'STARBAMCOLTERRA',\n",
       " 'SV_COLNAME',\n",
       " 'SV_COLRENAME',\n",
       " 'SV_FILENAME',\n",
       " 'TORAISE',\n",
       " 'TO_UPDATE',\n",
       " 'VCFCOLNAME',\n",
       " 'VCFDIR',\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_MODEL',\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_PR',\n",
       " 'VIRTUAL_FILENAMES_GUIDEMUT',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_PR',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_PR',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_PR',\n",
       " 'WESSETENTITY',\n",
       " 'WGSSETENTITY',\n",
       " 'WGSWORKSPACES',\n",
       " 'WORKING_DIR'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(dev.keys()).intersection(shared.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "46630618",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACH_CHOICE_TABLE_COLS',\n",
       " 'ACH_CHOICE_TABLE_NAME',\n",
       " 'Achillesreadme',\n",
       " 'BAMQC',\n",
       " 'CCLE_DELETERIOUS_COL',\n",
       " 'CHROM_COL',\n",
       " 'CIVIC_SCORE_COL',\n",
       " 'CNreadme',\n",
       " 'COLRENAMING',\n",
       " 'DATASETS',\n",
       " 'DATE_COL_DICT',\n",
       " 'DEFAULT_TABLE_COLS',\n",
       " 'DEFAULT_TABLE_NAME',\n",
       " 'DNA_CHANGE_COL',\n",
       " 'DUP_ARXSPANS',\n",
       " 'ENSEMBL_SERVER_V',\n",
       " 'EXTRACT_DEFAULTS',\n",
       " 'EXTRACT_TO_CHANGE',\n",
       " 'FPALLBATCHPAIRSETS',\n",
       " 'FUSION_COLNAME',\n",
       " 'FUSION_MAXFFPM',\n",
       " 'FUSION_MAXFREQ',\n",
       " 'FUSION_MINFFPM',\n",
       " 'FUSION_RED_HERRING',\n",
       " 'FUSIONreadme',\n",
       " 'GCS_PAYER_PROJECT',\n",
       " 'GENECHANGETHR',\n",
       " 'GUIDESBED',\n",
       " 'GUMBO_CLIENT_USERNAME',\n",
       " 'HESS_COL',\n",
       " 'HG38BAMCOL',\n",
       " 'HG38_CRAM_COLNAMES',\n",
       " 'HGNC_MAPPING',\n",
       " 'HOTSPOT_COL',\n",
       " 'HUGO_COL',\n",
       " 'IMMORTALIZED_THR',\n",
       " 'LEGACY_BAM_COLNAMES',\n",
       " 'LIKELY_LOF_COL',\n",
       " 'MAF_COL',\n",
       " 'MATCH',\n",
       " 'MAXAGE',\n",
       " 'MAXYCHROM',\n",
       " 'MC_TABLE_INDEX',\n",
       " 'MC_TABLE_NAME',\n",
       " 'MINFREQTOCALL',\n",
       " 'MINSIZES_BAM',\n",
       " 'MINSIZES_CRAM',\n",
       " 'MISC_SIG_TERRACOLS',\n",
       " 'MODEL_TABLE_INDEX',\n",
       " 'MODEL_TABLE_NAME',\n",
       " 'MUTATION_GROUPS',\n",
       " 'MUTCOL_DEPMAP',\n",
       " 'MYSTORAGE_ID',\n",
       " 'MY_ID',\n",
       " 'Mutationsreadme',\n",
       " 'POS_COL',\n",
       " 'PREV_VIRTUAL',\n",
       " 'PROCQC',\n",
       " 'PROFILE_TABLE_COLS',\n",
       " 'PROTEINEFILENAMES',\n",
       " 'PR_TABLE_INDEX',\n",
       " 'PR_TABLE_NAME',\n",
       " 'PURECN_COLRENAMING',\n",
       " 'PURECN_FAILED_COLNAME',\n",
       " 'PURECN_LOHVALUES',\n",
       " 'PURECN_LOH_COLNAME',\n",
       " 'PURECN_MAX_PLOIDY',\n",
       " 'PURECN_MIN_GOF',\n",
       " 'PURECN_SAMPLESET',\n",
       " 'README_changes',\n",
       " 'README_currentfolder',\n",
       " 'README_folder',\n",
       " 'RELEASE',\n",
       " 'RELEASE_PR_TABLE_NAME',\n",
       " 'REPLACE',\n",
       " 'RNAMINSIMI',\n",
       " 'RNASEQC_THRESHOLDS_FAILED',\n",
       " 'RNASEQC_THRESHOLDS_LOWQUAL',\n",
       " 'RNAWORKSPACES',\n",
       " 'RNAseqreadme',\n",
       " 'RSEMFILENAME',\n",
       " 'RSEMFILENAME_GENE',\n",
       " 'RSEMFILENAME_TRANSCRIPTS',\n",
       " 'RSEM_TRANSCRIPTS',\n",
       " 'RUN_NOTEBOOKS',\n",
       " 'SAMPLEID',\n",
       " 'SAMPLESETNAME',\n",
       " 'SAMPLE_TABLE_NAME',\n",
       " 'SEGMENTSTHR',\n",
       " 'SEQ_TABLE_INDEX',\n",
       " 'SEQ_TABLE_NAME',\n",
       " 'SHEETCREDS',\n",
       " 'SIGTABLE_BINARYCOLS',\n",
       " 'SIGTABLE_RENAMING',\n",
       " 'SIGTABLE_TERRACOLS',\n",
       " 'SKIP_UPLOADING_README',\n",
       " 'SOURCE_PRIORITY',\n",
       " 'SOURCE_RENAME',\n",
       " 'SSGSEAFILEPATH',\n",
       " 'STARBAMCOLTERRA',\n",
       " 'SV_COLNAME',\n",
       " 'SV_COLRENAME',\n",
       " 'SV_FILENAME',\n",
       " 'TORAISE',\n",
       " 'TO_UPDATE',\n",
       " 'VCFCOLNAME',\n",
       " 'VCFDIR',\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_MODEL',\n",
       " 'VIRTUAL_FILENAMES_GERMLINE_PR',\n",
       " 'VIRTUAL_FILENAMES_GUIDEMUT',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_CN_PR',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_EXP_PR',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_MODEL',\n",
       " 'VIRTUAL_FILENAMES_NUMMAT_MUT_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_CN_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_FUSION_PR',\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_MODEL',\n",
       " 'VIRTUAL_FILENAMES_TABLE_MUT_PR',\n",
       " 'WESSETENTITY',\n",
       " 'WGSSETENTITY',\n",
       " 'WGSWORKSPACES',\n",
       " 'WORKING_DIR'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(prod.keys()).intersection(shared.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c8ab7035",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in set(prod.keys()).intersection(dev.keys()):\n",
    "    if prod[k] == dev[k]:\n",
    "        shared[k] = prod[k]\n",
    "        del prod[k]\n",
    "        del dev[k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "aa83acb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['SHEETNAME', 'BAM_GCS_BUCKET', 'RNA_GCS_PATH_HG38', 'RNA_GCS_PATH', 'WGS_GCS_PATH', 'WGS_GCS_PATH_HG38', 'WES_GCS_PATH', 'REFSHEET_URL', 'TAIGA_ETERNAL_UPLOAD', 'TAIGA_MUTATION', 'TAIGA_CN', 'TAIGA_CN_ACHILLES', 'TAIGA_EXPRESSION', 'TAIGA_FUSION', 'VIRTUAL_FOLDER', 'RNAWORKSPACE', 'WGSWORKSPACE', 'WESCNWORKSPACE', 'WESMUTWORKSPACE', 'FPWORKSPACE', 'TAIGA_FP', 'LATEST2FN_NUMMAT_MODEL', 'LATEST2FN_TABLE_MODEL', 'LATEST2FN_NUMMAT_PR', 'LATEST2FN_TABLE_PR'])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prod.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e12dab2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['SAMPLESETNAME', 'RELEASE', 'SHEETCREDS', 'MY_ID', 'MYSTORAGE_ID', 'GCS_PAYER_PROJECT', 'SAMPLEID', 'SOURCE_PRIORITY', 'ENSEMBL_SERVER_V', 'HG38BAMCOL', 'LEGACY_BAM_COLNAMES', 'HG38_CRAM_COLNAMES', 'TORAISE', 'TO_UPDATE', 'MAXAGE', 'EXTRACT_TO_CHANGE', 'REPLACE', 'MATCH', 'EXTRACT_DEFAULTS', 'MINSIZES_BAM', 'MINSIZES_CRAM', 'DUP_ARXSPANS', 'RNAWORKSPACES', 'WGSWORKSPACES', 'WGSSETENTITY', 'WESSETENTITY', 'FPALLBATCHPAIRSETS', 'WORKING_DIR', 'PREV_VIRTUAL', 'DATASETS', 'RUN_NOTEBOOKS', 'PROCQC', 'BAMQC', 'VCFDIR', 'GUIDESBED', 'VCFCOLNAME', 'GENECHANGETHR', 'SEGMENTSTHR', 'MAXYCHROM', 'COLRENAMING', 'PURECN_MIN_GOF', 'PURECN_MAX_PLOIDY', 'PURECN_COLRENAMING', 'PURECN_LOH_COLNAME', 'PURECN_FAILED_COLNAME', 'PURECN_SAMPLESET', 'PURECN_LOHVALUES', 'SIGTABLE_TERRACOLS', 'MISC_SIG_TERRACOLS', 'SIGTABLE_BINARYCOLS', 'SIGTABLE_RENAMING', 'SOURCE_RENAME', 'MINFREQTOCALL', 'SV_COLNAME', 'SV_FILENAME', 'SV_COLRENAME', 'MUTATION_GROUPS', 'MAF_COL', 'HGNC_MAPPING', 'DNA_CHANGE_COL', 'CHROM_COL', 'POS_COL', 'HOTSPOT_COL', 'IMMORTALIZED_THR', 'HUGO_COL', 'LIKELY_LOF_COL', 'CCLE_DELETERIOUS_COL', 'CIVIC_SCORE_COL', 'HESS_COL', 'MUTCOL_DEPMAP', 'FUSION_COLNAME', 'FUSION_RED_HERRING', 'FUSION_MAXFREQ', 'FUSION_MINFFPM', 'FUSION_MAXFFPM', 'STARBAMCOLTERRA', 'RSEM_TRANSCRIPTS', 'RSEMFILENAME_GENE', 'PROTEINEFILENAMES', 'RSEMFILENAME_TRANSCRIPTS', 'RSEMFILENAME', 'SSGSEAFILEPATH', 'RNAMINSIMI', 'RNASEQC_THRESHOLDS_LOWQUAL', 'RNASEQC_THRESHOLDS_FAILED', 'README_folder', 'README_currentfolder', 'README_changes', 'SKIP_UPLOADING_README', 'FUSIONreadme', 'RNAseqreadme', 'CNreadme', 'Achillesreadme', 'Mutationsreadme', 'GUMBO_CLIENT_USERNAME', 'DATE_COL_DICT', 'MODEL_TABLE_NAME', 'MODEL_TABLE_INDEX', 'MC_TABLE_NAME', 'MC_TABLE_INDEX', 'PR_TABLE_NAME', 'PR_TABLE_INDEX', 'SEQ_TABLE_NAME', 'SEQ_TABLE_INDEX', 'SAMPLE_TABLE_NAME', 'ACH_CHOICE_TABLE_COLS', 'ACH_CHOICE_TABLE_NAME', 'DEFAULT_TABLE_COLS', 'DEFAULT_TABLE_NAME', 'PROFILE_TABLE_COLS', 'RELEASE_PR_TABLE_NAME', 'VIRTUAL_FILENAMES_NUMMAT_EXP_MODEL', 'VIRTUAL_FILENAMES_NUMMAT_CN_MODEL', 'VIRTUAL_FILENAMES_NUMMAT_MUT_MODEL', 'VIRTUAL_FILENAMES_GERMLINE_MODEL', 'VIRTUAL_FILENAMES_TABLE_FUSION_MODEL', 'VIRTUAL_FILENAMES_TABLE_CN_MODEL', 'VIRTUAL_FILENAMES_TABLE_MUT_MODEL', 'VIRTUAL_FILENAMES_GUIDEMUT', 'VIRTUAL_FILENAMES_NUMMAT_EXP_PR', 'VIRTUAL_FILENAMES_NUMMAT_CN_PR', 'VIRTUAL_FILENAMES_NUMMAT_MUT_PR', 'VIRTUAL_FILENAMES_GERMLINE_PR', 'VIRTUAL_FILENAMES_TABLE_FUSION_PR', 'VIRTUAL_FILENAMES_TABLE_CN_PR', 'VIRTUAL_FILENAMES_TABLE_MUT_PR', 'DEPMAP_PV', 'TAIGA_FP_FILENAME', 'TAIGA_LEGACY_CN', 'POTENTIAL_LIST', 'DEPMAP_TAIGA', 'TAIGA_ETERNAL', 'VIRTUAL'])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c8c0dcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_vars = set(shared.keys()) # set(prod.keys()).union(shared.keys()).union(dev.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "556a9b7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "32ab6ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rewrite(fn):\n",
    "    constant_var_pattern = re.compile(\"(\"+(\"|\".join(config_vars))+\")\")\n",
    "    env_var_pattern = re.compile(\"(\"+(\"|\".join(set(prod.keys()).union(dev.keys())))+\")\")\n",
    "    \n",
    "    print(\"rewriting\", fn)\n",
    "    content = open(fn, \"rt\").read()\n",
    "    assert \"constants\" not in content\n",
    "    with open(f\"{fn}\", \"wt\") as fd:\n",
    "        new_content = re.sub(constant_var_pattern, \"constants.\\\\1\", content)\n",
    "        new_content = re.sub(env_var_pattern, \"env_config.\\\\1\", new_content)\n",
    "        new_content = re.sub(\"from depmapomics.config import \\\\*\", \"\", new_content)\n",
    "        if new_content != content:\n",
    "            fd.write(\"from depmapomics import constants\\n\")\n",
    "            fd.write(\"from depmapomics import env_config\\n\")\n",
    "        fd.write(new_content)\n",
    "\n",
    "with open(\"constants.py\", \"wt\") as fd:\n",
    "    for k, v in shared.items():\n",
    "        fd.write(f\"{k} = {repr(v)}\\n\\n\")\n",
    "\n",
    "with open(\"config_dev.py\", \"wt\") as fd:\n",
    "    for k, v in dev.items():\n",
    "        fd.write(f\"{k} = {repr(v)}\\n\\n\")\n",
    "\n",
    "with open(\"config_prod.py\", \"wt\") as fd:\n",
    "    for k, v in prod.items():\n",
    "        fd.write(f\"{k} = {repr(v)}\\n\\n\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3223f43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fef55648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./vcf.py',\n",
       " './copynumbers.py',\n",
       " './config.py',\n",
       " './fingerprinting.py',\n",
       " './mutations.py',\n",
       " './config_prod.py',\n",
       " './constants.py',\n",
       " './expressions.py',\n",
       " './__init__.py',\n",
       " './terra.py',\n",
       " './config_global.py',\n",
       " './fusions.py',\n",
       " './sv.py',\n",
       " './config_dev.py',\n",
       " './dm_omics.py',\n",
       " './qc/conftest.py',\n",
       " './qc/config.py',\n",
       " './qc/test_plot_comparisons.py',\n",
       " './qc/__init__.py',\n",
       " './qc/rna.py',\n",
       " './qc/test_celllines.py',\n",
       " './qc/test_compare_to_ref_release.py',\n",
       " './qc/utils.py',\n",
       " './qc/create_tentative_virtual.py',\n",
       " './qc/cn.py',\n",
       " './qc/test_file_format.py']"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "python_files = []\n",
    "for root, subdirs, files in os.walk('.'):\n",
    "    for fn in files:\n",
    "        if fn.endswith(\".py\"):\n",
    "            python_files.append(os.path.join(root, fn))\n",
    "python_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6a53be65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rewriting ./vcf.py\n",
      "rewriting ./copynumbers.py\n",
      "rewriting ./fingerprinting.py\n",
      "rewriting ./mutations.py\n",
      "rewriting ./expressions.py\n",
      "rewriting ./__init__.py\n",
      "rewriting ./terra.py\n",
      "rewriting ./fusions.py\n",
      "rewriting ./sv.py\n",
      "rewriting ./dm_omics.py\n",
      "rewriting ./qc/conftest.py\n",
      "rewriting ./qc/config.py\n",
      "rewriting ./qc/test_plot_comparisons.py\n",
      "rewriting ./qc/__init__.py\n",
      "rewriting ./qc/rna.py\n",
      "rewriting ./qc/test_celllines.py\n",
      "rewriting ./qc/test_compare_to_ref_release.py\n",
      "rewriting ./qc/utils.py\n",
      "rewriting ./qc/create_tentative_virtual.py\n",
      "rewriting ./qc/cn.py\n",
      "rewriting ./qc/test_file_format.py\n"
     ]
    }
   ],
   "source": [
    "for fn in python_files:\n",
    "    if fn not in ['./config.py', \"./config_dev.py\", \"./config_global.py\", \"./config_prod.py\", \"./constants.py\"]:\n",
    "        rewrite(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2a10a0f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function sub in module re:\n",
      "\n",
      "sub(pattern, repl, string, count=0, flags=0)\n",
      "    Return the string obtained by replacing the leftmost\n",
      "    non-overlapping occurrences of the pattern in string by the\n",
      "    replacement repl.  repl can be either a string or a callable;\n",
      "    if a string, backslash escapes in it are processed.  If it is\n",
      "    a callable, it's passed the Match object and must return\n",
      "    a replacement string to be used.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(re.sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91b1cc1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
