This document summarizes the key components in DepMap omics' processing pipeline.

Note that input references, indices, and parameters used in the WDL workflows for the following pipelines in any given quarter can be found in `data/[release quarter]/[workflow name].json`.

### Copy Numbers and Somatic Mutations

We are currently running the following workflows to generate datasets from WGS data:

[WGS_pipeline](https://dockstore.org/workflows/github.com/broadinstitute/depmap_omics/WGS_pipeline:master?tab=info) runs the following sub-processes to generate relative and absolute copy number segments, mutation MAF file, structural variant (SV) calls, and various genomic features including loss of heterozygosity (LoH), LoH fraction, ploidy estimate, Whole Genome Doubling (WGD), Chromasomal Instability (CIN), and MSI score. This workflow runs the following subtasks:
- __gatk cnv__:
  - outputs relative segment and copy number from WES/WGS data
  - [https://software.broadinstitute.org/gatk/documentation/article?id=11682](https://software.broadinstitute.org/gatk/documentation/article?id=11682)
- __mutect2 (from `broadinstitute/gatk:4.2.6.1`)__:
  - outputs mutation calls from RNAseq/WES/WGS data
  - [https://gatk.broadinstitute.org/hc/en-us/articles/360036490432-Mutect2](https://gatk.broadinstitute.org/hc/en-us/articles/360036490432-Mutect2)
  - see our [documentation](https://storage.googleapis.com/shared-portal-files/Tools/%5BDMC%20Communication%5D%2022Q4%20Mutation%20Pipeline%20Update.pdf) on the mutation pipeline for details regarding filtering, annotation, and more
- __PureCN__:
  - computes absolute copy number, as well as features including loss of heterozygosity (LoH), LoH fraction, ploidy estimate, Whole Genome Doubling (WGD), Chromasomal Instability (CIN) from WES/WGS data
  - For some samples, PureCN requires manual curation to determine the most plausible solution. Details about this can be found in the `WGS_CCLE.ipynb` notebook
  - [Details on how PureCN is run for DepMap data](../WGS_pipeline/PureCN_pipeline/README.md)
  - [https://github.com/lima1/PureCN](https://github.com/lima1/PureCN)
- __MSIsensor2__:
  - computes Microsatellite Instability (MSI) score from WES/WGS data
  - [https://github.com/niu-lab/msisensor2](https://github.com/niu-lab/msisensor2)
- __Manta__:
  - calls structural variants from WES/WGS data
  - [https://github.com/Illumina/manta](https://github.com/Illumina/manta)
- __Manta SV annotator__:
  - annotates structural variants generated by Manta
  - https://github.com/acranej/MantaSVAnnotator

[Aggregate_CN_seg_files](https://dockstore.org/workflows/github.com/broadinstitute/depmap_omics/Aggregate_CN_seg_files:master?tab=info) aggregates copy number segment outputs.

WES:

The same subtasks above were run on our WES data. We used both Illumina ICE and Agilent intervals for our WES data. You can find their respective PON files and interval files as parameters in our workflow configurations in `data/[release quarter]/[workflow name].json`.


#### Panel of Normals (PONs)

For CN PONs are made from normals from the GTEx project as they were sequenced in the same fashion as CCLE samples with the same set of baits.
PONs for each bait set were created with XY only lines using the workflow `gatk/CNV_Somatic_Panel_Workflow`.

### Expression and Fusion

We are generating both expression and fusion datasets with RNAseq data. Specifically, we use the [GTEx pipeline](https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md) to generate the expression dataset, and [STAR-Fusion](https://github.com/STAR-Fusion/STAR-Fusion/wiki) to generate gene fusion calls. This task also contains a flag that lets you specify if you want to delete the intermediates (fastqs) that can be large and might cost a lot to store. The following two workflows are run in this order:

[RNA_pipeline](https://dockstore.org/workflows/github.com/broadinstitute/depmap_omics/RNA_pipeline:master?tab=info) imports and runs the following sub-processes to generate RNA expression and fusion data matrices.

- __star (from docker image `gcr.io/broad-cga-francois-gtex/gtex_rnaseq:V10`)__:
  - aligns RNAseq bam files for downstream processing
  - [https://www.ncbi.nlm.nih.gov/pubmed/23104886](https://www.ncbi.nlm.nih.gov/pubmed/23104886)
  - [https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)
  - STAR and RSEM indices are generated using GENCODE's "comprehensive gene annotations" GTF and the GRCh38 reference genome for RNA-seq alignment provided in [GTEx's pipeline](https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md), which includes ERCC spike-in and excludes ALT, HLA, and Decoy contigs. The STAR index is generated with flag `--sjdbOverhang 100`.
- __rsem (from docker image `gcr.io/broad-cga-francois-gtex/gtex_rnaseq:V10`)__: 
  - quantifes gene and isoform abundances from RNAseq data
  - [https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323)
- __star fusion (from docker image `trinityctat/starfusion:1.7.0`)__: 
  - generates fusion prediction from RNAseq data
  - [https://github.com/STAR-Fusion/STAR-Fusion/wiki](https://github.com/STAR-Fusion/STAR-Fusion/wiki)
  - [http://biorxiv.org/content/early/2017/03/24/120295](http://biorxiv.org/content/early/2017/03/24/120295)

[RNA_aggregate](https://dockstore.org/workflows/github.com/broadinstitute/depmap_omics/RNA_aggregate:master?tab=info) aggregates expression and fusion data files into their respective aggregated file.


__Finally, we save the workflow configurations used in the pipeline runs__

**Remarks:**
- for the copy number pipeline we have parametrized both an XX version and an XY version, we recommend using the XY version as it covers the entire genome
