



## firecloud documentation

[gdoc](https://docs.google.com/document/d/14KUdstQNLiPVq8jc4oE7ns3uwEakWoJuMh7WnUXbJtA/edit#heading=h.qmmoon4ai3il)

## problematic workspaces


- need to have access to this`gkugener/Rename_CN_Files/2`
- seems to contain a shit load of duplicate data `broad-firecloud-ccle/GATK4_CCLE_CNV`


broad-firecloud-ccle/DepMap_WGS_PON_hg38


### To merge

### WGS CN
- broad-firecloud-ccle/DepMap_WGS_PON_hg38: pon generation
- rmc-rnaseq/St_Jude_RMS: does pon generation as well
- single-cell-prism/NIH_RMS (also for PedDep CN (with stjude but WES) dependency on gene xB when copyloss of xA)



### WES PON
- broad-firecloud-ccle/DepMap_WES_PONS_hg38
- broad-firecloud-ccle/GATK4_CCLE_CNV


### data exom CCLE
__listed below__

### data rna CCLE 
__listed below__


### To Leave

- CCLE_WES_Pipelines: cannot access at all minoun athorization domain
- broad-genomics-delivery/CCLE_WES_and_RNAseq_data(read only): empty but cant delete
- CPTA_* : what are they for?? datastorage 
- TCGA_*_ControlledAccess_*_DATA: a set of 40 workspaces for storing data of different tissue types. It should not be of any use to me.
- nci-maruvka-broadinstitute-org/MSMuTect(read only): Unrelated to me



### BASE

#### data

__Exome__

[IBM](https://firecloud.terra.bio/#workspaces/broad-genomics-delivery/Getz_IBM_CellLines_Exomes/data)
[Broad](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/CCLE_DepMap_WES)
[Broad](https://firecloud.terra.bio/#workspaces/broad-genomics-delivery/CCLE_DepMap_WES)

__rna__

[IBM](https://firecloud.terra.bio/#workspaces/broad-genomics-delivery/Getz_IBM_CellLines_RNASeqData/data)
[Broad](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/CCLE_DepMap_RNAseq)
[Broad](https://firecloud.terra.bio/#workspaces/broad-genomics-delivery/CCLE_DepMap_RNAseq)

[other raw data](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/CCLE_raw_data/data)

#### processing

[CNV](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/DepMap_WES_CN_hg38/data)
[RNAseq](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/DepMap_hg38_RNAseq/data)
[PONS](https://firecloud.terra.bio/#workspaces/broad-firecloud-ccle/DepMap_WES_PONS_hg38/data)


### Current workspaces {#current-workspaces}


#### broad-firecloud-ccle/DepMap_hg38_RNAseq {#broad-firecloud-ccle-depmap_hg38_rnaseq}


##### Expression {#expression}

We use the GTEx pipeline ([https://github.com/broadinstitute/gtex-pipeline/blob/v9/TOPMed_RNAseq_pipeline.md](https://github.com/broadinstitute/gtex-pipeline/blob/v9/TOPMed_RNAseq_pipeline.md)).

To generate the expression dataset, run the following tasks on all samples that you need, in this order:



*   samtofastq_v1-0_BETA_cfg 

    (broadinstitute_gtex/samtofastq_v1-0_BETA Snapshot ID: 5)

*   star_v1-0_BETA_cfg

(broadinstitute_gtex/star_v1-0_BETA Snapshot ID: 7)



*   rsem_v1-0_BETA_cfg 

    (broadinstitute_gtex/rsem_v1-0_BETA Snapshot ID: 4)

*   rsem_aggregate_results_v1-0_BETA_cfg (broadinstitute_gtex/rsem_aggregate_results_v1-0_BETA Snapshot ID: 3)

The outputs to be downloaded will be saved under the sample set that you ran. The outputs we use for the release are:



*   rsem_genes_expected_count
*   rsem_genes_tpm
*   rsem_transcripts_tpm

****Make sure that you delete the intermediate files. These files are quite large so cost a lot to store. To delete, you can either write a task that deletes them or use gsutil rm*****


##### Fusions {#fusions}

We use STAR-Fusion [https://github.com/STAR-Fusion/STAR-Fusion/wiki](https://github.com/STAR-Fusion/STAR-Fusion/wiki). The fusions are generated by running the following tasks



*   hg38_STAR_fusion (gkugener/STAR_fusion Snapshot ID: 14)
*   Aggregate_Fusion_Calls (gkugener/Aggregate_files_set Snapshot ID: 2)

The outputs to be downloaded will be saved under the sample set you ran. The outputs we use for the release are: 



*   fusions_star

This task uses the same samtofastq_v1-0_BETA_cfg task as in the expression pipeline, although in the current implementation, this task will be run twice. It might be worth combing the expression/fusion calling into a single workflow. This task also contains a flag that lets you specify if you want to delete the intermediates (fastqs). 

There are several other tasks in this workspace. In brief:



*   Tasks prefixed with **EXPENSIVE** or **CHEAP** are identical to their non-prefixed version, except that they specify different memory, disk space, etc. parameters. These versions can be used when samples fail the normal version of the task due to memory errors.
*   The following tasks are part of the GTEx pipeline but we do not use them (we use RSEM exclusively): markduplicates_v1-0_BETA_cfg (broadinstitute_gtex/markduplicates_v1-0_BETA Snapshot ID: 2), rnaseqc2_v1-0_BETA_cfg (broadinstitute_gtex/rnaseqc2_v1-0_BETA Snapshot ID: 2)
*   **ExonUsage_hg38_fixed** (gkugener/ExonUsage_fixed Snapshot ID: 1): this task calculates exon usage ratios. The non-fixed version contains a bug in the script that is not able to handle chromosome values prefixed with ‘chr’. The ‘fixed’ version resolves this issue.
*   **AggregateExonUsageRObj_hg38** (ccle_mg/AggregateExonUsageRObj Snapshot ID: 2): combines the exon usage ratios into a matrices that are saved in an R object.


#### broad-firecloud-ccle/CCLE_implementation_CGA_WES_Characterization_TCGAControlledAccess {#broad-firecloud-ccle-ccle_implementation_cga_wes_characterization_tcgacontrolledaccess}

In order to run the CGA pipeline, you need to have TCGA access. To get TCGA access, follow the instructions here: [https://docs.google.com/document/d/1hwsjUypqUpse7IeMyBLKEXmdlXUzfBa4e4p9teCVtaw/edit](https://docs.google.com/document/d/1hwsjUypqUpse7IeMyBLKEXmdlXUzfBa4e4p9teCVtaw/edit)

The CGA mutations pipeline we have used in 19Q1 and 19Q2 is a copy of the CGA pipeline adapted to cell lines. CGA published their official pipeline recently on FireCloud, so I wanted to clone this version and run the samples through so that the samples are using the exact same pipeline as the CGA one. However, a subset of the samples failed when trying to use this implementation and I was not able to debug the reasons, so for now we are still using the cell line adapted version, but it may be worth looking into what caused the cell lines to fail this implementation of the pipeline, as it should be almost identical to the version we are currently using (main differences appear to be in output naming). 


#### broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline {#broad-firecloud-ccle-depmap_mutation_calling_cga_pipeline}

In order to run the CGA pipeline, you need to have TCGA access. To get TCGA access, follow the instructions here: [https://docs.google.com/document/d/1hwsjUypqUpse7IeMyBLKEXmdlXUzfBa4e4p9teCVtaw/edit](https://docs.google.com/document/d/1hwsjUypqUpse7IeMyBLKEXmdlXUzfBa4e4p9teCVtaw/edit)


##### Mutations {#mutations}

To generate the mutation dataset, run the following tasks:



*   For new samples in DepMap, run the ICE version of this task. CCLE2 samples used Agilent targets, so this pipeline should be used instead. The pipelines are identical in terms of their outputs, but the proper targets, baits, and pseudo normal should be used based on how the samples were sequenced.

    **ICE_CGA_Production_Analysis_Pipeline_Cell_Lines_copy** (cclf/CGA_Production_Analysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22) OR


    **AGILENT_CGA_Production_Analysis_Pipeline_Cell_Lines** (cclf/CGA_Production_Anablysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22)

*   **common_variant_filter** (breardon/common_variant_filterSnapshot ID: 3)
*   **filterMAF_on_CGA_pipeline** (gkugener/filterMAF_on_CGA_pipelineSnapshot ID: 8)
*   **aggregateMAFs_selectFields** (ccle_mg/aggregateMAFs_selectFieldsSnapshot ID: 1)

This outputs to be downloaded will be saved in the sample set that was run. The output we use for the release is:



*   **passedCGA_filteredMAF_aggregated** 

There are several other tasks in this workspace. In brief:



*   **CGA_Production_Analysis_Pipeline_Cell_Lines** (lelagina/CGA_Production_Analysis_Pipeline_Cell_LinesSnapshot ID: 12). This task is the same as the ICE and AGILENT prefixed version above, except that it relied on pulling the baits and targets to use from the metadata stored for the samples. Having AGILENT and ICE versions specified made the uploading and running process easier.
*   **SANGER_CGA_Production_Analysis_Pipeline_Cell_Lines** (cclf/CGA_Production_Analysis_Pipeline_Cell_Lines_debuggingSnapshot ID: 22). This task was trying to run the CGA pipeline on the Sanger WES data, using a Sanger pseudo normal. In its current implementation, this task fails to complete for the samples.
*   **UNFILTERED_aggregateMAFs_selectFields** (ccle_mg/aggregateMAFs_selectFieldsSnapshot ID: 1). Aggregates the MAF outputted by the CGA cell line pipeline prior to the common variant filter and germline filtering tasks. This can give us insight to which mutations are getting filtered out when. We may want to potentially include this MAF in the release so people can see why certain mutations of interest may be getting filtered out.
*   WES_DM_Mutation_Calling_Pipeline_(standard |expensive) (gkugener/WES_DM_Mutation_Calling_PipelineSnapshot ID: 2). This was a previous mutation calling pipeline implemented for CCLE. We do not use this pipeline any more as the CGA pipeline looks better.
*   aggregate_filterMAF_CGA (CCLE/aggregate_filterMAF_CGASnapshot ID: 1). An aggregation MAF task that we used in the past. We do not use this task anymore.
*   calculate_mutational_burden (breardon/calculate_mutational_burdenSnapshot ID: 21). This task can be used to calculate the mutational rate of the samples. We do not make use of this data in the release although it could be of interest.
*   summarizeWigFile (breardon/summarizeWigFileSnapshot ID: 5). CCLF ran this task (might be necessary for the mutational burden task). For our workflow, we do not run it.

#### broad-firecloud-ccle/DepMap_WES_CN_hg38 {#broad-firecloud-ccle-depmap_wes_cn_hg38}


##### Copy number {#copy-number}

To generate the copy number dataset:

*   **BamToUnmappedRGBams_MC** vdauwera/BamToUnmappedRGBamsSnapshot ID: 3
*   **Generate_uBAM_File_List** gkugener/ArrayOfFilesToTxtSnapshot ID: 1
*   **Realign_WES_GATK4** gatk/PreProcessingForVariantDiscovery_GATK4Snapshot ID: 7
*   **CNV_sample_XX** gatk/CNV_Somatic_Pair_WorkflowSnapshot ID: 9
*   **Aggregate_CN_seg_files** gkugener/Aggregate_CN_seg_filesSnapshot ID: 2

This output file for download will be saved under the sample set under the combined_seg_file attribute.

There are several other tasks in this workspace. In brief:

*   **CNV_Somatic_Panel_Workflow_Agilent_XX** gatk/CNV_Somatic_Panel_WorkflowSnapshot ID: 11. This task was used in this workspace to generate the Sanger PON. In the Sanger dataset, there is a set of 40 normal cell lines samples (cell lines derived from matched normal tissue). We can use these to generate a PON to normalize to rather than using the Agilent PON we use for the other CCLE cell lines. This leads to less noisy results. HOWEVER, results using the PON from this workflow should not use the X chromosome, as the sanger normals are not exclusively female or male (it is likely a mix).
*   **SANGER_PON_CNV_sample_XX** gatk/CNV_Somatic_Pair_WorkflowSnapshot ID: 9. Same as the CNV_sample_XX_gatk, except that is uses the Sanger based PON. Should be used only for the Sanger cell lines.
*   **Sanger_PON_Aggregate_CN_seg_files** gkugener/Aggregate_CN_seg_filesSnapshot ID: 2. Aggregates the segment files for the samples that were run using the Sanger PON based CNV workflow.


#### broad-firecloud-ccle/DepMap_WES_PONS_hg38 {#broad-firecloud-ccle-depmap_wes_pons_hg38}

This workspace was used to generate the PONs for the hg38 copy number pipeline for CCLE. In this workspace, normal samples from the different targeting strategies (ICE, Agilent) were realigned to hg38. Following realignment, samples were run through the GATKs PON generation workflow. Specifically, the tasks run in this workspace are:



*   **BamToUnmappedRGBams_MC** (vdauwera/BamToUnmappedRGBamsSnapshot ID: 3). Converts a bam into a uBAM and in the process splits the bam by read group.
*   **Generate_uBAM_File_List** (gkugener/ArrayOfFilesToTxtSnapshot ID: 1). Task implemented to generate a txt file that contains all the split uBAM paths.
*   **Relalign_WES_GATK4** (gatk/PreProcessingForVariantDiscovery_GATK4Snapshot ID: 7). Ailgns uBAM to genome (in this case, hg38).
*   **CNV_Somatic_Panel_Workflow_Agilent_XX** (gatk/CNV_Somatic_Panel_WorkflowSnapshot ID: 11)

    CNV_Somatic_Panel_Workflow_ICE_XX (gatk/CNV_Somatic_Panel_WorkflowSnapshot ID: 11). These workflows generate the PON for Agilent, ICE samples from the PON sample sets. The specific implementation of this task should only be run on female samples, as the interval list in this task includes chromosome X. 



#### broad-firecloud-ccle/DepMap_WGS_PON_hg38 {#broad-firecloud-ccle-depmap_wgs_pon_hg38}

This workspace was created to be a place where we could generate a PON for the WGS samples from GTEx. There is no data in this workspace but the plan was to transfer the WGS from GTEx from a requester pays workspaces generated for us by Francois Aguet ([https://console.cloud.google.com/storage/browser/gtex-v7-exchange-011618/wgs_cram_files](https://console.cloud.google.com/storage/browser/gtex-v7-exchange-011618/wgs_cram_files)) to this workspace, convert the cram files to bam, and then run the PON generation workflow. The samples provided by GTEx are already aligned using hg38. Due to quarterly priorities, we have not started this process.


#### broad-firecloud-ccle/ONE_OFFS_CCLE_WES {#broad-firecloud-ccle-one_offs_ccle_wes}

This workspace contains the task required to run mutation calling and copy number calling on cell line samples from WES that use the Illumina targets. I created this workspace so that it can be shared with others who want to run the pipeline. The workflow I use to share the pipelines:



*   Add the user as reader to this workspace
*   Add the user as reader to all the necessary methods in the workspace
*   Once the user has cloned the workspace, work with them to copy the needed reference files (for the copy number pipeline, this is the PON, common sites; for mutations, this will be the pseudo normal sample if they do not have a matched normal sample).
*   Users will need TCGA access to run the mutation calling pipeline.


### Legacy workspaces {#legacy-workspaces}


#### broad-firecloud-ccle/DepMap_CCLE_RNAseq_pipeline {#broad-firecloud-ccle-depmap_ccle_rnaseq_pipeline}

Used to generate the expression and fusion datasets for hg19 aligned RNAseq data. We no longer use this workspace as of 19Q2.


#### broad-firecloud-ccle/DepMap_Mutation_Calling {#broad-firecloud-ccle-depmap_mutation_calling}

Workspace used to generate mutations with a CCLE mutation calling pipeline implementation. We no longer use this workspace as we are now exclusively using the CGA pipeline. 


#### broad-firecloud-ccle/DepMap_CN_GATK4 {#broad-firecloud-ccle-depmap_cn_gatk4}

This workspace contains was used to generate the Agilent and ICE PONs for the copy number pipeline for the hg19 based alignment.

#### CCLE WGS sequences: https://console.cloud.google.com/storage/browser/fc-effb3f55-962a-4b1f-b41d-63234d7e5735/?authuser=2

### CCLE2 WGS MAF: /xchip/cle/analysis/mghandi/CCLE2/txt/mutations/ccle2maf_ExcludSangerDriftedSubset_20180820.txt

#### broad-firecloud-ccle/DepMap_WGS_CN {#broad-firecloud-ccle-depmap_wgs_cn}

This workspace contained an implementation of the copy number pipeline for WGS samples. This workspace was created to evaluate the GATK copy number pipeline on CCLE WGS samples. Because this was for internal testing only, the PON used for these samples were based on DIPG matched normal WGS that I had generated for another project. I ran the out-of-the-box GATK pipeline using appropriate interval lists and this PON on the CCLE WGS samples that also had WES to evaluate the similarity of the calls. This comparison was inlcuded in my Broad retreat poster in 2018.

The specific tasks run in this workspace:



*   CNV_Somatic_Workflow_on_Sample (gatk/CNV_Somatic_Pair_WorkflowSnapshot ID: 9). GATK copy number calling workflow adapted for WGS samples.
*   Rename_CN_Files (gkugener/Rename_CN_FilesSnapshot ID: 2). I used this workflow to rename the outputted files by DepMap ID (to remove any prefix). Later on, I decided this was a bad idea as sample sets can have multiple samples for the same cell line and this would cause them to be missed, so this task is no longer used.
*   Aggregate_CN_seg_files (gkugener/Aggregate_CN_seg_filesSnapshot ID: 2). Aggregates the copy number data across samples into a single file.


#### broad-firecloud-ccle/GATK4_CCLE_CNV {#broad-firecloud-ccle-gatk4_ccle_cnv}

Contains the normal samples used to generate PONs for the copy number pipeline for CCLE. The storage class for these samples has been set to nearline in an attempt to reduce the storage costs. We may consider removing them from cloud storage entirely as they are likely stored elsewhere. It is also unclear where these came from (they are the same set used by the CERES paper on that were used by Mahmoud Ghandi in Firehose for the older versions of the copy number pipeline). 


#### broad-firecloud-ccle/CCLE_WES_Pipelines {#broad-firecloud-ccle-ccle_wes_pipelines}

This workspace was created by Quintin Wessell when he interned for CCLE in the summer of 2018. He used this workspace to develop and test mutation calling pipelines for CCLE when the project was transitioning from firehose to FireCloud. We no longer use this workspace or these tasks, so it is likely safe to delete.


#### broad-firecloud-ccle/CCLE_RNAseq_pipeline {#broad-firecloud-ccle-ccle_rnaseq_pipeline}

This workspace was created by Quintin Wessell when he interned for CCLE in the summer of 2018. He used this workspace to test the GTEx pipeline on RNAseq. We do not use this workspace and it should be safe to delete, as all relevant RNAseq data is stored elsewhere.


#### broad-firecloud-ccle/CCLE_WES_test {#broad-firecloud-ccle-ccle_wes_test}

This workspace was created by Quintin Wessell when he interned for CCLE in the summer of 2018. He used this workspace to develop and test mutation calling pipelines for CCLE when the project was transitioning from firehose to FireCloud. We do not use this workspace or these tasks, so it is likely safe to delete.


#### fccredits-hassium-orange-5064/CCLE_WES_pipeline {#fccredits-hassium-orange-5064-ccle_wes_pipeline}

This workspace was created by Quintin Wessell when he interned for CCLE in the summer of 2018. He used this workspace to develop and test mutation calling pipelines for CCLE when the project was transitioning from firehose to FireCloud. We do not use this workspace or these tasks, so it is likely safe to delete.


#### broad-firecloud-ccle/hg38_dev {#broad-firecloud-ccle-hg38_dev}

I (Guillaume Kugener) created this workspace in order to develop pipelines for hg38 alignment. Many of the annotations used in the productions workspace are stored in the bucket associated with this workspace. 


#### 


## CCLE Data storage workspaces {#ccle-data-storage-workspaces}

The workspaces described in this section generally only contain data. Analysis should not (and usually, can not) be run in these workspaces.


### IBM {#ibm}


#### broad-genomics-delivery/Getz_IBM_CellLines_Exomes {#broad-genomics-delivery-getz_ibm_celllines_exomes}

Contains the WES data for cell lines screened by IBM. This workspace is updated when new cell line data is available.


#### broad-genomics-delivery/Getz_IBM_CellLines_RNASeqData {#broad-genomics-delivery-getz_ibm_celllines_rnaseqdata}

Contains the RNAseq data for cell lines screened by IBM. This workspace is updated when new cell line data is available.


### Sanger WES {#sanger-wes}


#### broad-firecloud-ccle/Sanger_WES_bams {#broad-firecloud-ccle-sanger_wes_bams}

This workspace contains files related to the analysis of Sanger WES data as part of the CCLE2 project.


#### broad-firecloud-ccle/Sanger_WES_recalib_bams {#broad-firecloud-ccle-sanger_wes_recalib_bams}

The bucket associated with this workspace contains the Sanger WES bams after recalibration was done at the Broad. These are the bams used by CCLE for copy number/mutation calling.


### Chordoma Foundation {#chordoma-foundation}


#### broad-firecloud-ccle/Chordoma_Foundation_DATA {#broad-firecloud-ccle-chordoma_foundation_data}

The bucket associated with this workspace contains WES data downloaded from the Chordoma foundation’s servers.


### CCLE2 {#ccle2}


#### broad-firecloud-ccle/CCLE_DepMap_WES {#broad-firecloud-ccle-ccle_depmap_wes}

The bucket associated with this workspace contains a subset of the CCLE WES samples.


#### broad-firecloud-ccle/DepMap_Omics_WES_DATA {#broad-firecloud-ccle-depmap_omics_wes_data}

The bucket associated with this workspace contains a subset of the CCLE WES samples. Specifically, it was used in a previous quarter as a place to upload samples from CCLE that have yet to be migrated to FireCloud. However, the genomics platform (GP) is expected to host these files, so we should be able to delete this workspace once we know for sure that all of the files currently stored here are stored elsewhere by GP.


#### broad-genomics-delivery/CCLE_DepMap_RNAseq {#broad-genomics-delivery-ccle_depmap_rnaseq}

The bucket associated with this workspace contains a subset of the CCLE RNAseq samples.


#### broad-firecloud-ccle/CCLE_DepMap_RNAseq {#broad-firecloud-ccle-ccle_depmap_rnaseq}

The bucket associated with this workspace contains a subset of the CCLE RNAseq samples.


#### broad-firecloud-ccle/DepMap_Omics_RNAseq_Data {#broad-firecloud-ccle-depmap_omics_rnaseq_data}

The bucket associated with this workspace contains a subset of the CCLE RNAseq samples. Specifically, it was used in a previous quarter as a place to upload samples from CCLE that have yet to be migrated to FireCloud. However, the genomics platform (GP) is expected to host these files, so we should be able to delete this workspace once we know for sure that all of the files currently stored here are stored elsewhere by GP.


#### firehose-to-firecloud/CCLE2 {#firehose-to-firecloud-ccle2}

The bucket with this workspace contains bams for cell lines screened in CCLE2. This includes both WES and RNAseq.


#### broad-firecloud-ccle/CCLE_raw_data {#broad-firecloud-ccle-ccle_raw_data}

The bucket with this workspace contains bams for CCLE2 across multiple screens. The annotations in this workspace also contain files paths for many of these samples.


#### broad-firecloud-ccle/CCLE_WGS_HC_RD_RRBS_TERT_bams {#broad-firecloud-ccle-ccle_wgs_hc_rd_rrbs_tert_bams}

The bucket with this workspace contains bams for CCLE2 across multiple screens. The annotations in this workspace also contain files paths for many of these samples.


#### broad-firecloud-ccle/CCLE_v2_data {#broad-firecloud-ccle-ccle_v2_data}

The bucket associated with this workspace is empty. The annotations in this workspace link CCLE2 samples to their sequencing files. The paths in the metadata may or may not be correct.




## PedDep workspaces {#peddep-workspaces}

The workspaces described in this section are related to several PedDep projects.


### RMS {#rms}


#### single-cell-prism/NIH_RMS {#single-cell-prism-nih_rms}

This workspace was created and used to attempt to process RMS tumor samples from the NIH that were sequenced using ABI Solid. We generated a PON from the matched normal samples in this cohort and then attempted to run the GATK copy number pipeline using matched tumor normal samples using this PON. We were able to run the pipeline but downstream analysis of the results suggests that the data quality of these samples is low (generally, low coverage).


#### rmc-rnaseq/St_Jude_RMS {#rmc-rnaseq-st_jude_rms}

This workspace contains was used to store all data and FireCloud analyses related to the St Jude RMS cohort. The sequencing data was downloaded using the DNAnexus command line tool from [https://www.stjude.cloud/](https://www.stjude.cloud/) (for more info on account details for the RMS dataset, follow up with Andrew Hong). RNAseq and WES were downloaded and processed through the expression, fusion, and copy number pipelines.


### RMC {#rmc}


#### rmc-rnaseq/RMC RNAseq {#rmc-rnaseq-rmc-rnaseq}

This workspace was created to process patient and cell line RNAseq data for a cohort of RMC tumors and cell lines. These samples were processed through the GTEx pipeline. See the broad-firecloud-ccle/DepMap_hg38_RNAseq workspace description for a description of the tasks and methods run.


#### single-cell-prism/RMC_copy {#single-cell-prism-rmc_copy}

This workspace was created to run mutation calling on RMC WES samples. As this was in the early days for FireCloud, several pipelines were testing to attempt to run mutation calling. We eventually went with Mutect2 for this project, but there are now much better pipelines that should be used (see the CGA pipeline). 


### DIPG {#dipg}


#### broad-firecloud-ccle/DIPG_WGS_CN {#broad-firecloud-ccle-dipg_wgs_cn}

This workspace was used to call copy number from WGS for a set of DIPG cell line models, to be used for downstream analysis (for input into CERES).


#### broad-firecloud-ccle/DIPG_WGS_PON {#broad-firecloud-ccle-dipg_wgs_pon}

This workspace was used to generate a PON from WGS to be used to call copy number for a set of DIPG cell lines. The standard out-of-the-box GATK copy number pipeline was used on a cohort of normal samples taken from DIPG patients. The tasks in this workflow are:



*   CNV_Somatic_Panel_WGS (gatk/CNV_Somatic_Panel_WorkflowSnapshot ID: 11). The near default GATK4 copy number pipeline workflow to generate a panel of normals from normal samples.
*   GATK-gCNV-Case-Mode (vruano-broad/GATK-gCNV-Case-ModeSnapshot ID: 3). This task was not used in the analysis but we were hoping to use it to obtain sex annotations for the samples (we eventually were provided these annotations).

 


## CDS workspaces {#cds-workspaces}

Workspaces for processing data related to CDS tasks (these could also fall under general DepMap needs).


#### single-cell-prism/Beat_AML {#single-cell-prism-beat_aml}

The workspace used to process the Beat AML RNAseq dataset. Briefly, fastqs were downloaded from synapse.org (**Joshua Dempster** has more info on how to get an account to view these data) using the python package. Tasks in this workspace concatenate the fastqs (reads from the same sample are split across fastqs) and then run them through the expression pipeline of CCLE (STAR + CCLE) for hg38.

Processing workflow:



*   synapse_combine_fastqs (gkugener/synapse_combine_fastqsSnapshot ID: 4). This task was used to combine the separated fastqs into a single fastq for each read pair (left, right) for each sample
*   STAR_RSEM_from_FASTQ (or EXPENSIVE_STAR_RSEM_from_FASTQ) (gkugener/STAR_RSEM_from_FASTQSnapshot ID: 1). These tasks are the same except for different memory parameter specifications (to deal with samples that are slightly too large for the defaults. This pipeline runs STAR and RSEM, taking the fastqs generated from the previous task. There is also a flag included so that the intermediates are deleted upon successful completion of the RSEM task. 
*   rsem_aggregate_results_v1-0_BETA_cfg (broadinstitute_gtex/rsem_aggregate_results_v1-0_BETASnapshot ID: 3). Aggregates the results into a single file.

Other tasks



*   rsem_v1-0_BETA_cfg (broadinstitute_gtex/rsem_v1-0_BETASnapshot ID: 4). The standalone RSEM task from GTEx. Used in the STAR_RSEM_from_FASTQ workflow
*   star_v1-0_BETA_cfg (broadinstitute_gtex/star_v1-0_BETASnapshot ID: 7).  The standalone STAR task from GTEx. Used in the STAR_RSEM_from_FASTQ workflow.
*   download_from_synapse (gkugener/download_from_synapseSnapshot ID: 9). I wrote this task to be able to download the fastqs directly from synapse and do the fastq combinations here. Works well on a single sample at a time. However, upon running this on the entire cohort, almost all of the samples failed (it looks like due to overwhelming the network).


#### single-cell-prism/DepMap_CCLE_RNAseq_pipeline_ONEOFFS {#single-cell-prism-depmap_ccle_rnaseq_pipeline_oneoffs}

This workspace was created to process samples for DepMap through the GTEx pipeline that are part of separate experiments unrelated to CCLE. In this workspace, the idea is to upload the samples, run the analysis, download the results, and upload the results to taiga. All files related to an analysis should then be deleted from this workspace (to prevent costs from such analyses from accruing overtime). One example use case was processing post perturbation RNAseq data from an experiment performed by Tsukasa Shibue.



## CCLF related workspaces {#cclf-related-workspaces}

Several of these workspaces were shared with me when Mimoun was transitioning from CCLF. They contain his analysis on CCLF samples as well as the path to the CCLF samples (WES, RNAseq). I have also cloned some of these workspaces in order to be able to run data aggregation tasks.


### Original workspaces {#original-workspaces}


#### nci-mimoun-bi-org/CCLF_RNA_2_0 {#nci-mimoun-bi-org-cclf_rna_2_0}

Contains RNAseq paths for CCLF samples. Also contains the results from Mimoun running the CCLE RNAseq pipeline on these samples.


#### nci-mimoun-bi-org/CCLF_WES {#nci-mimoun-bi-org-cclf_wes}

Contains the WES paths for CCLF samples and contains the results from Mimoun running the CGA mutation pipeline on these samples.


#### nci-mimoun-bi-org/CCLE_WES_Pipelines {#nci-mimoun-bi-org-ccle_wes_pipelines}

Workspace that Mimoun used when trying to run the cell line adapted CGA mutation calling pipeline on a subset of the CCLE samples.


### Cloned workspaces {#cloned-workspaces}


#### broad-firecloud-ccle/CCLF_RNA_2_0_copy {#broad-firecloud-ccle-cclf_rna_2_0_copy}

This workspace is a direct clone of CCLF’s RNAseq workspace (nci-mimoun/CCLF_RNA_2_0), where a previous CCLF ACB ran the RNAseq based expression pipeline from GTEx. I cloned it in order to add aggregation tasks to collect all of the results into a single file for download.



#### broad-firecloud-ccle/CCLF_WES_copy {#broad-firecloud-ccle-cclf_wes_copy}

This workspace is a direct clone of CCLF’s WES workspace (_nci-mimoun-bi-org/CCLF_WES_), where a previous CCLF ACB ran mutation calling workflows on all available CCLF samples at the time. I cloned this workspace in order to add some tasks to aggregate the results in a single file for further downstream analysis (specifically, to generate filtered and unfiltered MAFs of the CCLF samples and of a subset of the CCLF samples matched to a pseudo normal). The _broad-firecloud-ccle/CCLF_WES_UNTOUCHED _is better than this version as in this copy I reran some mutation calls but there were some errors at the time. This workspace is fairly safe to delete.


## Shared workspaces {#shared-workspaces}

The following workspaces were shared with me in the past to view methods/annotations for other pipelines.


#### broadtagteam/Fusion_Calling_Share_10_22_18 {#broadtagteam-fusion_calling_share_10_22_18}

Brian Granger had implemented workflows that ran several fusion calling pipelines. He shared this workspace with me so that I could see his implementations of these pipelines.


#### broad-fc-getzlab-workflows/CGA_WES_Characterization_TCGAControlledAccess {#broad-fc-getzlab-workflows-cga_wes_characterization_tcgacontrolledaccess}

This workspace contains the published vesion of the CGA mutation calling pipeline with a full description of the method.


#### vanallen-firecloud-dfci/vanallen_storage {#vanallen-firecloud-dfci-vanallen_storage}

Brendan Reardon, an ACB in the Van Allen lab, shared this workspace with Mahmoud Ghandi and me so that we would be able to have access to annotation files necessary when testing out parts of the Van Allen pipeline on CCLE samples.


## Created methods {#created-methods}

Below are all of the methods that I have created in FireCloud. 


#### gkugener/Aggregate_CN_seg_files {#gkugener-aggregate_cn_seg_files}

Simple workflow that aggregates segment files outputted from the GATK copy number pipeline.

For an example, see how it is used in broad-firecloud-ccle/DepMap_WES_CN_hg38


#### gkugener/Aggregate_files_set {#gkugener-aggregate_files_set}

More generalized workflow that aggregates samples outputted from a pipeline. It accepts a custom Rscript as an argument that will be used to aggregate the results.

For an example, see how it is used in broad-firecloud-ccle/DepMap_hg38_RNAseq


#### gkugener/ArrayOfFilesToTxt {#gkugener-arrayoffilestotxt}

This task was used to take a FireCloud list of file paths and convert it into a txt with a single file per line. This task was developed specifically for realignment to hg38.

For an example, see how it is used in broad-firecloud-ccle/DepMap_WES_CN_hg38


#### gkugener/download_from_synapse {#gkugener-download_from_synapse}

This method was developed to be able to download samples from synapse using the synapse command line client. Works for a single sample at a time, appears to overwhelm the server when done across many samples at once.

For an example, see how it is used in single-cell-prism/Beat_AML.


#### gkugener/ExonUsage_fixed {#gkugener-exonusage_fixed}

reads rna junction read count file and generates exon usage ratios. This is identical to Mahmoud Ghandi’s exon usage ratios script except that if is able to handle chromosomes prefixed with ‘chr’.

For an example, see how it is used in broad-firecloud-ccle/DepMap_hg38_RNAseq


#### gkugener/filterMAF_on_CGA_pipeline {#gkugener-filtermaf_on_cga_pipeline}

Runs the filterMAF (germline filtering task) after mutation calling but is able to handle the MAF outputted by the CGA pipeline.

For an example, see how it used in broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline


#### gkugener/RNA_Fusion_Analysis_v1.2.0 {#gkugener-rna_fusion_analysis_v1-2-0}

This is the same as the RNA_Fusion_Analysis_v1.2.0 developed by Brian Granger except that it allows you to specify memory and disk space parameters.


#### gkugener/STAR_fusion {#gkugener-star_fusion}

An implementation of STAR-Fusion. Uses GTEx’s samtofastq conversion task.

For an example, see how it is used in single-cell-prism/Beat_AML


#### gkugener/STAR_Fusion_after_STAR {#gkugener-star_fusion_after_star}

An implementation of STAR-Fusion that was to be run on existing outputs from STAR. Works however it will not work correctly if the annotations and references <span style="text-decoration:underline;">do not match appropriately</span> with the version of STAR that was run.


#### gkugener/STAR_RSEM_from_FASTQ {#gkugener-star_rsem_from_fastq}

Runs the STAR, RSEM tasks from GTEx starting from a fastq. It also contains the option to delete intermediate files.

For an example, see how it is used in single-cell-prism/Beat_AML


#### gkugener/synapse_combine_fastqs {#gkugener-synapse_combine_fastqs}

Concatenates fastqs based on a txt file of fastq paths of fastqs that should be concatenated. This was used to process the Beat AML data.

For an example, see how it is used in single-cell-prism/Beat_AML


#### gkugener/WES_DM_Mutation_Calling_Pipeline {#gkugener-wes_dm_mutation_calling_pipeline}

A clone of the CCLE mutation calling pipeline prior to using CGA’s pipeline.